---
title: "The Childhood Interstitial Lung Disease (chILD) Registry: A Key Instrument in the Study of Rare Childhood Lung Disease"
author: "Katie Tsukahara"
output: 
  html_document:
    toc: false 
    depth: 3 
    theme: paper 
    highlight: tango
bibliography: BMIN503_Final_bib.bib
csl: american-medical-association.csl
---
```{r set-options, echo=FALSE, cache=FALSE}
options(width = 400)
```  
***

### Overview

Childhood interstitial lung disease (chILD) is a heterogeneous group of rare childhood lung diseases affecting the interstitial region of the lung, leading to abnormal gas exchange and a range of morbidity and mortality. Research and development of therapies is challenging due to the low prevalence and heterogeneity of disease, thus the CHILD Research Network and Registry was established to collect data from multiple centers across the United States. Our goals are to compare patient demographics across subsets of chILD subjects by diagnosis, and to examine the diagnostic factors that contribute to diagnosis. Of particular interest is the use of invasive vs non-invasive diagnostics in the workup of chILD. Data for this project was from a subset of the national registry established by the Children’s Interstitial and Diffuse Lung Disease Research Network (ChILDRN). 

For this project, I met with Dr. Lisa Young, Chief of the Division of Pulmonary and Sleep Medicine at Children's Hospital of Philadelphia (CHOP), who is also the principal investigator for the Children’s Interstital Lung Disease (chILD) registry. We discussed prior work using the registry data and possible areas of investigation including: factors leading to diagnosis (genetic testing, imaging, pathology), factors leading to clinical suspicion for chILD, and factors associated with steroid-responsiveness. Due to the design and scope of the registry, we decided to focus on demographics and diagnostic factors. We decided to better characterize surfactant dysfunction disorders, as this is a well-defined group, and compare these subjects with others that would present at a similar age, i.e., in the first year of life. Devaney Camburn, Manager of the Pulmonary Clinical Research Programs at CHOP, helped introduce me to the REDCap platform and registry design, including navigation through the database, data report generation, and export of data in a manner in line with the IRB. I was introduced to Dr. Xiao Chang of The Center for Applied Genomics at CHOP through another project. We met to discuss the importance of accurate phenotyping in large studies and the challenges of obtaining this data from the electronic medical record. Fortunately, in this case, the registry is designed to capture and house the data of interest, which facilitates analysis with minimal data cleaning. Thus, for the purposes of this project, it was not necessary to obtain EMR data; however, this is something we may explore in the future.

The Rmd and html files for this project can be found [here](https://github.com/ktsukah/BMIN503_Final_Project). Due to the sensitive nature of the data, only summary statistics are provided and the raw data is not made available.


### Introduction 
Childhood interstitial lung disease (chILD) encompasses a variety of lung diseases that manifest via changes to the pulmonary interstitium. The pulmonary interstitium is, simply put, the lung tissue between the airspace and the pulmonary blood vessels, and excludes other lung structures such as the airways. These diseases are grouped together owing to similarities in the area of the lung affected and in clinical presentation. While both children and adults can develop interstitial lung disease (ILD), many causes of chILD are seen only in the pediatric age group. These diseases can be genetic in origin, or associated with abnormal lung development, immunologic dysfunction, rheumatologic disease, or infectious history. Some cases do not have a clear etiology. Certain diagnoses confer a poor prognosis of early lung transplant or mortality. Other diagnoses confer a more positive prognosis, including clinical improvement and resolution of disease over early childhood.
<br><br>To study this group of diseases, the Children’s Interstitial and Diffuse Lung Disease Research Network (ChILDRN) established a national registry in 2016, with participation by 13 sites across the United States. The original registry contains 436 subjects, enrolled from 2016 to 2020, with data on demographics, medical history, diagnostics, management and outcomes. With this project, we aim to examine differences in patient demographics, including age of onset and severity of presentation, among the more common chILD diagnoses. We also aim to examine the diagnostic and clinical factors that contribute to diagnosis, if possible examining the role that different diagnostic tests play in diagnosis and whether invasive testing can be avoided in these patients that are at higher risk for procedures due to their underlying pulmonary disease. The database and this subsequent project use the skills and expertise of clinicians across many fields (Pulmonology, Radiology, Pathology, Rheumatology, and others), clinical researchers, basic science researchers, and biomedical informaticians to collect, manage, and interpret data on this complex group of patients.


### Methods
The Children’s Interstitial and Diffuse Lung Disease Research Network (ChILDRN) established a national registry for chILD under an IRB for a longitudinal observational study with the participation of 13 centers across the United States [@YoungOA3786]. This has since expanded to 20 centers, with over 400 enrolled subjects. The IRB allows for both retrospective enrollment as well as prospective data collect, including clinical characteristics as well as diagnostic testing, longitudinal results, and outcomes such as supplemental oxygen use, mechanical ventilation or death. The data is housed by the REDCap (Research Electronic Data Capture) data platform. [@HARRIS2019103208; @HARRIS2009377] For the purposes of this project, data of interest was filtered using the REDCap Report feature, then de-identified and date-shifted data was exported for analysis, in accordance with the IRB. A CSV file contained the raw data, and the R file contained R code to read in the CSV, and appropriately label columns, data type, and factor levels using the *Hmisc* package. Further data cleaning was performed using the *tidyr* and *dplyr* packages.

```{r, include=TRUE, warning = FALSE, message = FALSE}
# Loading RedCAP data
setwd("C:/Users/tsukaharak/Documents/R files/BMIN503_Final_Project")
source("CHOPChildhoodILDNati-110620BMIN_R_2020-11-27_1739.R", local = knitr::knit_global())
```

A cursory check for duplicate entries was performed, showing that all entries have a unique chILD Registry ID. There are 436 total entries.

```{r, echo = TRUE, warning = FALSE, message = FALSE}
# A cursory check for duplicate entries, showing that all entries have a unique chILD Registry ID
print(paste(length(data$child_registry_id), "entries;", length(unique(data$child_registry_id)), "unique chILD Registry IDs."))

# Copying the data frame to create a working version, so that if going back to the original data is required, 
# it will not require reloading in all of the data
child <- data

```

After loading in the data, basic enrollment demographics were examined. This included center where the child was enrolled, age at enrollment, age at diagnosis, race and ethnicity. Graphics were built using *ggplot2*, *cowplot*, *maps* and *mapproj*. Additional statistical analysis was performed using *base* R.


```{r, echo = FALSE, warning = FALSE, message = FALSE}
# loading libraries for data cleaning and graphics
library(tidyr)
library(dplyr)
library(ggplot2)
library(cowplot)

```

### Results

#### Registry Characteristics

The initial phase of the project was an exploration of enrollment and demographic characteristics. In order to create a visualization of enrollment by center, a new data frame was created with with city and chILD Registry ID prefix corresponding to the enrollment center. Latitude and longitude were retrieved from the *us.cities* data frame in the *maps* package. Due to two centers being located in New York city boroughs, these coordinates were added manually as they were not included in the *us.cities* data frame. These were then plotted on a map of the continental United States. These geographic locations are representative only of the location of enrollment. It is not uncommon for children with rare disease to travel to a specialized center for, thus, it is difficult to infer any geographic distribution of the patient's location of residence from these numbers.

```{r, warning = FALSE, message = FALSE, eval = TRUE}
library(maps)
library(mapproj)

DAG.df <- data.frame(
  name = c("Boston MA", "Atlanta GA", "Philadelphia PA", "Cleveland OH", "Queens NY", "Aurora CO", "Indianapolis IN", "Kansas City MO", "Chicago IL", "Ann Arbor MI", "Columbia MO", "Bronx NY", "Portland OR", "Pittsburgh PA", "San Diego CA", "Seattle WA", "Palo Alto CA", "Chapel Hill NC", "Nashville TN", "Saint Louis MO"),
  prefix = c(704,720,712,709,702,703,718,711,717,716,715,708,713,705,706,710,707,719,701,714),
  stringsAsFactors = FALSE
)

DAG.df <- DAG.df %>% 
  arrange(prefix) %>%
  left_join(us.cities, by = "name") %>%
  mutate(prefix = as.character(prefix))

# Manually adding latitude and longitude for NY boroughs:
# Queens 40.728, -73.795
# Bronx 40.845, -73.865
DAG.df[which(grepl("Queens",DAG.df$name)), ]$lat <- 40.7282
DAG.df[which(grepl("Queens",DAG.df$name)), ]$long <- -73.7949
DAG.df[which(grepl("Bronx",DAG.df$name)), ]$lat <- 40.8448
DAG.df[which(grepl("Bronx",DAG.df$name)), ]$long <- -73.8648

# Grouping the subjects by enrollment center using the registry ID prefix, then joining with the geographic location data frame 
DAG.child <- child %>%
  select(child_registry_id) %>%
  mutate(prefix = substr(child_registry_id, 0, 3)) %>%
  count(prefix) %>%
  mutate(prefix = as.character(prefix)) %>%
  left_join(DAG.df, by = "prefix")

# Build the map using United States map data from the _maps_ package
USA <- map_data("world") %>% filter(region=="USA")
mybreaks <- c(20,40,60,80,100)
ggplot() +
    geom_polygon(data = USA, aes(x=long, y = lat, group = group), fill="grey", alpha=0.4) +
    geom_point(data = DAG.child, aes(x=long, y=lat, size=n), color = "#2b8cbe", shape=20, stroke=FALSE, alpha = 0.6) +
    scale_size_continuous(name="Number of Enrolled \nSubjects", trans="identity", range=c(0,12), breaks = mybreaks) +
    theme_void() +
    coord_map() + ylim(25,50) + xlim(-130, -60) +
    guides(colour = guide_legend()) +
    ggtitle("Subject Enrollment by Center in the chILD Registry") +
    theme(
      legend.position = c(0.85, 0.15),
      text = element_text(color = "#22211d"),
      plot.title = element_text(size= 16, hjust=0.1, color = "#4e4d47", margin = margin(b = -0.1, t = 0.4, l = 2, unit = "cm")),
    )

```

This study allowed for retrospective enrollment as well as prospective data collection. Using summary statistics and histograms, the age at enrollment in the Registry and age at diagnosis were examined. When the Registry was first established, it was expected that many subjects would be enrolled retrospectively. 

```{r, warning = FALSE, message = FALSE}
# Age at enrollment
summary(child$age_studyentry)
ggplot(data = child, aes(x = age_studyentry)) + 
  geom_histogram(color = "black", fill = "white", binwidth=3) +
  labs(x = "Age (in months)", title = "Age at Study Entry in the chILD Registry")
# Age at diagnosis
summary(child$age_diagnosis)
ggplot(data = child, aes(x = age_diagnosis)) + 
  geom_histogram(color = "black", fill = "white", binwidth=3) +
  labs(x = "Age (in months)", title = "Age at Diagnosis in the chILD Registry")
```

We were also interested to see if subjects are being enrolled closer to their time of diagnosis now that the Registry is more established. This was plotted as a scatter plot below, and showed that there continues to be a significant amount of retrospective enrollment, with many subjects being enrolled up to several years after their diagnosis. The reasons for this are unclear, and may reflect a growing awareness of the chILD Registry, the addition of new registry centers, a move towards providing care for chILD at specialized centers, or the need for further resources devoted to recruitment.

```{r, warning = FALSE, message = FALSE}
time.to.enroll <- child %>% 
  select(age_studyentry, age_diagnosis, cons_date) %>%
  mutate(age.diff = age_studyentry - age_diagnosis)
time.to.enroll <- time.to.enroll %>%
  mutate(cons_date = as.Date(time.to.enroll$cons_date))
ggplot(time.to.enroll, aes(x=cons_date, y=age.diff)) +
  geom_point() +
  ggtitle("Time Intervals between Diagnosis and Enrollment in the chILD Registry") +
  ylab("Time from Diagnosis to Enrollment (in Months)") +
  xlab("Approximate Time of Enrollment (Date-Shifted)") +
  theme(axis.text.x=element_text(angle=60, hjust=1)) +
  scale_x_date(date_breaks = "1 year", date_labels = "%Y")

```

Due to how the database was structured regarding race, significant data rearrangement was necessary in order to summarize the data. The database contained 6 race categories, each with a possible yes or no answer. This was synthesized into individual races, "two or more races," "other," or "unknown." 

```{r}
# Race across the registry - rearranging and cleaning data to better summarize racial characteristics
child.race <- child %>%
  select(child_registry_id, race___1:race___6, race___1.factor:race___6.factor, diagnosis.factor)
child.race <- child.race %>%
  mutate(race.sum = rowSums(select(child.race, race___1:race___6), na.rm = TRUE)) %>%
  mutate(race.text = NA)

table(child.race$race.sum)

N = nrow(child.race)
for(i in 1:N){
  if(child.race[i,"race.sum"] == 0){
    child.race[i,"race.text"] <- "Unknown"
  }else if(child.race[i,"race.sum"] == 2){
    child.race[i,"race.text"] <- "Two or more races"
  }else{
    if(child.race[i,"race___1"] == 1){
      child.race[i,"race.text"] <- "White"
    }else if(child.race[i,"race___2"] == 1){
      child.race[i,"race.text"] <- "Black"
    }else if(child.race[i,"race___3"] == 1){
      child.race[i,"race.text"] <- "Asian"
    }else if(child.race[i,"race___4"] == 1){
      child.race[i,"race.text"] <- "Native Hawaiian/Pacific Islander"
    }else if(child.race[i,"race___5"] == 1){
      child.race[i,"race.text"] <- "American Indian or Alaska Native"
    }else if(child.race[i,"race___6"] == 1){
      child.race[i,"race.text"] <- "Other"
    }else{print(i)}
  }
}
child.race <- child.race %>%
  mutate(race.text = factor(race.text, levels = c("White", "Black", "Asian", "Native Hawaiian/Pacific Islander", "American Indian or Alaska Native", "Other", "Two or more races", "Unknown")))
```


Overall demographic and clinical characteristics of the registry are shown below. Further analysis was then performed to compare clinical characteristics across diagnoses. 

```{r, warning = FALSE, message = FALSE}
# Table 1 - summarizing select demographic and clinical characteristics
# Was not able to download ethnicity data, so this is not included.
library(gtsummary)
library(gt)

table.data <- child %>%
  dplyr::select(child_registry_id, age_studyentry, age_diagnosis, patient_gender.factor,  relative_in_registry.factor, fibrosis.factor, airway.factor, home_o2.factor, mech_vent___1.factor, mech_vent___2.factor, transplant_yn.factor) 
levels(table.data$mech_vent___1.factor)=c("No","Yes")
levels(table.data$mech_vent___2.factor)=c("No","Yes")

table.data <- table.data %>%
  rename("Patient Gender" = patient_gender.factor) %>%
  rename("Relative in Registry" = relative_in_registry.factor) %>%
  rename("Pulmonary Fibrosis" = fibrosis.factor) %>%
  rename("Airway Disease" = airway.factor) %>%
  rename("Past or Current Home Supplemental Oxygen" = home_o2.factor) %>%
  rename("Past or Current Noninvasive Ventilation (>3 weeks)" = mech_vent___1.factor) %>%
  rename("Past or Current Invasive Ventilation (>3 weeks)" = mech_vent___2.factor) %>%
  rename("Lung Transplant" = transplant_yn.factor) %>%
  mutate(Death = ifelse(!is.na(child$death_age), "Yes", "No"))

table.data.race <- left_join(table.data, select(child.race, child_registry_id, race.text), by = "child_registry_id") %>%
  relocate(race.text, .after = "Patient Gender") %>%
  rename("Subject race" = race.text)

table.two <- tbl_summary(select(table.data.race, -child_registry_id),  missing = "no") %>%
as_gt() %>%
gt::tab_header("Table 1. Summary of Clinical Characteristics of the Study Cohort") %>%
opt_align_table_header(align = "left") %>%
gt::tab_options(table.font.size = '11px', table_body.hlines.style = "none", data_row.padding = px(1))
table.two

```

#### Characteristics by Diagnosis

Childhood interstitial lung disease includes a large number of disorders. Often, these children present similarly, with respiratory distress or failure, and failure to thrive (inadequate weight gain). However, different diagnoses confer very different prognoses, thus the workup is critical. After examining overall registry enrollment, I wanted to examine differences in demographics and clinical characteristics across primary diagnoses. The distribution of primary diagnoses is shown below.

```{r, warning = FALSE, message = FALSE}
# Distribution of primary diagnoses
ggplot(data = child, aes(x = diagnosis.factor)) + 
  geom_bar() + 
  theme(axis.text.x = element_text(angle = 45, hjust = 1)) +
  labs(x = "Primary Diagnosis", y = "Number of Subjects", title = "Primary Diagnoses in chILD Registry")
```

The overall database showed a small male predominance (56%), which had significant variability across diagnoses. Of note, those with a male predominance tend to coincide with diagnoses that are diagnosed earlier, as we will see in the next graph. In patients born prematurely, there are higher rates of bronchopulmonary dysplasia (chronic lung disease of prematurity) among male patients [@Mehta]. It would be interesting to explore whether there are intrinsic features of the young lung that lead to increased lung disease. It is possible that female patients have less severe presentations, have only subclinical disease, and thus do not undergo workup and diagnosis. 

```{r, warning = FALSE, message = FALSE}
# Analysis of gender across diagnoses - chi square test and visualization
p1chi <- chisq.test(table(child$diagnosis.factor, child$patient_gender.factor))
p1chi

ggplot(data = child, aes(x = diagnosis.factor, fill = patient_gender.factor)) + 
  geom_bar(position = "fill") + 
  theme(axis.text.x = element_text(angle = 45, hjust = 1)) +
  labs(x = "Primary Diagnosis", y = "Proportion of Total", title = "Gender by Diagnosis in chILD Registry", fill = "Gender",subtitle = paste("Chi-squared p value =", format(p1chi$p.value, digits = 3)))

```

As is known from the literature, certain diagnoses are prevalent at different ages. Disorders of growth or development, neuroendocrine cell hyperplasia of infancy (NEHI), pulmonary interstitial glycogenolysis (PIG), and surfactant dysfunction disorder tend to present early in life, as is shown below.

```{r, echo = TRUE, warning = FALSE, message = FALSE}
# Age across diagnosis
summary(child$age_diagnosis, na.rm = TRUE)

ggplot(data = filter(child, child$diagnosis < 10), aes(x = diagnosis.factor, y = age_diagnosis)) + 
  geom_boxplot() + 
  theme(axis.text.x = element_text(angle = 45, hjust = 1)) +
  labs(x = "Primary Diagnosis", y = "Age (in Months)", title = "Age at Diagnosis in chILD Registry")

library(kableExtra)
m <- child %>%
  group_by(diagnosis.factor) %>%
  dplyr::summarize(median(as.numeric(age_diagnosis), na.rm = TRUE))
knitr::kable(m, col.names = c("Primary Diagnosis", "Median Age of Diagnosis (in Months)"), format = "html", table.attr = "style='width:70%;'") %>%
    kable_styling(position = "center")

```


The predominant race of the registry is white (72%), which holds true across all diagnoses. This may be related to recruitment bias mores than true differences in disease prevalence, however, for the purposes of this project we do not have a comparison group to assess recruitment. Interestingly, for patients with pulmonary hemorrhage, there appears to be an increase in the proportion of black patients. The reasons for this are unclear.

```{r, echo = TRUE, warning = FALSE, message = FALSE}
# Analysis of race across diagnoses
# Adding the race as factor column if it has not already been added
if(!"race.text" %in% colnames(child)){
child <- left_join(child, select(child.race, child_registry_id, race.text), by = "child_registry_id")
}

ggplot(data = child, aes(x = diagnosis.factor, fill = race.text)) + 
  geom_bar(position = "fill") + 
  theme(axis.text.x = element_text(angle = 45, hjust = 1)) +
  labs(x = "Primary Diagnosis", y = "Proportion of Total", title = "Race by Diagnosis in chILD Registry", fill = "Race")

```



#### Diagnostic Factors

After characterizing differences in patient demographics among diagnoses, I was interested in the factors contributing to making the diagnosis. In childhood ILD, once a clinical suspicion is established, the next step is typically chest computed tomography (CT), which provides a much higher level of imaging resolution that chest radiograph (X-ray). From there, flexible bronchoscopy with bronchoalveolar lavage (BAL) is often performed to evaluate for certain disease processes, such as pulmonary hemorrhage or pulmonary alveolar proteinosis, and to rule out other causes. From there, the decision is made to proceed with genetic testing or lung biopsy or both. However, though there are  American Thoracic Society Clinical Practice Guidelines,[@Kurland23905526] this workup is not strictly standardized, and can be affected by many factors, including patient age, disease severity, and available resources.

The graphic below demonstrates which diagnostic modality made the diagnosis. Many patients required multiple diagnostic modalities to obtain the diagnosis, which, as the workup is often a stepwise process, is not surprising. Characteristic chest CT findings have been established for NEHI, which is consistent with the large number of cases the required only chest CT to make the diagnosis. It is interesting that there are also many cases that underwent multiple diagnostic modalities. It is possible that these represent earlier cases when lung biopsy with bombesin staining was the gold standard.[@OConnor26540427] The ATS clinical practice guidelines now recommend diagnosis of NEHI by imaging only.

```{r, warning = FALSE, message = FALSE}
# Looking at diagnostic modality by diagnosis

child.mod <- select(child, child_registry_id:diagnosis, fibrosis.factor:diagnosis.factor)
child.mod <- child.mod %>%
  mutate(modality = ifelse(child.mod$child_dia2___1 == 1 & child.mod$child_dia2___2 == 0 & child.mod$child_dia2___3 == 0 & child.mod$child_dia2___4 == 0, "CT only", NA)) %>%
  mutate(modality = ifelse(child.mod$child_dia2___2 == 1 & child.mod$child_dia2___1 == 0 & child.mod$child_dia2___3 == 0 & child.mod$child_dia2___4 == 0, "Biopsy only", modality)) %>%
  mutate(modality = ifelse(child.mod$child_dia2___3 == 1 & child.mod$child_dia2___1 == 0 & child.mod$child_dia2___2 == 0 & child.mod$child_dia2___4 == 0, "Genetic only", modality)) %>%
  mutate(modality = ifelse(child.mod$child_dia2___2 == 0 & child.mod$child_dia2___1 == 0 & child.mod$child_dia2___3 == 0 & child.mod$child_dia2___4 == 0, "No Diagnostic Modality Listed", modality)) %>%
  mutate(modality = ifelse(child.mod$child_dia2___4 == 1 & child.mod$child_dia2___1 == 0 & child.mod$child_dia2___2 == 0 & child.mod$child_dia2___3 == 0, "Other", modality)) %>%
  mutate(modality = ifelse(is.na(modality) == "TRUE", "Multiple", modality)) %>%
  mutate(modality.factor = as.factor(modality))
child.mod <- rename(child.mod, "Modality" = modality.factor)

tbl_summary(select(child.mod, "Modality"),  missing = "no") %>%
as_gt() %>%
gt::tab_header("Table 2. How was the diagnosis made?") %>%
opt_align_table_header(align = "left") %>%
gt::tab_options(table.font.size = '11px', table_body.hlines.style = "none", data_row.padding = px(1), table.width = px(300))


ggplot(data = filter(child.mod, !is.na(diagnosis)), aes(x = diagnosis.factor, fill = Modality)) +
  geom_bar(position = position_dodge(preserve = "single")) + 
  theme(plot.title = element_text(size=14, vjust=3),
  axis.text.x = element_text(angle = 45, hjust = 1, size = 8), 
  legend.title = element_text(size = 10), 
  legend.text = element_text(size = 8)) +
  labs(x = "Primary Diagnosis", fill = "Diagnostic Modality", title = "The chILD Registry: How was the diagnosis made?")

```

The following graph shows the frequency at which lung biopsy (the most invasive diagnostic test of the above options) contributed to the diagnosis across each diagnosis category. Lung pathology is necessary for several of the diagnoses, though it is now possible to diagnose a few of these diagnoses without biopsy. The data is consistent with this pattern.

```{r}
# looking at lung biopsy over diagnosis
ggplot(data = filter(child, !is.na(diagnosis.factor)), aes(x = diagnosis.factor, fill = child_dia2___2.factor)) + 
  geom_bar(position = "fill") + 
  theme(axis.text.x = element_text(angle = 45, hjust = 1)) +
  labs(x = "Primary Diagnosis", y = "Proportion of Total", title = "Lung Biopsy by Diagnosis in chILD Registry", fill = "Did lung biopsy make the diagnosis?")
```


Wheras NEHI can be diagnosed based on imaging findings alone, surfactant dysfunction disorder can be diagnosed by genetic testing. Both of these diseases tend to present early in life, during a period of time when the risks of anesthesia may be higher. The diagnostic modalities making the diagnosis of surfactant dysfunction disorder were examined more closely, showing that in the vast majority of cases, genetic testing played a role in diagnosis. Further studies may be pursued to characterize the types of genetic testing available over time and the limitations to early genetic testing, e.g., clinical severity, length of time to results, cost and/or insurance coverage, specialist availability, or provider level of comfort with interpreting results. (One subject was excluded from the analysis below due to having no diagnostic modality information coded.)

```{r, warning = FALSE, message = FALSE}
# A closer look at surfactant dysfunction disorder

child.surf <- filter(child, diagnosis.factor == "Surfactant dysfunction disorder")
print(paste0("In the chILD Registry, there are ", nrow(child.surf), " entries with a primary diagnosis of surfactant dysfunction disorder"))

ggplot(data = filter(child.mod, diagnosis.factor == "Surfactant dysfunction disorder"), aes(x = Modality)) +
  geom_bar() + 
  scale_x_discrete(drop=FALSE) +
  theme(axis.text.x = element_text(angle = 45, hjust = 1)) +
  labs(x = "Diagnostic Modality", title = "The chILD Registry: Diagnosis of Surfactant Dysfunction Disorders")


# colSums(select(child.surf, child_dia2___1:child_dia2___4), na.rm = TRUE)
# head(filter(child.surf, child_dia2___4 == 1) %>% select(child_registry_id:child_dia2___4)) - shows only 1 entry

library(VennDiagram)
library(magick)

venn.ct <- child.surf %>% filter(child_dia2___4 == 0) %>% select(child_registry_id, child_dia2___1) %>% filter(child_dia2___1 == 1) %>% mutate(venn.modality = "CT") %>% select(child_registry_id, venn.modality)
venn.bx <- child.surf %>% filter(child_dia2___4 == 0) %>% select(child_registry_id, child_dia2___2) %>% filter(child_dia2___2 == 1) %>% mutate(venn.modality = "Biopsy") %>% select(child_registry_id, venn.modality)
venn.gen <- child.surf %>% filter(child_dia2___4 == 0) %>% select(child_registry_id, child_dia2___3) %>% filter(child_dia2___3 == 1) %>% mutate(venn.modality = "Genetic Testing") %>% select(child_registry_id, venn.modality)
venn <- rbind(venn.ct, venn.bx, venn.gen)

venn.diagram(
  x = list(
    venn %>% filter(venn.modality == "CT") %>% select(child_registry_id) %>% unlist() , 
    venn %>% filter(venn.modality == "Biopsy")  %>% select(child_registry_id) %>% unlist() , 
    venn %>% filter(venn.modality == "Genetic Testing") %>% select(child_registry_id) %>% unlist()
    ),
  category.names = c("Chest Computed Tomography (CT)" , "Lung Biopsy" , "Genetic Testing"),
  filename = 'venn.png',
  output = TRUE ,
          imagetype="png" ,
          height = 600 , 
          width = 720 , 
          resolution = 300,
          compression = "lzw",
          lwd = 1,
          col=c("#440154ff", '#21908dff', '#fde725ff'),
          fill = c(alpha("#440154ff",0.3), alpha('#21908dff',0.3), alpha('#fde725ff',0.3)),
          cex = 0.5,
          fontfamily = "sans",
          cat.cex = 0.3,
          cat.default.pos = "outer",
          cat.pos = c(-27, 27, 120),
          cat.dist = c(0.070, -0.050, 0.055),
          cat.fontfamily = "sans",
          cat.col = c("#440154ff", '#21908dff', '#fde725ff'),
          rotation = 1, 
          main = "Diagnostic Modalities in Surfactant Dysfunction Disorder",
          main.fontfamily = "sans",
          main.cex = 0.4
        )

img <- magick::image_read('C:/Users/tsukaharak/Documents/R files/BMIN503_Final_Project/venn.png')
plot(img) # or print(img)

```

Since before the establishment of the chILD registry, genetic testing has continued to evolve. Tests that were initially available only through research studies are now routinely done as part of clinical care. To see how the frequency and yield of genetic testing has changed over time, this was plotted against the date of testing. There were a few data points that did not have a date provided for genetic testing; these points are excluded on the first graph.

```{r, warning = FALSE, message = FALSE}

library(lubridate)
# will look at genetic testing - which genes tested, if rate of genetic testing has increased over time
child.surf.gen <- child %>%
  filter(diagnosis.factor == "Surfactant dysfunction disorder") %>%
  filter(genetics.factor == "Yes")
child.surf.gen <- child.surf.gen %>%
  mutate(date_genetics.date = ymd(child.surf.gen$date_genetics))
if(!"race.text" %in% colnames(child.surf.gen)){
child.surf.gen <- left_join(child.surf.gen, select(child, child_registry_id, race.text), by = "child_registry_id")
}

# Looking at number of genetic tests over time in patients with surfactant dysfunction disorder
ggplot(child.surf.gen, aes(x=date_genetics.date)) +
  geom_histogram(binwidth = 365) +
  ggtitle("Genetic Testing over Time in the chILD Registry", subtitle = "Subjects with Surfactant Dysfunction Disorder who Had Genetic Testing") +
  ylab("Number of Subjects with Genetic Tests") +
  xlab("Approximate Time of Genetic Testing (Date-Shifted)") +
  theme(axis.text.x=element_text(angle=60, hjust=1)) +
  scale_x_date(date_breaks = "1 year", date_labels = "%Y")
```

To examine the role of genetic testing among total cases, we examined the data over the approximate date of diagnosis. In this way, we were able to see how the frequency and success of genetic testing has changed over time. The database does not include the date or age at which a patient first presented, so as a surrogate for the time period that the workup took place, an approximate date of diagnosis was used. Dates have been date-shifted up to 364 days, and in calculating the approximate diagnosis date this was rounded to the nearest month, however, this does provide some insight into overall trends over approximately 15 years. The last year of each data set likely represents less than a full year of data, thus the lower numbers are likely due to ongoing data collection rather than a true decrease in patient numbers or testing.

We initially looked at a smaller group of patients with diagnosed surfactant dysfunction disorder. Three genes involved in surfactant production and metabolism are known to cause surfactant dysfunction [@Wert]. As genetic testing has improved, the ability to detect mutations in these genes has become more clinically available. We wanted to examine how many of these patients with an eventual diagnosis of surfactant dysfunction disorder had had genetic testing as part of their diagnosis, what type, and whether the testing was able to detect a diagnostic variant. As we saw earlier, genetic testing contributed to diagnosis in the vast majority of cases. 

```{r}
# Looking at number of SDD diagnosed and proportion for which genetic testing gave the diagnosis
surf.time <- child %>% 
  filter(diagnosis.factor == "Surfactant dysfunction disorder") %>%
  mutate(age.diff = age_studyentry - age_diagnosis)
surf.time <- surf.time %>%
  mutate(cons_date = ymd(surf.time$cons_date))
surf.time <- surf.time %>%
  mutate(diag_date = surf.time$cons_date %m-% months(round(surf.time$age.diff)))
surf.time <- surf.time %>%
  mutate(genetics.factor_desc = factor(genetics.factor, levels = c("Unknown", "No", "Yes")))
N <- nrow(surf.time)
# Replacing NA with "Unknown"
for(i in 1:N){
  if(is.na(surf.time[i,"genetics.factor"])){
     surf.time[i,"genetics.factor_desc"] <- "Unknown"
  }
  }


ggplot(surf.time, aes(x=diag_date, fill = child_dia2___3.factor)) +
  geom_histogram(binwidth = 365, alpha = 0.8) +
  ggtitle("Genetic Testing over Time in the chILD Registry", subtitle = "Subjects with Diagnosed Surfactant Dysfunction Disorder") +
  ylab("Number of Subjects Diagnosed with \nSurfactant Dysfunction Disorder") +
  xlab("Approximate Time of Diagnosis (Date-Shifted)") +
  labs(fill = "Was the Diagnosis Made\nwith Genetic Testing?") +
  theme(axis.text.x=element_text(angle=60, hjust=1)) +
  scale_x_date(date_breaks = "1 year", date_labels = "%Y") +
  guides(fill = guide_legend(reverse = TRUE))
```

There are significant differences in cost and turnaround among different types of genetic tests. Targeted tests are typically cheaper and faster, though done at the risk of missing disease-causing genes that are not included on the panel. To see what kind of genetic testing was done, this was plotted below. Due to the high rate of genetic testing contributing to diagnosis, it is difficult to appreciate differences in yield based on type of genetic testing, as all 3 appear to have high rates of identifying disease-causing genes.


```{r,message=FALSE, warning=FALSE}
# Type of genetic testing performed in SDD
gg.gen <- child.surf.gen %>%
  select(child_registry_id, type_genetic_test___1.factor:type_genetic_test___4.factor) %>%
  rename("Single Gene" = type_genetic_test___1.factor) %>%
  rename("Gene Panel" = type_genetic_test___2.factor) %>%
  rename("Whole Exome Sequencing (WES)" = type_genetic_test___3.factor) %>%
  rename("Whole Genome Sequencing (WGS)" = type_genetic_test___4.factor)
gg.gen <- tidyr::pivot_longer(gg.gen, cols = "Single Gene":"Whole Genome Sequencing (WGS)", names_to = "type_genetic_test.long")
gg.gen <- filter(gg.gen, value == "Checked")
gg.gen <- left_join(gg.gen, select(child, child_registry_id, child_dia2___3.factor), by = "child_registry_id")


ggplot(gg.gen, aes(x = type_genetic_test.long, fill = child_dia2___3.factor)) +
  geom_bar() +
  ggtitle("Type of Genetic Testing in Surfactant Dysfunction Disorder") +
  ylab("Number of Subjects with Genetic Tests") +
  xlab("Type of Genetic Testing") +
  labs(fill = "Was the Diagnosis Made\nwith Genetic Testing?") +
  theme(axis.text.x=element_text(angle=60, hjust=1)) 


```



Though it is helpful to understand the path to diagnosis for patients with surfactant dysfunction disorder, most patients present un-differentiated. Thus, we wanted to understand the role for genetic testing in the infant period. We broadened our group of interest to all diagnoses, for all patients. 

```{r}
# Now looking at genetic testing for all entrants in the registry
child.gen <- child %>%
  mutate(date_genetics.date = ymd(child$date_genetics))%>%
  mutate(age.diff = age_studyentry - age_diagnosis) %>%
  mutate(cons_date = ymd(child$cons_date)) %>%
  filter(age.diff > 0)
child.gen <- child.gen %>%
  mutate(diag_date = child.gen$cons_date %m-% months(round(child.gen$age.diff)))
child.gen <- child.gen %>%
  mutate(genetics.factor_desc = factor(genetics.factor, levels = c("Unknown", "No", "Yes")))
N <- nrow(child.gen)
# Replacing NA with "Unknown"
for(i in 1:N){
  if(is.na(child.gen[i,"genetics.factor"])){
     child.gen[i,"genetics.factor_desc"] <- "Unknown"
  }
}

# For all patients in the chILD registry, how many received genetic testing? Measured by age at diagnosis, as the year they were undergoing diagnostic workup may have an association with the frequency of genetic testing given the advancing technology
ggplot(child.gen, aes(x=diag_date, fill = genetics.factor_desc)) +
  geom_histogram(binwidth = 365, alpha = 0.8) +
  ggtitle("Genetic Testing over Time in the chILD Registry") +
  ylab("Number of Subjects Diagnosed with chILD") +
  xlab("Approximate Time of Diagnosis\n(Date-Shifted and Rounded to Nearest Month)") +
  labs(fill = "Was Genetic Testing Done?") +
  theme(axis.text.x=element_text(angle=60, hjust=1)) +
  scale_x_date(date_breaks = "1 year", date_labels = "%Y") +
  guides(fill = guide_legend(reverse = TRUE))

# For all patients in the chILD registry who DID receive genetic testing, did it make the diagnosis?
ggplot(filter(child.gen, genetics.factor == "Yes"), aes(x=diag_date, fill = child_dia2___3.factor)) +
  geom_histogram(binwidth = 365, alpha = 0.8) +
  ggtitle("Genetic Testing over Time in the chILD Registry:", subtitle = "For patients who received genetic testing, did it make the diagnosis?") +
  ylab("Number of Subjects Diagnosed with chILD") +
  xlab("Approximate Time of Diagnosis\n(Date-Shifted and Rounded to Nearest Month)") +
  labs(fill = "Did Genetic Testing\nMake the Diagnosis?") +
  theme(axis.text.x=element_text(angle=60, hjust=1)) +
  scale_x_date(date_breaks = "1 year", date_labels = "%Y") +
  guides(fill = guide_legend(reverse = TRUE))

# For all patients in the chILD registry who DID receive genetic testing, what type of genetic testing did they receive?
gg.gen.full <- child %>%
  select(child_registry_id, type_genetic_test___1.factor:type_genetic_test___4.factor) %>%
  rename("Single Gene" = type_genetic_test___1.factor) %>%
  rename("Gene Panel" = type_genetic_test___2.factor) %>%
  rename("Whole Exome Sequencing (WES)" = type_genetic_test___3.factor) %>%
  rename("Whole Genome Sequencing (WGS)" = type_genetic_test___4.factor)
gg.gen.full <- tidyr::pivot_longer(gg.gen.full, cols = "Single Gene":"Whole Genome Sequencing (WGS)", names_to = "type_genetic_test.long")
gg.gen.full <- filter(gg.gen.full, value == "Checked")
gg.gen.full <- left_join(gg.gen.full, select(child, child_registry_id, child_dia2___3.factor), by = "child_registry_id")
gg.gen.full <- left_join(gg.gen.full, select(child.gen, child_registry_id, diag_date, age_diagnosis), by = "child_registry_id")
gg.gen.full <- gg.gen.full %>% 
  mutate(type_genetic_test.long = factor(type_genetic_test.long, levels = c("Gene Panel", "Single Gene", "Whole Exome Sequencing (WES)", "Whole Genome Sequencing (WGS)")))
gg.gen.full <- gg.gen.full %>% 
  mutate(type_genetic_test.long = factor(type_genetic_test.long, levels = c("Whole Genome Sequencing (WGS)", "Whole Exome Sequencing (WES)", "Single Gene", "Gene Panel")))

ggplot(gg.gen.full, aes(x=diag_date, fill = type_genetic_test.long)) +
  geom_histogram(binwidth = 365, alpha = 0.8) +
  ggtitle("Type of Genetic Testing over Time in the chILD Registry", subtitle = "All Subjects with Genetic Testing") +
  ylab("Number of Tests") +
  xlab("Approximate Time of Diagnosis (Date-Shifted)") +
  labs(fill = "Type of Genetic Testing") +
  theme(axis.text.x=element_text(angle=60, hjust=1)) +
  scale_x_date(date_breaks = "1 year", date_labels = "%Y")


```


To better compare our prior group of patients with surfactant dysfunction disorder with other undifferentiated patients that would present similarly, we then narrowed our focus to children who were diagnosed prior to 1 year old. As we saw earlier, age at entry into the study does not coincide with age of diagnosis, so would not likely be correlated to age at presentation either. So patients were filtered by age at diagnosis. If the age of diagnosis was less than 12 months, this is strongly suggestive that they presented earlier than 12 months (as typically these patients would not undergo such an extensive workup, including the risks of radiation from chest imaging, without symptoms).

```{r}

# Is genetic testing frequent among younger patients? Filtered by age of diagnosis at 12 months or less
ggplot(filter(child.gen, age_diagnosis <= 12), aes(x=diag_date, fill = genetics.factor_desc)) +
  geom_histogram(binwidth = 365, alpha = 0.8) +
  ggtitle("Genetic Testing over Time in the chILD Registry:\nSubjects Diagnosed under 1 Year Old") +
  ylab("Number of Subjects Diagnosed with chILD") +
  xlab("Approximate Time of Diagnosis\n(Date-Shifted and Rounded to Nearest Month)") +
  labs(fill = "Was Genetic Testing Done?") +
  theme(axis.text.x=element_text(angle=60, hjust=1)) +
  scale_x_date(date_breaks = "1 year", date_labels = "%Y") +
  guides(fill = guide_legend(reverse = TRUE)) +
  theme(plot.margin=unit(c(1.5,0,1.5,0),"cm"))

# For those that are under 1 year old, and did have genetic testing, did this lead to the diagnosis?
child.gen.1Y <- child.gen %>%
  filter(age_diagnosis <= 12) %>%
  filter(genetics.factor == "Yes")
ggplot(filter(child.gen.1Y, age_diagnosis <= 12 && genetics.factor == "Yes"), aes(x=diag_date, fill = child_dia2___3.factor)) +
  geom_histogram(binwidth = 365, alpha = 0.8) +
  ggtitle("Genetic Testing over Time in the chILD Registry:", subtitle = "Subjects diagnosed under 1 year old who had genetic testing") +
  ylab("Number of Subjects Diagnosed with chILD") +
  xlab("Approximate Time of Diagnosis\n(Date-Shifted and Rounded to Nearest Month)") +
  labs(fill = "Did Genetic Testing\nMake the Diagnosis?") +
  theme(axis.text.x=element_text(angle=60, hjust=1)) +
  scale_x_date(date_breaks = "1 year", date_labels = "%Y") +
  scale_y_continuous(breaks = c(0,2,4,6,8)) +
  guides(fill = guide_legend(reverse = TRUE)) +
  theme(plot.margin=unit(c(1.5,0,1.5,0),"cm"))


ggplot(filter(gg.gen.full, age_diagnosis <= 12), aes(x=diag_date, fill = type_genetic_test.long)) +
  geom_histogram(binwidth = 365, alpha = 0.8) +
  ggtitle("Type of Genetic Testing over Time in the chILD Registry", subtitle = "Subjects diagnosed under 1 year old who had genetic testing") +
  ylab("Number of Tests") +
  xlab("Approximate Time of Diagnosis (Date-Shifted)") +
  labs(fill = "Type of Genetic Testing") +
  theme(axis.text.x=element_text(angle=60, hjust=1)) +
  scale_x_date(date_breaks = "1 year", date_labels = "%Y") +
  theme(plot.margin=unit(c(1.5,0,1.5,0),"cm"))



```

We see that there is an increasing number of genetic tests over the past 15 years, and almost all cases of surfactant dysfunction disorder, a disorder with known genetic causes, used genetic testing in diagnosis. Over time, gene panels have become more common in the workup of chILD, and whole exome sequencing is much less common but also increasing in frequency. In the younger age group, whole exome sequencing has been a fairly new tool, as it has only been reported as part of the workup in the last 2-3 years. 

### Conclusions

Overall, we see that there are significant variations in demographics across diagnoses, as well as differences in diagnostic testing. Genetic testing has increased over time, with gene panels becoming the most common type. Though this project helps us understand the current trends in diagnostic workup, there is still much more work to be done to refine the best approach to workup of a patient presenting with suspected ILD. Much of this can be done using the database, however, some questions may require further information gathering to answer.

This project has several limitations. As this study is voluntary and requires each center to recruit patients, there is risk of recruitment biases. As each center is responsible for entering data in the database, there may be limitations to the availability of data, as it is not possible to go into the EMR or patient chart to verify or collect more data. There may be some variability in data entry, as each center is doing this individually. For example, when is a patient considered to have their diagnosis? This could be when they are determined to have ILD, or when they undergo BAL and discover pulmonary hemorrhage, or it could be weeks to months later when they undergo a lung biopsy to determine whether there is capillaritis; the choice of date may be center- or provider-dependent. The database is fairly granular, however, many fields are unstructured data and thus present a challenge when trying to synthesize data on a large scale. There are certain areas of data collection that may be beneficial to study that have not been included, for example, administration of systemic steroids, which can affect outcomes in certain disease processes. As chILD is a rapidly changing area of study, diagnostic tools and classification systems continue to evolve, which introduces increased variance in the dataset. Nevertheless, this is a large dataset on a group of diseases with significant morbidity and mortality, and it proves an invaluable tool in gaining further insight into workup and management.

The chILD Registry offers many future directions for study. Given concerns about recruitment biases, it would be interesting to do a single-center study comparing demographics of children diagnosed with ILD at our institution (likely pulled from the EMR) with those that have been enrolled in the registry. To further guide the workup algorithm, it would be helpful to develop a prediction model based on clinical characteristics to determine whether a patient needs to undergo invasive studies such as lung biopsy, or whether these can be avoided as our non-invasive testing capabilities continue to improve.


### Acknowledgements
We would like to thank the Children’s Interstitial and Diffuse Lung Disease Research Network (chILDRN) and Dr. Lisa Young for the opportunity to work on this project.

### References
<div id="refs"></div>
