---
title: "Investigating the effects of excess body weight on presentation, outcomes and gene signatures in multiple myeloma"
subtitle: "BMIN 5030 Final Project"
author: "Beatrice Razzo"
output: 
  prettydoc::html_pretty:
    toc: true
    number_sections: true
    theme: cayman
    highlight: tango
    
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE, warning = FALSE, message = FALSE)
```

# OVERVIEW: 
This study aims to identify any clinical, cytogenetic and outcome associations between newly diagnosed multiple myeloma (NDMM) and excess body weight (EBW) in participants of the MMRF CoMMpass study. Data include clinical information at diagnosis including ISS stage, BMI and ECOG performance status and, cytogenetics including whole genome, whole exome and bulk RNA sequencing. Outcome data is available over a 6 year follow-up period. We will use body mass index (BMI) as a surrogate marker for excess body weight though it is an imperfect measure of adiposity. Please find my Github Repository for the final project [here](https://github.com/bearazzo/BMIN503_Final_Project).

# INTRODUCTION:
Multiple myeloma (MM) is the second most common blood cancer diagnosis, with over 20,000 cases diagnosed each year in the United States. Age, race and genetic factors are known predisposing risk factors, but obesity is the only modifiable characteristic consistently implicated in the development of both MGUS and MM (Carson, Bates, and Tomasson 2014; Landgren et al. 2010; Wallin and Larsson 2011). Persons with excess body weight are 10-20% more likely to develop multiple myeloma, with an estimated number of 12,480 cases attributed to excess body weight in the United States between 2011-2015 (Islami et al. 2019). In several recent meta-analyses, the risk of mortality was increased in multiple myeloma patients with excess body weight (RR 1.16-1.22) (Abar et al. 2019; Teras et al. 2014; Wallin and Larsson 2011). In contrast, other analyses have failed to suggest this association (Beason et al. 2013; Vogl et al. 2011). Overall, the true impact of obesity on development and outcomes in MM is unknown, with most published data reflecting diagnosis and treatment between 1995-2010. 5-year mean overall survival of MM has changed from 34.5 to 49.6% during this period due to the advent of new therapeutic modalities (Fonseca et al. 2017). Therefore, further studies are warranted to reflect the clinical impact of excess body weight under today’s disease stratification and treatment strategies.

The myeloma-obesity association is of interest not only due to potential measures that address population-level disparities and cancer prevention, but also because of insights it may provide into the myeloma pre-cancerous state and process of malignant transformation. The pathophysiological basis of this association is not fully understood, with systemic inflammation and secondary immune dysfunction as putative driver mechanisms. Inflammation induces chronic cellular stressors that may adversely affect prognosis in patient with MM. Patients with obesity (based on BMI or increased waist to hip ratio) were reported to have increasing rates of clonal hematopoiesis of intederminate potential (CHIP)(Kumar ASH 2021 abstract). CHIP has been associated with adverse outcomes in multiple myeloma patients undergoing transplant (Mouhieddine Nature communications 2020). BMI is also negatively associated with telomere length (Gielen American Journal of Clinical Nutrition 2018). Short telomere length has been associated with adverse outcomes in a number of cancers including chronic lymphocytic leukemia and MM where it associated with high risk GEP defined group (ref 14,15). Lastly, the APOBEC3 (apolipoprotein B mRNA editing enzyme catalytic polypeptide 3) family of cytidine deaminases is responsible for two mutational signatures (SBS2 and SBS13) found in cancer genomes. APOBEC3 enzymes are activated in response to viral infection and causes DNA damage and promotes genomic instability in multiple myeloma (). 

As such, this report will include data analysis of known clinical and cytogenetic prognostic features in MM and also exploration of genetic features that may be mechanistically associated with MM as an obesity related cancer (ORC). I thank my mentors Dr. Eileen Boyle and Dr. Faith Davies at NYU Langone for their ongoing support in completing this project. 
 

# METHODS:

## Data Source  
The MMRF CoMMpass database (MMRF in short hereafter) is a study designed to prospectively profile newly diagnosed, treatment naive MM (NDMM) from more than 1000 patients with longitudinal clinical follow up. Tumor specimens collected at diagnosis and relapse are interrogated by whole-exome, modified low pass whole-genome, and or RNA sequencing. The longitudinal component required clinical follow-up of each patient with collection of clinical parameters four times annually over a period of 10 years. Furthermore, each patient participating in the study underwent an IMID and/or Proteasome inhibitor-based treatment regimen at diagnosis determined by the treating hematologist. 

## Data loading

All libraries loaded as follows. The .csv files containing data extracted between 2018-2020 by Dr. Eileen Boyle at NYU Langone were loaded and modified according to the following code. They are all available as .csv on github. Data from 758 of 1273 participants contained height and weight at the time of diagnosis and this subset of patients is primarily used in the following analyses given the focus of this report. Demographic distribution does not significantly differ from the overall cohort (data not shown, previously published in Boyle et. al Leukemia 2021).  

Of all samples available, 980 underwent germline WGS - all were analyzed for clonal hematopoiesis and 911 were analyzed for leukocyte telomere length. There were 848 matching tumor specimens available - 809 underwent tumor WGS (747 analyzed for complex rearrangements and 807 analyzed for tumor telemere length), 848 underwent tumor WES (included analyses for mutations, translocations and CNA) and 683 underwent bulk RNA sequencing. Most patients for which BMI information is available also contributed samples for genetic information as described above.  


```{r, eval = TRUE}
library(survival)
library(ranger)
library(ggplot2)
library(dplyr)
library(ggfortify)
library(tibble)
library(readr)
library(survminer)
library(RColorBrewer)
mycolors=brewer.pal(8,"Set2")
library(cowplot)
library(modelsummary)
library(viridis)
library(tidytidbits)
library(survivalAnalysis)
library(purrr)
library(stringr)

MMRFfull <- read.csv("~/Documents/BMIN 5030/Final Project/Raw Files/Razzo_MMRF raw csv.csv", header = T, fill = TRUE, row.name = NULL)
MMRFfull <- rename(MMRFfull, Patient_ID = Patient)

MMRFclean <- read.csv("~/Documents/BMIN 5030/Final Project/Raw Files/Razzo_MMRF clean csv.csv", header = T, fill = TRUE, row.name = NULL)
MMRFclean <- rename(MMRFclean, Sex = D_Gender, Race = D_Race, AgeDx = D_Age, BMIDx = BMI, ISSDx = D_ISS) %>%
  mutate(BMI_cat_4 = factor(BMI_cat_4, levels = c(0,1,2,3), labels = c("<18.5", "18.5-25", "25-30", ">30"))) %>%
  mutate(BMI_cat_6 = factor(BMI_cat_6, levels = c(0,1,2,3,4,5), labels = c("<18.5", "18.5-25", "25-30", "30-35", "35-40", ">40"))) %>%
  mutate(FU_status = factor(FU_status, levels = c(0,1), labels = c("Not censored", "Censored"))) %>%
  mutate(OS_status = factor(OS_status, levels = c(0,1), labels = c("Alive", "Dead"))) %>%
  mutate(ISSDx = factor(ISSDx, level = c(1,2,3)))

MMRFclean$Translocation_Summary <- as.factor(MMRFclean$Translocation_Summary)
MMRFclean$HRD <- as.factor(MMRFclean$HRD)
MMRFclean$Sex <- as.factor(MMRFclean$Sex)
MMRFclean$RNA_high_risk <- as.factor(MMRFclean$RNA_high_risk)
MMRFclean$OS_status_num <- as.numeric(MMRFclean$OS_status)

#Merging full and clean datasets
MMRFadd <- read.csv("~/Documents/BMIN 5030/Final Project/BMIN503_Final_Project/MMRF clinical add on.csv", header = T, fill = TRUE, row.name = NULL)
MMRFadd <- rename(MMRFadd, Patient_ID = Patient)
MMRFcleanadd <- right_join(MMRFclean, MMRFadd, by = "Patient_ID")
MMRFcleanadd <- rename(MMRFcleanadd, HxTransplant = TRT_1_bmtx_n) 
MMRFcleanadd <- MMRFcleanadd %>% mutate(HxTransplant = factor(HxTransplant, levels = c(0,1), labels = c("No", "Yes")))
MMRFcleanadd$ECOG <- as.factor(MMRFcleanadd$ECOG)

#Selecting participants with BMI data for further analyses
MMRFcleanbmionly <- MMRFcleanadd %>%
  filter(!is.na(BMIDx))
MMRFcleanbmionly$age_group <- cut(MMRFclean$AgeDx,
                       breaks=c(0, 54, 64, 74, 110),
                       labels=c("<55", "55-64", "65-74", ">74"))
MMRFfullcgsumm <- MMRFfull[1:1273, c("Patient_ID","TP53_Loss","CKS1B_Gain")]
MMRFcleanbmionly <- inner_join(MMRFcleanbmionly, MMRFfullcgsumm, by = "Patient_ID")
MMRFcleanbmionly$CKS1B_Gain_binary <- case_when(MMRFcleanbmionly$CKS1B_Gain == -2 ~ 'No',
                                          MMRFcleanbmionly$CKS1B_Gain == 1 ~ 'No',
                                          MMRFcleanbmionly$CKS1B_Gain == 2 ~ 'No',
                                          MMRFcleanbmionly$CKS1B_Gain == 3 ~ '1q gain',
                                          MMRFcleanbmionly$CKS1B_Gain == 4 ~ '1q gain')

MMRFcleanbmionly$TP53_Loss_binary <- case_when(MMRFcleanbmionly$TP53_Loss == -2 ~ '17p del',
                                          MMRFcleanbmionly$TP53_Loss == 1 ~ '17p del',
                                          MMRFcleanbmionly$TP53_Loss == 2 ~ 'No',
                                          MMRFcleanbmionly$TP53_Loss == 3 ~ 'No',
                                          MMRFcleanbmionly$TP53_Loss == 4 ~ 'No')                                          
MMRFcleanbmionly$t414_binary <- case_when(MMRFcleanbmionly$Translocation_Summary == "t(4;14)" ~ 't(4;14)',
                                          TRUE ~ 'No') 
MMRFcleanbmionly$t1416_binary <- case_when(MMRFcleanbmionly$Translocation_Summary == "t(14;16)" ~ 't(14;16)',
                                          TRUE ~ 'No') 
MMRFcleanbmionly$EBW_binary <- case_when(MMRFcleanbmionly$BMIDx >= 30 ~ '>30',
                                         TRUE ~ '<30')

MMRFcleanbmionly$CKS1B_Gain_binary <- as.factor(MMRFcleanbmionly$CKS1B_Gain_binary)
MMRFcleanbmionly$CKS1B_Gain_binary <- factor(MMRFcleanbmionly$CKS1B_Gain_binary, levels = c("No", "1q gain"))
MMRFcleanbmionly$TP53_Loss_binary <- as.factor(MMRFcleanbmionly$TP53_Loss_binary)
MMRFcleanbmionly$TP53_Loss_binary <- factor(MMRFcleanbmionly$TP53_Loss_binary, levels = c("No", "17p del"))
MMRFcleanbmionly$t414_binary <- as.factor(MMRFcleanbmionly$t414_binary)
MMRFcleanbmionly$t1416_binary <- as.factor(MMRFcleanbmionly$t1416_binary)
MMRFcleanbmionly$EBW_binary <- as.factor(MMRFcleanbmionly$EBW_binary)
MMRFcleanbmionly$OS_status_num <- as.numeric(MMRFcleanbmionly$OS_status)

#Adding in apobec data
MMRFapobec <-read.csv("~/Documents/BMIN 5030/Final Project/BMIN503_Final_Project/MMRF apobec values.csv", header = T, fill = TRUE, row.name = NULL)
MMRFapobec <- rename(MMRFapobec, Patient_ID = X)

#Adding in CHIP data
MMRFfullchip <- MMRFfull[1:1273, c("Patient_ID","DNMT3A_Loss","NS_TET2")]
MMRFfullchip$DNMT3A_Loss_binary <- case_when(MMRFfullchip$DNMT3A_Loss == -2 ~ 'DNMT3A loss',
                                          MMRFfullchip$DNMT3A_Loss == 1 ~ 'DNMT3A loss',
                                          MMRFfullchip$DNMT3A_Loss == 2 ~ 'No',
                                          MMRFfullchip$DNMT3A_Loss == 3 ~ 'No',
                                          MMRFfullchip$DNMT3A_Loss == 4 ~ 'No') 
MMRFfullchip$DNMT3A_Loss_binary <- as.factor(MMRFfullchip$DNMT3A_Loss_binary)
MMRFcleanbmionly <- inner_join(MMRFcleanbmionly, MMRFfullchip, by = "Patient_ID")

#Adding in differential gene expression results
MMRFdge <-read.csv("~/Documents/BMIN 5030/Final Project/BMIN503_Final_Project/DGE by BMI group.csv", header = T, fill = TRUE, row.name = NULL)

```


# Conventional clinical and genetic features and their relationship with BMI distribution

## Demographic and clinical data according to BMI

The histogram below depicts the BMI distribution at the time of diagnosis of MMRF database participants. The numeric and percent breakdown of some of the clinical and cytogenetic information has been included in the following summary table for reference.

```{r, eval = TRUE}

ggplot(data = MMRFclean, aes(x = BMIDx)) + 
  geom_histogram(aes(y= ..density..), fill = "white", alpha = 0.7, binwidth = 0.8)+
  geom_density() +
  labs(title="BMI distribution in MMRF dataset", x ="BMI distribution at diagnosis", y = "Density") +
  theme_minimal ()

#Baseline Patient Characteristics, BMI 4 categories
patientcharacteristics <- MMRFclean %>% gtsummary::tbl_summary()
MMRFclean %>% 
  dplyr::select(c(Sex, Race, AgeDx, BMI_cat_4, ISSDx, IMWG_Risk, Translocation_Summary)) %>%
  gtsummary::tbl_summary(by = BMI_cat_4) %>%
  gtsummary::add_p()
 
```

The scatter plot shows a small but statistically significant decline in BMI with older age at diagnosis. Later analyses will need to take this relationship into account, because age has been shown to have significant prognostic implications as previously published by our group (ref). Although there was a trend toward increased mean BMI in males, this was not a statistically significant trend. Similarly, there were no significant variations in BMI distribution according to race. Mean BMI was also not significantly different according to ISS stage at diagnosis. ECOG performance status was distributed evenly though patients with BMI >18.5. There was an increased proportion of ECOG >0 in underweight patients suggesting a sicker patient population. Transplant status did not significantly change according to BMI group.


```{r, eval = TRUE}
ggplot1<- ggplot(MMRFclean, aes(x = AgeDx, y = BMIDx)) +
    geom_point(color = "darkseagreen3") + 
    geom_smooth(method = "lm", color = "green4") +
    labs(x = "Age at diagnosis", y = "BMI at diagnosis") +
    stat_cor(method = "pearson", label.x = 60, label.y = 60) +
    theme_minimal ()

#KRUSKAL-WALLIS, BMI DISTRIBUTION BY SEX
KWsex<- kruskal.test(Sex~BMIDx, MMRFclean)
mylabel1 <- bquote(p == .(format(KWsex$p.value, digits = 5)))
mylabel2 <- bquote(Chi2 == .(format(KWsex$statistic, digits = 2)))

ggplot2 <- ggplot(data = MMRFclean, aes(x = BMIDx, color = Sex, na.rm = TRUE)) + 
  geom_histogram(aes(y= ..density.., color=Sex), na.rm=TRUE, Male = "blue", Female = "red", fill = "white", alpha = 0.7, binwidth=1)+
  geom_density(Male = "blue", Female = "red", na.rm=TRUE) +
  theme_minimal() +
  labs(x = "Biological sex", y = "BMI distribution") +
  annotate("text", x=50, y=0.26, label=mylabel2) +
  annotate("text", x=50, y=0.22, label=mylabel1)

ggplot3 <- ggplot(data = MMRFclean, aes(x = factor(Race), y = BMIDx)) +
  geom_violin(draw_quantiles = c(0.25, 0.5, 0.75)) +
  stat_summary(fun.data=mean_sdl, mult=1, 
                 geom="pointrange", color="red") +
  theme_minimal() +
  stat_compare_means (label.x = 2.2, label.y = 61) +
  labs(x = "Race", y = "BMI distribution")
  
ggplot4 <- ggplot(data = MMRFclean, aes(x = factor(ISSDx), y = BMIDx)) +
  geom_violin(draw_quantiles = c(0.25, 0.5, 0.75)) +
  stat_summary(fun.data=mean_sdl, 
                 geom="pointrange", color="red") +
  theme_minimal() +
  stat_compare_means (label.x = 2.8, label.y = 61) +
  labs(x = "ISS Stage at Diagnosis", y = "BMI distribution")

#ECOG status distribution - proportion stacked barplot
BMIECOGdf <- MMRFcleanbmionly[1:758, c("BMI_cat_4","ECOG")] 
BMIECOGdf <- BMIECOGdf %>% filter(!is.na(ECOG))
                              
ggplot6 <- ggplot(BMIECOGdf, aes(fill=ECOG, y=1, x=BMI_cat_4)) + 
  geom_bar(position="fill", stat="identity") +
  labs(x = "BMI WHO category", y = "Proportion of patients") +
  theme_minimal()

#Transplant distribution - proportion stacked barplot
BMItranspdf <- MMRFcleanbmionly[1:758, c("HxTransplant","BMI_cat_4")] 
BMItranspdf <- BMItranspdf %>% filter(!is.na(HxTransplant))
                              
ggplot7 <- ggplot(BMItranspdf, aes(fill=HxTransplant, y=1, x=BMI_cat_4)) + 
  geom_bar(position="fill", stat="identity") +
  labs(x = "BMI WHO category", y = "Proportion of patients") +
  theme_minimal()

plot_grid(ggplot1, ggplot2, ggplot3, ggplot4, ncol = 2)
plot_grid(ggplot6, ggplot7, ncol = 2, nrow =2)
  
```

## Characterization of BMI at diagnosis with respect to MM related cytogenetic changes and RNA risk category

In regards to the BMI distribution of the most common MM related translocations, most did not segregate with BMI. Although t(8;14) did appear to have statistically significant distribution with respect to weight, it is a rare translocation often associated with later clonal evolution so unlikely to be a clinically significant finding in regards to newly diagnosed genetic drivers (Tomas LML abstract 2022). 1q gain and 17p deletion are included in newer MM staging systems (R-ISS and R2-ISS) and were obtained in this subset by tumor NGS. Neither appear to change in frequency according to BMI.

Finally, the BMI distribution is also exactly the same across individuals with and without the clinically established high risk RNA risk profiling GEP70 (70 risk genes with high risk defined as score > ≥45.2). 

```{r, eval = TRUE}

#Translocation Summary 
ggplot(data = MMRFclean, aes(x = factor(Translocation_Summary), y = BMIDx)) +
    geom_violin(draw_quantiles = c(0.25, 0.5, 0.75)) +
    stat_summary(fun.data=mean_sdl, mult=1, 
                 geom="pointrange", color="red") +
    geom_jitter(shape=16, position=position_jitter(0.2))

lm4<- lm(BMIDx ~ Translocation_Summary, data = MMRFclean)
modelsummary(list(lm4), coef_omit = "Intercept",
             statistic = c("p = {p.value}"),
             estimate = "{estimate} [{conf.low}, {conf.high}]") 

#1q gain
KWhighrisk17pdel <- kruskal.test(TP53_Loss_binary~BMIDx, MMRFcleanbmionly)
mylabel7 <- bquote(p == .(format(KWhighrisk17pdel$p.value, digits = 5)))
mylabel8 <- bquote(Chi2 == .(format(KWhighrisk17pdel$statistic, digits = 2)))
  
ggplot9 <- ggplot(data = MMRFcleanbmionly, aes(x = factor(TP53_Loss_binary), y = BMIDx), na.rm = TRUE) +
  geom_boxplot() +
  theme_minimal() +
  labs(x = "17p del", y = "BMI distribution") +
  annotate("text", x=3, y=60, label=mylabel8) +
  annotate("text", x=3, y=54, label=mylabel7)

#17p del
KWhighrisk1q <- kruskal.test(CKS1B_Gain_binary~BMIDx, MMRFcleanbmionly)
mylabel5 <- bquote(p == .(format(KWhighrisk1q$p.value, digits = 5)))
mylabel6 <- bquote(Chi2 == .(format(KWhighrisk1q$statistic, digits = 2)))
  
ggplot8 <- ggplot(data = MMRFcleanbmionly, aes(x = factor(CKS1B_Gain_binary), y = BMIDx), na.rm = TRUE) +
  geom_boxplot() +
  theme_minimal() +
  labs(x = "1q gain", y = "BMI distribution") +
  annotate("text", x=3, y=60, label=mylabel6) +
  annotate("text", x=3, y=54, label=mylabel5) 

#High risk GEP70 does not occur more frequently in patients with excess body weight
KWhighriskRNA <- kruskal.test(RNA_high_risk~BMIDx, MMRFclean)
mylabel3 <- bquote(p == .(format(KWhighriskRNA$p.value, digits = 5)))
mylabel4 <- bquote(Chi2 == .(format(KWhighriskRNA$statistic, digits = 2)))

ggplot5 <- ggplot(data = MMRFclean, aes(x = factor(RNA_high_risk), y = BMIDx), na.rm = TRUE) +
  geom_boxplot() +
  theme_minimal() +
  labs(x = "High risk GEP70 (≥45.2)", y = "BMI distribution") +
  annotate("text", x=3, y=60, label=mylabel4) +
  annotate("text", x=3, y=54, label=mylabel3) 

plot_grid(ggplot9, ggplot8, ggplot5, ncol = 2)

```

## Outcome Summary Statistics

In order to first validate our cohort containing BMI data, I first demonstrated that genetic and clinical data used in NDMM staging (ISS, R-ISS and R2-ISS staging/prognostic systems) are indeed prognostic of survival in this cohort. Additionally, trial participants who were older, male and who had worse ECOG performance status did worse. There were no statistically significant race-related outcomes in this population.  

There were no survival differences in outcomes with BMI stratified into the 4 WHO defined discrete categories or as a binary variable. Univariate cox proportional hazard regression also did not show an association with BMI as a continuous variable (not shown here). Given the confounding effects of age and possibly sex on BMI, I also performed a multivariate cox regression to assess a BMI-survival association and again this was not demonstrated.



*Data validation with clinically established prognostic features*
```{r, eval = TRUE}
splots1 <- list()

#ISS
km_fit_BMIISS <- survfit(Surv(OS_time, OS_status_num) ~ ISSDx, data=MMRFcleanbmionly)
splots1[[1]] <- ggsurvplot(km_fit_BMIISS,
           conf.int=TRUE,
           pval=TRUE,
           risk.table=FALSE,
           legend.labs=c("1", "2", "3"), 
           legend.title="ISS Stage at diagnosis", 
           palette=c("#66C2A5", "#8DA0CB","#A6D854", "#E5C494"),
           risk.table.height=0.1)

#High GEP70
km_fit_BMIRNA <- survfit(Surv(OS_time, OS_status_num) ~ RNA_high_risk, data=MMRFclean)
splots1[[2]] <- ggsurvplot(km_fit_BMIRNA,
           conf.int=TRUE,
           pval=TRUE,
           risk.table=FALSE,
           legend.labs=c("False", "True"), 
           legend.title="High risk GEP70", 
           palette=c("#66C2A5", "#8DA0CB","#A6D854", "#E5C494"),
           risk.table.height=0.1)

#t(4;14)
km_fit_BMI414 <- survfit(Surv(OS_time, OS_status_num) ~ t414_binary, data=MMRFcleanbmionly)
splots1[[3]] <- ggsurvplot(km_fit_BMI414,
           conf.int=TRUE,
           pval=TRUE,
           risk.table=FALSE,
           legend.labs=c("No", "t(4;14)"), 
           legend.title="t(4;14) gain status", 
           palette=c("#66C2A5", "#8DA0CB"),
           risk.table.height=0.1)

#1q gain - PENDING
km_fit_BMI1q <- survfit(Surv(OS_time, OS_status_num) ~ CKS1B_Gain_binary, data=MMRFcleanbmionly)
splots1[[4]] <- ggsurvplot(km_fit_BMI1q,
           conf.int=TRUE,
           pval=TRUE,
           risk.table=FALSE,
           legend.labs=c("1q gain","No"), 
           legend.title="1q gain status", 
           palette=c("#66C2A5", "#8DA0CB"),
           risk.table.height=0.1)

#Delete once you figure out the issue
km_fit_BMI17p <- survfit(Surv(OS_time, OS_status_num) ~ TP53_Loss_binary, data=MMRFcleanbmionly)
ggsurvplot(km_fit_BMI17p,
           conf.int=TRUE,
           pval=TRUE,
           risk.table=FALSE,
           legend.labs=c("17p del","No"), 
           legend.title="17p del status", 
           palette=c("#66C2A5", "#8DA0CB"),
           risk.table.height=0.1)

arrange_ggsurvplots(splots1, print = TRUE,
  ncol = 2, nrow = 2)

#OBS: Other high risk translocations omitted due to infrequency 
```

*Clinical factors and respective survival curves in this population subset*
```{r, eval = TRUE}

splots3 <- list()

#Survival by Age Ranges - will use same as Age paper - PENDING
km_fit_BMIage <- survfit(Surv(OS_time, OS_status_num) ~ age_group, data=MMRFcleanbmionly)
splots3[[1]] <- ggsurvplot(km_fit_BMIage,
           conf.int=TRUE,
           pval=TRUE,
           risk.table=FALSE,
           legend.labs=c("<55", "55-64", "65-74", ">74"), 
           legend.title="Age Range", 
           palette=c("#66C2A5", "#8DA0CB","#A6D854", "#E5C494"),
           risk.table.height=0.1)

#Survival by Sex - 
km_fit_BMIsex <- survfit(Surv(OS_time, OS_status_num) ~ Sex, data=MMRFclean)
splots3[[2]] <- ggsurvplot(km_fit_BMIsex,
           conf.int=TRUE,
           pval=TRUE,
           risk.table=FALSE,
           legend.labs=c("Female", "Male"), 
           legend.title="Biological sex", 
           palette=c("#66C2A5", "#8DA0CB","#A6D854", "#E5C494"),
           risk.table.height=0.1)

#Survival by ECOG PS
km_fit_BMIECOG <- survfit(Surv(OS_time, OS_status_num) ~ ECOG, data=MMRFcleanbmionly)
splots3[[3]] <- ggsurvplot(km_fit_BMIECOG,
           conf.int=TRUE,
           pval=TRUE,
           risk.table=FALSE,
           legend.labs=c("0", "1", "2", "3","4"), 
           legend.title="ECOG PS", 
           palette=c("#66C2A5", "#8DA0CB","#A6D854", "#E5C494", "#B3B3B3"),
           risk.table.height=0.1)

MMRFcleanbmionly$Race <- as.factor(MMRFcleanbmionly$Race)
km_fit_BMIrace <- survfit(Surv(OS_time, OS_status_num) ~ Race, data=MMRFclean)
splots3[[4]] <- ggsurvplot(km_fit_BMIrace,
           conf.int=TRUE,
           pval=TRUE,
           risk.table=FALSE,
           legend.labs=c("Asian", "Black or African", "Other", "White"), 
           legend.title="Race", 
           palette=c("#66C2A5", "#8DA0CB","#A6D854", "#E5C494"),
           risk.table.height=0.1)

splots2 <- list()

#Survival by BMI 4 discrete categories 
km_fit_BMIcensor <- survfit(Surv(OS_time, OS_status_num) ~ BMI_cat_4, data=MMRFclean)
splots2[[1]] <- ggsurvplot(km_fit_BMIcensor,
           conf.int=FALSE,
           pval=TRUE,
           risk.table=FALSE,
           legend.labs=c("<18.5", "18.5-25","25-30",">30"), 
           legend.title="BMI WHO category", 
           palette=c("#66C2A5", "#8DA0CB","#A6D854", "#E5C494"),
           risk.table.height=0.1)

#Survival + or no EBW
km_fit_BMIEBW_binary <- survfit(Surv(OS_time, OS_status_num) ~ EBW_binary, data=MMRFcleanbmionly)
splots2[[2]] <- ggsurvplot(km_fit_BMIEBW_binary,
           conf.int=TRUE,
           pval=TRUE,
           risk.table=FALSE,
           legend.labs=c("<30", ">30"), 
           legend.title="Presence of EBW", 
           palette=c("#66C2A5", "#8DA0CB"),
           risk.table.height=0.1)

arrange_ggsurvplots(splots3, print = TRUE,
  ncol = 2, nrow = 2)

arrange_ggsurvplots(splots2, print = TRUE,
  ncol = 2, nrow = 2)

# UNIVARIATE with BMI as continuous and binary to confirm lack of association suggested with survplot of BMI as 4 discrete categories 
unisurv1 <- coxph(Surv(OS_time, OS_status_num) ~ BMIDx, data = MMRFcleanbmionly)
summary(unisurv1)
unisurv2 <- coxph(Surv(OS_time, OS_status_num) ~ EBW_binary, data = MMRFcleanbmionly)
summary(unisurv2)

survmulti4 <- analyse_multivariate(MMRFcleanbmionly, vars(OS_time, OS_status_num), covariates = vars(BMIDx, AgeDx, Sex))
print(survmulti4$summaryAsFrame)

```

Yet although no significant overall survival differences were dependent on BMI, I also wondered whether the established prognostic features are weighed differently in the obese and non-obese populations. 

As such I performed multivariate cox regressions in patients with BMI <30 and BMI >30 as shown on in the following forest plots. In the 509 patients with BMI <30, most variables had at least a tendency toward HR >1. There was a small increment in HR/year of age gained and ISS was highly prognostic in this patient population. BMI again did not show any tendencies with respect to survival. In terms of patients with BMI >30, BMI also did not contribute toward mortality and age again demonstrated a small association. GEP70 high risk RNA expression profiling and ISS risk had lower HR compared to the non-obese group. The contribution of high-risk cytogenetic features was similar.


*Cox Regression models in patients with and without excess body weight using established prognostic features*
```{r, eval = TRUE}

#MULTIVARIATE OBESE (clinical and clinically established prognostic features), n = 249
MMRFcleanbmionlyEBW <- MMRFcleanbmionly %>%
  filter(BMIDx >= 30)
survmultiobese <- analyse_multivariate(MMRFcleanbmionlyEBW, vars(OS_time, OS_status_num), covariates = vars(BMIDx, AgeDx, Sex, ISSDx, RNA_high_risk, t414_binary, t1416_binary, CKS1B_Gain_binary, TP53_Loss_binary))
forest_plot(
  survmultiobese,
  use_one_hot = FALSE,
  factor_labeller = identity,
  endpoint_labeller = identity,
  orderer = ~order(HR),
  categorizer = NULL,
  relative_widths = c(1, 1, 1),
  ggtheme = theme_bw(),
  labels_displayed = c("factor"),
  label_headers = c(endpoint = "Endpoint", factor = "Subgroup", n = "n"),
  values_displayed = c("HR", "CI", "p"),
  value_headers = c(HR = "HR", CI = "CI", p = "p", n = "n", subgroup_n = "n"),
  HRsprintfFormat = "%.2f",
  psprintfFormat = "%.3f",
  p_lessthan_cutoff = 0.001,
  log_scale = TRUE,
  HR_x_breaks = seq(0, 10),
  HR_x_limits = NULL,
  factor_id_sep = ":",
  na_rm = TRUE,
  title = "Cox HR in BMI >30, n = 249",
  title_relative_height = 0.1,
  title_label_args = list(),
  base_papersize = dinA(4)
)

#MULTIVARIATE NON-OBESE (clinical and clinically established prognostic features), n = 509
MMRFcleanbmionlynoEBW <- MMRFcleanbmionly %>%
  filter(BMIDx < 30)
survmultinonobese<- analyse_multivariate(MMRFcleanbmionlynoEBW, vars(OS_time, OS_status_num), covariates = vars(BMIDx, AgeDx, Sex, ISSDx, RNA_high_risk, t414_binary, t1416_binary, CKS1B_Gain_binary, TP53_Loss_binary))
forest_plot(
  survmultinonobese,
  use_one_hot = FALSE,
  factor_labeller = identity,
  endpoint_labeller = identity,
  orderer = ~order(HR),
  categorizer = NULL,
  relative_widths = c(1, 1, 1),
  ggtheme = theme_bw(),
  labels_displayed = c("factor"),
  label_headers = c(endpoint = "Endpoint", factor = "Subgroup", n = "n"),
  values_displayed = c("HR", "CI", "p"),
  value_headers = c(HR = "HR", CI = "CI", p = "p", n = "n", subgroup_n = "n"),
  HRsprintfFormat = "%.2f",
  psprintfFormat = "%.3f",
  p_lessthan_cutoff = 0.001,
  log_scale = TRUE,
  HR_x_breaks = seq(0, 10),
  HR_x_limits = NULL,
  factor_id_sep = ":",
  na_rm = TRUE,
  title = "Cox HR in BMI <30, n = 509",
  title_relative_height = 0.1,
  title_label_args = list(),
  base_papersize = dinA(4)
)

```


# Exploratory genetic features and relationship with BMI distribution in this population

This section serves to explore whether any additional genetic features in patients with NDMM that may be associated with BMI at the time of diagnosis. The following are pathways are significantly implicated in inflammatory drivers and are putative mechanisms of obesity related cancers as introduced before.

## NMF for Apobec signatures

*Methods (provided to and not performed by me):* Mutational signatures were called using non-negative matrix factorization (NMF) with counts per sample calculated for the six possible SNV types and the 16 possible 3-base sequence contexts, creating a table with 96 columns. The R package “NMF” was used for all calculations (CRAN - Package NMF). The number of signatures was determined by running 50 iterations of the algorithm for 2-7 signatures. A number of signatures was chosen that maximized the cophenetic distance and dispersion values. One thousand (1,000) iterations of the algorithm were run for that number of signatures. Cosine similarity was used to determine the Sanger signatures that were closest to the detected signatures.

*Results:*

```{r, eval = TRUE}

#MMRFapobec (B) signatures with respect to BMI distribution
KWNMF2 <- kruskal.test(NMF2~BMIDx, MMRFcleanbmionly)
KWNMF2lab1 <- bquote(Chi2 == .(format(KWNMF2$statistic, digits = 2)))
KWNMF2lab2 <- bquote(p == .(format(KWNMF2$p.value, digits = 5)))

ggplot(data = MMRFcleanbmionly, aes(x = factor(NMF2), y = BMIDx), na.rm = TRUE) +
  geom_boxplot() +
  theme_minimal() +
  labs(x = "APOBEC signature", y = "BMI distribution") +
  annotate("text", x=2, y=60, label=KWNMF2lab1) +
  annotate("text", x=2, y=54, label=KWNMF2lab2) 

#Cleaning duplicate apobec reads, adding df to MMRFcleanbmionly
MMRFapobec[c('ID1','ID2','samplenum','samplesource')] <- str_split_fixed(MMRFapobec$Patient_ID, "_", n=4) 
MMRFapobec <- rename(MMRFapobec, Patient_ID_long = Patient_ID)
MMRFapobec$Patient_ID <- str_c(MMRFapobec$ID1, "_", MMRFapobec$ID2)
MMRFapobecclean <- MMRFapobec %>% distinct(Patient_ID, .keep_all = TRUE) 
MMRFapobecclean <- MMRFapobecclean[1:968, c("Patient_ID","APOBEC.2","Background")]
MMRFcleanbmionly <- left_join(MMRFcleanbmionly, MMRFapobecclean, by = "Patient_ID")

#APOBEC and BMI distribution
ggplot(MMRFcleanbmionly, aes(x = BMIDx, y = APOBEC.2)) +
    geom_point(color = "darkseagreen3") + 
    geom_smooth(method = "lm", color = "green4") +
    labs(x = "BMI at diagnosis", y = "Proportion of APOBEC mutation signatures") +
    stat_cor(method = "pearson", label.x = 37, label.y = 0.7) +
    theme_minimal ()

#Apobec mutation signatures are known to cluster by high risk translocations, the following codes for a *** 
##PENDING


```


## CHIP
*Methods (provided to and not performed by me):* In order to identify mutations within the uninvolved peripheral blood DNA as evidence of clonal hematopoesis we took the approach of strictly filtering a set of ultra-sensitive called variants that 1)fell within a low variant allele frequency range (<50%) and 2)occurred within a gene that has been linked to hematopoietic and lymphoid cancers in the Catalogue of Somatic Mutations in Cancer. (Genovese G, Kähler AK, Handsaker RE, et al. Clonal Hematopoiesis and Blood-Cancer Risk Inferred from Blood DNA Sequence. N. Engl. J. Med. 2014;371(26):2477–2487)

*Results:* 74 patients (9.8%) of 753 had concomitant clonal hematopoiesis of indeterminate potential (CHIP). Surprisingly, ***. 

Because of the known association with CHIP and age, I performed a multivariate logistic regression to assess for associations between CHIP and age, BMI or sex. Increased OR for CHIP was again demonstrated with age as previously demonstrated in this dataset (Boyle ref). There was also a small but significantly changed OR toward CHIP at 0.95 (CI: 0.90-0.99, p = 0.026) in relationship with BMI. The relationship is redemonstrated when analyzing presence of excess body weight (BMI >30) as a logical variable and is depicted in the barplot graph that follows.    

Next, we inquired whether this difference may be driven by distribution of the most common CHIP-related mutations: DNMT3A, TET2, TP53, ASXL1 and PPM1D. Data was available regarding presence or lack of DNMT3A, TET2 and TP53 mutations with frequency of distribution not associated with BMI as shown below.

```{r, eval = TRUE}

BMIchip <- MMRFcleanbmionly[1:758, c("BMIDx","BMI_cat_4","EBW_binary","CHIP","AgeDx","Sex","Race")] 
BMIchip <- BMIchip %>% filter(!is.na(CHIP))

BMIchip$EBW_binary <- as.numeric(BMIchip$EBW_binary)
BMIchip$CHIP <- as.numeric(BMIchip$CHIP)
wilcox.test(CHIP~EBW_binary, BMIchip)
KWCHIP <-kruskal.test(CHIP~BMIDx, BMIchip)
bquote(Chi2 == .(format(KWCHIP$statistic, digits = 2)))
bquote(p == .(format(KWCHIP$p.value, digits = 5)))

ggplotchipbmi <- ggplot(data = BMIchip, aes(x = factor(CHIP), y = BMIDx), na.rm = TRUE) +
  geom_boxplot() +
  theme_minimal() +
  labs(x = "Presence of CHIP", y = "BMI distribution") +
  stat_compare_means()

#Both as continuous variables
ggplotchipage <- ggplot(data = MMRFcleanbmionly, aes(x = factor(CHIP), y = AgeDx), na.rm = TRUE) +
  geom_boxplot() +
  theme_minimal() +
  labs(x = "Presence of CHIP", y = "BMI distribution")+
  stat_compare_means()

plot_grid(ggplotchipbmi, ggplotchipage, ncol = 2)

#Logistic regression for CHIP using age, BMI and sex with BMI as a continuous then with EBW as a binary variable
MMRFcleanbmionly$CHIP <- as.factor(MMRFcleanbmionly$CHIP)
MMRFcleanbmionly$Race <- as.factor(MMRFcleanbmionly$Race)

glm1 <- glm(CHIP ~ AgeDx + Sex + BMIDx,
                  data = MMRFcleanbmionly, 
                  family = binomial())
summary(glm1)
exp(cbind(OR = coef(glm1), CI = confint(glm1)))

glm2 <- glm(CHIP ~ AgeDx + Sex + EBW_binary,
                  data = MMRFcleanbmionly, 
                  family = binomial())
summary(glm2)
exp(cbind(OR = coef(glm2), CI = confint(glm2)))

ggplot(MMRFcleanbmionly, aes(fill=CHIP, y=1, x=BMI_cat_4)) + 
  geom_bar(position="fill", stat="identity") +
  labs(x = "BMI WHO category", y = "Proportion of patients") +
  theme_minimal()

##CHIP mutations:
#TP53
ggplottp53<- ggplot(data = MMRFcleanbmionly, aes(x = factor(TP53_Loss_binary), y = BMIDx), na.rm = TRUE) +
  geom_boxplot() +
  theme_minimal() +
  labs(x = "TP53 loss", y = "BMI distribution") +
  stat_compare_means()

#DNMT3A
ggplotdnmt3a <- ggplot(data = MMRFcleanbmionly, aes(x = factor(DNMT3A_Loss_binary), y = BMIDx), na.rm = TRUE) +
  geom_boxplot() +
  theme_minimal() +
  labs(x = "DNMT3A loss", y = "BMI distribution") +
  stat_compare_means()

#TET2
ggplottet2 <- ggplot(data = MMRFcleanbmionly, aes(x = factor(NS_TET2), y = BMIDx), na.rm = TRUE) +
  geom_boxplot() +
  theme_minimal() +
  labs(x = "TET2 loss", y = "BMI distribution") +
  stat_compare_means()

plot_grid(ggplottp53, ggplotdnmt3a, ggplottet2, ncol = 2)

```

## TLT and TTL
*Methods (provided to and not performed by me)*: Telomere length was estimated in both tumors and germline using Telomerecat. (Farmery JHR, Smith ML, Lynch AG. Telomerecat: A ploidy-agnostic method for estimating telomere length from whole genome sequencing data. Sci. Rep. 2018;8(1):1300)

*Results*:

```{r, eval = TRUE}

#TTL, BMI discrete
MMRFcleanbmionly$TTL <- as.numeric(MMRFcleanbmionly$TTL)
ylim1 = boxplot.stats(MMRFcleanbmionly$TTL)$stats[c(1, 5)]
TL1 <- ggplot(data = MMRFcleanbmionly, aes(x = factor(BMI_cat_4), y = TTL)) +
    geom_violin(draw_quantiles = c(0.25, 0.5, 0.75)) +
    stat_summary(fun.data=mean_sdl, mult=1, 
                 geom="pointrange", color="red") +
  coord_cartesian(ylim = ylim1*2) +
  stat_compare_means(label.x = 1.8, label.y = 25)

#TTL, BMI continuous
TL2 <- ggplot(MMRFcleanbmionly, aes(x = BMIDx, y = TTL)) +
    geom_point(color = "darkseagreen3") + 
    geom_smooth(method = "lm", color = "green4") +
    labs(x = "BMI at diagnosis", y = "Tumor telomere length") +
    stat_cor(method = "pearson", label.x = 40, label.y = 40) +
    theme_minimal ()

#LTL, BMI discrete
MMRFcleanbmionly$LTL <- as.numeric(MMRFcleanbmionly$LTL)
ylim2 = boxplot.stats(MMRFcleanbmionly$LTL)$stats[c(1, 5)]
TL3 <- ggplot(data = MMRFcleanbmionly, aes(x = factor(BMI_cat_4), y = LTL)) +
    geom_violin(draw_quantiles = c(0.25, 0.5, 0.75), ylim = range(0:6)) +
    stat_summary(fun.data=mean_sdl, mult=1, 
                 geom="pointrange", color="red") +
    geom_jitter(shape=16, position=position_jitter(0.2)) +
    stat_compare_means(label.x = 1, label.y = 12)+
  coord_cartesian(ylim = ylim1*1)

kruskal.test(TTL~EBW_binary, MMRFcleanbmionly)
lm1<- lm(LTL~BMIDx, MMRFcleanbmionly)
summary(lm1)

#LTL, BMI continuous
TL4 <- ggplot(MMRFcleanbmionly, aes(x = BMIDx, y = LTL)) +
    geom_point(color = "darkseagreen3") + 
    geom_smooth(method = "lm", color = "green4") +
    labs(x = "BMI at diagnosis", y = "Tumor telomere length") +
    stat_cor(method = "pearson", label.x = 40, label.y = 8) +
    theme_minimal() +
    ylim(0,10)

plot_grid(TL1, TL2, TL3, TL4, ncol = 2)

#To interrogate whether age relationship with telomere length is affecting outcome:
###How best to do though it TTL and TLT are not binary variables?


```

## FSGEA for genes related to ***metabolomics*** other ***obesity related cancer pathways***

*Methods (provided to and not performed by me)*: A subset of 683 samples from the CoMMpass study had RNA sequencing available.(10) Data were aligned to hg38 using STAR (2.5.1b)(11) and quality controlled using QoRTS (v1.2.42) 12 with alignment and quantification of gene read count with Salmon (v0.7.2).13 Normalization of counts and differential gene expression analysis was performed using DESeq2 (v1.14.1).(14) 

*Results*:

```{r, eval = TRUE}
#Changes to DGE data.frame
MMRFdge$sig <- rep("insignificant", nrow(MMRFdge))
MMRFdge$sig[which(MMRFdge$ANOVA<0.05)] <- "significant"
MMRFdge[c('Gene','num')] <- str_split_fixed(MMRFdge$Gene_name, "_", n=2)
MMRFsigdge <- MMRFdge %>%
  filter(sig == "significant") 
MMRFsigdge <- MMRFsigdge[1:2069, c("Gene","foldchange","ANOVA", "BMI_1", "BMI_3")]
MMRFsigdge <- rename(MMRFsigdge, p.value = ANOVA)

#Volcano Plot
ggplot(MMRFdge, aes(x = foldchange, y = -log10(ANOVA), color = sig)) + 
  geom_point() +
  theme_bw() +
  ggtitle("Volcano plot") +
  xlab("LogFC")+
  ylab("-Log10(q-value)") +
  scale_color_manual(values = c("black", "red")) +
  theme(legend.position = "none")

#Enhanced volcano using signficant values only, although it does a beautiful job with insignificants: https://github.com/kevinblighe/EnhancedVolcano
#?Normalized enrichment score

#FSGEA
library(fgsea)
kegg <- gmtPathways("~/Documents/BMIN 5030/Assignment 6/c2.cp.kegg.v7.4.symbols.gmt")
reactome <- gmtPathways("~/Documents/BMIN 5030/Assignment 6/c2.cp.reactome.v7.4.symbols.gmt")
pathways <- c(kegg, reactome)

#Ranking genes by p-value at the t-statistic was not available to me
MMRFsigdge <- MMRFsigdge %>% 
  group_by(Gene) %>%
  arrange(p.value) %>%
  filter(row_number() == 1) %>%
  data.frame()

#List, however ranking by the logfoldchange? ##Is the foldchange reported as LogFC already?
gene_list <- MMRFsigdge$foldchange
names(gene_list) <- MMRFsigdge$Gene
gene_list <- sort(gene_list, decreasing = T)

#FGSEA
fgseaRes <- fgsea(pathways = pathways, stats = gene_list, minSize = 15, 
                  maxSize = 500, nperm = 10000, gseaParam = 1)
fgseaRes <- fgseaRes[order(fgseaRes$pval), ]

#Will not collapse pathways - although only 26 pathways came up, only one with NES >2 with sig adjusted p-value, that is, KEGG_CYTOKINE_CYTOKINE_RECEPTOR_INTERACTION 
#Here is the enrichment pathway  
plotEnrichment(pathway = pathways[["KEGG_CYTOKINE_CYTOKINE_RECEPTOR_INTERACTION"]], stats = gene_list) + 
  labs(title="Kegg cytokine cytokine receptor interaaction")

#Which is enriched by the following differentially expressed genes
fgseaRes %>% 
  filter(pathway == "KEGG_CYTOKINE_CYTOKINE_RECEPTOR_INTERACTION") %>% 
  dplyr::select(leadingEdge) %>% 
  unlist() %>% 
  unname()

#Otherwise if I repeat the whole thing including the insignificant pvalues, I get more pathways, but again, only one with NES >2: REACTOME_RHO_GTPASES_ACTIVATE_IQGAPS
MMRFdge <- MMRFdge %>% 
  group_by(Gene) %>%
  arrange(ANOVA) %>%
  filter(row_number() == 1) %>%
  data.frame()

gene_list1 <- MMRFdge$foldchange
names(gene_list1) <- MMRFdge$Gene
gene_list1 <- sort(gene_list1, decreasing = T)

fgseaRes1 <- fgsea(pathways = pathways, stats = gene_list1, minSize = 15, 
                  maxSize = 500, nperm = 10000, gseaParam = 1)
fgseaRes1 <- fgseaRes1[order(fgseaRes1$pval), ]

#In this case, we might consider doing collapse pathways
collapsedPathways <- collapsePathways(fgseaRes = fgseaRes1[padj < 0.05], 
                                      pathways = pathways, stats = gene_list1)
mainPathways <- fgseaRes1[pathway %in% collapsedPathways$mainPathways] # keep results of independent pathways
head(mainPathways)

plotEnrichment(pathway = pathways[["REACTOME_RHO_GTPASES_ACTIVATE_IQGAPS"]], stats = gene_list1) + 
  labs(title="Reactome Rho-GTPases")

```

However, given a single pathway with a borderline NES score in both, sought to validate whether a significant finding by also running an overrepresentation analysis. 

```{r, eval = TRUE}
DE_genes <- as.character(MMRFdge$Gene)[which(MMRFdge$ANOVA < 0.05)]

#Creation of 2x2 table
PT <- as.character(MMRFdge$Gene)
PH <- intersect(PT, pathways[["REACTOME_RHO_GTPASES_ACTIVATE_IQGAPS"]])
LT <- DE_genes
LH <- intersect(PH, LT)
PH_LH <- setdiff(PH, LH) #PH-LH
LT_LH <- setdiff(LT, LH) #LT-LH
PT_LT_PH_LH <- setdiff(setdiff(PT, LT), PH_LH)
mat <- matrix(
  c(length(LH), length(PH_LH), length(LT_LH), length(PT_LT_PH_LH)), 
  ncol = 2, nrow = 2, 
  dimnames = list(c("in pathway", "not in pathway"), c("DE genes", "background genes")), byrow = T)
mat

#EASE score
mat <- matrix(
  c(length(LH) - 1, length(PH_LH) + 1, 
    length(LT_LH), length(PT_LT_PH_LH)), 
  ncol = 2, nrow = 2, 
  dimnames = list(c("in pathway", "not in pathway"), 
                  c("DE genes", "background genes")), byrow = T)
mat

#FISHER exact test
fisher.test(mat, alternative = "greater")
```

#DISCUSSION:
*This is still pending*

# CONCLUSIONS:
*this is still pending*


# REFERENCES
(1)
(2)
(3)
(4)
(5)
(6)


