---
title: "BMIN503/EPID600 Project: A Comparison of the Nasal Microbiome in Patients with Granulomatosis with Polyangiitis Versus Healthy Controls"
author: "Rennie Rhee"
output: 
  html_document:
    toc: false 
    depth: 3 
    theme: paper 
    highlight: tango
---

### Overview
The goal of this [project](https://github.com/rheer1/BMIN503_Final_Project) is to examine the nasal microbiota in patients with granulomatosis with polyangiitis (GPA) compared to healthy controls. With the help of collaborators at Penn, we collected nasal swabs on 60 patients with GPA and 40 healthy controls along with clinical metadata. 

I met with 3 faculty/staff to discuss my project: (1) Elizabeth Grice, PhD, is a microbiologist in the Dept of Dermatology who focuses on the skin microbiome. She helped me develop an analytic plan to address my study question and gave me some background on performing microbiome research; (2) Peter Merkel, MD, MPH, is Chief of Rheumatology at Penn and expert in vasculitis (including GPA) and patient-oriented research. He guided me on study design, tissue collection, and important factors to consider when studying patients with GPA; and (3) Chunyu Zhao, PhD, is a bioinformatician with expertise in using R to analyze metagenomic data. Dr. Zhao helped substantially with processing data (e.g., aligning sequences, taxonomic assignment using reference database) and advised me on software programs available to analyze my data. 

### Introduction 
The objective of this project is to compare the nasal microbial composition between patients with granulomatosis with polyangiitis (GPA; Wegener's) and healthy controls. GPA is a systemic granulomatous vasculitis which often affects the upper respiratory tract and frequently relapses. Prior studies demonstrated that microbes in the nose are associated with relapse in GPA but these studies relied on culture-dependent approaches. Advances in high-throughput sequencing now allow for a better understanding of the complex relationship between the dynamic community of microbes that occupy our mucosal tissues and the host's immune system. We hypothesize that patients with GPA have greater dysbiosis (disruption in indigenous microbiota) in the nares compared to controls and dysbiosis is associated with disease activity in GPA. A better understanding of the role of microbes in the disease process of GPA has the potential to: 1) redefine patient classification and identify new phenotypic subsets, 2) be used as a biomarker of disease activity, disease severity, treatment response, and patient-oriented outcomes, and 3) identify novel therapeutic agents.

This interdisciplinary project will require expertise on GPA pathogenesis and management, patient-oriented research, microbiology, and analysis of metagenomic data. I met with Dr. Peter Merkel, my primary mentor and Chief of Rheumatology Division, to discuss patient selection and addressing confounders. I met with Dr. Elizabeth Grice, a microbiologist and dermatologist, to formulate an analytic plan. I also met with Dr. Chunyu Zhao, a bioinformatician, to review R packages used for microbiome research. She provided source code which, for purpose of this project, was just used to generate the heatmaps and also helped substantially with performing tests for differential abundance. Together these faculty/staff gave critical guidance in performing this project.

### Methods
We collected nasal swabs and clinical data on 60 patients with GPA and 40 healthy controls who were seen at the Penn Vasculitis Center. Processing and DNA sequencing was performed in the CHOP Microbiome Core. DNA was extracted using the QIAGEN QIAamp Ultraclean kit and bacterial 16S ribosomal RNA was sequenced with Illumina HiSeq and primers annealing to the V1V2 region were used for amplification. Sequence data was processed with Quantitative Insights Into Microbial Ecology (QIIME). Operational Taxonomic Units (OTUs) – which are surrogates for bacterial taxa defined by 16S rRNA homology – were selected by clustering reads at 97% sequence similarity. Taxa refers to a group of organisms (at the level of species, genus, family, etc) which have similar sequences. Taxonomic assignments were generated by comparison to the Greengenes reference database, using the consensus method implemented in QIIME. A phylogenetic tree was inferred from the OTU data using FastTree.

First, descriptive statistics were performed with clinical metadata. Means and proportions were used to summarize data. Boxplots and barplots were used to visualize co-variates by group and other variables of interest. 

Definition of GPA variables:
- Birmingham Vasculitis Activity Score (BVAS) is a validated disease activity measure which was coded as binary yes/no. A severe flare is defined by major organ involvement (e.g. abnormal kidney function, lung hemorrhage) that is new or worsening in past 28 days, limited flare is without major organ involvement (e.g., skin, sinusitis, joint pain) that is new or worsening in past 28 days, and persistent disease is ongoing disease activity (not new or worsening) beyond 28 days.
- Vasculitis Damage Index (VDI) is a measure of damage from vasculitis or its therapies. For this project, I focused only on sinonasal damage.
- Antineutrophil cytoplasmic antibody (ANCA) is a blood test used for the diagnosis of GPA and ANCA type (MPO vs PR3) is associated with prognosis.

Data files from QIIME were analyzed in the R environment for statistical computing, using the QIIMER library. From the sequencing above, I obtained tables of OTUs to analyze outcomes of interest. Beta diversity, the degree of difference between bacterial communities measured as a continuous distance, was visualized using principal coordinates analysis (PCoA) of unweighted and weighted UniFrac distance (distance measurement which incorporates phylogenetic relationships, weighted by abundance). Permutational multivariate analysis of variance (PERMANOVA) is a non-parametric multivariate test to test the null hypothesis that centroids and dispersion of groups are equivalent. Alpha diversity, measured by the Shannon diversity index, will also be examined. Alpha diversity measures the number (richness) and distribution (evenness) of taxa expected within a single population. The Shannon index, a continuous measure, will be calculated from OTUs. The relative abundance of taxa in each group (percent composition of particular taxa relative to the total number of taxa) was determined for the 5 most abundant taxa present in the samples.  


```{r load scripts, eval=TRUE, echo=FALSE, message=FALSE, warning=FALSE}
library(qiimer)
library(ape)
library(vegan)
library(ggplot2)
library(dplyr)
library(reshape2)
library(tidyr)
library(broom)
library(pheatmap)

library(magrittr)
library(pander)
library(ggbeeswarm)

library(colorRamps)
library(ggthemes)
library(zoo)

library(kylemisc)
library(subfunc)
```

```{r}
#I obtained source code from Dr. Zhao which primarily helped with generating clustered heatmaps for this project
source("helper.R")
```


Read in clinical data and clean data.
```{r eval=TRUE}
c <- read.table("J:/Backup/BMI Certificate/Data Science in Biomedical Informatics/Final Project/clinical_metadata_withmeds.txt", sep = "\t", header = TRUE, stringsAsFactors = TRUE)

#Remove data from negative controls (environmental controls)
c <- subset(c, HostSpecies=="Human")

#Change "not answered" string to "NA"
c[c=="not answered"] <- NA

#Examine data frame and structure
str(c)

#Change age from factor to numeric
c$age <- as.numeric(levels(c$age))[c$age]
class(c$age)

#Change certain factor variables to characters
c$SampleID <- as.character(c$SampleID)
c$current_antibiotic_name <- as.character(c$current_antibiotic_name)
c$past6mon_antibiotic_name <- as.character(c$past6mon_antibiotic_name)
c$otherimmunoname_current <- as.character(c$otherimmunoname_current)
c$otherimmunoname_in6months <- as.character(c$otherimmunoname_in6months)

str(c)

#Clean variables with typos
c$race[c$race=="White "] <- "White"
c$match[c$match=="Umatched"] <- "Unmatched"

#Clean dates
c$date_diagnosis <- as.character(c$date_diagnosis)
c$date_diagnosis <- as.Date(as.yearmon(c$date_diagnosis))
format(c$date_diagnosis, "%m/%Y")

c$date_collected <- as.character(c$date_collected)
c$date_collected <- as.Date(c$date_collected, "%d-%b-%y")

#Replace NAs to No in VDI variables if group is GPA (this is artifact from data collection)
c$vdi_nasalblockade[is.na(c$vdi_nasalblockade) & c$study_group=="GPA"] <- "No"
c$vdi_nasalbridgecollapse[is.na(c$vdi_nasalbridgecollapse) & c$study_group=="GPA"] <- "No"
c$vdi_chronicsinusitis[is.na(c$vdi_chronicsinusitis) & c$study_group=="GPA"] <- "No"

#Create new variable indicating any sinonasal damage (any VDI item)
c$vdi_any <- ifelse(c$vdi_nasalblockade=="Yes" | c$vdi_nasalbridgecollapse=="Yes" | c$vdi_chronicsinusitis=="Yes", "Yes", "No")

#Create new variable indicating any disease activity (combine limited disease, severe disease, and persistent disease into 1 category)
c$bvas_yes <- ifelse(c$bvas_diseasestatus=="limited disease-flare" | c$bvas_diseasestatus=="severe disease-flare" | c$bvas_diseasestatus=="persistent disease", "Yes", "No")


```


Read in OTU data and merge with clinical data
```{r}
genus_props1 <- read.csv("J:/Backup/BMI Certificate/Data Science in Biomedical Informatics/Final Project/genus_props1.csv", sep = ",", header = TRUE, stringsAsFactors = FALSE, row.names=1)

genus_cts1 <- read.csv("J:/Backup/BMI Certificate/Data Science in Biomedical Informatics/Final Project/genus_cts1.csv", sep = ",", header = TRUE, stringsAsFactors = FALSE, row.names=1)

genus_cts_top_df <- read.csv("J:/Backup/BMI Certificate/Data Science in Biomedical Informatics/Final Project/genus_cts_top_df.csv", sep = ",", header = TRUE, stringsAsFactors = TRUE, row.names=1)

s1 <- read.csv("J:/Backup/BMI Certificate/Data Science in Biomedical Informatics/Final Project/s1.csv", sep = ",", header = TRUE, stringsAsFactors = FALSE, row.names=1)

s <- read.csv("J:/Backup/BMI Certificate/Data Science in Biomedical Informatics/Final Project/s.csv", sep = ",", header = TRUE, stringsAsFactors = FALSE, row.names=1)

genus_props <- read.csv("J:/Backup/BMI Certificate/Data Science in Biomedical Informatics/Final Project/genus_props.csv", sep = ",", header = TRUE, stringsAsFactors = FALSE)

genus_props_df <- read.csv("J:/Backup/BMI Certificate/Data Science in Biomedical Informatics/Final Project/genus_props_df.csv", sep = ",", header = TRUE, stringsAsFactors = FALSE)

genus_cts <- read.csv("J:/Backup/BMI Certificate/Data Science in Biomedical Informatics/Final Project/genus_cts.csv", sep = ",", header = TRUE, stringsAsFactors = FALSE, row.names=1)

s1 <- mutate(s1, Group = ifelse(grepl("Case", study_group), "GPA", "Control"))

genus_cts_top_df$StudyGroup <- as.factor(genus_cts_top_df$StudyGroup)

#Create new variable indicating any sinonasal damage (any VDI item)
s$vdi_any <- ifelse(s$vdi_nasalblockade=="Yes" | s$vdi_nasalbridgecollapse=="Yes" | s$vdi_chronicsinusitis=="Yes", "Yes", "No")
s$vdi_any <- as.factor(s$vdi_any)

#Create new variable indicating any disease activity (combine limited disease, severe disease, and persistent disease into 1 category)
s$bvas_yes <- ifelse(s$bvas_diseasestatus=="limited disease-flare" | s$bvas_diseasestatus=="severe disease-flare" | s$bvas_diseasestatus=="persistent disease", "Yes", "No")
s$bvas_yes <- as.factor(s$bvas_yes)
s$age <- as.numeric(s$age)

#Create subsetted dataframe of just GPA group
gpa <- subset(s, StudyGroup=="GPA")

```

Read QIIME files and use R package qiimer to obtain unweighted and weighted UniFrac distances
```{r}
uu_fp <- file.path("J:", "Backup", "BMI Certificate", "Data Science in Biomedical Informatics", "Final Project", "run3","beta_diversity", "unweighted_unifrac_dm.txt")
wu_fp <- file.path("J:", "Backup", "BMI Certificate", "Data Science in Biomedical Informatics", "Final Project", "run3","beta_diversity", "weighted_normalized_unifrac_dm.txt")

uu <- read_qiime_distmat(uu_fp)
uu <- dist_subset(uu, s$BarcodeID)
uu_gpa <- dist_subset(uu, gpa$BarcodeID)

wu <- read_qiime_distmat(wu_fp)
wu <- dist_subset(wu, s$BarcodeID)
wu_gpa <- dist_subset(wu, gpa$BarcodeID)
```


### Results

#### Descriptive Statistics of Patient Characteristics
```{r eval=TRUE}
#Evaluate summary of data
summary(c)

#Obtain means (sd) and counts/proportions for patient characteristics by study group
c %>% group_by(study_group) %>% summarise_if(is.numeric, mean, na.rm=TRUE)
c %>% group_by(study_group) %>% summarise_if(is.numeric, sd, na.rm=TRUE)

#I know there must be a faster way to do this but couldn't figure it out!
c %>%
  group_by(study_group, race) %>%
  summarise(n=n()) %>%
  mutate(percent = paste0(round(100 * n/sum(n), 0), "%"))

c %>%
  group_by(study_group, sex) %>%
  summarise(n=n()) %>%
  mutate(percent = paste0(round(100 * n/sum(n), 0), "%"))

c %>%
  group_by(study_group, anca_elisa) %>%
  summarise(n=n()) %>%
  mutate(percent = paste0(round(100 * n/sum(n), 0), "%"))

c %>%
  group_by(study_group, vdi_any) %>%
  summarise(n=n()) %>%
  mutate(percent = paste0(round(100 * n/sum(n), 0), "%"))

c %>%
  group_by(study_group, bvas_diseasestatus) %>%
  summarise(n=n()) %>%
  mutate(percent = paste0(round(100 * n/sum(n), 0), "%"))

c %>%
  group_by(study_group, pred_current) %>%
  summarise(n=n()) %>%
  mutate(percent = paste0(round(100 * n/sum(n), 0), "%"))

c %>%
  group_by(study_group, pred_in6months) %>%
  summarise(n=n()) %>%
  mutate(percent = paste0(round(100 * n/sum(n), 0), "%"))

c %>%
  group_by(study_group, otherimmuno_current) %>%
  summarise(n=n()) %>%
  mutate(percent = paste0(round(100 * n/sum(n), 0), "%"))

c %>%
  group_by(study_group, otherimmuno_in6months) %>%
  summarise(n=n()) %>%
  mutate(percent = paste0(round(100 * n/sum(n), 0), "%"))

c %>%
  group_by(study_group, antibiotics_current) %>%
  summarise(n=n()) %>%
  mutate(percent = paste0(round(100 * n/sum(n), 0), "%"))

c %>%
  group_by(study_group, antibiotics_in6months) %>%
  summarise(n=n()) %>%
  mutate(percent = paste0(round(100 * n/sum(n), 0), "%"))

#Determine percent missing for each variable by study group
c %>%
  group_by(study_group) %>%
  summarise_all(function(x) sum(is.na(x))/length(x))
```

In summary, age, sex, and race was similar between 2 groups. Majority of patients were PR3-ANCA positive, had sinonasal damage, were receiving prednisone or other immunosuppressive and patients wigh GPA were more likely to have received antibiotics within past 6 months (but not currently).

#### Descriptive plots of characteristics to visualize data
```{r eval=TRUE}
#Histogram of Age for All Patients
ggplot(data=c, aes(x=age)) +
  geom_histogram(binwidth=3)

#Histogram of Age by Group (GPA vs control)
ggplot(data=c, aes(x=age, fill=study_group)) +
  geom_histogram(binwidth=3)

#Bar plot of Race and Sex by Group
ggplot(data=c, aes(x=race, fill=study_group)) +
  geom_bar(position="dodge")

ggplot(data=c, aes(x=sex, fill=study_group)) +
  geom_bar(position="dodge")

#Bar plot of ANCA type within GPA Group
ggplot(data=subset(c, study_group=="GPA"), aes(x=anca_elisa)) +
  geom_bar()

#Histogram of Disease Duration Within GPA Group
ggplot(data=subset(c, study_group=="GPA"), aes(x=disease_duration_yrs)) +
  geom_histogram()

#Histogram of SNOT22 Score by Group
ggplot(data=c, aes(x=snot22, fill=study_group)) +
  geom_histogram()

#Bar plot of Disease status Based on ANCA type
p <-ggplot(data=subset(c, study_group=="GPA"), aes(x=bvas_yes, fill=anca_elisa)) +
  geom_bar(position="fill") +
  ggtitle("Disease Activity Stratified by ANCA Type") +
  labs(y="Proportion", x="Disease Activity")
ggsave(filename="p.jpg", plot=p)

#Box plot of disease duration by sinonasal damage (VDI_any) 
ggplot(data=subset(c, study_group=="GPA"), aes(x=vdi_any, y=disease_duration_yrs)) +
  geom_boxplot()

#Bar plot of sinonasal damage (vdi_any) and prior flare of sinonasal disease
ggplot(data=subset(c, study_group=="GPA"), aes(x=vdi_any, fill=ever_flare_sinonasal)) +
  geom_bar(position="fill")

#Bar plot of sinonasal damage and ANCA type
ggplot(data=subset(c, study_group=="GPA"), aes(x=anca_elisa, fill=vdi_any)) +
  geom_bar(position="fill")

#Bar plot of prior flare of sinonasal disease and ANCA type
ggplot(data=subset(c, study_group=="GPA"), aes(x=anca_elisa, fill=ever_flare_sinonasal)) +
  geom_bar(position="fill")

#Bar plot of Medication use by Disease Status
pc <- ggplot(data=subset(c, study_group=="GPA"), aes(x=bvas_yes, fill=pred_current)) +
  geom_bar(position="fill")+
  ggtitle("Current Prednisone") +
  labs(y="Proportion", x="Disease Activity") +
  theme(plot.title = element_text(size=36)) +
  theme(axis.title=element_text(size=22))
ggsave(filename="pc.jpg", plot=pc)
p6 <- ggplot(data=subset(c, study_group=="GPA"), aes(x=bvas_yes, fill=pred_in6months)) +
  geom_bar(position="fill")+
  ggtitle("Prednisone in Past 6 Months") +
  labs(y="Proportion", x="Disease Activity") +
  theme(plot.title = element_text(size=32)) +
  theme(axis.title=element_text(size=22))
ggsave(filename="p6.jpg", plot=p6)
oc <- ggplot(data=subset(c, study_group=="GPA"), aes(x=bvas_yes, fill=otherimmuno_current)) +
  geom_bar(position="fill")+
  ggtitle("Current Immunosuppressive") +
  labs(y="Proportion", x="Disease Activity") +
  theme(plot.title = element_text(size=32)) +
  theme(axis.title=element_text(size=22))
ggsave(filename="oc.jpg", plot=oc)
o6 <- ggplot(data=subset(c, study_group=="GPA"), aes(x=bvas_yes, fill=otherimmuno_in6months)) +
  geom_bar(position="fill")+
  ggtitle("Immunosuppressive in Past 6 Months") +
  labs(y="Proportion", x="Disease Activity") +
  theme(plot.title = element_text(size=24)) +
  theme(axis.title=element_text(size=28))
ggsave(filename="o6.jpg", plot=o6)
ggplot(data=subset(c, study_group=="GPA"), aes(x=bvas_yes, fill=in_steroids_current)) +
  geom_bar(position="fill")
ac <- ggplot(data=subset(c, study_group=="GPA"), aes(x=bvas_yes, fill=antibiotics_current)) +
  geom_bar(position="fill")+
  ggtitle("Current Antibiotic") +
  labs(y="Proportion", x="Disease Activity") +
  theme(plot.title = element_text(size=36)) +
  theme(axis.title=element_text(size=22))
ggsave(filename="ac.jpg", plot=ac)
a6 <- ggplot(data=subset(c, study_group=="GPA"), aes(x=bvas_yes, fill=antibiotics_in6months)) +
  geom_bar(position="fill")+
  ggtitle("Antibiotic in Past 6 Months") +
  labs(y="Proportion", x="Disease Activity") +
  theme(plot.title = element_text(size=32)) +
  theme(axis.title=element_text(size=22))
ggsave(filename="a6.jpg", plot=a6)

#Boxplot of SNOT22 score based on Disease Status
snot <- ggplot(data=subset(c, study_group=="GPA"), aes(x=bvas_diseasestatus, y=snot22)) +
  geom_boxplot()+
  ggtitle("SNOT22") +
  labs(y="Proportion", x="Disease Activity") +
  theme(plot.title = element_text(size=36)) +
  theme(axis.title=element_text(size=22))
ggsave(filename="snot.jpg", plot=snot)
```

Most interesting observations from these plots are the differences in medication between GPA patients with and without active disease. Those with active disease have similar rate of prednisone use, are less likely to be on other immunosuppressive therapies, and more likely to have received an antibiotic in past 6 months.


#### Heatmap depicting relative abundance of taxa within each sample
```{r eval=TRUE}
#Heatmap clustered by most abundant taxa (all GPA and controls)
heatmap_clustered(genus_props1, s1, grps = c("Group"), option=1, thre = 0.6, fname="J:/Backup/BMI Certificate/Data Science in Biomedical Informatics/Final Project/heatmap_abundant.pdf")

#Heatmap clustered by group (GPA vs control)
heatmap_grouped(genus_props1, s1, grps = c("Group"), option=1, thre = 0.6, fname="J:/Backup/BMI Certificate/Data Science in Biomedical Informatics/Final Project/heatmap_grouped_abundant.pdf")
```

In summary, the heatmaps show clusters of abundant taxa, specifically Staphylococcus, Corynebacterium, Propionibacterium, and Bacillales.


#### Beta diversity: Comparison of microbial communities using UniFrac distance which incorporates phylogenetic relationships

- Unweighted UniFrac: No weighting by abundance (therefore, rare taxa has equal weight as abundant taxa)
```{r eval=TRUE}
#Unweighted UniFrac test between GPA and Controls
adonis(uu ~ StudyGroup, data = s)

plot(make_pcoa_plot_discrete_barcodeID(uu, s, "StudyGroup", "StudyGroup", 'PCoA plot based on unweighted unifrac distances'))

#Unweighted UniFrac test between Active Disease and Remission within GPA
adonis(uu_gpa ~ bvas_yes, data=gpa)

plot(make_pcoa_plot_discrete_barcodeID(uu_gpa, gpa, "bvas_yes", "bvas_yes", 'PCoA plot based on unweighted unifrac distances'))

```
In summary, no significant difference in unweighted UniFrac between GPA and controls and between active and inactive GPA.

- Weighted UniFrac: Weighted by abundance
```{r eval=TRUE}
#Weighted UniFrac test Between GPA and Control
adonis(wu ~ StudyGroup, data = s)

plot(make_pcoa_plot_discrete_barcodeID(wu, s, "StudyGroup", "StudyGroup", 'PCoA plot based on weighted unifrac distances'))

plot(make_pcoa_plot_discrete_barcodeID(wu, s, "StudyGroup", "otherimmunoname_current", 'Weighted unifrac distances by current immunosuppressive'))

plot(make_pcoa_plot_discrete_barcodeID(wu, s, "StudyGroup", "otherimmunoname_current", 'Weighted unifrac distances by current immunosuppressive'))

plot(make_pcoa_plot_discrete_barcodeID(wu, s, "StudyGroup", "otherimmunoname_in6months", 'Weighted unifrac distances by immunosuppressive in past 6 months'))

adonis(wu ~ StudyGroup + otherimmuno_in6months, data = s)
adonis(wu ~ StudyGroup + sex + otherimmuno_in6months + pred_in6months +antibiotics_in6months, data = s)

#Weighted UniFrac test between Active Disease and Remission within GPA
adonis(wu_gpa ~ bvas_yes, data = gpa)
adonis(wu_gpa ~ bvas_yes  + sex + otherimmuno_in6months + pred_in6months +antibiotics_in6months, data = gpa)

plot(make_pcoa_plot_discrete_barcodeID(wu_gpa, gpa, "bvas_yes", "bvas_yes", 'PCoA plot based on weighted unifrac distances'))

```
In summary, a significant difference found between weighted UniFrac between GPA and controls even after adjusting for medications and sex. No difference in weighted UniFrac distance between active and inactive GPA even after adjusting for sex and medications. This indicates there is a significant difference in overall microbial composition between GPA and controls.


#### Alpha Diversity (Shannon diversity) - within sample diversity

```{r eval=TRUE}
#Between GPA and control
ggplot(s, aes(x=StudyGroup, y=Shannon)) +
  geom_boxplot(fill="lightblue") +
  labs(y="Shannon diversity", x="StudyGroup") +
  geom_jitter(width=.10, size=1) +
  theme_bw() +
  scale_color_brewer(palette = "Set2")

aov(Shannon ~ StudyGroup, data=s) %>% summary()

#Between Active and Inactive Disease within GPA
ggplot(gpa, aes(x=bvas_yes, y=Shannon)) +
  geom_boxplot(fill="lightblue") +
  labs(y="Shannon diversity", x="Disease Status") +
  geom_jitter(width=.10, size=1) +
  theme_bw() 

aov(Shannon ~ bvas_yes, data=gpa) %>% summary()

```
In summary, no difference in Shannon Index between groups of interest.


#### Relative Abundance of 5 Most Abundant Taxa

```{r eval=TRUE}
## Identify 5 most abundant taxa

#Remove blank controls from data fram
genus_props <- genus_props[, -c(2:17)]

#average proportion of each taxa and put in descending order 
genus_props <- transform(genus_props, avg = rowMeans(genus_props[,-1], na.rm=TRUE))
genus_top <- genus_props[order(genus_props$avg, decreasing=TRUE),]
head(genus_top)
print(genus_top$avg)

#keep top 5 taxa
topfive <- genus_top[1:5,]
topfive$avg <- NULL

#combine all other taxa into "Other" category???
topfive <- rbind(topfive, c("Other",apply(topfive[,-1],2,sum, na.rm=TRUE)))
topfive[6, 2:ncol(topfive)] <- sapply(topfive[6, 2:ncol(topfive)], function(x) 1-as.numeric(x))

#melt data
topfive_melt <- melt(topfive, id.var="X")
colnames(topfive_melt) <- c("Taxa", "BarcodeID", "Proportion")

#Merge with clinical metadata
merged <- topfive_melt %>%
  merge(s, by="BarcodeID")
merged$BarcodeID <- reorder(merged$BarcodeID, merged$Proportion)
merged$Proportion <- as.numeric(merged$Proportion)

#Make barplot of relative abundance
g <- ggplot(data=merged, aes(x=BarcodeID, y=Proportion, fill=Taxa)) +
  geom_bar(stat="identity") +
  xlab("Samples") + ylab("Relative Abundance") +
  theme_bw()
g + facet_grid(. ~ StudyGroup, scales = "free_x")

#Stacked barplot just among patients with active disease, according to disease activity
da <- ggplot(data=subset(merged, subset = bvas_diseasestatus %in% c("severe disease-flare", "limited disease-flare", "persistent disease")), aes(x=BarcodeID, y= Proportion), fill=Taxa) +
  geom_bar(stat="identity") +
  xlab("Samples") + ylab("Relative Abundance") +
  theme_bw()
daplot <- da + facet_grid(. ~ bvas_diseasestatus, scales = "free_x")
ggsave(filename="stackedplot.jpg", plot=daplot)

```


#### Differential Abundance for top 10 Most Abundant Taxa using R package pander
```{r eval=TRUE}
grps <- c("StudyGroup")

panderOptions('knitr.auto.asis', FALSE)
for (grp in grps){
  if (nlevels(genus_cts_top_df[[grp]]) == 2){
    options(scipen=1, digits=3)
    pander(tidy_rank_test(genus_cts_top_df, grp,color_by=NA), caption=paste0("rank sum test for ",grp))
  }
  else{
    options(scipen=1, digits=3)
    pander(tidy_rank_test_multiple(genus_cts_top_df, grp,color_by=NA), caption=paste0("rank sum test for ",grp))
  }
}
```
In summary, there is a lower abundance of Propionibacterium and Bacillales in GPA patients compared to controls.

#### Conclusion
In conclusion, there is a significant difference in overall microbial composition between patients with GPA compared to healthy controls and a lower proportion of Propionibacterium and Bacillales. No differences in microbial composition were found among patients with GPA who had active vs inactive disease. Additional analyses will include examining the effects of medications as well as differences between other GPA subgroups (such as ANCA type). I also plan to compare characteristics and outcomes among the different clusters of abundant bacteria (as shown in heatmap). Currently we are also collecting nasal swabs longitudinally in these patients to better examine associations between microbial composition and outcomes.A better understanding of the role of nasal microbes in GPA has the potential to improve managmenet and outcomes in patients with this disease.
