---
title: " Differential Gene Expression analysis using data from a on Alveolar Macrophage gene expression on cannabis exposure"
author: "Nilanjana Samanta"
output: 
  html_document:
    toc: true
    toc_float: 
        collapsed: true
        smooth_scroll: true
    depth: 3
    theme: paper 
    highlight: tango
---
```{r set-options, echo=FALSE, cache=FALSE}
options(width = 400)
``` 

```{r global options, include = FALSE}
knitr::opts_chunk$set(warning = FALSE, message = FALSE, fig.pos = "center", fig.height = 8, fig.width = 9)
``` 
***

# Overview
Literature on Gene expression studies on cigarette smokers are abundant and most have identified several dysregulated molecular pathways that can alter respiratory health of a smoker or increase their risk of developing chronic conditions. However, there is a lack of literature on similar and/or parallel studies in marijuana smokers, partly because there are high chances of concurrent cigarette smoking whcich acts as a confounder in determining its effect. S. Sharma and C. Liu et al in their unpublished [study](https://www.atsjournals.org/doi/pdf/10.1164/ajrccm-conference.2018.197.1_MeetingAbstracts.A4161) hypothesized that studying the difference in expression of alveolar macrophage (AM) between Marijuana and cigarette smokers can identify disease pathways that may influence respiratory disease development among those exposed to cannabis smoke. This group is also the first group to perform genome-wide gene expression analysis from the lungs of marijuana smokers. 

Marijuana use has recently been politicized a lot due to several US states legalizing its recreational use while many states either keep it limited to medical usage or criminalizeit. Further study of gene expression among marijuana smokers, could help understand associated respiratory disease pathways, if detected and aid in more informed policies around its consumption habits. This study is at the intersection of public health, Bioinformatics and Biomedical genetics and should draw insight from these fields for a deeper understanding of the effects of cannabis smoking on a molecular level in the general population.

# Introduction 

S. Sharma and C. Liu et al, is the first one to to genome wide gene expression analysis from the lungs of marijuana smokers to test their hypothesis. Hence it is a unqique dataset on marijuana exposure publicly available!

My goal is not to reproduce their results, which is difficult because not only is it an unpublished work but we don't have acess to their statistical analysis methods to compare results with. Besides I will be utilizing the processed data provided by them  hence hard to conclude a result and if my results are reliable. But my goal is to use a snapshot of their comrehensive dataset, to perform Difrerential gene expresion analysis to see if indeed there are differentially expressed genes in the samples with marijuana exposure, learn the challenges of using publicly available data and maybe compare two diffrent methods of doing DE analysis. I also wanted to test if limma is a better tool to do DE analysis where covariates need to be adjusted for. 

## Study design 
1. In this study, there are three main exposure groups of interest **a) Marijuana smokers (n = 15) b) Tobacco smokers (n = 16) and c) Controls or non-smokers (n = 10)**. 
2. All subjects were matched on the basis of their gender. 
3. Smoking history of subjects was acquired from their medical records and presence of cannabis was assessed using toxicology screening. 
3. To recruit non-smokers, participants had to report no active use of either substance for no more than a year in total.
4. The reads were mapped to **Homo sapiens GRCh37 (Ensembl v79)**
5. The sequecing for this study was performed in two separate batches i.e. the RNA library creation and sequencing were performed in two batches: Batch 1 (TruSeq Stranded mRNA Library Prep Kit and **Illumina HiSeq4000**) and Batch 2 (**Illumina TruSeq Stranded Total RNA kit, HiSeq2500**) using manufacturer directions. This is provided as a variable in the sample metadata



# Methods
I am be utilizing the raw count data for this study available in the GEO database, accession [GSE155213](https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE155213). Unlike the microarray data we saw in class, this is RNA-seq data available in the form of a feature count matrix with reads of each gene tag for each biological sample which in our case are the subjects in each exposure grops. Here columsn are the samples and rows are the gene ids with there counts. 

I downloaded the data locally and read it into Rstudio instead of using GEOquery function to access it.

## Library and package installation

I will be using the **EdgeR** package for this analysis, as it is the tool suggested and used by the authors of the above study, whic requires a raw count matrix as well phenotypic group as input. I am using **dplyr** for data cleaning and manipulation to prepare the dataset for DE analysis, **data.table** to read and write (save) tables. 
 
```{r,eval=TRUE, message=FALSE}
library(GEOquery)
#install.packages('R.utils')
library(R.utils)
library(data.table)
library(oligo)
library(limma)
library(viridis)
#BiocManager::install('biomaRt')
library(biomaRt)
#BiocManager::install('edgeR')
library(edgeR)
library(tidyr)
library(dplyr)
```

## Set working directory

In order to read the downloaded dataset from our working directory of interest and to save all files generated there, we need to set R's working directory. We can check our current working directory with the command **`getwd()`** and change it using **`setwd("path/to/file")`**

```{r,eval=TRUE,message=FALSE}
getwd() #satisfied with my working directory, hence not using setwd() to change it

```

## Loading and Reading the dataset from the GEO database into R

 The count table was downloaded into local directory and read into Rstudio using the **read.table()** function to unzip it for analysis

```{r,eval=TRUE, message=FALSE}

feature.table <- read.table(gzfile("GSE155213\\GSE155213_count_table.txt.gz"), header=TRUE, sep="\t")

#creating a safe copy of feature.table
copy.main <- feature.table

head(feature.table)
tail(feature.table)
colnames(feature.table)

```

> Immediately after importing the file I saw that the gene counts were flots and not integers as a raw gene count should be. This was an indication that the counts provided on the GEO omnibus could have already been normalized, which is not ideal. The authors had mentioned in their method section that the count table was built as an input to edgeR object

## Acquiring phenotypic matrix from eset

Here I will be adding some important phenotypic information for each sample that I will be adjusting for in my model later. They are-
1. The three  exposure groups i.e Marijuana, tobacco, control 
2. Age
3. Sex, M or F
3. Technical batch (tech.batch), batch 1 & batch 2

**NOTE:** I obtained the sample metadata (phenotype data) from eset using **`getGEO()`** function of GEOquery package. However due to some administrative permission related issue on my current system, I had to run the program on a different computer to obtain the files, saved them and then imported them  into Rstudio. Hence the specific codes to acquire and save the files have been commented out from haulting the Knitting step.   
```{r,eval=TRUE, message=FALSE}
#Acquiring phenotypic matrix in 2 parts

#part 1
#gse <- getGEO("GSE155213", destdir="data", GSEMatrix = TRUE)[[1]]
#raw.pheno <- pData(phenoData(gse))

#saving the eset file #1 locally
#write.table(raw.pheno, "GSE155213_raw_pheno.txt", row.names = F, col.names = T, sep="\t", quote=F)

#Part 2
#gse2 <- getGEO("GSE155213", destdir="data", GSEMatrix = TRUE)[[2]]
#raw.pheno2 <- pData(phenoData(gse2))

#saving the eset file #2 locally
#write.table(raw.pheno2, "GSE155213_raw_pheno2.txt", row.names = F, col.names = T, sep="\t", quote=F)


#reading both phenotypic files into rstudio
pheno.table1 <- read.table( gzfile("GSE155213\\GSE155213_raw_pheno.txt"), header=TRUE, sep="\t")

pheno.table2 <- read.table( gzfile("GSE155213\\GSE155213_raw_pheno2.txt"), header=TRUE, sep="\t")

#Merging the two phenotype datasets by rows, doing a vertical join
pheno.data <- rbind(pheno.table1, pheno.table2)
head(pheno.data)

#cleaning and subsetting pheno data
pheno.data.final <- pheno.data %>%
  select(title, characteristics_ch1, characteristics_ch1.1, characteristics_ch1.2, characteristics_ch1.5) %>%
  rename(status = title, group = characteristics_ch1, age = characteristics_ch1.1, sex = characteristics_ch1.2, tech.batch = characteristics_ch1.5) %>%
  mutate(group = gsub("group: ", "", group), age = gsub("age: ", "", age), sex = gsub("Sex: ", "", sex), tech.batch = gsub("technical_batch: ", "batch", tech.batch))

#changing group text to reflect Marijuana, Control and Tobacco groups
pheno.data.final$group <- recode(pheno.data.final$group,
                                 C = "control",
                                 `T` = "tobacco",
                                 M = "marijuana") 


head(pheno.data.final)
```

## Data exploration and Filtration

The important quality check steps that I performed were to first create a count only table,by replacing row index with unique gene-ids and removing non-count columns. Then to remove the underexpressed or lowly expressed genes from the dataset I chose a cutoff based on median log~2-transformed counts per gene per million mapped reads (cpm). edgeR provides the function, `cpm`, to compute the counts per million. However, since it produced some negative values, not ideal for DE analysis, I added a pseudo-number to the log-2 transformed values. I decided to stick with this filtration based on the boxplots that showed how scalable my transformed counts were. It is also important to transorm before fitting linear models.

Next we are only keeping a gene if it has a cpm of 0.5 or greater for at least two samples. Ideally one should know the library size of the samples to set such thresholds but sinceI did not have that information, I chose a random cutoff to only keep genes with reasonable expression. I skipped normalization because theraw data online seemed alredy normalized. 

I used the count table and the phenotype data to prepare the DGEList object that edgeR uses for storing the data from a differential expression experiment. I also Visualized sample relationships with multidimensional scaling (MDS).

```{r,eval=TRUE, message=FALSE}

#creating a count only table for DE analysis
count.table <- feature.table
rownames(count.table) <- feature.table$Features
count.table$Features <- NULL

#Creating DGElist object with "Groups"
mydata <- DGEList(count.table, group=factor(pheno.data.final$group))

dim(mydata)
head(mydata$counts)
head(mydata$samples)

#calculatin counts per million uing cpm() funcion for each gene
head(cpm(mydata))

#calculating total gene count per sample
apply(mydata$counts, 2, sum)

#Keeping genes atleast a count of 0.5 cpm for atleat 2samples and removing the rest from our dataset
keep.genes <- rowSums(cpm(mydata)>0.5) >= 2 #keep genes is a logical vector with TRUE and FALSE values
table(keep.genes)
mydata <- mydata[keep.genes,]
dim(mydata)

#log transformation
cpm_log <- cpm(mydata$counts, log = TRUE)
head(cpm_log)
```


### Boxplot

Boxplot of each gene tag to visualize effect of transformation and if it is meaningful

```{r,eval=TRUE, message=FALSE}

#boxplot showing transformed count data
boxplot(count.table, las = 2, main = "Before Log2 transformation")
boxplot(cpm_log, las = 2, main = "After log2 transformation")

```

>I plotted the box plots of gene tags to assess if the log2 transformation improves the data and normalizes it a bit, because we were dealing with really small values/integers. The plots show that it mads sense to use transformed values.   



```{r,eval=TRUE, message=FALSE}

#The log transformation, produced negative counts for values <0, so i used the below trick
#adding pseudo numbr to avoid negative values
count_log2 <- log2(mydata$counts + 1)
head(count_log2)

#boxplot of transformed values
boxplot(count_log2, las = 2, main = "After log2+1 transformation")

#creating DGElist with the transformed counts
mydata.copy <- DGEList(count_log2, group=factor(pheno.data.final$group))
dim(mydata.copy)
head(mydata.copy$counts)
```


### Multidimensional scaling plot of distances between  gene expression profiles

```{r,eval=TRUE, message=FALSE}

#sample relationship on multidimendionl scale based on all covariates in the list 
plotMDS(mydata.copy)

#Plotting sample relationship on multidimendionl scale based on exposure 
plotMDS(mydata.copy, method="bcv", col=as.numeric(mydata.copy$samples$group))
legend("bottomleft", as.character(unique(mydata.copy$samples$group)), col=1:3, pch=20)


#Plotting sample relationship on multidimendionl scale based on batch effect alone
mydata.batch <- DGEList(count_log2, group=factor(pheno.data.final$tech.batch))
mydata.batch

dim(mydata.batch)

keep.batch <- rowSums(cpm(mydata.batch)>0.5) >= 4
table(keep.batch)
mydata.batch <- mydata.batch[keep.batch,]
dim(mydata.batch)

plotMDS(mydata.batch, method="bcv", col=as.numeric(mydata.batch$samples$group))
legend("bottomleft", as.character(unique(mydata.batch$samples$group)), col=1:3, pch=20)

```


> based on the two plots, there is no clear separation of samples based on the exposure alone but some separatio based on the technical batch. But mostly we see that the smokers tent to cluster together while the non-smokers and marijuana smokers cluster nearby as well. This could be because there are not many differentially expressed genes between these exposure groups. 

# Results

## Differential Expression (DE) Analysis using Generalized Linear Model (GLM) 

Now we start preparing the data for the the test of differential expression. We create three vectors a) `group`, that labels each of the columns as belonging to group "marijuana", "tobacco" or "control" b) 'batch', labels the columns belonging to "batch 1" or "batch 2" c) age, labels thecolumns belonging to various age groups, its a continuous integer. We then use this vector and the gene counts to create a design matrix. 

I chose a Generalized Linear model (GLM) framework, because this is more than a two group comparison and has covariates like age that I am adjusting for. First step is to prepare the design matrix and then decide the contrast groups I will be comparing for gene expression levels. The linear model we construct will not have an intercept. This is referred to as the group-means parametrization 

```{r, eval=TRUE, message=FALSE}

#creating group vectors for pasing into design matrix 
group <- c(rep("marijuana", 15), rep("tobacco", 16), rep("control", 10))
batch <- c(rep("batch1", 30), rep("batch2", 11))
age <- as.numeric(c("22", "23", "24", rep("25",2), "29", rep("30",5), rep("31",2), "32", "33", rep("34",2), rep("35",2), rep("36",2), rep("37",2), rep("38",3), rep("39",3), rep("40",2), rep("41",3), "47", rep("50",3), "53", "54", "55"))

#design matrix to describe the statistical model
design <- model.matrix(~0 + group + batch + age)
colnames(design) <- gsub("group", "", colnames(design))
design
design.copy <- design #copy of design in case we make any mistakes

#This is not necessary if your GLM does not have an intercept, but I made it to visualize the contrast groups explicitly
contrast.matrix <- makeContrasts(mari_vs_cont = marijuana - control,
                              tobacco_vs_cont = tobacco - control,
                              mari_vs_tobacco = marijuana - tobacco,
                              levels = colnames(design))
contrast.matrix

#gene-specific (referred to in edgeR as tagwise) dispersion estimates are used in the test for differential expression
mydata.copy <- estimateDisp(mydata.copy, design.copy)
head(mydata.copy)

plotBCV(mydata.copy)


#fitting a GLM
fit <- glmFit(mydata.copy, design.copy)

#fit the contrasts
lrt_mc <- glmLRT(fit, coef = 2)
lrt_tc <- glmLRT(fit, coef = 3)
lrt_mt <- glmLRT(fit, contrast = c(0,1,-1,0,0))


#use toptags() to rank top genes for each contrast groups
marijuana <- topTags(lrt_mc, n=Inf, adjust.method = "BH", sort.by = "PValue", p.value = .05)
topTags(lrt_mc, n=10)

tobacco <- topTags(lrt_tc, n=Inf, adjust.method = "BH", sort.by = "PValue", p.value = .05)
topTags(lrt_tc, n=10)

mari_vs_tobacco <- topTags(lrt_mt, n=Inf, adjust.method = "BH", sort.by = "PValue", p.value = .05)
topTags(lrt_mt, n=10)

#saving the top genes list to use later
write.table(tobacco, file="top-mc.txt")
write.table(tobacco, file="top-tc.txt")

#identify DE genes, upregulated, non-expressed and downregulated in each group
mc_DE <- decideTestsDGE(lrt_mc, adjust.method="BH", p.value = .05)
summary(mc_DE)

tc_DE <- decideTestsDGE(lrt_tc, adjust.method="BH", p.value = .05)
summary(tc_DE)

mt_DE <- decideTestsDGE(lrt_mt, adjust.method="BH", p.value = .05)
summary(mt_DE)

#boxplot of one random top expressed gene in the "mari_vs_control" group
boxplot(as.numeric(count.table["ENSG00000138061", ]) ~ group)


#edgeR tests for differential expression between two classes using exactTest() function
et <- exactTest(mydata.copy)
results_edgeR <- topTags(et, n = nrow(mydata.copy), sort.by = "none")
head(results_edgeR$table)

sum(results_edgeR$table$FDR < 0.05)
plotSmear(et, de.tags = rownames(results_edgeR)[results_edgeR$table$FDR < 0.05])
abline(h = c(-2, 2), col = "blue")

```


### Testing DE analysis using Limma-voom packages

When the library sizes are quite variable between samples, then the voom approach is theoretically more powerful than limma-trend. I am passing the filtered DGElist to Voom because voom convert the read counts to log2-cpm, with associated weights, ready for linear modelling, and my **mydata.copy** list is already log-transformed. This could make the comparison beiween the two DE analysis mrhthods little difficult but overall the steps follwed are similar. 

```{r,eval=TRUE,message=FALSE}

v_DE <- voom(mydata, design.copy, plot = TRUE)
dim(v_DE)

#Fit a linear model per gene
fit_limma <- lmFit(v_DE, design.copy)

## Calculate the statistics for our specific contrasts of interest
vfit <- contrasts.fit(fit_limma, contrasts = contrast.matrix)

#Use an empirical Bayes algorithm to shrink the gene-specific variances towards the average variance across all genes
efit <- eBayes(vfit)

#diagnostic plot to check if voom removed mean-variance relationship
plotSA(efit, main = "Final model: Meanâˆ’variance trend")

#Tabulate the results
summary(decideTests(efit))

#If the magnitude of the effect size is important for downstream analysis, we can specify a minimal log-fold-change with the function `treat`
tfit <- treat(vfit, lfc = .05)

dt <- decideTests(tfit)
summary(dt)

# Identify top DE genes for Marijuana group, just to compare result with edgeR 
mari_tobacco <- topTreat(tfit, coef = 3, n = 10)

```



## Gene Annotation

Annotating the Ensemble Gene IDs in the `mari_tobacco` file, i.e. output of top 10 DE genes in the Marijuana_vs_tobacco group with gene symbols. This is just a step to prepare datasets for GSEA. or interpret results better, But here I am simply merging the files toattempt gene annotation and interpret the result. The `connection` to ensemble.org is temporarily unavailable, hence I have commented the line, in order to avoid haulting of knitting process. mart package has been haulted 

```{r,eval=TRUE, message=FALSE}

# interrogate the BioMart database to get gene symbol and description for these genes 
us_mart <- useEnsembl(biomart="ENSEMBL_MART_ENSEMBL", host="https://grch37.ensembl.org", dataset="hsapiens_gene_ensembl", mirror = "uswest")

genes <- biomaRt::getBM(attribute=c('ensembl_gene_id', 'hgnc_symbol'), values=feature.table$Features, mart=us_mart)

#removing any duplicate gene ids
genes.matrix <- genes[-which(duplicated(genes$ensembl_gene_id)),]

summary(genes.matrix)
length(unique(genes.matrix$hgnc_symbol))
length(unique(genes.matrix$ensembl_gene_id))

#Replacing blank cells with NAs
genes.matrix[genes.matrix == ""] <- NA

genes_id <- as.character(rownames(mari_tobacco))
length(genes_id)

# select genes of interest only
rownames(genes.matrix)<-genes.matrix$ensembl_gene_id
gene.annot <- genes.matrix[as.character(genes_id),]

# join the two tables
rownames(mari_tobacco) <- mari_tobacco$ID
merged.feature.table <- cbind(gene.annot,mari_tobacco)

head(merged.feature.table)

#Removing features column
merged.feature <- subset(merged.feature.table, select = -c(ensembl_gene_id))

#checking NA rows for symbol in merged.feature table before deleting them
row.has.na <- apply(merged.feature, 1, function(x){any(is.na(x))})
sum(row.has.na)

#deleting gene-ids with no corresponding symbols
merged.feature <- merged.feature[complete.cases(merged.feature),]

```


# Conclusion

The result of DE genes were simlar between limma and edgeR and hence I cannot say which package is better. perhaps a venndiagram of how many DE genes each method grabbed could help conclude this issue better. edgegR seemed to be less sensitive.

There are indeed differentially expressed genes that are clearly upregulated and down regulated in each exposure group but with some overlaps. I tried to validate my result by trying to check for one of the "smoking positive" henes "CYP1B!" or "CYP1A1" in either the Marijuana_vs_control or tobacco_vs_control groups because they are known to be upregulated on cigarette exposure. I did find CYP1B1 (see boxplot above) but not CYP1A1, maybe because it was not well covered by sequencing reads. I was also able to find several genes mentioned by the authors to be differenially expressed in the marijuana and tobaco groups. A good next step would be to use the gene annotated dataset to perform a GSEA or Gene enrichment analysis to see what pathways are up or down-regulated in these exposure groups. I skipped normalization because theraw data online seemed alredy normalized. 

However, one of the greatest limitation was the publicly available dataset that was pre-normalized, which is not ideal for DE analysis and can bias the results. There is no way to know if the log2 transformation inut some bias in my dataset. Ideally for research purposes, in such cases getting the raw data from the SRA database and starting from the fastqs and mapping them to a genome is a better idea than the count table. But its beyond the scope of this class. 


# References

1. Bioconductor EdgeR manual: https://bioconductor.org/packages/devel/bioc/manuals/edgeR/man/edgeR.pdf
2. Tutorial on DE analysis with edgeR:https://web.stanford.edu/class/bios221/labs/rnaseq/lab_4_rnaseq.html  
3. Forum on CPM cutoff: https://support.bioconductor.org/p/101838/
4. Case studies and user guide on edgeR: https://bioconductor.org/packages/release/bioc/vignettes/edgeR/inst/doc/edgeRUsersGuide.pdf
5. Limma manual: https://bioconductor.org/packages/release/bioc/vignettes/limma/inst/doc/usersguide.pdf


# Setup 

The analysis was conducted on-

```{r info}

sessionInfo()

```