---
title: "Efficient Translation of Autism Genetics Research for Clinical Decision Support""
author: "Olivia J. Veatch"
output: 
  html_document:
    toc: false 
    depth: 3 
    theme: paper 
    highlight: tango
---
```{r set-options, echo=FALSE, cache=FALSE}
options(width = 400)
```  
***

### Overview
The goal of this project was to develop an automated method for prioritizing likely clinically actionable results from genetic studies of autism spectrum disorders (ASD). It is expected that annotating genes implicated in ASD with evidence that is functionally relavent to the known etiology of this disorder will be useful to identifying risk genes that are more likely to be clinically relevant. The ultimate goal is to use this information to help clinicians be more informed when making key decisions regarding the benefits of ordering genetic testing for a patient with ASD.

### Introduction 
Autism Spectrum Disorders (ASD) are neurodevelopmental conditions with well-established influences from genetic factors. Rapid advancements in genomics technologies have helped generate massive amounts of knowledge on the genetic factors contributing to ASD. Although discovery continues to be important, there are already hundreds of genes and thousands of genomic variants implicated in risk for ASD. As such, it is also important to understand how currently implicated genes can help immediately inform diagnosis and treatment. To make strides toward this goal, it is crucial to identify efficient ways to sort through the sea of association study results to determine which ASD-associated genes are more likely to be clinically-actionable.

While there are numerous challenges limiting the clinical utility of information from genetic studies in ASD, after speaking with Ben Voight, Associate Professor of Genetics, I was able to refine the specific challenges I am interested in addressing. This project addresses two key issues: 1) the 'immediacy' gap which will be bridged by developing an efficient way to sort through genetic data to identify information about known and potentially clinically actionable genes, and 2) the 'processing' gap which will be bridged by working to validate and automate the prioritization method so that the approach can be adapted for future applications in other complex disorders and diseases. As incorporating genomic technology into patient care is largely dependent on clinicians’ perspectives of its utility, it is important to identify ways to better inform clinicians about relevant genetic findings beneficial to optimizing treatment. Following discussions with Allan Pack, John Miclot Professor of Medicine and Assistant Dean and Director of Global Sleep Programs, I identified the specific clinical decision I want to support with the approach—to help clinicians determine if genetic testing will be useful to informing their decisions when planning treatment. Therefore, clinically-actionable genetic information will be defined as any functionally relevant (e.g., expressed in brain, causes an ASD-related phenotype when knocked-out in mice) ASD-associated gene, with evidence indicating that variation in the gene may influence how a patient will respond to a drug. There are a number of excellent resources available that allow for identifying clinically useful ways to interpret information from genetic studies. Dr. Jason Moore, Edward Rose Professor of Informatics, Director of the Penn Institute for Biomedical Informatics, Senior Associate Dean for Informatics and Director of the Division of Informatics in the Department of Biostatistics and Epidemiology, helped me to identify the databases that would be useful to the automation of the approach and offered advice on how to validate the approach. Dr. Blanca Himes, Assistant Professor of Biostatistics, Epidemiology, and Informatics also offered advice regarding approaches that could be used to help validate the gene prioritization pipeline. This project focused on automating a gene prioritization approach using data from DisGeNET (http://www.disgenet.org), the Genotype Tissue Expression (GTEx) project, the International Mouse Phenotyping Consortium (IMPC; http://www.mousephenotype.org/), and the Pharos Drug Ontology database (https://pharos.nih.gov/idg/dto) to identify clinically-actionable candidate genes for ASD. Please see the following for more details: https://github.com/veatcho/BMIN503_Final_Project.git.

### Methods
To identify an initial list of candidate genes for ASD, data available in DisGeNET was queried. The benefits of this database include that the reported gene-disease associations are identified by text-mining data from multiple sources and the data can be automatically pulled via the RCurl package in R. All of the Unified Medical Language System (UMLS) Concept Unique Identifiers (CUI) that relate to ASD (e.g., Autism Spectrum Disorder, Atypical autism, Autistic Behavior, and Autistic Disorder) were pulled from the disease mappings file provided by the database. An R function provided by DisGeNET was modified to create a data frame of all genes with any level of evidence for a relationship with CUIs representing ASD. These genes were then annotated based on evidence of human brain expression, as this is the tissue most relevant to ASD etiology. A summary of baseline expression results from all available brain regions, expressed at the default level of Transcripts Per Million (TPM)>=0.5, based on RNA-seq data from the GTEx project were downloaded from the Expression Atlas (https://www.ebi.ac.uk/gxa/home). Genes were then annotated with any associated mouse phenotype term (p<0.05) that reflects symptoms in ASD when manipulated in mice. To do this, the genotype-phenotype representational state transfer (REST) application program interface (API) from the IMPC was queried. To identify currently approved drugs that target proteins encoded by genes implicated in ASD, a software request was sent via RCurl to the Pharos API. Targets labeled Tclin (i.e., currently approved) were considered. Please see the following url for more details: https://github.com/veatcho/BMIN503_Final_Project
 
+The first step was to pull all genes with evidence for an association with Autism Spectrum Disorders from DisGeNET. Evidence in the queried database (DisGeNET) includes results from genome-wide association studies, candidate gene studies and analyses of whole-exome sequence data focused on identifying genes housing 'de novo' variants.   
``` {r eval=TRUE}
library(RCurl)
library(tidyverse)
#Download UMLS CUI to several disease categories file directly from http://www.disgenet.org/web/DisGeNET/menu/downloads#r and read into R environment
diseasecuis<-read.table(url("https://github.com/veatcho/BMIN503_Final_Project/blob/master/disease_mappings.tsv?raw=true"), header = TRUE, sep = "\t", fill = TRUE, stringsAsFactors = FALSE, quote="")

#Pull Autism-related CUIs from DisGeNET gene-disease association mappings file
ASDcuis<-diseasecuis %>%
  select(diseaseId, name) %>%
  filter(grepl("^Autis|MACROCEPHALY/AUTIS|^AUTIS", name)) %>%
  distinct(diseaseId)

#Pulled original function from http://www.disgenet.org/ds/DisGeNET/scripts/disgenet.R and modified to suit the current pipeline. Modifications are denoted with the initials 'OJV'.

# Copyright [2010-2017] Integrative Biomedical Informatics Group, Research Programme on Biomedical Informatics (GRIB) IMIM-UPF 
# http://ibi.imim.es/
# contact for technical questions support@disgenet.org
# creator: janet.pinero@upf.edu  
# Script to query disgenet using a list of genes or diseases
# requires as input the gene or disease list in a file 
# the output file name
# the type of entity (gene or disease)
# the type of identifier 
###############################################################################


# main
###############################################################################
# load packages  	
#require(RCurl)

###############################################################################
# subs
###############################################################################

doQuery_modifiedOJV = function(inputFile, outFile, entity, identifier){
  #print(inputFile)
  #print(outFile)
  #print(entity)
  #print(identifier)
  
  
  # read in all data
  #inFile = read.csv(file=paste(getwd(), inputFile, sep="/"), sep="\t", header=F)
  inFile = inputFile ###OJV altered to address the object from the previous command (ASDcuis), instead of loading a file that was saved on disk
  dataFin <- data.frame(matrix(nrow=0, ncol=14)) 
  
  STR = "";
  if (entity == "gene"){
    if (identifier == "entrez"){
      STR = "c2.geneId = '"
    }
    else  if (identifier == "hgnc"){
      STR = "c2.symbol = '"
    }
    else{
      stop ( "the type of identifier must be entrez gene identifiers or gene symbols \n")
    }
  }
  else if (entity == "disease"){
    if (identifier == "cui"){
      STR = "c1.diseaseId = '"
    }
    else  if (identifier == "mesh"){
      STR = "c1.MESH = '"
    }
    else  if (identifier == "omim"){
      STR = "c1.OMIM = '"
    }
    
    else{
      stop  ("the type of identifier must be cui or mesh or omim identifiers\n")
    }
  }
  else{
    stop ("the type of entity must be disease or gene \n");
  }
  #for (ent in inFile$V1 ){
  for (ent in inFile[,1] ){ ### OJV altered to be universal and pull 1st column by index as opposed to column name
    url <- "http://www.disgenet.org/oql"
    oql <- paste( "DEFINE
   	c0='/data/gene_disease_summary',
	c1='/data/diseases',
	c2='/data/genes',
	c4='/data/sources'
    ON
    'http://www.disgenet.org/web/DisGeNET'
    SELECT
	c1 (diseaseId, name, STY, MESH, OMIM, type),
	c2 (geneId, symbol,uniprotId, description, DPI, DSI, pantherName),
	c0 (score, EI, Npmids, Nsnps)

    FROM
    c0
    WHERE
    (
    c4 = 'ALL'
    AND ", STR, ent , "' )
    ORDER BY
    c0.score DESC" , sep = "")

    dataTsv <-  getURLContent(url, readfunction =charToRaw(oql), upload = TRUE, customrequest = "POST")
    #dataTsv <- rawToChar( getURLContent(url, readfunction =charToRaw(oql), upload = TRUE, customrequest = "POST"))
    myTextConnection <- textConnection( dataTsv ) 
    data <-  read.csv( myTextConnection, header = TRUE, sep = "\t" )
    close(myTextConnection)

    if (dim(data)[1] == 0 ){
      print ( paste (entity , ent, " is not in DisGeNET ", sep = " "))
    }
    else  {
      data$c0.EI <- ifelse(data$c0.EI == "null", NA, as.character(data$c0.EI) )
      dataFin <- rbind(dataFin, data)
    }
  
  }
  
  #Create outFile if it does not exist
  address <-  paste(getwd(), outFile, sep="/")

  print(address)
  
  write.table(dataFin,  address, sep="\t", row.names = F,dec = ".", quote = FALSE)
  return(dataFin) ###OJV altered to create dataframe in R environment
}

#Query DisGeNET for all genes with evidence for an association with ASD
ASDcuigenes<-doQuery_modifiedOJV(ASDcuis, "ASDcuigenes.txt", "disease", "cui") ### NOTE: input file is object from previous command prior to doQuery function call
```

+Next, the list of ASD candidate genes was evaluated to identify genes that are specifically expressed in the human brain. As the direct pull for all GTEx data is large, only data for genes expressed at the GTEx default cut-off of TPM>=0.5 in human brain regions were downloaded. Details for identifying genes with baseline expression in human tissue from GTEx are provided here: https://www.ebi.ac.uk/gxa/help/r-data-objects.html, here http://www.bioconductor.org/packages/release/bioc/html/ExpressionAtlas.html, and here https://www.ebi.ac.uk/arrayexpress/experiments/E-MTAB-5214/. 
``` {r eval=TRUE}
# To download GTEx data in entirety:
#if (!requireNamespace("BiocManager", quietly = TRUE))
    #install.packages("BiocManager")
#BiocManager::install("ExpressionAtlas", version = "3.8")
#library('ExpressionAtlas')
#GTExdata<-getAtlasData("E-MTAB-5214")
GTExbraindata <- read.table(url("https://github.com/veatcho/BMIN503_Final_Project/blob/master/E-MTAB-5214-query-resultsbrain_TPM.tsv?raw=true"), header = TRUE, sep = "\t", fill = TRUE, stringsAsFactors = FALSE)

#Select variables of interest for ASD risk genes from DisGeNET query and prepare for merge with GTEx data
ASDcuigenes <- ASDcuigenes %>%
  mutate(Gene.Name = as.character(c2.symbol)) %>%
  select(EntrezID = c2.geneId, Gene.Name, UniProtID = c2.uniprotId, DisGeNETScore = c0.score)

#Pull only unique genes with evidence of encoding a protein product based on Uniprot ID
ASDcuigenesdistinct <- ASDcuigenes %>%
  filter(!UniProtID=='null') %>%
  distinct(UniProtID, .keep_all = TRUE)

#Create dataframe with ASD candidate protein-encoding genes expressed in human brain
ASDGenes.BrainExp <- left_join(ASDcuigenesdistinct, GTExbraindata, by='Gene.Name')
```

+Next, ASD risk genes were queried to identify genes that when knocked out (KO'd) of mice have an associated phenotype that is comparable to symptoms of ASD that are observed in humans (e.g., Abnormal Nervous System Development=MP:0003631, Abnormal Behavior/Neurological Phenotype=MP:0004924). Note that it is currently necessary to pull 'top level terms' from IMPC. Details for requesting mouse-phenotype data can be found here: http://www.mousephenotype.org/data/documentation/data-access.
``` {r eval=TRUE}
library('jsonlite')
#Pull genes from IMPC that when knocked out of mice are associated with Abnormal Nervous System Development=MP:0003631 and Abnormal Behavior/Neurological Phenotype=MP:0004924 (p<0.05)
AbnormalNervousSystemMouse0.05 <- fromJSON('https://www.ebi.ac.uk/mi/impc/solr/genotype-phenotype/select?q=top_level_mp_term_id:%22MP:0003631%22AND%20p_value:%5b0%20TO%200.05%5d&rows=100000&wt=json&indent=1', flatten = TRUE)

AbnormalBehaviorNeuroMouse0.05 <- fromJSON('https://www.ebi.ac.uk/mi/impc/solr/genotype-phenotype/select?q=top_level_mp_term_id:%22MP:0005386%22AND%20p_value:%5b0%20TO%200.05%5d&rows=100000&wt=json&indent=1', flatten = TRUE)

#Reformat lists and create dataframe of mouse gene names and phenotypic consequences when KO'd
AbnormalNervousSystemMouseGenes<-data.frame(phenotype=AbnormalNervousSystemMouse0.05$response$docs$mp_term_name[c(1:532)],  genes=AbnormalNervousSystemMouse0.05$response$docs$marker_symbol[c(1:532)], stringsAsFactors = FALSE)

AbnormalBehaviorNeuroMouseGenes<-data.frame(phenotype=AbnormalBehaviorNeuroMouse0.05$response$docs$mp_term_name[c(1:7025)],  genes=AbnormalBehaviorNeuroMouse0.05$response$docs$marker_symbol[c(1:7025)], stringsAsFactors = FALSE)

MouseKOASDtraits<-distinct(full_join(AbnormalNervousSystemMouseGenes, AbnormalBehaviorNeuroMouseGenes, by='genes'))

#To identify the most likely human orthologs of mouse genes identified above, the mouse gene list was exported and batch searched using all prediction tools available via the DRSC Integrative Ortholog Prediction Tool (DIOPT; Version 7.1 March 2018; https://www.flyrnai.org/diopt) with filtering to return only best match when there is more than one match per input gene or protein
MouseKOASDtraitsgenes <- distinct(MouseKOASDtraits, genes)
write(MouseKOASDtraitsgenes$genes, file="MouseKOASDtraitsgenes.txt")

#Read results from mouse/human ortholog query using DIOPT back into R environment
MouseKOASDtraitsgenes.hsortho <- read.table(url("https://github.com/veatcho/BMIN503_Final_Project/blob/master/MouseKO_ASDtraitsgenes_hsortho.xls?raw=true"), header = TRUE, sep = "\t", fill = TRUE, stringsAsFactors = FALSE)

#Identify mouse orthologs for human ASD risk genes
ASDGenes.BrainExp <- mutate(ASDGenes.BrainExp, Human.GeneID = EntrezID)
ASDGenes.BrainExp.MouseGenes <- left_join(ASDGenes.BrainExp, MouseKOASDtraitsgenes.hsortho[, c(1, 6)], by='Human.GeneID')

#Identify ASD risk genes that have relevant ASD-related phenotypes when knocked out of mice
MouseKOASDtraits <- mutate(MouseKOASDtraits, Mouse.GeneID=MouseKOASDtraits$genes)

#Join Mouse KO data to DisGeNET and GTEx brain expression data
ASDGenes.BrainExp.MouseKOPheno <- left_join(ASDGenes.BrainExp.MouseGenes, MouseKOASDtraits, by='Mouse.GeneID')
```

+Next, ASD risk genes were annotated to determine which genes encode proteins that are currently FDA-approved drug targets. Details for pulling data from Pharos can be found here: https://pharos.nih.gov/idg/api
``` {r eval=TRUE}
#Pull proteins included in the Pharos database at any drug development level (i.e., tdark, tbio, tchem, tclin)
#This step can be time-consuming the first pull (i.e. good time for a coffee/bathroom break)
targets_pharos <- fromJSON("https://pharos.nih.gov/idg/api/v1/targets?facet=IDG+Development+Level&top=25000&wt=json&indent=1")
drugtargets<-data.frame(genes=targets_pharos$content$gene[c(1:20244)], idgTDL=targets_pharos$content$idgTDL[c(1:20244)], uniprotId=targets_pharos$content$accession[c(1:20244)], stringsAsFactors = FALSE)

#Pull tclin level (i.e. FDA-aproved) targets
#To double-check the expected number of Tclin targets queried go to the following website: https://pharos.nih.gov/idg/targets?facet=IDG+Development+Level/Tclin
tclintargets <- filter(drugtargets, grepl("Tclin", idgTDL))

#Identify ASD risk genes that encode FDA-approved drug targets and join with DisGeNET, GTEx brain expression, and Mouse KO data
tclintargets<-mutate(tclintargets, Gene.Name=genes)
ASDGenes.BrainExp.MouseKOPheno.DrugTarget <- left_join(ASDGenes.BrainExp.MouseKOPheno, tclintargets, by='Gene.Name')
```

+Next, we evaluated if the annotations observed for the 'ASD gene set' were different for ASD risk genes by evaluating the same annotations in a random sample of human protein coding genes. The random gene set was equal to the number of ASD risk genes pulled from DisGeNET that encode protein products in UniProt (n=916).
```{r eval=TRUE}
#Pull names for all protein coding genes known in humans that are included in the European Bioinformatics Institute database (https://www.ebi.ac.uk/)
hsproteincodinggenes<-read.table(url('ftp://ftp.ebi.ac.uk/pub/databases/genenames/new/tsv/locus_types/gene_with_protein_product.txt'), header=TRUE, sep="\t", fill=TRUE, quote="")

hsproteincodinggenes <- hsproteincodinggenes %>%
  mutate(Gene.Name = as.character(symbol), Human.GeneID = as.integer(entrez_id)) %>%
  select(Gene.Name, Human.GeneID)

#Select random sample of genes
set.seed(1234)
randomhsgeneset1 <- hsproteincodinggenes %>%
  sample_n(916)

#Rerun annotation steps described above for ASD genes using random subset protein coding genes
#Brain expression in GTEx
Randomgenes1.BrainExp <- left_join(randomhsgeneset1, GTExbraindata, by='Gene.Name')

#Mouse KO phenotype related to ASD
Randomgenes1.BrainExp.MouseKOgenes <- left_join(Randomgenes1.BrainExp, MouseKOASDtraitsgenes.hsortho[, c(1, 6)], by='Human.GeneID')
Randomgenes1.BrainExp.MouseKOPheno <- left_join(Randomgenes1.BrainExp.MouseKOgenes, MouseKOASDtraits, by='Mouse.GeneID')

#Approved Drug Target
Randomgenes1.BrainExp.MouseKOPheno.DrugTarget <- left_join(Randomgenes1.BrainExp.MouseKOPheno, tclintargets, by='Gene.Name')

#Determine the proportion of ASD candidate genes pulled from DisGeNET that fall into categories of interest and compare to Random gene set
#Generate proportions for the number of ASD protein coding genes that 1) are expressed in brain, 2) have a phenotype when KO'd of mice, 3) encode approved drug target
ASDGeneAttributes <- ASDGenes.BrainExp.MouseKOPheno.DrugTarget %>%
  select(ASDRiskGene=Gene.Name, BrainGene=Gene.ID, MouseKOGene=Mouse.GeneID, DrugTarget=genes.y) %>%
  mutate(nASDriskgenes=n_distinct(ASDRiskGene)) %>%
  mutate(nBraingenes=n_distinct(BrainGene, na.rm = TRUE)) %>%
  mutate(nMouseKOgenes=n_distinct(MouseKOGene, na.rm = TRUE)) %>%
  mutate(nDrugtargets=n_distinct(DrugTarget, na.rm = TRUE)) %>%
  distinct(nASDriskgenes, nBraingenes, nMouseKOgenes, nDrugtargets) %>%
  mutate(Brain.Expression = (nBraingenes/nASDriskgenes)) %>%
  mutate(Mouse.KO.Phenotype = (nMouseKOgenes/nASDriskgenes)) %>%
  mutate(Drug.Target = (nDrugtargets/nASDriskgenes)) %>%
  select(nASDriskgenes, nBraingenes, Brain.Expression, nMouseKOgenes, Mouse.KO.Phenotype, nDrugtargets, Drug.Target)

#Generate proportions for the number of random protein coding genes that 1) are expressed in brain, 2) have a phenotype when KO'd of mice, 3) encode approved drug target
RandomGenesSet1Attributes <- Randomgenes1.BrainExp.MouseKOPheno.DrugTarget %>%
  select(ProteinCodingGene=Gene.Name, BrainGene=Gene.ID, MouseKOGene=Mouse.GeneID, DrugTarget=genes.y) %>%
  mutate(nRandomgenes=n_distinct(ProteinCodingGene)) %>%
  mutate(nBraingenes=n_distinct(BrainGene, na.rm = TRUE)) %>%
  mutate(nMouseKOgenes=n_distinct(MouseKOGene, na.rm = TRUE)) %>%
  mutate(nDrugtargets=n_distinct(DrugTarget, na.rm = TRUE)) %>%
  distinct(nRandomgenes, nBraingenes, nMouseKOgenes, nDrugtargets) %>%
  mutate(Brain.Expression = (nBraingenes/nRandomgenes)) %>%
  mutate(Mouse.KO.Phenotype = (nMouseKOgenes/nRandomgenes)) %>%
  mutate(Drug.Target = (nDrugtargets/nRandomgenes)) %>%
  select(nRandomgenes, nBraingenes, Brain.Expression, nMouseKOgenes, Mouse.KO.Phenotype, nDrugtargets, Drug.Target)
 
#Determine if proportions of annotations are different
ASDvsRandom1.BrainExp<-prop.test(x=c(ASDGeneAttributes$nBraingenes, RandomGenesSet1Attributes$nBraingenes), n=c(ASDGeneAttributes$nASDriskgenes, RandomGenesSet1Attributes$nRandomgenes))

ASDvsRandom1.MouseKO<-prop.test(x=c(ASDGeneAttributes$nMouseKOgenes, RandomGenesSet1Attributes$nMouseKOgenes), n=c(ASDGeneAttributes$nASDriskgenes, RandomGenesSet1Attributes$nRandomgenes))

ASDvsRandom1.Drugs<-prop.test(x=c(ASDGeneAttributes$nDrugtargets, RandomGenesSet1Attributes$nDrugtargets), n=c(ASDGeneAttributes$nASDriskgenes, RandomGenesSet1Attributes$nRandomgenes))

#Adjust p-values for multiple comparisons
ps <- c(ASDvsRandom1.BrainExp$p.value, ASDvsRandom1.MouseKO$p.value, ASDvsRandom1.Drugs$p.value)
newps<-p.adjust(ps, method = "BH")

#Determine if distribution of mouse phenotypes associated with KOing ASD genes is different from the random set of protein coding genes
#Pull mouse phenotype data to determine number of genes with the same phenotype
ASDgenemousekos <- ASDGenes.BrainExp.MouseKOPheno.DrugTarget %>%
    filter(!is.na(Mouse.GeneID)) %>%
    select(as.numeric(c(2, 21, 23)))

#Reshape to count number of ASD genes that show the same phenotype in KO mice
ASDgenemousekosmelt <- melt(as.data.frame(ASDgenemousekos), id.vars = "Gene.Name", na.rm = TRUE)

ASDgenemousekosmelt <- ASDgenemousekosmelt %>%
  mutate(nASDKopheno=n_distinct(Gene.Name)) %>%
  group_by(value) %>%
  mutate(KOphenoinASDGenes=n_distinct(Gene.Name))

#Pull, reshape and count number of random genes that show the same phenotype in KO mice
Randomgenemousekos <- Randomgenes1.BrainExp.MouseKOPheno.DrugTarget %>%
    filter(!is.na(Mouse.GeneID)) %>%
    select(as.numeric(c(1, 18, 20)))

Randomgenemousekosmelt <- melt(as.data.frame(Randomgenemousekos), id.vars = "Gene.Name", na.rm = TRUE)

Randomgenemousekosmelt <- Randomgenemousekosmelt %>%
  mutate(nRandomKopheno=n_distinct(Gene.Name)) %>%
  group_by(value) %>%
  mutate(KOphenoinRandomGenes=n_distinct(Gene.Name)) #Note that the number of distinct mouse genes is 71 above when counting MouseGeneIDs, but the number of distinct Human Genes is 70! Will look into issues with ortholog mapping, possibly one mouse gene mapped to 2 human orthologs?

#Compare proportion of random genes with phenotype to proportion of ASD genes
ASDMousekosinRandomgenes <- inner_join(ASDgenemousekosmelt, Randomgenemousekosmelt, by='value')

ASDMousekosinRandomgenes <- ASDMousekosinRandomgenes %>%
  distinct(value, .keep_all = TRUE) %>%
  select(value, nASDKopheno, KOphenoinASDGenes, nRandomKopheno, KOphenoinRandomGenes) %>%
  mutate(KOproportionASD=(KOphenoinASDGenes/nASDKopheno)) %>%
  mutate(KOproportionRandom=(KOphenoinRandomGenes/nRandomKopheno))

#Create empty dataframe for results from for_loop
proptest_results <- data.frame(matrix(NA, ncol = 3, nrow = nrow(ASDMousekosinRandomgenes)))
colnames(proptest_results) <- c("KOPhenotype", "Chi.square", "p")

#Run tests for each phenotype  
for (i in 1:nrow(ASDMousekosinRandomgenes)) {
  
  res <- prop.test(x=c(ASDMousekosinRandomgenes$KOphenoinASDGenes[i],
                       ASDMousekosinRandomgenes$KOphenoinRandomGenes[i]),
                   n=c(ASDMousekosinRandomgenes$nASDKopheno[i],
                       ASDMousekosinRandomgenes$nRandomKopheno[i]))
  
  proptest_results$KOPhenotype[i] <- ASDMousekosinRandomgenes$value[i]
  proptest_results$Chi.square[i] <- as.numeric(res$statistic)
  proptest_results$p[i] <- as.numeric(res$p.value)
  
}

#Adjust p-values and add to dataframe populated with results of Chi-square tests
proptest_results$padj <- p.adjust(proptest_results$p, method = "BH")
```

### Results
More ASD Risk Genes are Expressed in the Human Brain Compared to Other Evaluated Annotations

There were 916 unique protein coding genes with evidence for a relationship with ASD based on data available in DisGeNET. Of these, 847 (92.5%) were observed to meet a default cut-off level of  TPM>=5 in at least one brain region based on RNA-seq of human brain tissue. In addition, there were 128 ASD risk genes (14.0%) that encoded a protein that is an FDA-approved drug target. Furthermore, 98 ASD risk genes (10.7%) were associated with either a nervous system or behavioral phenotype when knocked out of mouse models.
``` {r eval=TRUE, fig.width=12, fig.height=10}
library(reshape2)
library(ggplot2)

#Melt data to reshape for bar plots
ASDGeneAttributes.melt <- ASDGeneAttributes %>%
  select(nASDriskgenes, Brain.Expression, Mouse.KO.Phenotype, Drug.Target)
ASDGeneAttributes.melt <- melt(as.data.frame(ASDGeneAttributes.melt), id.vars = "nASDriskgenes")

#Plot ASD Gene Attributes
ggplot(data=ASDGeneAttributes.melt, aes(x=variable, y=value, fill=variable)) + 
  geom_bar(position="dodge", stat="identity") +
  scale_fill_brewer(palette="Set1") +
  ggtitle("ASD Candidate Gene Attributes\n(n=916)") +
  theme(plot.title = element_text(hjust = 0.5, face = "bold", size=15), legend.title=element_text(size=12, face="bold"), legend.text = 
        element_text(size=12, face="bold"), axis.title = element_text(size=14, face="bold"), axis.text.x = element_text(size=12, angle = 45, 
        vjust = 0.5), axis.text.y = element_text(size=12)) + 
  labs(fill = "Functional Annotation") + ylab("Proportion of ASD Genes") + 
  scale_y_continuous(breaks=seq(0,1,0.05)) + xlab("")
```

Compared to a Random Gene Set, ASD Risk Genes are More Often Expressed in Brain, have a Mouse KO Phenotype, and Encode Drug Targets

After comparing the annotations that were selected to prioritize ASD risk genes for a subset of 916 genes that were randomly selected from all known human protein coding genes, it was observed that a higher proportion of ASD risk genes were expressed in the human brain (0.92 ASD genes vs. 0.83 random genes, p=4.22x10-9). In addition, more ASD risk genes were associated with a nervous system or behavioral phenotype when knocked out of mice (0.11 ASD genes vs 0.08 random genes, p=3.58x10-2). Furthermore, more ASD risk genes encoded approved drug targets (0.14 ASD genes vs 0.03 random genes, p=3.48x10-16).
``` {r eval=TRUE, fig.width=12, fig.height=10}
#Melt data to reshape for bar plots
RandomGene1Attributes.melt <- RandomGenesSet1Attributes %>%
    select(nRandomgenes, Brain.Expression, Mouse.KO.Phenotype, Drug.Target)
RandomGene1Attributes.melt <- melt(as.data.frame(RandomGene1Attributes.melt), id.vars = "nRandomgenes")

#Bind ASD risk gene and random gene set proportions into one dataframe
RandomGene1Attributes.compare <- RandomGene1Attributes.melt %>%
  mutate(Gene = recode(nRandomgenes, '916'=as.factor('ProteinCodingGene'))) %>%
  select(c(2:4))

ASDGeneAttributes.compare <- ASDGeneAttributes.melt %>%
  mutate(Gene = recode(nASDriskgenes, '916'=as.factor('ASDRiskGene'))) %>%
  select(c(2:4))

ASDvsRandomGeneAttributes <- rbind(ASDGeneAttributes.compare, RandomGene1Attributes.compare)

#p-values for plot pulled from p.adjust command above and need to be modified to be visually acceptable on plot
newps<-formatC(newps, format = "e", digits = 2)

#Plot comparisons of attributes between ASD vs random protein coding genes
ggplot(data=ASDvsRandomGeneAttributes, aes(x=variable, y=value, fill=Gene)) + 
  geom_bar(position=position_dodge(), stat="identity") +
  scale_fill_brewer(palette="Set1") +
  ggtitle("Functional Annotations\nASD Gene Set vs Random Gene Set") +
  theme(plot.title = element_text(hjust = 0.5, face = "bold", size=15), legend.title=element_text(size=12, face="bold"), legend.text = 
          element_text(size=12, face="bold"), axis.title = element_text(size=14, face="bold"), axis.text.x = element_text(size=12, angle = 45, vjust = 0.5), axis.text.y = element_text(size=12)) +
  xlab("Functional Annotation") + ylab("Proportion of Genes") +
  labs(fill = "Gene Set") + scale_y_continuous(breaks=seq(0,1,0.05)) +
  annotate("text",x = 1, y = 0.98, label=paste0('p=', newps[1]), cex = 5, parse=F) +
  annotate("text",x = 2, y = 0.15, label=paste0('p=', newps[2]), cex = 5, parse=F) +
  annotate("text",x = 3, y = 0.20, label=paste0('p=', newps[3]), cex = 5, parse=F)
```

ASD Risk Gene Expression Patterns in the Hippocampus and Hypothalamus are Distinct from a Random Gene Set

Based on results form hierarchical clustering of gene expression patterns across all available brain regions in GTEx, it appears that the major differences between ASD risk genes and the random gene set is that ASD risk genes have more similar levels of expression in the hippocampus and hypothalamus while gene expression of the random genes in these regions does not cluster together.
```{r eval=TRUE, fig.width=12, fig.height=10}
#Identify the specific human brain regions that are enriched for ASD candidate gene expression
library(limma)
library(gplots)

#Compare distribution of brain regions with expression of ASD genes to random set of protein coding genes
#Generate data matrices for ASD and random gene sets
ASDgenebrainexpression<-ASDGenes.BrainExp %>%
    filter(!is.na(Gene.ID)) %>%
    select(as.numeric(c(6:18)))

ASDgenebrainexpressionm <-as.matrix(ASDgenebrainexpression)

#Plot the human brain regions with the strongest expression of ASD candidate genes
#To better compare results from this plot to those of the random gene set, I considered the suggestion of imputing RNA-seq data and found the 'Linnorm.DataImput' function from the 'Linorm' package but instead chose to replace missing values with a zero assuming that if the TPM for the gene is missing this means it was not expressed, or expressed at very low levels (TPM<0.5).

ASDgenebrainexpressionm[is.na(ASDgenebrainexpressionm)] <- 0

ASDexpmap<-coolmap(ASDgenebrainexpressionm, cluster.by = "de pattern", linkage.row = "average", show.dendrogram = "both", main = "Brain Region Expression Normalized by ASD Gene\n(n=847)", xlab="Brain Region", ylab="ASD Candidate Genes", keysize=0.75,  offsetCol = 0, margins = c(12, 5))
```

``` {r eval=TRUE, fig.width=12, fig.height=10}
Randomgenebrainexpression<-Randomgenes1.BrainExp %>%
    filter(!is.na(Gene.ID)) %>%
    select(as.numeric(c(4:16)))

Randomgenebrainexpressionm <- as.matrix(Randomgenebrainexpression)
Randomgenebrainexpressionm[is.na(Randomgenebrainexpressionm)] <- 0

RandomGeneexpmap<-coolmap(Randomgenebrainexpressionm, cluster.by = "de pattern", linkage.row = "average", show.dendrogram = "both", main = "Brain Region Expression Normalized by Random Gene\n(n=763)", xlab="Brain Region", ylab="Random Gene Set", keysize=0.75,  offsetCol = 0, margins = c(12, 5))
```  

Distinct Traits are Associated with Knocking Out ASD Genes in Mice

A number of phenotypes associated with knocking out an ASD risk gene were not observed when knocking out genes in the random gene set. These included traits that reflect common core symptoms and comorbidities observed in individuals with ASD (e.g., abnormal vocalization, decreased exploration in a new environment, convulsive seizures). While a higher proportion of ASD risk genes were associated with the 'top level terms' from IMPC for Abnormal Nervous System Development (MP:0003631) and Abnormal Behavior/Neurological Phenotypes (MP:0004924), the proportions of genes associated with the more specific phenotypes in these categories, that were also observed for genes in the random gene set, were not significantly increased.

``` {r eval=TRUE, fig.width=12, fig.height=10}
#Order by phenotype associated with the most ASD genes
ASDgenemousekosmelt <- arrange(ASDgenemousekosmelt, desc(KOphenoinASDGenes))
#Set levels for plotting
kophenos <- unique(ASDgenemousekosmelt$value)
ASDgenemousekosmelt$value <- factor(ASDgenemousekosmelt$value, levels = kophenos)

#Generate plot showing distribution of kophenos by ASD gene  
ggplot(data=ASDgenemousekosmelt, aes(x=value, y=KOphenoinASDGenes, fill=value)) + 
  geom_bar(position="dodge", stat="identity") +
  ggtitle("Most Common Phenotypes of Interest in ASD Gene Mouse KOs\n(n=98)") +
      theme(plot.title = element_text(hjust = 0.5)) +
      xlab("Phenotypes of Interest") + ylab("Number of ASD Genes with Phenotype in KO") +
      labs(fill = "Phenotypes of Interest") +
      theme(axis.text.x=element_blank()) +
      scale_y_continuous(breaks=seq(0,40,5))
```

```{r eval=TRUE, fig.width=12, fig.height=10}
#Plot comparisons of mouse phenotypes associated with KOing ASD genes to the random set of protein coding genes
#Join ASD risk gene and random gene set proportions into one dataframe
ASDMousekosinRandomgenes <- left_join(ASDgenemousekosmelt, Randomgenemousekosmelt, by='value')

ASDMousekosinRandomgenes <- ASDMousekosinRandomgenes %>%
  distinct(value, .keep_all = TRUE) %>%
  select(value, nASDKopheno, KOphenoinASDGenes, nRandomKopheno, KOphenoinRandomGenes) %>%
  mutate(ASDGenes=(KOphenoinASDGenes/nASDKopheno)) %>%
  mutate(RandomGenes=(KOphenoinRandomGenes/nRandomKopheno)) %>%
  select(Phenotype=value, ASDGenes, RandomGenes) %>%
  arrange(desc(ASDGenes))

#Replace NAs with zeros for plotting comparison
ASDMousekosinRandomgenes$RandomGenes[is.na(ASDMousekosinRandomgenes$RandomGenes)] <- 0

#Ensure that phenotypes are plotted in descending order from most frequent
ASDMousekosinRandomgenesmelt <- melt(ASDMousekosinRandomgenes, id.vars = 'Phenotype')
ASDMousekosinRandomgenesmelt <- arrange(ASDMousekosinRandomgenesmelt, desc(value))

correct_order <- unique(ASDMousekosinRandomgenesmelt$Phenotype)

ASDMousekosinRandomgenesmelt$Phenotype <- factor(ASDMousekosinRandomgenesmelt$Phenotype, levels=correct_order)

#Generate plot comparing ASD Gene phenotypes to Random genes
ggplot(data=ASDMousekosinRandomgenesmelt, aes(x=Phenotype, y=value, fill=variable)) + 
  geom_bar(position="dodge", stat="identity") +
  ggtitle("Mouse KO Phenotypes of Interest\nASD Compared to Protein Coding Genes") +
  theme(plot.title = element_text(hjust = 0.5)) +
  xlab("Phenotypes of Interest") + ylab("Proportion of Genes with Phenotype in KO") +
  labs(fill = "Gene.Set") +
  scale_y_continuous(breaks=seq(0,1,0.1)) +
  theme_bw() +
  theme(plot.title = element_text(hjust = 0.5, face = "bold", size=15), legend.title=element_text(size=12, face="bold"), legend.text = element_text(size=12, face="bold"), axis.title = element_text(size=14, face="bold"), axis.text.x = element_text(size=12, angle = 90, vjust = 0.5, hjust = 1), axis.text.y = element_text(size=12))
  
```    
``` {r eval=TRUE}
proptest_results <- proptest_results %>%
  arrange(p)
proptest_results
```

The ASD Gene Prioritization Approach Identified 14 Genes

Of the original set of 916 genes pulled from DisGeNET with evidence for an association with ASD, 14 genes were prioritized for having all functional attributes of interest. All of these genes are expressed in the brain, associated with an ASD-related trait when knocked out of mice, and encode FDA-approved drug targets. These genes are of particular interest for follow-up studies that will be focused on evaluating current evidence for containing clinically-relevant variants (e.g., pathogenic, pharmacogenomic).
``` {r eval=TRUE}
#How many/which ASD Risk genes have evidence in all three functional categories
ASDallannotations<-ASDGenes.BrainExp.MouseKOPheno.DrugTarget %>%
  filter(!is.na(Gene.ID)) %>%
  filter(!is.na(Mouse.GeneID)) %>%
  filter(!is.na(genes.y)) %>%
  distinct(Gene.Name, .keep_all=TRUE) %>%
  select(Gene.Name)

ASDallannotations
```

Overall, the approach was capable of capturing functionally-relevant information related to currently implicated ASD risk genes. There were 847 ASD risk genes expressed in human brain and 98 with associated ASD-related phenotypes in mouse knock-outs. In addition, 129 ASD risk genes encode currently approved drug targets, indicating these are potentially clinically actionable. Notably, to ultimately translate results from human genetic studies into the clinic it is often necessary to conduct extensive research focused on deciphering the mechanisms underlying expression of symptoms in ASD in relation to individual genes. Considering the large number of genes implicated in ASD, functional characterization for each gene would limit the ability to rapidly deliver evidence useful to informing personalized approaches to treatment. From the larger list of 916 candidate genes, our method identified a small subset (n=14) that meet all of the chosen functional criteria and have strong potential to be clinically-relevant. We expect these results may have helped pinpoint the ASD candidate genes that should be prioritized for future work aimed at more detailed functional characterization.