---
title: "Efficient Translation of Autism Genetics Research for Clinical Decision Support""
author: "Olivia J. Veatch"
output: 
  html_document:
    toc: false 
    depth: 3 
    theme: paper 
    highlight: tango
---
```{r set-options, echo=FALSE, cache=FALSE}
options(width = 400)
```  
***

### Overview
This project will develop an automated method for prioritizing clinically actionable results from genetic studies of autism spectrum disorders (ASD). The goal is to rapidly deliver important information in such a manner that the clinician can be more informed when making key decisions regarding the benefits of ordering genetic testing for a patient.


### Introduction 
Autism Spectrum Disorder (ASD) is a neurodevelopmental disorder with well-established influences from genomic factors. Rapid advancements in genomics technologies have helped generate massive amounts of new knowledge on the genetic factors contributing to ASD. Although discovery continues to be important, it is crucial to understand how currently implicated genes can help immediately inform diagnosis and treatment. There are currently hundreds of genes and thousands of genomic variants implicated in risk for ASD and the list is continually growing. In order to understand how genetic information can be useful to informing treatment, it is important to identify efficient ways to sort through the sea of association study results to determine clinically-actionable genes.

Integration of genetics and genomics into precision medicine has been successful in cancer prediction and therapy; however, the need to incorporate these data into non-oncological practice of medicine has been noted as an area that requires particular attention. Following discussions with Allan Pack, Professor of Medicine in the Center for Sleep and Circadian Neurobiology, I was able to identify the specific clinical decision I want to support with the development of the approach (i.e., help clinicians to determine if genetic testing will be useful to informing their decisions when planning treatment). As incorporating genomic technology into patient care is largely dependent on clinicians’ perspectives of its utility, it is important to identify ways to better inform clinicians about relevant genetic findings beneficial to optimizing treatment. Although the cost of generating new data is steadily decreasing, there remain barriers in interpreting and translating these data. These include limited access to necessary resources and knowledge gaps among clinicians regarding the benefits/limitations of genetic testing for many complex disorders. While there are numerous challenges limiting the clinical utility of information from genetic studies in ASD, after speaking with Ben Voight, Associate Professor of Genetics, I was able to refine the specific challenges I am interested in addressing. The proposed project addresses two key issues: 1) the 'immediacy' gap which will be bridged by developing an efficient way to sort through genetic data to identify information about known and potentially clinically actionable genes, and 2) the 'processing' gap which will be bridged by working to validate and automate the prioritization method so that the validated approach can be adapted for future applications in other complex disorders and diseases. There are excellent resources available that allow for identifying clinically useful ways to interpret information from genetic studies. Dr. Jason Moore, Professor of Biostatistics and Director of the Institute for Informatics, helped me to identify the specific databases that would be useful to the automation of the approach and offered advice on how to validate the approach. I will use data form DisGeNET, Genotype-Tissue Expression database, International Mouse Phenotyping Consortium, and Pharos Drug Ontology to automate a gene prioritization pipeline useful to identification of clinically-actionable candidate genes for ASD. Clinically-actionable genetic information will be defined as any functionally relevant (e.g., expressed in brain, causes an ASD-related phenotype when knocked-out in mice), ASD-associated gene with evidence indicating that variation in the gene may influence how a patient will respond to a drug. https://github.com/veatcho/BMIN503_Final_Project.git

### Methods
In the first paragraph, describe the data used and general methodological approach. Subsequently, incorporate full R code necessary to retrieve and clean data, and perform analysis. Be sure to include a description of code so that others (including your future self) can understand what you are doing and why.

The first step is to pull all genes with evidence for influencing risk for Autism Spectrum Disorders from DisGeNET.
``` {r eval=TRUE}
library(RCurl)
library(tidyverse)
#download UMLS CUI to several disease categories file from http://www.disgenet.org/web/DisGeNET/menu/downloads#r and read into R environment
diseasecuis<-read.table(url("https://github.com/veatcho/BMIN503_Final_Project/blob/master/disease_mappings.tsv?raw=true"), header = TRUE, sep = "\t", fill = TRUE, stringsAsFactors = FALSE)

#Pull Autism-related CUIs
ASDcuis<-diseasecuis %>%
  select(diseaseId, name) %>%
  filter(grepl("^Autis|MACROCEPHALY/AUTIS|^AUTIS", name)) %>%
  distinct(diseaseId)

#Pulled original function from http://www.disgenet.org/ds/DisGeNET/scripts/disgenet.R and modified to suit my pipeline_OJV

# Copyright [2010-2017] Integrative Biomedical Informatics Group, Research Programme on Biomedical Informatics (GRIB) IMIM-UPF 
# http://ibi.imim.es/
# contact for technical questions support@disgenet.org
# creator: janet.pinero@upf.edu  
# Script to query disgenet using a list of genes or diseases
# requires as input the gene or disease list in a file 
# the output file name
# the type of entity (gene or disease)
# the type of identifier 
###############################################################################


# main
###############################################################################
# load packages  	
require(RCurl)

###############################################################################
# subs
###############################################################################

doQuery_modifiedOJV = function(inputFile, outFile, entity, identifier){
  #print(inputFile)
  #print(outFile)
  #print(entity)
  #print(identifier)
  
  
  # read in all data
  #inFile = read.csv(file=paste(getwd(), inputFile, sep="/"), sep="\t", header=F)
  inFile = inputFile ###OJV altered to address the object from the previous command (ASDcuis), instead of loading a file that was saved on disk
  dataFin <- data.frame(matrix(nrow=0, ncol=14)) 
  
  STR = "";
  if (entity == "gene"){
    if (identifier == "entrez"){
      STR = "c2.geneId = '"
    }
    else  if (identifier == "hgnc"){
      STR = "c2.symbol = '"
    }
    else{
      stop ( "the type of identifier must be entrez gene identifiers or gene symbols \n")
    }
  }
  else if (entity == "disease"){
    if (identifier == "cui"){
      STR = "c1.diseaseId = '"
    }
    else  if (identifier == "mesh"){
      STR = "c1.MESH = '"
    }
    else  if (identifier == "omim"){
      STR = "c1.OMIM = '"
    }
    
    else{
      stop  ("the type of identifier must be cui or mesh or omim identifiers\n")
    }
  }
  else{
    stop ("the type of entity must be disease or gene \n");
  }
  #for (ent in inFile$V1 ){
  for (ent in inFile[,1] ){ ### OJV altered to be universal and pull 1st column by index as opposed to column name
    url <- "http://www.disgenet.org/oql"
    oql <- paste( "DEFINE
   	c0='/data/gene_disease_summary',
	c1='/data/diseases',
	c2='/data/genes',
	c4='/data/sources'
    ON
    'http://www.disgenet.org/web/DisGeNET'
    SELECT
	c1 (diseaseId, name, STY, MESH, OMIM, type),
	c2 (geneId, symbol,uniprotId, description, DPI, DSI, pantherName),
	c0 (score, EI, Npmids, Nsnps)

    FROM
    c0
    WHERE
    (
    c4 = 'ALL'
    AND ", STR, ent , "' )
    ORDER BY
    c0.score DESC" , sep = "")

    dataTsv <-  getURLContent(url, readfunction =charToRaw(oql), upload = TRUE, customrequest = "POST")
    #dataTsv <- rawToChar( getURLContent(url, readfunction =charToRaw(oql), upload = TRUE, customrequest = "POST"))
    myTextConnection <- textConnection( dataTsv ) 
    data <-  read.csv( myTextConnection, header = TRUE, sep = "\t" )
    close(myTextConnection)

    if (dim(data)[1] == 0 ){
      print ( paste (entity , ent, " is not in DisGeNET ", sep = " "))
    }
    else  {
      data$c0.EI <- ifelse(data$c0.EI == "null", NA, as.character(data$c0.EI) )
      dataFin <- rbind(dataFin, data)
    }
  
  }
  
  #Create outFile if it does not exist
  address <-  paste(getwd(), outFile, sep="/")

  print(address)
  
  write.table(dataFin,  address, sep="\t", row.names = F,dec = ".", quote = FALSE)
  return(dataFin) ###OJV altered to create dataframe in R environment
}

ASDcuigenes<-doQuery_modifiedOJV(ASDcuis, "ASDcuigenes.txt", "disease", "cui") ### NOTE: input file is object from previous command prior to doQuery function call
```
Next, we are interested in identifying genes that are specifically expressed in the tissue of most interest to ASD etiology (i.e., the brain). Details for identifying genes with baseline expression in human tissue from GTEx are provided here: https://www.ebi.ac.uk/gxa/help/r-data-objects.html, here http://www.bioconductor.org/packages/release/bioc/html/ExpressionAtlas.html, and here https://www.ebi.ac.uk/arrayexpress/experiments/E-MTAB-5214/ and optional code is provided below if the user prefers. As the direct pull for all GTEx data is large and we are only interested in brain regions for this project, data for genes expressed at the GTEx default cut-off of FPKM≥0.5 in all brain regions available were downloaded directly and are made availalbe via github.
``` {r eval=TRUE}
# To download GTEx data in entirety
#if (!requireNamespace("BiocManager", quietly = TRUE))
    #install.packages("BiocManager")
#BiocManager::install("ExpressionAtlas", version = "3.8")
#library('ExpressionAtlas')
# GTExdata<-getAtlasData("E-MTAB-5214")

GTExbraindata<-read.table(url("https://github.com/veatcho/BMIN503_Final_Project/blob/master/E-MTAB-5214-query-results.tsv?raw=true"), header = TRUE, sep = "\t", fill = TRUE, stringsAsFactors = FALSE)
summary(GTExbraindata)

#Create dataframe with ASD candidate genes expressed in human brain
ASDcuigenes<-ASDcuigenes %>% 
      select(c2.geneId, c2.symbol, c2.uniprotId, c0.score) %>%
      mutate(Gene.Name = as.character(c2.symbol))

ASDbraingenes<-inner_join(ASDcuigenes, GTExbraindata[, c(1:2)], by='Gene.Name')
ASDbraingenes<-distinct(ASDbraingenes, Gene.ID, .keep_all = TRUE)
```


### Results
Describe your results and include relevant tables, plots, and code/comments used to obtain them. End with a brief conclusion of your findings related to the question you set out to address. You can include references if you'd like, but this is not required.
