---
title: "BMIN503/EPID600 Vaccination Trends and PBMC Gene Expression Responses to Ebola Vaccination"
author: "David Mai"
output: 
  html_document:
    toc: false 
    depth: 3 
    theme: paper 
    highlight: tango
---
```{r set-options, echo=FALSE, cache=FALSE}
options(width = 400)
```  
***

### Overview

The invention of vaccines has arguably changed the course of human history by drastically decreasing the incidence of many infectious diseases, saving countless lives. However, recent years have seen an increase in reported cases of diseases thought to have been effectively eliminated since the advent and implementation of vaccination. This project seeks to analyze historic and recent trends between vaccination coverage and disease incidence across various infectious diseases and explore the genetic profile of immune cells as a response against vaccination using Ebola as a case study to highlight factors associated with immunity.

Follow the link [here][repository] to my final project GitHub repository.

<!-- Links -->
[repository]: https://github.com/davemai/BMIN503_Final_Project

### Introduction

The first reported vaccine in the western world was invented by Edward Jenner, who observed that milkmaids did not suffer the severe disfiguring seen by others during the contraction of smallpox. It turned out that milkmaids contracted cowpox from contact with the cowâ€™s udders, leading to a milder form of the pox that provided protection against smallpox. Since then, the concept of vaccination has been applied to many different diseases such as measles, mumps, and polio, effectively eliminating them from the western world. However, recent trends show an increase in parents refusing to vaccinate their children due to religious reasons or the belief that vaccines cause impairment. This trend coincides with the resurfacing of diseases thought to be long eliminated since the implementation of comprehensive vaccination policies.

This problem can be best approached with interdisciplinary expertise from clinicians, public health experts, as well as health surveillance officials. However, stakeholders in social economics, vaccine developers, and public health workers would benefit from increased understanding about why particular diseases are resurfacing. I am using datasets provided from the WHO and unicef for vaccination coverage and disease incidence by country over time (1980s - 2018). Their datasets also cover various vaccine antigens on a per country basis. Analysis of these datasets will lead to a greater understanding of how vaccines contributed to the precipitous decline in once-dreaded diseases, and hopefully insight into their recent recurrence. Additionally, I take Ebola vaccination data (sequencing and microwell data of PBMCs) as a case study to highlight what genetic factors are upregulated upon vaccination and identify key differences that lead to effective immunity.


### Methods
In the first paragraph, describe the data used and general methodological approach. Subsequently, incorporate full R code necessary to retrieve and clean data, and perform analysis. Be sure to include a description of code so that others (including your future self) can understand what you are doing and why. 

Package and library installation:
```{r, eval = TRUE, message = FALSE}


```

Import datasets:
```{r, eval = TRUE, message = FALSE}

# Dataset on vaccination coverage
coverage_series <- read.csv("C:/Users/DMAI/Downloads/BMIN503/BMIN503_Final_Project/data/coverage_series.csv")

# Datasets on various disease incidences
crs_incidence <- read.csv("C:/Users/DMAI/Downloads/BMIN503/BMIN503_Final_Project/data/crs_incidence.csv")
diptheria_incidence <- read.csv("C:/Users/DMAI/Downloads/BMIN503/BMIN503_Final_Project/data/diptheria_incidence.csv")
japenc_incidence <- read.csv("C:/Users/DMAI/Downloads/BMIN503/BMIN503_Final_Project/data/japenc_incidence.csv")
measles_incidence <- read.csv("C:/Users/DMAI/Downloads/BMIN503/BMIN503_Final_Project/data/measles_incidence.csv")
mumps_incidence <- read.csv("C:/Users/DMAI/Downloads/BMIN503/BMIN503_Final_Project/data/mumps_incidence.csv")
ntetanus_incidence <- read.csv("C:/Users/DMAI/Downloads/BMIN503/BMIN503_Final_Project/data/ntetanus_incidence.csv")
pertussis_incidence <- read.csv("C:/Users/DMAI/Downloads/BMIN503/BMIN503_Final_Project/data/pertussis_incidence.csv")
polio_incidence <- read.csv("C:/Users/DMAI/Downloads/BMIN503/BMIN503_Final_Project/data/polio_incidence.csv")
rubella_incidence <- read.csv("C:/Users/DMAI/Downloads/BMIN503/BMIN503_Final_Project/data/rubella_incidence.csv")
ttetanus_incidence <- read.csv("C:/Users/DMAI/Downloads/BMIN503/BMIN503_Final_Project/data/ttetanus_incidence.csv")
yfever_incidence <- read.csv("C:/Users/DMAI/Downloads/BMIN503/BMIN503_Final_Project/data/yfever_incidence.csv")
regional_global_incidence <- read.csv("C:/Users/DMAI/Downloads/BMIN503/BMIN503_Final_Project/data/regional_global_incidence.csv")

# Datasets on Ebola factors
# https://www.ncbi.nlm.nih.gov/pmc/articles/PMC5011782/
# https://www.ncbi.nlm.nih.gov/pmc/articles/PMC5497219/

```

Clean datasets:
```{r, eval = TRUE, message = FALSE}


```

Analysis:
```{r, eval = TRUE, message = FALSE}


```


### Results
Describe your results and include relevant tables, plots, and code/comments used to obtain them. End with a brief conclusion of your findings related to the question you set out to address. You can include references if you'd like, but this is not required.
