---
title: "Interferon-driven immunotherapy resistance and response signatures in human hepatocellular carcinoma"
author: "Caitlin Foley"
output: 
  html_document:
    toc: false 
    depth: 3 
    theme: paper 
    highlight: tango
---
```{r set-options, echo=FALSE, cache=FALSE}
options(width = 400)
```  
***
### Overview
Anti-PD1 therapy has revolutionized cancer treatment however, in hepatocellular carcinoma (HCC) only 20% of respond to treatment.  The mechanisms underlying HCC resistance to anti-PD1 are presently unknown.  This work analyzes HCC RNA sequencing data from the Cancer Genome Atlas (TCGA) to assess whether signatures predictive of anti-PD1 therapy resistance in other tumor types are also present in HCC.

https://github.com/cfoley04/BMIN503_Final_Project


### Introduction 
Hepatocellular carcinoma (HCC) is a primary malignancy of liver hepatocytes and the third leading cause of cancer mortality world wide.  In recent years, new systemic therapies have been approved for the treatment of advanced HCC, including immune checkpoint blockade (ICB) targeting programmed cell death protein 1 (PD1).  Unfortunately, only a minority of HCC patients (20%) respond to anti-PD1 therapy and there is little understanding of how anti-PD1 resistance occurs.  Our laboratory has previosly shown in melanoma, breast, and lung cancers that a gene expression signature composed of a subset of interferon (IFN) target genes, termed ISG.RS, predicts acquired resistance to ICB.  Re-analyzing the RNA sequencing data from the TCGA HCC project will offer insights as to whether HCC ICB resistance is mediated by this ISG.RS signature.

Despite precision medicine advances in other tumor types, molecular phenotyping has been limited in HCC as this cancer is typically diagnosed based on radiographic criteria and not biopsied.  However, in 2017, TCGA released molecular profiling including RNA sequencing.  This profiling has confirmed the involvement of previously known pathways in HCC pathogenesis, including TP53, beta-catenin, and TERT, however predictive molecular phenotypes of treatment response remain elusive.  Re-analysis of these data crosses multiple disciplines, utilizing gene expression signatures initially described in human data sets and then confirmed mechanistically in mouse models, leveraging RNA sequencing analysis pipelines developed by data scientists, and correlating these findings to clinical care of HCC.  A significant challenge in this analysis is the heterogenity of HCC tumors in the data set in terms of etiology of HCC (hepatitis B, hepatitis C, alcoholic cirrhosis, etc), tumor grade, and purity of tumor sample.  Additionally, these specimens were obtained from surgical resections and may not be representative of tumors which would be exposed to anti-PD1 therapy. However, HCC molecular phenotyping is presently accelerating with a Chinese cohort released in October so additional data sets can be added to this analysis to improve power. 

### Methods
TCGA hepatocellular carcinoma dataset includes somatic mutation and DNA copy number analysis from 363 patient samples (surgical resections) and full analysis of DNA methylation, mRNA expression, microRNA expression, and protein expression in 196 patients and was published in Cell in 2017.  The National Cancer Institute Genomic Data Commons (GDC) is the repository for all TCGA data in addition to other data sources.  Any de-identified data (including RNA sequencing) is open access.  GDC currently contains RNA sequecning for 371 hepatocellular carcinomas.  For a different project, Jingya Qui (GCB PhD candidate) had already done some of the initial data processing which she shared with me.  The HTseq-counts files were downloaded for each sample using the GDC download client.  This data was then loaded into a table (genes by sample (file name)).  Using PMACS, DESeq2 was performed and low count genes (<1% samples with counts >3) were filtered out.  Counts were then normalized using VST.

I obtained the normalized RNA counts table from Jingya.  I then modified the table to include gene name (HGNC) instead of ENSMBL ID (removing any genes without HGNC or duplicates).  To interface with sample meta-data, I converted the sample name to GDC UUID based on the manifest which had been used to download the original files.  I then filtered the expression data into two subsets of interferon-target genes of interest, the treatment-resistance signature (ISG.RS) and the treatment-responsive signature (IFNG.GS).  Given that there appeared to be a subset of samples with high expression of these signatures, I then generated a "score" to be able to exam samples with a high signature vs low signature.  Because all of the genes in the signatures are targets for upregulation following interferon signaling, I created a simple score by scaling the expression of each gene to avoid highly expressed genes dominating the signature and summed the scaled values for all genes across each sample.  This score was then used to divide the data into quartiles so first and fourth quartiles could be correlated with survival data.  

Survival and other meta data was obtained from the supplemental tables of Liu, J. et al, Cell (2018).

 
```{r}
library(tidyverse)
library(biomaRt) #enables retrieval from biomart
library(rjson) #reads JSON file (TCGA metadata)
library(pheatmap) #generates clustered heatmaps
library(TCGAutils) #From bioconductor to convert UUIDs
library(RColorBrewer)
library(survminer)
library(ggplot2)
```

Manipulating HCC sequencing counts data frame to include patient UUID
```{r}
#setwd("~/repo/BMIN503_Final_Project")

#load table of RNA counts from TCGA HCC data set that Jingya Qiu combined into one matrix and normalized by DESeq2.  Check.names=F necessary so R does not add z to unsupported file names
HCCseq <- read.table("TCGA_LIHC_RNAseq_normalized_counts_DESeq2_vst_FULL.txt", check.names = FALSE)

#FOR ANALYSIS OFF GITHUB, load subset (40 smaples) of data that is small enough file to upload to github
#HCCseq <- read.table("TCGA_LIHC_RNAseq_normalized_counts_DESeq2_vst.txt", check.names = FALSE)


#Add HGNC column to Ensembl ID
ensembl = useMart("ensembl",dataset="hsapiens_gene_ensembl")
ensembl_gene_id <- c(rownames(HCCseq))
bm <- getBM(attributes=c("ensembl_gene_id","hgnc_symbol"), values=rownames(HCCseq), mart=ensembl) #generate data frame of ensembl ID and HGNC
HCCseq <- mutate(HCCseq, ensembl_gene_id = rownames(HCCseq)) #makes ensembl ID a column instead of row names
HCCs <- cbind(ensembl_gene_id, HCCseq)
HCCseq_hgnc <- inner_join(bm, HCCseq, by = "ensembl_gene_id") #add HGNC symbol, removes any genes without HGNC

#having difficulty converting hgnc to row names due to duplicates
hgnc_occur <- data.frame(table(HCCseq_hgnc$hgnc_symbol))
hgnc_occur[hgnc_occur$Freq > 1,] #six HGNC names map to 2 ensembl id
#ensmbl_occur <- data.frame(table(HCCseq_hgnc$ensembl_gene_id))
#ensmbl_occur[ensmbl_occur$Freq>1,] #three ensembl have 2 rows

#remove duplicated hgnc rows so I can use them for row names
HCCseq_hgnc <- HCCseq_hgnc[!duplicated(HCCseq_hgnc$hgnc_symbol),]
hgnc_occur2 <- data.frame(table(HCCseq_hgnc$hgnc_symbol))
hgnc_occur2[hgnc_occur2$Freq > 1,] #Fixed
rownames(HCCseq_hgnc) <- HCCseq_hgnc[,2]
HCCseq_hgnc <- HCCseq_hgnc[,-1] #remove ensmbl column
#HCCseq_hgnc <- HCCseq_hgnc[,-1] #remove hgnc column

```

To merge with meta data, need to convert col names (derived from file names) to UUID using manifest that downloaded original data as a key
```{r}
manifest <- read.table("gdc_manifest_LIHC.txt", sep="\t", check.names = FALSE, header=T, stringsAsFactors=F)
df_fileUUIDtoCaseUUID <- UUIDtoUUID(manifest$id) #TCGAutils file-->case UUID

rna.dat_fileUUID <- manifest$id[match(colnames(HCCseq_hgnc), sapply(strsplit(manifest$filename, split="\\."), function(x) x[[1]]))] #truncates filename to remove extensions, translate HCCseq colnames (filenames) to file UUID from manifest
rna.dat_caseUUID <- df_fileUUIDtoCaseUUID$cases.case_id[match(rna.dat_fileUUID, df_fileUUIDtoCaseUUID$file_id)] #translate HCCseq file UUID to case UUID
colnames(HCCseq_hgnc) <- rna.dat_caseUUID
#dupcols <- data.frame(table(colnames(HCCseq_hgnc)))
#dupcols[dupcols$Freq>1,] #Converting UUIDtoUUID has generated 52 columns with duplicate names
#dupfileid <- data.frame(table(df_fileUUIDtoCaseUUID$cases.case_id)) 
#going back to GDC portal, appears these samples are patients with multiple HTseq.counts files, will ignore for now
#dupfile <- data.frame(table(colnames(HCCseq_hgnc))) 
```

Import metadata using database compiled by Liu, et al (Cell (2018) 173:2) from TCGA (easier format to work with than downloading each individual patient JSON file)
```{r}
library("readxl")
HCCmeta <- read_excel("metadata_Liu2018.xlsx") #loads spreadsheet limited to HCC meta data
HCCmeta_UUID <- barcodeToUUID(HCCmeta$bcr_patient_barcode)
HCCmeta_UUID <- HCCmeta_UUID %>% rename(bcr_patient_barcode = submitter_id)
HCCmeta2 <- inner_join(HCCmeta_UUID, HCCmeta, by = "bcr_patient_barcode")
rownames(HCCmeta2)<-HCCmeta2$case_id
```


Examine gene sets of interferon target genes that predict resistance to cancer treatment (ISG.RS) or response (IFNG.GS)
```{r}
#load ISG.RS and IFNG.GS gene lists
ISG.RS <- read.delim("ISG.RS.txt", sep="", header=F, stringsAsFactors=F)
ISG.RS <- ISG.RS$V1
IFNG.GS <- read.delim("IFNG.GS.txt", sep="", header=F, stringsAsFactors = F)
IFNG.GS <- IFNG.GS$V1

#Filter data set down to 37 genes in resistance signature
HCCseq_ISG.RS <- HCCseq_hgnc[as.numeric(na.omit(match(ISG.RS, HCCseq_hgnc[,1]))), 1:(ncol(HCCseq_hgnc)-1)]
#Remove hgnc column.
HCCseq_ISG.RS <- HCCseq_ISG.RS[,-1]


#Filter data set down to 176 genes in IFNG responsive signature
HCCseq_IFNG.GS <- HCCseq_hgnc[as.numeric(na.omit(match(IFNG.GS, HCCseq_hgnc[,1]))), 1:(ncol(HCCseq_hgnc)-1)]
#Remove ensmbl and hgnc columns. make rownames hgnc symbol.
HCCseq_IFNG.GS <- HCCseq_IFNG.GS[,-1]
```

Generate scaled average of gene expression across each signature to divide data into signature high and low quartiles.

```{r}
#Resistance Score
RSmeta <- t(HCCseq_ISG.RS) #scale works on columns so need to transpose 
RSmeta <- scale(RSmeta) #scale so highly expressed genes dont dominate signature
RSmeta <- t(RSmeta)
RSmeta <- as.tibble(RSmeta)
Resis_Score <- summarise_all(RSmeta, mean) #take mean scaled expression of all genes
Resis_Score <- t(Resis_Score)

#Sensitivity Score
SSmeta <- t(HCCseq_IFNG.GS) #scale works on columns so need to transpose 
SSmeta <- scale(SSmeta) #scale so highly expressed genes dont dominate signature
SSmeta <- t(SSmeta)
SSmeta <- as.tibble(SSmeta)
Sens_Score <- summarise_all(SSmeta, mean) #take mean scaled expression of all genes
Sens_Score <- t(Sens_Score)
```

Merge resistance and sensitivity scores with patient meta data table
```{r}
Resis_Scorerow <- rownames(Resis_Score)
Resis_Score <- cbind(Resis_Scorerow, Resis_Score)
Resis_Score <- as.tibble(Resis_Score)
Resis_Score <- Resis_Score %>% rename(case_id = Resis_Scorerow) %>% rename (Resistance_Score = V2)

case_id <- rownames(Sens_Score)
Sens_Score <- cbind(case_id, Sens_Score)
Sens_Score <- as.tibble(Sens_Score)
Sens_Score <- Sens_Score %>% rename(Sensitivity_Score = V2)

HCCmetaScored <- inner_join(Resis_Score, HCCmeta2, by = "case_id")
HCCmetaScored <- inner_join(Sens_Score, HCCmetaScored, by = "case_id")

```



### Results
Consistent with other cancer types our group has studied, the hepatocellular carcinoma samples display heterogenity in expression of the interferon target signatures with some patients demonstrating high expression of the resistance signature and some patients having high response signatures.

```{r}
 ## 'Overwrite' default draw_colnames with your own version 
assignInNamespace(x="draw_colnames", value="draw_colnames_45",
ns=asNamespace("pheatmap"))

p_theme <- theme_classic(base_size = 12) +
	theme(axis.ticks = element_line(color = "black"),
		axis.text = element_text(color = "black", size = 12),
		axis.title.x=element_blank(),
		legend.position="none")

color_palette <- rev(brewer.pal(11, "RdBu"))

#Graph IFN signatures
draw_colnames_45 <- function (coln, gaps, ...) {
    coord = pheatmap:::find_coordinates(length(coln), gaps)
    x = coord$coord - 0.5 * coord$size
    res = textGrob(coln, x = x, y = unit(1, "npc") - unit(3,"bigpts"), vjust = 0.5, hjust = 1, rot = 45, gp = gpar(...))
    return(res)}


p <- pheatmap(t(scale(t(HCCseq_IFNG.GS))), color=color_palette, border_color=NA, fontsize_row=3, show_colnames=F, cluster_cols=T, cluster_rows=T, main = "Response gene set")

#pr <- pheatmap(t(scale(t(HCCseq_ISG.RS))), color=color_palette, border_color=NA, fontsize_row=6, show_colnames=F, cluster_cols=T, cluster_rows=T)
#a few random genes in random samples are skewing the color scale and not centering on 0, rescale data to bring white to zero
ISG_meta <- as.numeric(colMeans(t(scale(t(HCCseq_ISG.RS)))))
quantile(ISG_meta)
mat_scaled <- t(scale(t(HCCseq_ISG.RS)))
mat_scaled[which(mat_scaled>5)] <- 5 #cap any scale values >5
p2 <- pheatmap(mat_scaled, color=color_palette, border_color=NA, fontsize_row=6, show_colnames=F, cluster_cols=T, cluster_rows=T, main = "Resistance gene set")

```

Survival for patients with high versus low expression of the interferon response signature (high expression should be improved survival).
```{r}
quantile(as.numeric(HCCmetaScored$Sensitivity_Score))
high <- HCCmetaScored %>% filter(as.numeric(HCCmetaScored$Sensitivity_Score) > 0.25)
low <- HCCmetaScored %>% filter(as.numeric(HCCmetaScored$Sensitivity_Score) < -0.35)
```

Survival for patients with high versus low expression of the interferon resistance signature (high expression should be worse survival).
```{r}



```

