---
title: "BRCA1/2 Associated Gastric Cancer - Kole Buckley"
format: html
editor: visual
theme: yeti
embed-resources: true
execute: 
  warning: false
  message: false
---

```{r}
knitr::opts_chunk$set(warning = FALSE)
```

## Overview:

The overarching goal of this project is to explore potential mechanisms for an increased risk of gastric cancer among *BRCA1* and *BRCA2* pathogenic germline variant (PGV) carriers. Herein, we detail one route of exploration involving the use of *BRCA1* and *BRCA2* PGV patient-derived gastric organoids (PDGOs). Using PDGOs we look at differences in the number of organoids formed, growth, and morphology between PDGOs generated from *BRCA1* and *BRCA2* PGV carriers as well as compared to PDGOs derived from patients without a known gene mutation.

## Introduction:

Breast cancer susceptibility gene one (*BRCA1*) and two (*BRCA2*), are well established tumor suppressor genes that play a pivotal role in promoting homologous recombination in response to DNA damage (1-3). Indeed, pathogenic germline variants (PGVs) in either of these genes leads to an increase in non-homologous end joining, which can in turn promote genomic instability and tumorigenesis (2-5). While *BRCA1*/*2* PV carriers are well established to be at risk for breast and ovarian cancer, there is mounting evidence that *BRCA1/2* PGV carriers also have an elevated risk of gastric cancer (GC) (6-11). Indeed, a recent study, among others, showed a cumulative risk as high as 21.3% for *BRCA1* and 19.3% for *BRCA2* PV carriers by age 85, however this study was performed in a Japanese population where risk of GC is increased at baseline (7). Importantly, despite this potential increased GC risk, there are currently no recommended GC surveillance guidelines for *BRCA1/2* PGV carriers. Furthermore, mechanistic implications of gastric carcinogenesis remain obscure as little to no studies have attempted to explore this in the context of a *BRCA1/2* PGV.

For this project we utilized organoids. Organoids are miniature three-dimensional (3D) cellular structures that resemble the architecture and functionality of the organs from which they were derived (12,13). These lab-grown models are created by cultivating stem cells or tissue-specific cells in a controlled environment that allows these cells to self-organize and differentiate into various cell types (12-14) . One of the key advantages of organoids is their ability to recapitulate human biology more closely than traditional two-dimensional (2D) cell cultures (12-14). In particular, human organoids have been shown to maintain the genetic diversity of their tissue of origin (14-16). Organoids offer a unique opportunity to study human organ development, model diseases, and test potential therapeutics in a controlled laboratory setting. Furthermore, organoids can be derived from individual patient samples, enabling personalized medicine approaches and the potential development of individualized treatments (14,17,18).

Researchers have used human gastric organoids to investigate various aspects of gastric biology and pathology. Prominent examples include the use of patient-derived gastric organoids (PDGOs) to predict gastric cancer chemotherapy responses (19-21) and model the epithelial response to *Helicobacter pylori* infection (22-24). Human gastric organoids consist of various cell types found in the stomach, including neck cells, pit cells, and other supporting cells (22,25). Gastric organoids can either be generated from induced pluripotent stem cells (iPSCs) or stem cells directly isolated from gastric tissue obtained via biopsies or from gastric resection specimens (22,25).

To begin to elucidate mechanisms associated with GC carcinogenesis in *BRCA1/2* PGV carriers we utilized PDGOs generated from gastric biopsies acquired during an upper endoscopic procedure. Using PDGOs we compare the number of organoids formed, growth, and morphology between *BRCA1* and *BRCA2* PGV carriers as well as those from patients that do not harbor a *BRCA1/2* PGV. This work intersects with several different fields, including gastroenterology, genetics, and microbiology. The goal of this work is to identify a mechanistic avenue to further explore in future studies.

## Methods:

*Tissue Acquisition:*

This research has been approved by the Institutional Review Board (IRB) at the University of Pennsylvania. All human tissue utilized in this protocol was collected from individuals who provided informed consent for tissue collection. Patient gastric biopsies were obtained from the Hospital of the University of Pennsylvania via upper endoscopy. For each patient, gastric biopsies were take from both the body and antral regions of the stomach. In total biospies were collected from both the gastric body and antrum from 10 different *BRCA1* PGV carriers, 6 *BRCA2* PGV carriers, and 8 control patients. These biopsies were used to generate PDGOs and which were analyzed separately based on the location of the biopsy (body vs. antrum).

*Patient-derived Gastric Organoid Generation and Data Collection:*

Gastric biopsy tissue from individual patients were enzymatically digested to liberate single cells. Single cells were embedded in a basement membrane extract and seeded in 24-well cell culture plates at a standardized density of 100,000 cells per a well. The entirety of each well was imaged at 10-, 15-, and 20-days post initiation. From these images, organoid number, size, and morphology were measured using an image analysis software (cellSens, Olympus). Data were imported and organized using Excel. All the data shown in this project was collected by myself.

*Importing the Data Set from Excel:*

Here I import the data set from an Excel file. I do not do any "cleaning" as I had already organized this data in a comprehensive manner as I collected it, prior to this project.

```{r}
# intall readxl, if needed
#install.packages("readxl")

# load readxl
library(readxl)

# import my exel dataset
excel_data_path <- "/Users/kolebuckley/Desktop/BMIN_503/final_project/Data/final_project_data.xlsx"
excel_data <- read_excel(excel_data_path)

# here i show the first 5 rows as an example of the what the data looks like
head(excel_data, n = 5)

```

## Results:

\*I have elected not to run statistical analysis on this data as it is not a complete data set yet. I still have a few samples yet to analyze, and don't want to bias myself based on whatever statistical findings may be found with the current data shown in this document.

*Representative Images:*

Here we show representative images of organoids at 10-, 15-, and 20-days post initiation for *BRCA1* and *BRCA2* PGV carriers as well as controls (Figure 1).

![](images/BMIN_org_pics.jpg)

Figure 1: Representative images of organoid growth from day 10 to 20 post initiation. Organoids shown were derived from the gastric body. Scale bars = 500 um.

*Gastric Organoid Number:*

We first looked to see if there are differences in the number of organoids generated from gastric biopsies of the gastric body and antrum among *BRCA1* and *BRCA2* PGV carriers as well as controls (no PGVs). We measure organoid number at 10-, 15-, and 20-days post initiation.

```{r}
# here i create bar graphs to show differences betwen organoid number at 10, 15, and 20 days post organoid initiation between organoids from BRCA1 carriers, BRCA2 carriers, and control

# load necessary packages
library(ggplot2)
library(tidyverse)

# Filter the data so as to only examine the organoids dervied from biopsies of the gastric body first
filtered_data_body <- excel_data %>%
  filter(Location == "Body")

# Reshape the data to long format for easier usage
tidy_data_body <- filtered_data_body %>%
  gather(key = "Timepoint", value = "OrganoidNumber", Organoid_num_D10, Organoid_num_D15, Organoid_num_D20)

# Create a bar plot
ggplot(tidy_data_body, aes(x = Timepoint, y = OrganoidNumber, fill = Genotype)) +
  geom_bar(stat = "summary", fun.y = "mean", position = "dodge") +
  labs(title = "Average Number of Organoids Derived From Biopsies of the Gastric Body",
       y = "Number of Organoids",
       x = "Timepoint") +
  theme_minimal() +
  scale_fill_manual(values = c("blue", "black", "gray"))

# for organoids dervied from biopsies of the antral region of the stomach
filtered_data_antral <- excel_data %>%
  filter(Location == "Antral")

tidy_data_antral <- filtered_data_antral %>%
  gather(key = "Timepoint", value = "OrganoidNumber", Organoid_num_D10, Organoid_num_D15, Organoid_num_D20)

ggplot(tidy_data_antral, aes(x = Timepoint, y = OrganoidNumber, fill = Genotype)) +
  geom_bar(stat = "summary", fun.y = "mean", position = "dodge") +
  labs(title = "Average Number of Organoids Derived From Biopsies of the Gastric Antrum",
       y = "Number of Organoids",
       x = "Timepoint") +
  theme_minimal() +
  scale_fill_manual(values = c("blue", "black", "gray"))
```

From this data we see that organoids derived from both *BRCA1* and *BRCA2* PGV carriers grow more numerous than those derived from controls, at all timepoints. While there does not appear to be any differences in the number of organoids between *BRCA1* and *BRCA2* PGV carriers when organoids are generated from gastric body, it does appear that *BRCA2* organoids generated from the gastric antrum grow more numerous than *BRCA1*.

*Gastric Organoid Growth:*

We next looked at the size of organoids generated from gastric biopsies of the gastric body and antrum among *BRCA1* and *BRCA2* PGV carriers as well as controls. This was done by aquiring images of the entire well in which the organoids grew in and analyzed via an image analysis software. Organoid size is depicted as area (um\^2).

```{r}
# here i create a bar graphs to show differences betwen organoid size as measured by area at 10, 15, and 20 days post organoid initiation between organoids from BRCA1 carriers, BRCA2 carriers, and control

# Reshape the data to long format for easier usage, we will look at organoid area derived from biopsies of the gastric body first
tidy_data_body_size <- filtered_data_body %>%
  gather(key = "Timepoint", value = "OrganoidSize", Area_D10, Area_D15, Area_D20)

# Create a bar plot
ggplot(tidy_data_body_size, aes(x = Timepoint, y = OrganoidSize, fill = Genotype)) +
  geom_bar(stat = "summary", fun.y = "mean", position = "dodge") +
  labs(title = "Average Area of Organoids Derived From Biopsies of the Gastric Body",
       y = "Organoid Area (um^2)",
       x = "Timepoint") +
  theme_minimal() +
  scale_fill_manual(values = c("blue", "black", "gray"))

# for organoids derived from biopsies of the gastric antrum
tidy_data_antral_size <- filtered_data_antral %>%
  gather(key = "Timepoint", value = "OrganoidSize", Area_D10, Area_D15, Area_D20)

ggplot(tidy_data_antral_size, aes(x = Timepoint, y = OrganoidSize, fill = Genotype)) +
  geom_bar(stat = "summary", fun.y = "mean", position = "dodge") +
  labs(title = "Average Area of Organoids Derived From Biopsies of the Gastric Antrum",
       y = "Organoid Area (um^2)",
       x = "Timepoint") +
  theme_minimal() +
  scale_fill_manual(values = c("blue", "black", "gray"))
```

As it relates to organoids derived from the gastric body, it appears that organoids from *BRCA1* PGV carriers and controls grow at a similar pace, as measured by organoid area. Although, by day 20 controls are the largest. Meanwhile, body organoids from *BRCA2* PGV carriers show little growth. For organoids derived from the gastric antrum, we see that those from *BRCA1* PGV carriers exhibit a slightly greater growth compared to *BRCA2* PGV carriers and controls.

Next we will look at organoid diameter.

```{r}
#here i create a bar graphs to show differences betwen organoid size as measured by diameter at 10, 15, and 20 days post organoid initiation between organoids from BRCA1 carriers, BRCA2 carriers, and control

# Reshape the data to long format for easier usage, we will look at organoid diameter derived from biopsies of the gastric body first
tidy_data_body_diam <- filtered_data_body %>%
  gather(key = "Timepoint", value = "OrganoidDiameter", Diameter_D10, Diameter_D15, Diameter_D20)

# Create a bar plot
ggplot(tidy_data_body_diam, aes(x = Timepoint, y = OrganoidDiameter, fill = Genotype)) +
  geom_bar(stat = "summary", fun.y = "mean", position = "dodge") +
  labs(title = "Average Diameter of Organoids Derived From Biopsies of the Gastric Body",
       y = "Organoid Diameter (um)",
       x = "Timepoint") +
  theme_minimal() +
  scale_fill_manual(values = c("blue", "black", "gray"))

#for organoids derived from biopsies of the gastric antrum
tidy_data_antral_diam <- filtered_data_antral %>%
  gather(key = "Timepoint", value = "OrganoidDiameter", Diameter_D10, Diameter_D15, Diameter_D20)

ggplot(tidy_data_antral_diam, aes(x = Timepoint, y = OrganoidDiameter, fill = Genotype)) +
  geom_bar(stat = "summary", fun.y = "mean", position = "dodge") +
  labs(title = "Average Diameter of Organoids Derived From Biopsies of the Gastric Antrum",
       y = "Organoid Diameter (um)",
       x = "Timepoint") +
  theme_minimal() +
  scale_fill_manual(values = c("blue", "black", "gray"))
```

As can seen been here, organoids diameter largely follows the same trends seen when measuring organoid area. However for organoids derived from the gastric antrum, there is a bit more separation for organoids for *BRCA1* PGV carriers and controls compared to *BRCA2* PGV carriers.

Lastly we look a organoid perimeter.

```{r}
#here i create a bar graphs to show differences betwen organoid size as measured by perimeter at 10, 15, and 20 days post organoid initiation between organoids from BRCA1 carriers, BRCA2 carriers, and control

# Reshape the data to long format for easier usage, we will look at organoid perimeter derived from biopsies of the gastric body first
tidy_data_body_perim <- filtered_data_body %>%
  gather(key = "Timepoint", value = "OrganoidPerimeter", Perimeter_D10, Perimeter_D15, Perimeter_D20)

# Create a bar plot
ggplot(tidy_data_body_perim, aes(x = Timepoint, y = OrganoidPerimeter, fill = Genotype)) +
  geom_bar(stat = "summary", fun.y = "mean", position = "dodge") +
  labs(title = "Average Perimeter of Organoids Derived From Biopsies of the Gastric Body",
       y = "Organoid Perimeter (um)",
       x = "Timepoint") +
  theme_minimal() +
  scale_fill_manual(values = c("blue", "black", "gray"))

#for organoids derived from biopsies of the gastric antrum
tidy_data_antral_perim <- filtered_data_antral %>%
  gather(key = "Timepoint", value = "OrganoidPerimeter", Perimeter_D10, Perimeter_D15, Perimeter_D20)

ggplot(tidy_data_antral_perim, aes(x = Timepoint, y = OrganoidPerimeter, fill = Genotype)) +
  geom_bar(stat = "summary", fun.y = "mean", position = "dodge") +
  labs(title = "Average Perimeter of Organoids Derived From Biopsies of the Gastric Antrum",
       y = "Organoid Perimeter (um)",
       x = "Timepoint") +
  theme_minimal() +
  scale_fill_manual(values = c("blue", "black", "gray"))
```

As can seen been here, organoids perimeter largely follows the same trends seen when measuring organoid area and diameter. However, like the unlike the area findings, organoids derived from the gastric antrum show a bit more separation for organoids from *BRCA1* PGV carriers and controls compared to *BRCA2* PGV carriers, when measuring perimeter.

*Gastric Organoid Morphology:*

Organoid morphology as measured by sphericity (where a value of 1 equals a perfect sphere) was accessed for *BRCA1* and *BRCA2* PGV carriers as well as controls.

For any given well of organoids, we saw a diversity of morphologies. An example of different organoid morphologies are shown in Figure 2.

![](images/morph_images.jpg)

Figure 2: Organoid morphology. Scale bars = 200 um.

```{r}
# here i create a bar graph to show differences betwen organoid morphology as measured via Sphericity at 10, 15, and 20 days post organoid initiation between organoids from BRCA1 carriers, BRCA2 carriers, and control

# Reshape the data to long format for easier usage, we will look at organoid sphericity derived from biopsies of the gastric body first
tidy_data_body_morph <- filtered_data_body %>%
  gather(key = "Timepoint", value = "OrganoidMorphology", Sphericity_D10, Sphericity_D15, Sphericity_D20)

# Create a bar plot
ggplot(tidy_data_body_morph, aes(x = Timepoint, y = OrganoidMorphology, fill = Genotype)) +
  geom_bar(stat = "summary", fun.y = "mean", position = "dodge") +
  labs(title = "Average Spheriicity of Organoids Derived From Biopsies of the Gastric Body",
       y = "Organoid Sphericity",
       x = "Timepoint") +
  theme_minimal() +
  scale_fill_manual(values = c("blue", "black", "gray"))

# for organoids derived from the gastric antrum
tidy_data_antral_morph <- filtered_data_antral %>%
  gather(key = "Timepoint", value = "OrganoidMorphology", Sphericity_D10, Sphericity_D15, Sphericity_D20)

ggplot(tidy_data_antral_morph, aes(x = Timepoint, y = OrganoidMorphology, fill = Genotype)) +
  geom_bar(stat = "summary", fun.y = "mean", position = "dodge") +
  labs(title = "Average Sphericity of Organoids Derived From Biopsies of the Gastric Antrum",
       y = "Organoid Sphericity",
       x = "Timepoint") +
  theme_minimal() +
  scale_fill_manual(values = c("blue", "black", "gray"))
```

Despite observing a variety of morphologies in a given well, on average, the results here show little to no differences in sphericity amongst *BRCA1* and *BRCA2* PGV carriers as well as controls.

Next we will look at organoid shape factor (where a value of 1 is equal to a circle with an equal perimeter).

```{r}
# here i create a bar graph to show differences betwen organoid morphology as measured via Shape Factor at 10, 15, and 20 days post organoid initiation between organoids from BRCA1 carriers, BRCA2 carriers, and control

# Reshape the data to long format for easier usage, we will look at organoid shape factor derived from biopsies of the gastric body first
tidy_data_body_sf <- filtered_data_body %>%
  gather(key = "Timepoint", value = "OrganoidShapeFactor", ShapeFactor_D10, ShapeFactor_D15, ShapeFactor_D20)

# Create a bar plot
ggplot(tidy_data_body_sf, aes(x = Timepoint, y = OrganoidShapeFactor, fill = Genotype)) +
  geom_bar(stat = "summary", fun.y = "mean", position = "dodge") +
  labs(title = "Average Shape Factor of Organoids Derived From Biopsies of the Gastric Body",
       y = "Organoid Shape Factor",
       x = "Timepoint") +
  theme_minimal() +
  scale_fill_manual(values = c("blue", "black", "gray"))

# for organoids derived from the gastric antrum
tidy_data_antral_sf <- filtered_data_antral %>%
  gather(key = "Timepoint", value = "OrganoidShapeFactor", ShapeFactor_D10, ShapeFactor_D15, ShapeFactor_D20)

ggplot(tidy_data_antral_sf, aes(x = Timepoint, y = OrganoidShapeFactor, fill = Genotype)) +
  geom_bar(stat = "summary", fun.y = "mean", position = "dodge") +
  labs(title = "Average Shape Factor of Organoids Derived From Biopsies of the Gastric Antrum",
       y = "Organoid Shape Factor",
       x = "Timepoint") +
  theme_minimal() +
  scale_fill_manual(values = c("blue", "black", "gray"))
```

While there are some slight changes to this data compared to organoid sphericity, the overall trends are similar. That is, there does not appear to be any differences among organoids derived from *BRCA1* and *BRCA2* PGV carriers and controls when evaluated via shape factor.

Lastly, we look at organoid elongation (the higher the value, the more elongated the organoid is shaped).

```{r}
# here i create a bar graph to show differences betwen organoid morphology as measured via elongation at 10, 15, and 20 days post organoid initiation between organoids from BRCA1 carriers, BRCA2 carriers, and control

# Reshape the data to long format for easier usage, we will look at organoid elongation derived from biopsies of the gastric body first
tidy_data_body_elong <- filtered_data_body %>%
  gather(key = "Timepoint", value = "OrganoidElongation", Elongation_D10, Elongation_D15, Elongation_D20)

# Create a bar plot
ggplot(tidy_data_body_elong, aes(x = Timepoint, y = OrganoidElongation, fill = Genotype)) +
  geom_bar(stat = "summary", fun.y = "mean", position = "dodge") +
  labs(title = "Average Elongation of Organoids Derived From Biopsies of the Gastric Body",
       y = "Organoid Elongation",
       x = "Timepoint") +
  theme_minimal() +
  scale_fill_manual(values = c("blue", "black", "gray"))

# for organoids derived from the gastric antrum
tidy_data_antral_elong <- filtered_data_antral %>%
  gather(key = "Timepoint", value = "OrganoidElongation", Elongation_D10, Elongation_D15, Elongation_D20)

ggplot(tidy_data_antral_elong, aes(x = Timepoint, y = OrganoidElongation, fill = Genotype)) +
  geom_bar(stat = "summary", fun.y = "mean", position = "dodge") +
  labs(title = "Average Elongation of Organoids Derived From Biopsies of the Gastric Antrum",
       y = "Organoid Elongation",
       x = "Timepoint") +
  theme_minimal() +
  scale_fill_manual(values = c("blue", "black", "gray"))
```

Again, there are minute differences between this data and that for sphericity and shape factor, however, the overall trends look similar. That is, there does not appear to be any differences among organoids derived from *BRCA1* and *BRCA2* PGV carriers and controls when evaluated via elongation.

## Discussion:

The results from this project are one aspect of a larger study. However, I will only discuss the results based on data presented in this document.

The number of organoids generated from biopsies of the gastric body and antrum from *BRCA1/2 PGV* carriers are clearly elevated over controls. This may imply that cells harbored within the gastric epithelium of *BRCA1/2* carriers have greater proliferation or at least a greater proliferative capacity. In which case, increased proliferation of the gastric epithelium may be an early driver of gastric cancer carcinogenesis among *BRCA1/2 PGV* carriers.

As it pertains to PDGO size, as measured by organoid area, diameter, and perimeter, there did not appear to be a robust growth advantage for *BRCA1/2 PGV* carriers compared to controls. Specifically, PDGOs derived from the gastric body of *BRCA2* PGV carriers exhibited limited growth, compared to *BRCA1* PGV carriers and controls. While body PDGO size was similar between *BRCA1* PGV carriers and controls at day 10 and 15, by day 20 PDGOs from controls were the largest. For PDGOs generated from the gastric antrum, there were no distinct differences amongst *BRCA1/2 PGV* carriers and controls, however, PDGOs from *BRCA1* PGV carriers were slightly larger than the others at all time points. Taken together, this data indicates that there is little to no difference in growth among PDGOs generated from *BRCA1* PGV carriers and controls. This is true for PDGOs generated from both the gastric body and antrum.

While any single well of PDGOs will show a diversity of mophologies, we show, on average, there are little to no differences in PDGO morphology as measured via sphericity, shape factor, and elongation amongst *BRCA1/2 PGV* carriers and controls. This true for PDGOs derived from both the gastric body and antrum of patients.

In conclusion, it appears that increased proliferation of the gastric epithelium may be one possible mechanism of gastric carcinogenesis for *BRCA1/2 PGV* carriers. Furthermore, we are currently evaluating other potential mechanisms (e.g. increased DNA damage) to explain the elevated gastric cancer risk amongst *BRCA1/2 PGV* carriers.

## References:

1.  Prakash, R.; Zhang, Y.; Feng, W.; Jasin, M. Homologous recombination and human health: The roles of BRCA1, BRCA2, and associated proteins. Cold Spring Harb. Perspect. Biol. **2015**, 7, a016600.

2.  Powell, S.N.; Kachnic, L.A. Roles of BRCA1 and BRCA2 in homologous recombination, DNA replication fidelity and the cellular response to ionizing radiation. Oncogene **2003**, 22, 5784--5791.

3.  Walsh, C.S. Two decades beyond BRCA1/2: Homologous recombination, hereditary cancer risk and a target for ovarian cancer therapy. Gynecol. Oncol. **2015**, 137, 343--350.

4.  Chang, H.H.Y.; Pannunzio, N.R.; Adachi, N.; Lieber, M.R. Non-homologous DNA end joining and alternative pathways to double-strand break repair. Nat. Rev. Mol. Cell Biol. **2017**, 18, 495--506.

5.  Snouwaert, J.N.; Gowen, L.C.; Latour, A.M.; Mohn, A.R.; Xiao, A.; DiBiase, L.; Koller, B.H. BRCA1 deficient embryonic stem cells display a decreased homologous recombination frequency and an increased frequency of non-homologous recombination that is corrected by expression of a BRCA1 transgene. Oncogene **1999**, 18, 7900--7907.

6.  Li, S.; Silvestri, V.; Leslie, G.; Rebbeck, T.R.; Neuhausen, S.L.; Hopper, J.L.; Nielsen, H.R.; Lee, A.; Yang, X.; McGuffog, L.; et al. Cancer Risks Associated with BRCA1 and BRCA2 Pathogenic Variants. J. Clin. Oncol. **2022**, 40, 1529--1541.

7.  Momozawa, Y.; Sasai, R.; Usui, Y.; Shiraishi, K.; Iwasaki, Y.; Taniyama, Y.; Parsons, M.T.; Mizukami, K.; Sekine, Y.; Hirata, M.; et al. Expansion of Cancer Risk Profile for BRCA1 and BRCA2 Pathogenic Variants. JAMA Oncol. **2022**, 8, 871--878.

8.  Sun, P.; Li, Y.; Chao, X.; Li, J.; Luo, R.; Mei, L.; He, J. Clinical characteristics and prognostic implications of BRCA-associated tumors in males: A pan-tumor survey. BMC Cancer **2020**, 20, 994.

9.  Ichikawa, H.; Wakai, T.; Nagahashi, M.; Shimada, Y.; Hanyu, T.; Kanu, Y.; Muneoka, Y.; Ishikawa, T.; Takizawa, K.; Tajima, Y.; et al. Pathogenic germline BRCA1/2 mutations and familial predisposition to gastric cancer. JCO Precis. Oncol. **2018**, 2, 18--00097.

10. Moran, A.; O'Hara, C.; Khan, S.; Shack, L.; Woodward, E.; Maher, E.R.; Lalloo, F.; Evans, D.G. Risk of cancer other than breast or ovarian in individuals with BRCA1 and BRCA2 mutations. Fam. Cancer **2012**, 11, 235--242.

11. Jakubowska, A.; Nej, K.; Huzarski, T.; Scott, R.J.; Lubinski, J. BRCA2 gene mutations in families with aggregations of breast and stomach cancers. Br. J. Cancer **2002**, 87, 888--891.

12. Drost, J. & Clevers, H. Organoids in cancer research. *Nature Reviews Cancer.* **18** (7), 407-418, (2018).

13. Corrò, C., Novellasdemunt, L. & Li, V. S. A brief history of organoids. *American Journal of Physiology-Cell Physiology.* **319** (1), C151-C165, (2020).

14. Zhao, Z. *et al.* Organoids. *Nature Reviews Methods Primers.* **2** (1), 94, (2022).

15. Weeber, F. *et al.* Preserved genetic diversity in organoids cultured from biopsies of human colorectal cancer metastases. *Proceedings of the National Academy of Sciences.* **112** (43), 13308-13311, (2015).

16. Boretto, M. *et al.* Patient-derived organoids from endometrial disease capture clinical heterogeneity and are amenable to drug screening. *Nature cell biology.* **21** (8), 1041-1051, (2019).

17. Lo, Y.-H., Karlsson, K. & Kuo, C. J. Applications of organoids for cancer biology and precision medicine. *Nature Cancer.* **1** (8), 761-773, (2020).

18. Grönholm, M. *et al.* Patient-derived organoids for precision cancer immunotherapy. *Cancer research.* **81** (12), 3149-3155, (2021).

19. Yan, H. H. *et al.* A comprehensive human gastric cancer organoid biobank captures tumor subtype heterogeneity and enables therapeutic screening. *Cell stem cell.* **23** (6), 882-897. e811, (2018).

20. Yoon, C. *et al.* Patient-Derived Organoids from Locally Advanced Gastric Adenocarcinomas Can Predict Resistance to Neoadjuvant Chemotherapy. *Journal of Gastrointestinal Surgery.* **27** (4), 666-676, (2023).

21. Miao, X. *et al.* Establishment of gastric cancer organoid and its application in individualized therapy. *Oncology Letters.* **24** (6), 1-8, (2022).

22. Pompaiah, M. & Bartfeld, S. Gastric organoids: an emerging model system to study Helicobacter pylori pathogenesis. *Molecular Pathogenesis and Signal Transduction by Helicobacter pylori.* 149-168, (2017).

23. Schlaermann, P. *et al.* A novel human gastric primary cell culture system for modelling Helicobacter pylori infection in vitro. *Gut.* **65** (2), 202-213, (2016).

24. Bartfeld, S. *et al.* In vitro expansion of human gastric epithelial stem cells and their responses to bacterial infection. *Gastroenterology.* **148** (1), 126-136. e126, (2015).

25. Seidlitz, T., Koo, B.-K. & Stange, D. E. Gastric organoids---an in vitro model system for the study of gastric development and road to personalized medicine. *Cell Death & Differentiation.* **28** (1), 68-83, (2021).
