---
title: "BMIN503: Conceptualizing Actionability in Clinical Genomics"
author: "Kellie Owens"
output: 
  html_document:
    toc: false 
    depth: 3 
    theme: paper 
    highlight: tango
---
```{r set-options, echo=FALSE, cache=FALSE}
options(width = 400)
```  
***
Use this template to complete your project throughout the course. Your Final Project presentation in class will be based on the contents of this document. Replace the title/name and text below with your own, but keep the headers.

### Overview
In this section, give a brief a description of your project and its goal, what data you are using to complete it, and what three faculty/staff in different fields you have spoken to about your project with a brief summary of what you learned from each person. Include a link to your final project GitHub repository.

> The primary questions driving my research are: How and when did "actionability" become the primary framework to assess when genomic data is useful for patients? How can we conceptualize a more sophisticated, ethical, and effective method for determining when genomic tests are useful in a clinical setting? To being answering this question (alongside qualitative and conceptual work), I will perform a natural language processing analysis of all scientific abstracts mentioning "actionability," "genomics," and related terms that are archived in PubMed. This natural language processing analysis can help answer the following pieces of this research: When did "actionability" first appear in genomics literature, and how did its use spread over time? What types of fields and journals use the concept of actionability? What words and concepts appear most frequently alongside "actionability?"


### Introduction 
In the first paragraph, describe the problem addressed, its significance, and some background to motivate the problem.

> Clinical genomic sequencing (CGS) is becoming more common, even in primary care settings (eg, Geisinger 2018, Vassi et al. 2017). By identifying pathogenic and highly penetrant variants in healthy individuals, providers can suggest treatments or lifestyle interventions to manage health risks and prevent poor health outcomes. But CGS is also expensive and not without potential harms (Wilson 2017, Vassy et al. 2017). Clinical genomic sequencing (CGS) produces large amounts of data, much of which is hard to characterize or may have a negligible or negative influence on health. The concept of actionability is commonly used to help separate genomic information that may be useful from information that is likely irrelevant for patients (Berg et al 2013, Amendola et al 2015, Berg et al 2011, Jarvik et al 2014, Webber et al. 2018). Definitions of actionability generally include the following factors: the level of evidence regarding pathogenicity and penetrance of a variant; the efficacy, burden, and availability of interventions; and the severity of potential disease (Berg et al. 2016). Even with widely referenced guidelines on actionability from the American College of Medical Genetics and Genomics (ACMG) (Green et al. 2013, Kalia et al. 2017), there are still ongoing debates about what actionability means (eg. Chae et al. 2017, Burke et al 2013, Goddard et al 2013, Grove et al 2014, McCormick et al. 2014, Moret et al. 2017, Ramos et al. 2014, Institute of Medicine 2014) and significant variability in how laboratories and clinicians interpret the ACMG guidelines (eg. Ackerman and Koenig 2018, O’Daniel et al. 2017, Scheuner et al. 2018, Lázaro-Muñoz et al. 2017, Bland et al. 2017). Actionability directs attention to whether genomic information warrants action and reflects its initial development as a strategy to augment diagnosis and treatment in sick patients. As CGS expands towards healthy populations in primary care settings, actionability is still widely embraced without consensus regarding its definition and use. Despite its centrality in clinical genetics/genomics, there has been little research examining what actionability means and how it is used in published literature. This project will perform a natural language processing analysis to examine how actionability emerged and proliferated in published genomics literature archived in PubMed.

In the second paragraph, explain why your problem is interdisciplinary, what fields can contribute to its understanding, and incorporate background related to what you learned from meeting with faculty/staff.

> This project is interdisciplinary because it draws upon methodological and theoretical tools from a number of fields including: clinical genomics, bioethics, sociology, science and technology studies, and biomedical informatics. Social scientists and humanities scholars (like myself) have many theoretical tools to assess how a concept is taken up by a scientific or medical field and how this concept shapes clinical decision-making and practice. This theory can help guide policy related to the translation of research products into the clinic. Most of this social science research is qualitative and based on small sample sizes. Biomedical informatics methods, such as natural language processing, can help analyze large amounts of textual data related to a given concept. This project will bridge the strengths of natural language processing with normative and conceptual theory from the social sciences.


### Methods

>The data for this project are scientific abstracts archived in PubMed and Web of Science that include the words "actionable" or "actionability." I conducted searches on the PubMed and Web of Science websites and saved results as text files (or, when using the RISmed package, searched PubMed abstracts directly in RStudio). I used three R packages to read and analyze this data: pubmed.mineR, bibiometrix, and RISmed. These tools allowed me to identify words associated with actionability (in a variety of ways), graph patterns in the dataset, and create wordclouds. [Still adding more details to methods section!]

In the first paragraph, describe the data used and general methodological approach. Subsequently, incorporate full R code necessary to retrieve and clean data, and perform analysis. Be sure to include a description of code so that others (including your future self) can understand what you are doing and why. 

```{r eval = TRUE}
library(pubmed.mineR)
getwd()
# Read PubMed Abstracts, Print info and first/last abstract
pubmed.actionab = readabs("/Users/knowens/Desktop/BMIN503_Final_Project/Data/PubMed.Actionab.txt")
printabs(pubmed.actionab)
```

```{r eval = TRUE}
# Extract top frequency words
actionab.words = word_atomizations(pubmed.actionab)
actionab.words[1:50,]
```

```{r eval = TRUE}
# Find the sentence where "actionable" occurs
sentence = Give_Sentences("actionable", abs = pubmed.actionab)
sentence[1:5]
```

```{r eval = TRUE}
# Find words before and after "actionable"
associations = word_associations("actionable", pubmed.actionab)
associations[1:10]
```

```{r eval = TRUE}
# Word clusters
clusters = wordscluster(5, 10)
wordsclusterview(clusters)
```

```{r eval = TRUE}
#Themes
themes <- searchabsL(pubmed.actionab, include = c("Genomic", "Genetic", "Gene", "Cancer"))
```

```{r eval = TRUE}
#Journals
journals <- pubmed.actionab@Journal
journals[1:10]
```


```{r eval = TRUE}
#Web of Science Analysis
library(bibliometrix)

WebOfScience <- readFiles("file1.txt", "file2.txt", "file3.txt", "file4.txt", "file5.txt", "file6.txt", "file7.txt", "file8.txt", "file9.txt", "file10.txt")
WS <- convert2df(WebOfScience, dbsource = "isi", format = "plaintext")
WSresults <- biblioAnalysis(WS, sep = ";")

#general summaries of Web of Science
options(width=100)
summary <- summary(object = WSresults, k=10, pause = FALSE)
plot(x = WSresults, k=10, pause = FALSE)

#Keyword co-occurences
NetMatrix <- biblioNetwork(WS, analysis = "co-occurrences", network = "keywords", sep = ";")
net=networkPlot(NetMatrix, normalize="association", weighted=T, n=30, Title = "Keyword Co-occurences", type = "fruchterman", size=T,edgesize = 5,labelsize=0.7)

#Source co-occurences
NetMatrixSource <- biblioNetwork(WS, analysis = "co-occurrences", network = "sources", sep = ";")
net=networkPlot(NetMatrixSource, normalize="association", weighted=T, n=30, Title = "Source Co-occurences", type = "fruchterman", size=T,edgesize = 5,labelsize=0.7)

#Co-word analysis
CS <- conceptualStructure(WS, field="ID", method="CA", minDegree=4, clust=5, stemming=FALSE, labelsize=10, documents=10)

#RISmed Analysis
library(RISmed)
res1 <- EUtilsSummary("actionab*", 
                      type = "esearch", 
                      db = "pubmed",
                      datetype = "pdat",
                      retmax = 12000,
                      mindate = 1960, 
                      maxdate = 2019)
fetch <- EUtilsGet(res1, type = "efetch", db = "pubmed")
RISmed_abstracts <- data.frame(title = fetch@ArticleTitle,
                        abstract = fetch@AbstractText, 
                        journal = fetch@Title,
                        DOI = fetch@PMID, 
                        year = fetch@YearPubmed)
## ensure abstracts are character fields (not factors)
RISmed_abstracts <- RISmed_abstracts %>% mutate(abstract = as.character(abstract))
RISmed_abstracts %>%
  head()

#Abstracts by Year
library(ggplot2)
RISmed_abstracts %>%
group_by(year) %>%
count() %>%
ggplot(aes(year, n)) +
geom_point() +
geom_line() +
labs(title = "Pubmed Articles by Year", hjust = 0.5,
y = "Articles")

#Journal Word Cloud
journal_wordcloud <- RISmed_abstracts %>%
  select(journal) %>%
  group_by(journal) %>%
  count(sort = TRUE)
journal_wordcloud %>%
  with(wordcloud(journal, n, min.freq = 10, random.order = FALSE, max.words = 80, colors = brewer.pal(9, "Set1")), rot.per = .6)
```

### Results
Describe your results and include relevant tables, plots, and code/comments used to obtain them. End with a brief conclusion of your findings related to the question you set out to address. You can include references if you'd like, but this is not required.

```{r eval = TRUE}
#Subset of abstracts related to genetics/genomics
gene_abstracts = searchabsL(pubmed.actionab, include = c("Genomic", "Genetic", "Gene", "Genomics", "Genetics"))
gene_abstracts = cleanabs(gene_abstracts)

#Top frequency words for genetics/genomics dataset
actionab.gene.words = word_atomizations(gene_abstracts)
actionab.gene.words[1:50,]

#Abstracts by Year
library(ggplot2)
RISmed_abstracts %>%
group_by(year) %>%
count() %>%
ggplot(aes(year, n)) +
geom_point() +
geom_line() +
labs(title = "Pubmed Articles by Year", hjust = 0.5,
y = "Articles")
```

```{r eval = TRUE}
#Word Cloud for whole sample
install.packages("tm")
install.packages("wordcloud")
library(tm)
library(wordcloud)
library(RColorBrewer)

wordcloud_text <- WS$AB #convert list entries to text
wordcloud_corpus <- Corpus(VectorSource(wordcloud_text))
wordcloud_corpus <- tm_map(wordcloud_corpus,
                              content_transformer(function(x) iconv(x, to = 'UTF-8-MAC', sub =' byte'))
)

wordcloud_corpus <- tm_map(wordcloud_corpus, content_transformer(tolower))
wordcloud_corpus <- tm_map(wordcloud_corpus, removePunctuation)
wordcloud_corpus <- tm_map(wordcloud_corpus, removeNumbers)
myStopwords <- c(stopwords(), "actionable", "actionability")
wordcloud_corpus <- tm_map(wordcloud_corpus, function(x) removeWords(x, myStopwords))

wordcloud(wordcloud_corpus, min.freq = 10, max.words = 100, scale = c(3,.1), random.order = FALSE, colors = brewer.pal(8, "Dark2"))
```-
