---
title: " Review on a brief analysis of Alveolar Macrophage gene expression on cannabis exposure"
author: "Nilanjana Samanta"
output: 
  html_document:
    toc: true
    toc_float: 
        collapsed: true
        smooth_scroll: true
    depth: 3
    theme: paper 
    highlight: tango
---
```{r set-options, echo=FALSE, cache=FALSE}
options(width = 400)
```  
***

### Overview
Literature on Gene expression studies in cigarette smokers are abundant and most have identified several dysregulated molecular
pathways that can alter respiratory health of a smoker or increase their risk of developing chronic conditions. However, there is a lack of literature on similar and/or parallel studies in marijuana smokers, partly because there are high chances of concurrent cigarette smoking whcih acts as a confounder in determining effect. S. Sharma and C. Liu et al hypothesized that studying difference in expression of alveolar macrophage (AM) 
between Marijuana and cigarette smokers can identify disease pathways that may influence respiratory disease development among those exposed to cannabis smoke.


### Introduction 
Recently as many US states legalized recreational marijuana use, while policy makers are trying to restrict access of cannabis vaping among teenagers because of evidence of respiratory implications of cannabis vaping. Hence it has become increasingly important to study effect of cannabis smoke exposure and its association with potential respiratory disease related gene expression. 

[This study](https://www.atsjournals.org/doi/pdf/10.1164/ajrccm-conference.2018.197.1_MeetingAbstracts.A4161) by S. Sharma and C. Liu et al, could potentially identify genes and pathways which would give additional insight into mechanisms that influence cannabis-associated respiratory diseases and if it differs greatly from those exposed to cigarette smoke or no smoke.

This study is at the intersection of public health, Bioinformatics and Biomedical genetics and should draw insight from these fields for a deeper understanding of the effects of cannabis smoking on a molecular level in the general population.


### Methods
I will be utilizing the raw data for this study available in the GEO database, accession [GSE155213](https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE155213) 

```{r,eval=FALSE, message=FALSE}
library(GEOquery)
install.packages('R.utils')
library(R.utils)
library(data.table)
library(oligo)
library(limma)
library(viridis)
BiocManager::install('biomaRt')
library(biomaRt)
BiocManager::install('edgeR')
library(edgeR)
BiocManager::install('DGElist')
library(DGElist)
library(tidyr)
library(dplyr)
```

## Set working directory
```{r}
getwd()
setwd("")
```

## Loading and Reading the dataset from the GEO database into R
> The dataset was downloaded into local directory and read into Rstudio for analysis

```{r,eval=TRUE, message=FALSE}

feature.table <- read.table( gzfile("C:\\Users\\nsamanta\\Desktop\\HPAP Files\\GitHub\\BMIN503_Final_Project\\GSE155213\\GSE155213_count_table.txt.gz"), header=TRUE, sep="\t")

#creating a safe copy of feature.table
copy3 <- feature.table

head(feature.table)
tail(feature.table)
colnames(feature.table)

#Acquiring phenotypic gset
pheno.table1 <- read.table( gzfile("C:\\Users\\nsamanta\\Desktop\\HPAP Files\\GitHub\\BMIN503_Final_Project\\GSE155213\\GSE155213_raw_pheno.txt"), header=TRUE, sep="\t")

pheno.table2 <- read.table( gzfile("C:\\Users\\nsamanta\\Desktop\\HPAP Files\\GitHub\\BMIN503_Final_Project\\GSE155213\\GSE155213_raw_pheno2.txt"), header=TRUE, sep="\t")

#Merging the two phenotype datasets by rows
pheno.data <- rbind(pheno.table1, pheno.table2)

#cleaning and subsetting pheno data
pheno.data.final <- pheno.data %>%
  select(title, characteristics_ch1, characteristics_ch1.1, characteristics_ch1.2, characteristics_ch1.5) %>%
  rename(status = title, group = characteristics_ch1, age = characteristics_ch1.1, sex = characteristics_ch1.2, tech.batch = characteristics_ch1.5) %>%
  mutate(group = gsub("group: ", "", group), age = gsub("age: ", "", age), sex = gsub("Sex: ", "", sex), tech.batch = gsub("technical_batch: ", "batch", tech.batch))

#changing group text to reflect Marijuana, Control and Tobacco groups
pheno.data.final$group <- recode(pheno.data.final$group,
                                 C = "control",
                                 `T` = "tobacco",
                                 M = "marijuana") 


#creating a count only table
count.table <- feature.table
rownames(count.table) <- feature.table$Features
count.table$Features <- NULL

#log transformation
cpm_log <- cpm(count.table, log = TRUE)

boxplot(count.table, las = 2, main = "Before Log2 transformation")
boxplot(cpm_log, las = 2, main = "After log2 transformation")

#adding pseudo numbr to avoid negative values
count_log2 <- log2(count.table + 1)

#checking if count table has any missing values
row.has.na <- apply(count.table, 1, function(x){any(is.na(x))})
sum(row.has.na)

#Creating DGElist with "Groups"
mydata <- DGEList(count_log2, group=factor(pheno.data.final$group))
mydata
dim(mydata)
head(mydata$counts)
head(mydata$samples)

#creating a copy of mydata
mydata.copy <- mydata

#calculatin counts per million uing cpm() funcion for each gene
head(cpm(mydata.copy))

#calculating total gene count per sample
apply(mydata.copy$counts, 2, sum)

#Keeping genes atleast a count of 0.5 cpm for atleat 2samples and removing the rest from our dataset
keep.genes <- rowSums(cpm(mydata.copy)>0.5) >= 2 #keep genes is a logical vector with TRUE and FALSE values
table(keep.genes)
mydata.copy <- mydata.copy[keep.genes,]
dim(mydata.copy)

#Plotting sample relationship on multidimendionl scale based on exposure 
plotMDS(mydata.copy, method="bcv", col=as.numeric(mydata.copy$samples$group))
legend("bottomleft", as.character(unique(mydata.copy$samples$group)), col=1:3, pch=20)


#Plotting sample relationship on multidimendionl scale based on batch effect
mydata.batch <- DGEList(count_log2, group=factor(pheno.data.final$tech.batch))
mydata.batch

dim(mydata.batch)

keep.batch <- rowSums(cpm(mydata.batch)>0.5) >= 4
table(keep.batch)
mydata.batch <- mydata.batch[keep.batch,]
dim(mydata.batch)

plotMDS(mydata.batch, method="bcv", col=as.numeric(mydata.batch$samples$group))
legend("bottomleft", as.character(unique(mydata.batch$samples$group)), col=1:3, pch=20)

```


```{r}
#testing DE without log transformation

mydata.x <- DGEList(count.table, group=factor(pheno.data.final$group))
mydata.x
dim(mydata)
head(mydata$counts)
head(mydata$samples)


#calculatin counts per million uing cpm() funcion for each gene
head(cpm(mydata.x))

#calculating total gene count per sample
apply(mydata.x$counts, 2, sum)

#Keeping genes atleast a count of 0.5 cpm for atleat 2samples and removing the rest from our dataset
keep.genes_x <- rowSums(cpm(mydata.x)>0.5) >= 2 #keep genes is a logical vector with TRUE and FALSE values
table(keep.genes_x)
mydata.x <- mydata.x[keep.genes_x,]
dim(mydata.x)

#design matrix to describe the statistical model
design.x <- model.matrix(~0 + group + batch + age)
colnames(design) <- gsub("group", "", colnames(design))
design
design.copy <- design

contrast.matrix <- makeContrasts(mari_vs_cont = marijuana - control,
                              tobacco_vs_cont = tobacco - control,
                              mari_vs_tobacco = marijuana - tobacco,
                              levels = colnames(design))
contrast.matrix

#gene-specific (referred to in edgeR as tagwise) dispersion estimates are used in the test for differential expression
mydata.x <- estimateDisp(mydata.x, design.x)
mydata.copy

#fitting a GLM
fitx <- glmFit(mydata.x, design.x)

lrt_mcx <- glmLRT(fitx, contrast=makeContrasts(mari_vs_cont = marijuana - control, levels = colnames(design)))
lrt_tcx <- glmLRT(fitx, contrast=makeContrasts(tobacco_vs_cont = tobacco - control, levels = colnames(design)))
lrt_mtx <- glmLRT(fitx, contrast=makeContrasts(mari_vs_tobacco = marijuana - tobacco, levels = colnames(design)))

topTags(lrt_mcx, n=Inf, adjust.method = "BH", sort.by = "PValue", p.value = 0.05)
topTags(lrt_tc, n=10, adjust.method = "BH", sort.by = "PValue", p.value = 1)
topTags(lrt_mt, n=10, adjust.method = "BH", sort.by = "PValue", p.value = 1)

mc_DEx <- decideTestsDGE(lrt_mcx, adjust.method="BH", p.value = 0.05)
summary(mc_DE)

tc_DE <- decideTestsDGE(lrt_tc, adjust.method="BH", p.value = 1)
summary(tc_DE)

mt_DE <- decideTestsDGE(lrt_mt, adjust.method="BH", p.value = 1)
summary(mt_DE)

```

## DE Analysis using Generalized Linear Model (GLM)

```{r, eval=TRUE, message=FALSE}
#creating group vectors for pasing into design matrix 
group <- c(rep("marijuana", 15), rep("tobacco", 16), rep("control", 10))
batch <- c(rep("batch1", 30), rep("batch2", 11))
age <- as.numeric(c("22", "23", "24", rep("25",2), "29", rep("30",5), rep("31",2), "32", "33", rep("34",2), rep("35",2), rep("36",2), rep("37",2), rep("38",3), rep("39",3), rep("40",2), rep("41",3), "47", rep("50",3), "53", "54", "55"))

#design matrix to describe the statistical model
design <- model.matrix(~0 + group + batch + age)
colnames(design) <- gsub("group", "", colnames(design))
design
design.copy <- design

contrast.matrix <- makeContrasts(mari_vs_cont = marijuana - control,
                              tobacco_vs_cont = tobacco - control,
                              mari_vs_tobacco = marijuana - tobacco,
                              levels = colnames(design))
contrast.matrix

#gene-specific (referred to in edgeR as tagwise) dispersion estimates are used in the test for differential expression
mydata.copy <- estimateDisp(mydata.copy, design.copy)
mydata.copy

#fitting a GLM
fit <- glmFit(mydata.copy, design.copy)

lrt_mc <- glmLRT(fit, contrast=makeContrasts(mari_vs_cont = marijuana - control, levels = colnames(design)))
lrt_tc <- glmLRT(fit, contrast=makeContrasts(tobacco_vs_cont = tobacco - control, levels = colnames(design)))
lrt_mt <- glmLRT(fit, contrast=makeContrasts(mari_vs_tobacco = marijuana - tobacco, levels = colnames(design)))


topTags(lrt_mc, n=10, adjust.method = "BH", sort.by = "PValue", p.value = 1)
topTags(lrt_tc, n=Inf, adjust.method = "BH", sort.by = "PValue", p.value = 1)
topTags(lrt_mt, n=Inf, adjust.method = "BH", sort.by = "PValue", p.value = 1)

mc_DE <- decideTestsDGE(lrt_mc, adjust.method="BH", p.value = )
summary(mc_DE)

tc_DE <- decideTestsDGE(lrt_tc, adjust.method="BH", p.value = 1)
summary(tc_DE)

mt_DE <- decideTestsDGE(lrt_mt, adjust.method="BH", p.value = 1)
summary(mt_DE)

#boxplot of one top gene
boxplot(as.numeric(count.table["ENSG00000284369", ]) ~ group)

de2tagsmc <- rownames(mydata)[as.logical(mc_DE)]
plotSmear(lrt_mc, de.tags=de2tagsmc)
abline(h = c(-2, 2), col = "blue")

#heat map
library("gplots")
library(viridis)
top_results <- smoking_results[1:200, ]
top.eset <- count.table[row.names(exprs(count.table)) %in% row.names(top_mc)]
status.colors <- unlist(lapply($group, 
                               function (x) {if (x == "marijuana") "red"
                                 else if (x == "tobacco") "blue"
                                 else "navy"}))
heatmap.2(exprs(top.eset), col = viridis(256, option = "B"),
          trace = "none", keysize = 1.5, key.title = NA,
          ColSideColors = status.colors)
legend("topright", legend = c("marijuana", "tobacco", "control"), fill = c("red", "blue", "navy")) 

#testing
et <- exactTest(mydata.copy)
results_edgeR <- topTags(et, n = nrow(mydata.copy), sort.by = "none")
head(results_edgeR$table)

sum(results_edgeR$table$FDR < 1)
plotSmear(et, de.tags = rownames(results_edgeR)[results_edgeR$table$FDR < 1])
abline(h = c(-2, 2), col = "blue")

```


## Assigning Gene Annotation

```{r,eval=TRUE, message=FALSE}

#Acquiring gene symbols for Ensmble Ids
mart <- useMart(biomart="ENSEMBL_MART_ENSEMBL", host="https://grch37.ensembl.org", path="/biomart/martservice", dataset="hsapiens_gene_ensembl")

genes <- biomaRt::getBM(attribute=c('ensembl_gene_id', 'hgnc_symbol'), values=main.table$Features, mart=mart)

#removing any duplicate gene ids
genes.matrix <- genes[-which(duplicated(genes$ensembl_gene_id)),]

#converting the 2 columns to factors before merge
genes.matrix$ensembl_gene_id <- as.factor(genes.matrix$ensembl_gene_id)
genes.matrix$hgnc_symbol <- as.factor(genes.matrix$hgnc_symbol)

summary(genes.matrix)
length(unique(genes.matrix$hgnc_symbol))


#Replacing blank cells with NAs
genes.matrix[genes.matrix == ""] <- NA


# select genes of interest only
rownames(genes.matrix)<-genes.matrix$ensembl_gene_id
features.annot <- genes.matrix[as.character(copy3$Features),]

# join the two tables
rownames(copy3) <- copy3$Features
merged.feature.table <- cbind(features.annot,copy3)

head(merged.feature.table)

#Removing features column
merged.feature <- subset(merged.feature.table, select = -c(Features))

#checking NA rows for symbol in merged.feature table before deleting them
sum(is.na(merged.feature$hgnc_symbol))

#deleting gene-ids with no corresponding symbols
merged.feature %>%
filter(merged.feature, rowSums(is.na(merged.feature)) != ncol(merged.feature))

```



```{r, eval=TRUE, message=FALSE}

library(GEOquery)

family.table <- read.table(gzfile("C:\\Users\\nsamanta\\Desktop\\HPAP Files\\GitHub\\BMIN503_Final_Project\\GSE155213\\GSE155213_family.soft.gz"), header=TRUE, sep="\t")

gse <- getGEO("GSE155213", destdir=".", GSEMatrix = TRUE)[[1]]

gse <-getGEO(filename='GSE155213/GSE155213-GPL16791_series_matrix.txt.gz',GSEMatrix=TRUE)
raw.pheno <- pData(phenoData(gse))

gse <- getGEO("GSE155213", AnnotGPL = TRUE)

BiocManager::install('hgu95av2.db')
BiocManager::install('affycoretools')
library(hgu95av2.db)

eset <- annotateEset(gse, hgu95av2.db)
```

### Results- to be finalized
Describe your results and include relevant tables, plots, and code/comments used to obtain them. End with a brief conclusion of your findings related to the question you set out to address. You can include references if you'd like, but this is not required.

```{r, eval=TRUE}
#ROUGH WORK
#referenced articles
1. Forum on CPM cutoff: https://support.bioconductor.org/p/101838/
2. Tutorial on DE analysis by edgeR:  https://web.stanford.edu/class/bios221/labs/rnaseq/lab_4_rnaseq.html  
3. https://bioconductor.org/packages/devel/bioc/manuals/edgeR/man/edgeR.pdf

#copy of main table
copy.feature.table <- feature.table

#Reading in unzipped count-table to ensure counts are correct
count.table <- read.table(file.choose(), sep ="\t", header = TRUE, stringsAsFactor = FALSE)

#Reading in Series-matrix supplementary files for phenotypic info
series.matrix_1 <- read.table(gzfile("C:\\Users\\nsamanta\\Desktop\\HPAP Files\\GitHub\\BMIN503_Final_Project\\GSE155213\\GSE155213-GPL16791_series_matrix.txt.gz"), header=TRUE, fill=TRUE, sep="\t")

series.matrix_2 <- read.table(gzfile("C:\\Users\\nsamanta\\Desktop\\HPAP Files\\GitHub\\BMIN503_Final_Project\\GSE155213\\GSE155213-GPL20301_series_matrix.txt.gz"), header=TRUE, fill=TRUE, sep="\t")


# trying to remove genes with cpm <1, not working
copy.feature.table$Features <- as.numeric(copy.feature.table$Features)

gene.expr1 <- rowSums(cpm(merged.feature.table)>1) >= 41

table(gene.expr1)

sapply(feature.table, class)

mycounts <- main.table[gene.expr1,]
mygenes <- rownames(mycounts)

summary(copy3)
class(copy3$Features)

head(main.table)
```