---
title: "BMIN503/EPID600 Project: A Comparison of Nasal Microbiota in Patients with Granulomatosis with Polyangiitis Compared to Healthy Controls"
author: "Rennie Rhee"
output: 
  html_document:
    toc: false 
    depth: 3 
    theme: paper 
    highlight: tango
---
```{r set-options, echo=FALSE, cache=FALSE}
options(width = 400)
```  
***
Use this template to complete your project throughout the course. Your Final Project presentation in class will be based on the contents of this document. Replace the title/name and text below with your own, but leave the headers.
#FYI, I left headers in for now but will remove when I submit final project.

### Overview
In this section, give a brief a description of your project and its goal, what data you are using to complete it, and what three faculty/staff in different fields you have spoken to about your project with a brief summary of what you learned from each person. Include a link to your final project GitHub repository.

The goal of this project is to examine the nasal microbiota in patients with granulomatosis with polyangiitis (GPA) compared to healthy controls. With the help of collaborators at Penn, we collected nasal swabs on 60 patients with GPA and 41 healthy controls along with clinical metadata. 

I met with 3 faculty/staff to discuss my project: (1) Elizabeth Grice, PhD, is a microbiologist in the Dept of Dermatology who focuses on the skin microbiome. She helped me develop an analytic plan to address my study question and gave me some background on performing microbiome research; (2) Peter Merkel, MD, MPH, is Chief of Rheumatology at Penn and expert in vasculitis (including GPA) and patient-oriented research. He guided me on study design, tissue collection, and important factors to consider when studying patients with GPA; and (3) Chunyu Zhao, PhD, is a bioinformatician with expertise in using R to analyze metagenomic data. Dr. Zhao advised me on software programs available to analyze my data.

### Introduction 
In the first paragraph, describe the problem addressed, its significance, and some background to motivate the problem.

In the second paragraph, explain why your problem is interdisciplinary, what fields can contribute to its understanding, and incorporate background related to what you learned from meeting with faculty/staff.

The objective of this project is to compare the nasal microbial composition between patients with granulomatosis with polyangiitis (GPA; Wegener's) and healthy controls. Granulomatosis with polyangiitis (GPA; Wegener's) is a systemic granulomatous vasculitis which often affects the upper respiratory tract and frequently relapses. Prior studies demonstrated that microbes in the nose are associated with relapse in GPA but these studies relied on culture-dependent approaches. Advances in high-throughput sequencing now allow for a better understanding of the complex relationship between the dynamic community of microbes that occupy our mucosal tissues and the host's immune system. We hypothesize that patients with GPA have greater dysbiosis (disruption in indigenous microbiota) in the nares compared to controls.

This interdisciplinary project will require expertise on GPA pathogenesis and management, patient-oriented research, microbiology, and analysis of metagenomic data. I met with Dr. Peter Merkel, my primary mentor and Chief of Rheumatology Division, to discuss patient selection and addressing confounders. I met with Dr. Elizabeth Grice, a microbiologist and dermatologist, to formulate an analytic plan. I also met with Dr. Chunyu Zhao, a bioinformatician, to review R packages used for microbiome research. Together these faculty/staff will guide me as I complete this project.

### Methods
In the first paragraph, describe the data used and general methodological approach. Subsequently, incorporate full R code necessary to retrieve and clean data, and perform analysis. Be sure to include a description of code so that others (including your future self) can understand what you are doing and why. 

We collected nasal swabs and clinical data on 60 patients with GPA and 41 healthy controls who were seen at the Penn Vasculitis Center. Processing and DNA sequencing was performed in the CHOP Microbiome Core. DNA was extracted using the QIAGEN QIAamp Ultraclean kit and bacterial 16S ribosomal RNA was sequenced with Illumina HiSeq and primers annealing to the V1V2 region were used for amplification. Sequence data was processed with Quantitative Insights Into Microbial Ecology (QIIME). Operational Taxonomic Units (OTUs) – which are surrogates for bacterial taxa defined by 16S rRNA homology – were selected by clustering reads at 97% sequence similarity. Taxonomic assignments were generated by comparison to the Greengenes reference database, using the consensus method implemented in QIIME. A phylogenetic tree was inferred from the OTU data using FastTree.

First, descriptive statistics were performed with clinical metadata. Means and proportions were used to summarize data. Continuous variables were compared using t-test and categorical variables compared with chi-square test. 

Data files from QIIME were analyzed in the R environment for statistical computing, using the QIIMER library. From the sequencing above, I obtained tables of OTUs to analyze outcomes of interest. Beta diversity, the degree of difference between bacterial communities measured as a continuous distance, was compared between groups using unweighted and weighted UniFrac distance (distance measurement which incorporates phylogenetic relationships, weighted by abundance). Global differences in bacterial community composition were visualized using Principal Coordinates Analysis. Alpha diversity, measured by the Shannon diversity index, will also be examined. Alpha diversity measures the number (richness) and distribution (evenness) of taxa expected within a single population. The Shannon index, a continuous measure, will be calculated from OTUs. The relative abundance of taxa in each group (percent composition of particular taxa relative to the total number of taxa) was determined for the 10 most abundant taxa present in the samples. Differential abundance of specific taxa between groups was assessed using nonparametric rank tests. 

<R code to be placed here>
```{r load scripts, echo=FALSE, message=FALSE, warning=FALSE}
library(qiimer)
library(ape)
library(vegan)
library(ggplot2)
library(dplyr)
library(reshape2)
library(kylemisc)
library(tidyr)
library(broom)
library(pheatmap)

library(lme4)
library(lmerTest)

library(subfunc)
library(forcats)
library(lubridate)
library(magrittr)
library(pander)
library(ggbeeswarm)

library(scales)
library(colorRamps)
library(RColorBrewer)
library(ggthemes)
```

```{r eval=TRUE}
###Read in data and then clean data
setwd("E:/Backup/BMI Certificate/Data Science in Biomedical Informatics/Final Project")
meta_file_fp <- "clinical_metadata.txt"
c <- read.table(meta_file_fp, sep = "\t", header = TRUE, stringsAsFactors = FALSE)

#Remove data from negative controls (environmental controls)
c <- subset(c, HostSpecies=="Human")

#Change "not answered" string to "NA"
c[c=="not answered"] <- NA

#Examine data frame and structure
str(c)

#Change age from character to numeric
c$age <- as.numeric(as.character(c$age))
class(c$age)

#Change character variables to factors
c %>% mutate_if(is.character, as.factor)
str(c)
```

```{r eval=TRUE}
###Descriptive Statistics of Clinical Data

#Obtain means and counts/proportions for patient characteristics by study group
c %>% group_by(study_group) %>% summarise_if(is.numeric, mean, na.rm=TRUE)

c %>% group_by(study_group) %>% count(sex) %>% mutate(percent = n/sum(n))
c %>% group_by(study_group) %>% count(race) %>% mutate(percent = n/sum(n))

#to be continued

```

### Results
Describe your results and include relevant tables, plots, and code/comments used to obtain them. End with a brief conclusion of your findings related to the question you set out to address. You can include references if you'd like, but this is not required.

Outline of results:
Table 1: patient characteristics by study group (GPA vs controls)
Figure 1A/B: PCoA plot of unweighted and weighted uniFrac
Figure 2: Shannon diversity index boxplot
Figure 3+: Results of differential abundance