---
title: "Efficient Translation of Autism Genetics Research for Clinical Decision Support""
author: "Olivia J. Veatch"
output: 
  html_document:
    toc: false 
    depth: 3 
    theme: paper 
    highlight: tango
---
```{r set-options, echo=FALSE, cache=FALSE}
options(width = 400)
```  
***

### Overview
This project will develop an automated method for prioritizing clinically actionable results from genetic studies of autism spectrum disorders (ASD). The goal is to rapidly deliver important information in such a manner that the clinician can be more informed when making key decisions regarding the benefits of ordering genetic testing for a patient.


### Introduction 
Autism Spectrum Disorder (ASD) is a neurodevelopmental disorder with well-established influences from genomic factors. Rapid advancements in genomics technologies have helped generate massive amounts of new knowledge on the genetic factors contributing to ASD. Although discovery continues to be important, it is crucial to understand how currently implicated genes can help immediately inform diagnosis and treatment. There are currently hundreds of genes and thousands of genomic variants implicated in risk for ASD and the list is continually growing. In order to understand how genetic information can be useful to informing treatment, it is important to identify efficient ways to sort through the sea of association study results to determine clinically-actionable genes.

Integration of genetics and genomics into precision medicine has been successful in cancer prediction and therapy; however, the need to incorporate these data into non-oncological practice of medicine has been noted as an area that requires particular attention. Following discussions with Allan Pack, John Miclot Professor of Medicine and Assistant Dean and Director of Global Sleep Programs, I was able to identify the specific clinical decision I want to support with the development of the approach (i.e., help clinicians to determine if genetic testing will be useful to informing their decisions when planning treatment). As incorporating genomic technology into patient care is largely dependent on cliniciansâ€™ perspectives of its utility, it is important to identify ways to better inform clinicians about relevant genetic findings beneficial to optimizing treatment. Although the cost of generating new data is steadily decreasing, there remain barriers in interpreting and translating these data. These include limited access to necessary resources and knowledge gaps among clinicians regarding the benefits/limitations of genetic testing for many complex disorders. While there are numerous challenges limiting the clinical utility of information from genetic studies in ASD, after speaking with Ben Voight, Associate Professor of Genetics, I was able to refine the specific challenges I am interested in addressing. The proposed project addresses two key issues: 1) the 'immediacy' gap which will be bridged by developing an efficient way to sort through genetic data to identify information about known and potentially clinically actionable genes, and 2) the 'processing' gap which will be bridged by working to validate and automate the prioritization method so that the validated approach can be adapted for future applications in other complex disorders and diseases. There are excellent resources available that allow for identifying clinically useful ways to interpret information from genetic studies. Dr. Jason Moore, Edward Rose Professor of Informatics, Director of the Penn Institute for Biomedical Informatics, Senior Associate Dean for Informatics and Director of the Division of Informatics in the Department of Biostatistics and Epidemiology, helped me to identify the specific databases that would be useful to the automation of the approach and offered advice on how to validate the approach. I will use data form DisGeNET, Genotype-Tissue Expression database, International Mouse Phenotyping Consortium, and Pharos Drug Ontology to automate a gene prioritization pipeline useful to identification of clinically-actionable candidate genes for ASD. Clinically-actionable genetic information will be defined as any functionally relevant (e.g., expressed in brain, causes an ASD-related phenotype when knocked-out in mice), ASD-associated gene with evidence indicating that variation in the gene may influence how a patient will respond to a drug. https://github.com/veatcho/BMIN503_Final_Project.git

### Methods
To identify an initial list of candidate genes for ASD, data available in DisGeNET (http://www.disgenet.org) was queried. The benefits of this database include that the reported gene-disease associations' are identified by text-mining data from multiple sources and the data can be automatically pulled via the RCurl package in R. All of the Unified Medical Language System (UMLS)? Concept Unique Identifiers (CUI) that relate to ASD (e.g., Autism Spectrum Disorder, Atypical autism, Autistic Behavior, and utistic Disorder) were identified from the disease mappings file provided by the database. An R function provided by DisGeNET was used to create a data frame of all genes with any level of evidence for a relationship with CUIs relating to ASD. These genes were then prioritized based on whether or not they are expressed in the brain, as this is the tissue most relevant to ASD etiology. A summary of baseline expression results from all available brain regions, expressed at the default level of FPKM>=0.5, included in the Genotype Tissue Expression (GTEx) project were downloaded from the Expression Atlas (https://www.ebi.ac.uk/gxa/home) and the resulting GTEx data frame was filtered for ASD risk genes. These genes were then be prioritized based on an associated (p<0.05) mouse phenotype term that reflects symptoms in ASD when manipulated in mice, by querying the genotype-phenotype representational state transfer (REST) application program interface (API) from the International Mouse Phenotyping Consortium (IMPC; http://www.mousephenotype.org/). To identify currently approved drugs that target or have drug ontology to proteins encoded by the prioritized genes implicated in ASD, a software request was sent via RCurl to the Pharos API (https://pharos.nih.gov/idg/dto). Only targets labeled Tclin (i.e., currently approved) were considered. 

+The first step was to pull all genes with evidence for influencing risk for Autism Spectrum Disorders from DisGeNET.
``` {r eval=TRUE}
library(RCurl)
library(tidyverse)
#download UMLS CUI to several disease categories file from http://www.disgenet.org/web/DisGeNET/menu/downloads#r and read into R environment
diseasecuis<-read.table(url("https://github.com/veatcho/BMIN503_Final_Project/blob/master/disease_mappings.tsv?raw=true"), header = TRUE, sep = "\t", fill = TRUE, stringsAsFactors = FALSE, quote="")

#Pull Autism-related CUIs
ASDcuis<-diseasecuis %>%
  select(diseaseId, name) %>%
  filter(grepl("^Autis|MACROCEPHALY/AUTIS|^AUTIS", name)) %>%
  distinct(diseaseId)

#Pulled original function from http://www.disgenet.org/ds/DisGeNET/scripts/disgenet.R and modified to suit my pipeline_OJV

# Copyright [2010-2017] Integrative Biomedical Informatics Group, Research Programme on Biomedical Informatics (GRIB) IMIM-UPF 
# http://ibi.imim.es/
# contact for technical questions support@disgenet.org
# creator: janet.pinero@upf.edu  
# Script to query disgenet using a list of genes or diseases
# requires as input the gene or disease list in a file 
# the output file name
# the type of entity (gene or disease)
# the type of identifier 
###############################################################################


# main
###############################################################################
# load packages  	
#require(RCurl)

###############################################################################
# subs
###############################################################################

doQuery_modifiedOJV = function(inputFile, outFile, entity, identifier){
  #print(inputFile)
  #print(outFile)
  #print(entity)
  #print(identifier)
  
  
  # read in all data
  #inFile = read.csv(file=paste(getwd(), inputFile, sep="/"), sep="\t", header=F)
  inFile = inputFile ###OJV altered to address the object from the previous command (ASDcuis), instead of loading a file that was saved on disk
  dataFin <- data.frame(matrix(nrow=0, ncol=14)) 
  
  STR = "";
  if (entity == "gene"){
    if (identifier == "entrez"){
      STR = "c2.geneId = '"
    }
    else  if (identifier == "hgnc"){
      STR = "c2.symbol = '"
    }
    else{
      stop ( "the type of identifier must be entrez gene identifiers or gene symbols \n")
    }
  }
  else if (entity == "disease"){
    if (identifier == "cui"){
      STR = "c1.diseaseId = '"
    }
    else  if (identifier == "mesh"){
      STR = "c1.MESH = '"
    }
    else  if (identifier == "omim"){
      STR = "c1.OMIM = '"
    }
    
    else{
      stop  ("the type of identifier must be cui or mesh or omim identifiers\n")
    }
  }
  else{
    stop ("the type of entity must be disease or gene \n");
  }
  #for (ent in inFile$V1 ){
  for (ent in inFile[,1] ){ ### OJV altered to be universal and pull 1st column by index as opposed to column name
    url <- "http://www.disgenet.org/oql"
    oql <- paste( "DEFINE
   	c0='/data/gene_disease_summary',
	c1='/data/diseases',
	c2='/data/genes',
	c4='/data/sources'
    ON
    'http://www.disgenet.org/web/DisGeNET'
    SELECT
	c1 (diseaseId, name, STY, MESH, OMIM, type),
	c2 (geneId, symbol,uniprotId, description, DPI, DSI, pantherName),
	c0 (score, EI, Npmids, Nsnps)

    FROM
    c0
    WHERE
    (
    c4 = 'ALL'
    AND ", STR, ent , "' )
    ORDER BY
    c0.score DESC" , sep = "")

    dataTsv <-  getURLContent(url, readfunction =charToRaw(oql), upload = TRUE, customrequest = "POST")
    #dataTsv <- rawToChar( getURLContent(url, readfunction =charToRaw(oql), upload = TRUE, customrequest = "POST"))
    myTextConnection <- textConnection( dataTsv ) 
    data <-  read.csv( myTextConnection, header = TRUE, sep = "\t" )
    close(myTextConnection)

    if (dim(data)[1] == 0 ){
      print ( paste (entity , ent, " is not in DisGeNET ", sep = " "))
    }
    else  {
      data$c0.EI <- ifelse(data$c0.EI == "null", NA, as.character(data$c0.EI) )
      dataFin <- rbind(dataFin, data)
    }
  
  }
  
  #Create outFile if it does not exist
  address <-  paste(getwd(), outFile, sep="/")

  print(address)
  
  write.table(dataFin,  address, sep="\t", row.names = F,dec = ".", quote = FALSE)
  return(dataFin) ###OJV altered to create dataframe in R environment
}

ASDcuigenes<-doQuery_modifiedOJV(ASDcuis, "ASDcuigenes.txt", "disease", "cui") ### NOTE: input file is object from previous command prior to doQuery function call
```

+Next, the list of ASD candidate genes was evaluated to identify genes that are specifically expressed in the tissue of most interest to ASD etiology (i.e., the human brain). Details for identifying genes with baseline expression in human tissue from GTEx are provided here: https://www.ebi.ac.uk/gxa/help/r-data-objects.html, here http://www.bioconductor.org/packages/release/bioc/html/ExpressionAtlas.html, and here https://www.ebi.ac.uk/arrayexpress/experiments/E-MTAB-5214/. Optional code is provided below if the user prefers. As the direct pull for all GTEx data is large and we were only interested in brain regions, data for genes expressed at the GTEx default cut-off of TPM>=0.5 in all human brain regions available were downloaded directly.
``` {r eval=TRUE}
# To download GTEx data in entirety
#if (!requireNamespace("BiocManager", quietly = TRUE))
    #install.packages("BiocManager")
#BiocManager::install("ExpressionAtlas", version = "3.8")
#library('ExpressionAtlas')
# GTExdata<-getAtlasData("E-MTAB-5214")

GTExbraindata <- read.table(url("https://github.com/veatcho/BMIN503_Final_Project/blob/master/E-MTAB-5214-query-resultsbrain_TPM.tsv?raw=true"), header = TRUE, sep = "\t", fill = TRUE, stringsAsFactors = FALSE)
summary(GTExbraindata)

#Create dataframe with ASD candidate genes expressed in human brain
ASDcuigenes <- ASDcuigenes %>% 
      select(c2.geneId, c2.symbol, c2.uniprotId, c0.score) %>%
      mutate(Gene.Name = as.character(c2.symbol))
ASDcuigenesdistinct <- distinct(ASDcuigenes, c2.uniprotId, .keep_all = TRUE)
ASDbraingenes <- left_join(ASDcuigenesdistinct, GTExbraindata, by='Gene.Name')
```

+Once all ASD candidate genes expressed in the brain were identified, the list was queried to identify genes that when knocked out of mice have a phenotypic consequence that is comparable to symptoms of ASD observed in humans (e.g., Abnormal Nervous System Development=MP:0003631, Abnormal Behavior/Neurological Phenotype=MP:0004924). Details for requesting mouse-phenotype data can be found here: http://www.mousephenotype.org/data/documentation/data-access
``` {r eval=TRUE}
library('jsonlite')
#Identify genes that when knocked out of mice are associated (p<0.05) with ASD-related symptoms
#Note: must pull 'top level terms' from IMPC
AbnormalNervousSystemMouse0.05 <- fromJSON('https://www.ebi.ac.uk/mi/impc/solr/genotype-phenotype/select?q=top_level_mp_term_id:%22MP:0003631%22AND%20p_value:%5b0%20TO%200.05%5d&rows=100000&wt=json&indent=1', flatten = TRUE)

AbnormalBehaviorNeuroMouse0.05 <- fromJSON('https://www.ebi.ac.uk/mi/impc/solr/genotype-phenotype/select?q=top_level_mp_term_id:%22MP:0005386%22AND%20p_value:%5b0%20TO%200.05%5d&rows=100000&wt=json&indent=1')

#Reformat lists to pull mouse gene names and phenotypic consequences in KOs 
AbnormalNervousSystemMouseGenes<-data.frame(phenotype=AbnormalNervousSystemMouse0.05$response$docs$mp_term_name[c(1:532)],  genes=AbnormalNervousSystemMouse0.05$response$docs$marker_symbol[c(1:532)], stringsAsFactors = FALSE)

AbnormalBehaviorNeuroMouseGenes<-data.frame(phenotype=AbnormalBehaviorNeuroMouse0.05$response$docs$mp_term_name[c(1:7025)],  genes=AbnormalBehaviorNeuroMouse0.05$response$docs$marker_symbol[c(1:7025)], stringsAsFactors = FALSE)

MouseKOASDtraits<-distinct(full_join(AbnormalNervousSystemMouseGenes, AbnormalBehaviorNeuroMouseGenes, by='genes'))

#Identify the most likely human orthologs of mouse genes identified above
#The mouse gene list was exported and batch searched using all prediction tools available via the DRSC Integrative Ortholog Prediction Tool (DIOPT; Version 7.1 March 2018; https://www.flyrnai.org/diopt) with filtering to return only best match when there is more than one match per input gene or protein

MouseKOASDtraitsgenes <- distinct(MouseKOASDtraits, genes)
write(MouseKOASDtraitsgenes$genes, file="MouseKOASDtraitsgenes.txt")

#Pull results from mouse/human ortholog query using DIOPT
MouseKOASDtraitsgenes.hsortho <- read.table(url("https://github.com/veatcho/BMIN503_Final_Project/blob/master/MouseKO_ASDtraitsgenes_hsortho.xls?raw=true"), header = TRUE, sep = "\t", fill = TRUE, stringsAsFactors = FALSE)

#Identify ASD risk genes expressed in human brain that also have a relevant ASD-related phenotype when knocked out of mice
ASDbraingenes <- mutate(ASDbraingenes, Human.GeneID = c2.geneId)
ASDbraingenes.mousegenes <- left_join(ASDbraingenes, MouseKOASDtraitsgenes.hsortho[, c(1, 6)], by='Human.GeneID')
MouseKOASDtraits <- mutate(MouseKOASDtraits, Mouse.GeneID=MouseKOASDtraits$genes)
ASDbraingenes.mouseKOpheno <- left_join(ASDbraingenes.mousegenes, MouseKOASDtraits, by='Mouse.GeneID')
```

+Next, we determined if any ASD risk genes were currently FDA-approved drug targets
``` {r eval=TRUE}
#Details for pulling from Pharos can be found here: https://pharos.nih.gov/idg/api

targets_pharos <- fromJSON("https://pharos.nih.gov/idg/api/v1/targets?facet=IDG+Development+Level&top=25000&wt=json&indent=1")
drugtargets<-data.frame(genes=targets_pharos$content$gene[c(1:20244)], idgTDL=targets_pharos$content$idgTDL[c(1:20244)], uniprotId=targets_pharos$content$accession[c(1:20244)], stringsAsFactors = FALSE)

#Pull tclin level (i.e. FDA-aproved) targets
#To identify expected number of Tclin targets queried https://pharos.nih.gov/idg/targets?facet=IDG+Development+Level/Tclin

tclintargets <- filter(drugtargets, grepl("Tclin", idgTDL))

#Identify ASD risk genes expressed in human brain, have a relevant ASD-related phenotype when knocked out of mice and are FDA-approved drug targets
tclintargets<-mutate(tclintargets, Gene.Name=genes)
ASDbraingenes.mousepheno.drugtarget <- left_join(ASDbraingenes.mouseKOpheno, tclintargets, by='Gene.Name')
```

### Results
Describe your results and include relevant tables, plots, and code/comments used to obtain them. End with a brief conclusion of your findings related to the question you set out to address. You can include references if you'd like, but this is not required.
``` {r eval=TRUE}
library(reshape2)
library(ggplot2)
#Determine the proportion of ASD candidate genes pulled from DisGeNET that fall into categories of interest: 
#1) expressed in human brain, #2) have relevant mouse phenotype when knocked out #3) are FDA-approved drug targets?

ASDGeneAttributes<- ASDbraingenes.mousepheno.drugtarget %>%
    select(ASDRiskGene=c2.symbol, BrainGene=Gene.ID, MouseKOGene=Mouse.GeneID, DrugTarget=genes.y) %>%
    mutate(nASDriskgenes=n_distinct(ASDRiskGene)) %>%
    mutate(nBraingenes=n_distinct(BrainGene)) %>%
    mutate(Brain.Expression = (sum(nBraingenes)/sum(nASDriskgenes))) %>%
    mutate(nMouseKOgenes=n_distinct(MouseKOGene)) %>%
    mutate(Mouse.KO.Phenotype = (sum(nMouseKOgenes)/sum(nASDriskgenes))) %>%
    mutate(nDrugtargets=n_distinct(DrugTarget)) %>%
    mutate(Drug.Target = (sum(nDrugtargets)/sum(nASDriskgenes))) %>%
    select(ASDRiskGene, Brain.Expression, Mouse.KO.Phenotype, Drug.Target) %>%
    distinct(ASDRiskGene, .keep_all = TRUE)

ASDGeneAttributes.melt<-melt(as.data.frame(ASDGeneAttributes), id.vars = "ASDRiskGene")

ggplot(data=ASDGeneAttributes.melt, aes(x=variable, y=value, fill=variable)) + 
      geom_bar(position="dodge", stat="identity") +
      scale_fill_brewer(palette="Set1") +
      ggtitle("ASD Candidate Gene Attributes\n(n=917) ", ) +
      theme(plot.title = element_text(hjust = 0.5)) +
      xlab("Functional Annotation") + ylab("Proportion of ASD Genes") +
      labs(fill = "Functional Annotation")

```
```{r eval=TRUE, fig.width=12, fig.height=10}

#Identify the specific human brain regions that are enriched for ASD candidate gene expression
library(limma)
library(gplots)

ASDgenebrainexpression <- inner_join(ASDbraingenes.mousepheno.drugtarget[, c(1, 5)], GTExbraindata, by="Gene.Name")
ASDgenebrainexpression <- ASDgenebrainexpression %>%
    select(as.numeric(c(4:16)))
ASDgenebrainexpressionm<-as.matrix(ASDgenebrainexpression)

ASDgenebrainexpressionTEST<-ASDbraingenes.mousepheno.drugtarget %>%
    filter(!is.na(Gene.ID)) %>%
    select(as.numeric(c(4:16)))

#Determine the human brain regions with the strongest expression of ASD candidate genes
coolmap(ASDgenebrainexpressionm, cluster.by = "de pattern", linkage.row = "none", linkage.col = "none", show.dendrogram = "none", main = "Brain Region Expression Normalized by ASD Gene", xlab="Brain Region", ylab="ASD Candidate Genes", keysize=0.75,  offsetCol = 0, margins = c(12, 5))

#Determine the ASD candidate genes wtih the strongest expression in the human brain
coolmap(ASDgenebrainexpression, cluster.by = "expression level", linkage.row = "none", linkage.col = "none", show.dendrogram = "none", col="redblue", main = "Brain Region Expression Normalized by ASD Gene", xlab="Brain Region", ylab="ASD Candidate Genes", keysize=0.75,  offsetCol = 0, margins = c(12, 5))

```  

``` {r eval=TRUE}
#Determine the most common mouse nervous system and behavioral/neurological phenotypes that are observed when ASD candidate genes are knocked out

ASDgenemousekos <- inner_join(ASDbraingenes.mousepheno.drugtarget[, c(1, 5)], GTExbraindata, by="Gene.Name")
  select(, na.exclude=TRUE)
  

```    

``` {r eval=TRUE}
#Determine how many genes prioritized in this pipeline are ACMG recommendations
#PPV, NPV
```

``` {r eval=TRUE}
#Determine confidence level of prioritized genes compared to expert review via SFARI
```

``` {r eval=TRUE}
#Determine how many genes encode targets of drugs used to treat symptoms in ASD & how many have evidence for pharmacogenomic variants
```
#917 random protein coding genes (houekeeping??), same ids and rerun pipeline perform bootstrapping, impute missing values and look at clustering