---
title: "BMIN503/EPID600 Project Template"
author: "Your Name"
output: 
  html_document:
    toc: false 
    depth: 3 
    theme: paper 
    highlight: tango
---
```{r set-options, echo=FALSE, cache=FALSE}
options(width = 400)
```  
***
Use this template to complete your project throughout the course. Your Final Project presentation in class will be based on the contents of this document. Replace the title/name and text below with your own, but leave the headers.

### Overview
In this section, give a brief a description of your project and its goal, what data you are using to complete it, and what three faculty/staff in different fields you have spoken to about your project with a brief summary of what you learned from each person. Include a link to your final project GitHub repository.

Draft1

Cell type identification and categorization is crucial to our understanding of cell function and diseases. Utilizing single cell sequencing and clustering cells allows for the identification and validation of disease-specific biomarkers. By determining biomarkers for a specific cell type, a biomarker can be used to identify a specific cell type in other studies provided the biomarker does not overlap too heavily with other cell types present in the cell population. It is worthwhile to verify numerous biomarkers for an individual cell identity, as those cell types can be more readily recognized in diverse cell populations.

### Introduction 
In the first paragraph, describe the problem addressed, its significance, and some background to motivate the problem.

In the second paragraph, explain why your problem is interdisciplinary, what fields can contribute to its understanding, and incorporate background related to what you learned from meeting with faculty/staff.

Draft1

This analysis will seek to utilize data from the 25,000 mouse bipolar cells sequenced by Dr. Shekhar to reproduce the identification of biomarkers for the 15 cell types presented within the paper. The analysis will be done using Seurat, and an R toolkit designed for single cell genomics, specifically to preform QC, analyze, and explore single-cell RNA data. Seurat comes with functions designed to import data, filter cells, preform PCA/CCA, cluster cells, and determine differentially expressed genes between the clusters. By potentially reproducing the biomarkers identified by Dr. Shekhar this may help to further authenticate these biomarkers, as well as validating the methodologies of the Seurat package.

### Methods
In the first paragraph, describe the data used and general methodological approach. Subsequently, incorporate full R code necessary to retrieve and clean data, and perform analysis. Be sure to include a description of code so that others (including your future self) can understand what you are doing and why. 

Draft1

See ShekharAnalysis.Rmd in repository, contains current methods/code used thus far

### Results
Describe your results and include relevant tables, plots, and code/comments used to obtain them. End with a brief conclusion of your findings related to the question you set out to address. You can include references if you'd like, but this is not required.

Draft1

See Shekhar Analysis.Rmd in repository, contains a few revelavant results. Still a work in progress.
