---
title: 'Effect of Airborne Patriculate Matter in Gene Expression of BEAS-2B Cells'
author: 'Nisha Narayanan'
output: 
  html_document:
    toc: true
    toc_float: 
        collapsed: true
        smooth_scroll: true
    depth: 3 
    theme: flatly
    highlight: pygments
---

```{r global options, include = FALSE}
knitr::opts_chunk$set(warning = FALSE, message = FALSE)
```   

***
## Overview

Airborne particulate matter is considered to be a hazard due to its effects in human health and air inside workplaces/classrooms/household buildings can be more polluted than the outside air.(Valavidinis et al., 2008 and Jovanovic et al., 2014). It is well-known that undiagnosed respiratory disorders caused due to air pollutants can lead to short-term (frequent hospital visits) and long-term health effects (morbidity, lung cancer, cardiovascular  and cardiopulmonary diseases). However, recent studies have shown that the relationships between indoor air quality index and respiratory health of students in school are insufficiently explored (Annesei-Masinao et al., 2013). To try to bridge this knowledge gap, this project aims to analyze an existing dataset (GSE34607)  available in the GEO public repository that contains the gene expression data taken from classrooms with high concentrations of PM10 to study how air pollutants influence gene transcription in airway cells. The hope is that this results will help to understand the mechanisms involved in air pollution induced respiratory disorders and thereby lead to the understanding of whether interventions are required for vulnerable people. 

The project was discussed with the below three faculties of Penn Medicine and the objectives of the project were designed based on their inputs.
1. Dr. Blanca E. Himes, Associate Professor of Informatics, Department of Biostatistics, Epidemiology and Informatics 
2. Dr. Mengyuan Kan, Postdoctoral Fellow, Department of Biostatistics, Epidemiology and Informatics 
3. Dr. Aimin Chen, Professor of Epidemiology, Department of Biostatistics, Epidemiology and Informatics

## Introduction

Air pollution is the contamination of the indoor or oudoor environment by any chemical, physical or biological agent. According to World Health Organization, air pollution is a growing environmental concern that is responsible for around 7 million  deaths every year (World Health Organization, 2008). Among the six common air pollutants, Particulate Matter (PM) is gaining wide popularity among researchers. PM is defined as a mixture of solid and liquid particles in air. PM that less that 10 μm or smaller in diameter are inhalable by human beings and it has the ability to penetrate deep into the lungs (Utah Dept of Health). Several studies have indicated that the exposure to high PM10 concentrations is associated to cause short- and long-term effects such as increased respiratory symptoms, alterations in tissue and lung structure and premature death (de Gennaro et al., 2014). PM is found in both indoor and outdoor environments but only the effects of outdoor PM on human health have been extensively researched. Even though the Environmental Protection Agency (EPA) has indicated that interior PM levels have the potential to exceed outdoor PM levels and can produce detrimental health consequences in humans, little is known regarding the impact of indoor PM on human health (US EPA). The impact of indoor PM in the school environment on pupils' respiratory health is still mostly unknown. Pollens, spores, industrial pollutants, cooking, and smoking are all common PM sources. Despite the fact that these aren't prevalently found in schools, PM10 concentrations are shown to be higher. The majority of respiratory illnesses can be treated with glucocorticoids, which alter gene transcription directly, although their cell- and tissue-specific effects are poorly known (Kan et al., 2018).
Exploring the gene expression of airway cells in an indoor and outdoor environments when exposed to PM10 can aid in deciphering the nuances of altered genes and pathways, and thus help in the development of effective therapies. The goal of this project is to investigate how airborne PM influences gene expression in BEAS-2B (bronchial epithelial) cells and to suggest viable intervention approaches for reducing the impact of indoor PM on students' respiratory health.

We need an inter-disciplinary methodology to execute the data analysis and analyze the outcomes. To begin, possessing a fundamental knowledge of air pollution and its impact on the human respiratory system is crucial. Furthermore, it is clear that understanding how indoor pollution differs from outdoor pollution is significant for the research. Second, having a better grasp of the biology and disease mechanism of asthma and allergic reactions would aid in better identifying the therapy options available. Thirdly, insights on genes and gene pathways (especially those involved in respiratory illnesses) is critical for interpreting how air pollutants affect gene transcription in airway cells. Finally, statistics and R will assist with data analysis and visualization, allowing reasonable and reliable judgments to be made. As a result, the subject addressed in this project covers multiple domains, making it highly inter-disciplinary.

## Methods

The data used in this project was accessed from Gene Expression Omnibus (GEO), a public functional genomics data repository. The microarray based data used in this study can be found here - https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE34607. An overview of the experiment done is essential to comprehend the data. BEAS-2B cells, passage 41, were treated with PM10 (10 µg/ml) in six indoor and outdoor school environments, and samples were tested at time intervals of 4, 10, and 24 hours in triplicates.Genome-wide gene expression in human bronchial BEAS-2B epithelial cells were analysed and regulated genes were verified by quantitative PCR. "Affymetrix" is the platform that was used (Human Genome U133A 2.0 Array).  

Extensive data clenasing was necessary to begin the analysis process. This was done with the use of a reproducible tool developed by Himes Lab at the University of Pennsylvania called "RAVED" (Himesgroup-RAVED). To understand the structure of the data and execute data cleansing, a Quality Control (QC) study was performed initially. This resulted in the creation of a phenotypic file comprising GEO ID, Tissue, Disease, and Treatment information. The Differential Expression (DE) analysis was performed using the generated report as an input for six different combinations, Indoor vs Control 4h, Indoor vs Control 10h, Indoor vs Control 24h, Outdoor vs Control 4h, Outdoor vs Control 10h, and Outdoor vs Control 24h. The results of the DE analysis were saved as.csv files and utilized as input for the rest of the project.
Note: The corresponding and raw rmd files used to generate the phenotype and DE results are available in the Github Repository and can be accessed here (https://github.com/Narayanan-Nisha/BMIN503_Final_Project/tree/master/RAVED-QC%26DE). 

***

To begin the data analyis, the required packages were loaded in R, which can be done through the below R chunk.

```{r packages, eval = FALSE, echo=TRUE, message=FALSE, warning=FALSE}
if (!requireNamespace("BiocManager", quietly = TRUE)) #install BiocManager if not available
    install.packages("BiocManager")
#To install packages from BiocManager
for (pkg in c("GEOquery", "oligo", "ArrayExpress", "limma", "viridis", "fgsea", "affy")) {
  if (!requireNamespace(pkg, quietly = TRUE)) {
    BiocManager::install(pkg)
  }
}
#To install base packages from R
for (pkg in c("dplyr", "gplots", "ggplots", "devtools", "pander", "tidyr", "ggpubr", "flextable", "ftExtra")) {
  if (!requireNamespace(pkg, quietly = TRUE)) {
    install.packages(pkg)
  }
}
```

Loading the libraries. 
NOTE: "dplyr" and "affy" packages will be loaded if and when required as it will mask the other library functions.

```{r lib, eval=TRUE, echo=TRUE, message=FALSE, warning=FALSE}
library(GEOquery) #to access and download raw data and supplementary files from GEO
library(ArrayExpress)
library(oligo)
library(viridis) # heatmap colour
library(ggplot2) #for plots
library(gplots) # heatmap.2 plot
library(devtools) # compute PCs
library(pander) #for tables
library(limma)
library(flextable) #for spanning tables
library(ftExtra) #to add features to flextable
library(tidyr)
library(ggpubr) #to arrange plots
```

The raw and supplementary data required for this project is first downloaded from GEO through the below steps. 

```{r raw and supplementary data, eval=TRUE, echo=TRUE, message=FALSE, warning=FALSE}
if (!file.exists("./GSE34607/GSE34607_RAW.tar"))
  getGEOSuppFiles("GSE34607")
untar("./GSE34607/GSE34607_RAW.tar", exdir = "./GSE34607/data")
```

This chunk is used to define some of the base variables required for further analysis. - including platform used, whether the data  is normalized, shortname function to shorten variable names, whether supplementary data exists, GEO ID and datadir (note this needs to be changed depending upon where the required files are stored in the user's machine).

```{r var2, eval=T, echo=T}
geo_id="GSE34607" #GEOID, as defined in GEO
platform="Affymetrix" #platform used for analysis, as mentioned in GEO
geo_GPL=""
normdata=FALSE #Data normalized

# The shortname_func function shortens the sample name shown in the plots. To start, define shortname_func <- function(x){x}
shortname_func <- function(x){gsub("^(.*)\\.(cel|CEL).gz","\\1",x)}

suppldata = TRUE #supplementary data present
datadir="/Users/nishanarayanan/Downloads/" #this needs to be changed depending upon where the required files are stored in the user's machine
```

The raw probe intensity distributions for each sample is visualized by displaying them as boxplots. Next, Robust Multi-array Average (RMA) is used to normalize and transform the intensities to a comparable scale. Looking at the raw expression, notice that the intensities for each sample vary widely. This does not reflect the biological variation we are interested in, which is why a function like RMA is used.

```{r}
celFiles <- list.celfiles("./GSE34607/data", full.names = TRUE, listGzipped = TRUE) #to list all celFiles from the doe=wnloaded location
raw.data <- read.celfiles(celFiles) #reading the celfiles and storing it to a variable called rawdata
boxplot(raw.data, col = "red", main = "Raw Probe Intensities") #to visualize raw probe intensity
raw.data.rma <- oligo::rma(raw.data) #normalizing raw data through rma
boxplot(raw.data.rma, col = "blue", main = "RMA Expression Values") #to visualize normalized raw probe intensity
```

Note: All the raw files that need to be manually downloaded and loaded to R for this project can be accessed at this github repo - https://github.com/Narayanan-Nisha/BMIN503_Final_Project/tree/master/RequiredFilesforMainCode 

To obtain the raw phenotype file, as already mentioned, Quality Control steps from RAVED was followed. A summary table was generated to understand the data using RAVED. The file is loaded below and displayed for further analysis. 

```{r pheno_raw, eval=T, echo=T, message=F, warning=F}
pheno.raw <- read.delim("RawPhenoFile_GEO") 
pandoc.table(head(pheno.raw[,1:10], 5)) #visualize first five rows and ten columns of raw phenotype
```

It is evident that not all the variables are informative. Hence, the raw phenotype is modified to generate a "pheno" (dataframe) with the required variables from the raw phenotype file for further analysis. Two columns, scan date group and filename is then included to the pheno dataframe and the corresponding data is extracted from the raw.data expression feature set and added to the respective columns in the dataframe. In this step, the donor names are modified to a more readable format; treatment, treatment time and tissue variables are also defined.

```{r pheno_new, eval=T, echo=T, message=F, warning=F}
library(dplyr)
cols <- c("title","geo_accession","source_name_ch1") #list with required original column names from pheno.raw required for data cleansing
pheno <- pheno.raw %>%
  dplyr::select(cols) %>%
  dplyr::mutate(GEO_ID=geo_accession) %>% #assigning GEOID
  dplyr::mutate(Donor=gsub("\\S+_", "D", gsub(".rma-Signal","",title))) %>% #extracting the donor/replicate details from the tile column
  dplyr::mutate(Sample=paste(geo_accession,Donor,sep="_")) %>% #defining treatment method
  dplyr::mutate(Tissue="BEAS-2B") %>% #tissue is default for all experiments
  dplyr::mutate(treatment_time=gsub("^.*for (\\d+.*)$","\\1",source_name_ch1)) %>% #extracting the exposure time from the source_name_ch1 column
  dplyr::mutate(treatment_time=gsub("PM10_Indoor_|PM10_Outdoor_|PM10_Untreated_|untreated_|PM_10_indoor_|PM10_ourtdoor_|PM_10_outdoor_","",treatment_time, ignore.case=T)) %>% #making the treatment time more intuitive
  dplyr::mutate(treatment_drug=ifelse(grepl("indoor", ignore.case = T, source_name_ch1),"IndoorExposure",ifelse(grepl("outdoor", ignore.case = T, source_name_ch1),"OutdoorExposure",ifelse(grepl("ourtdoor", ignore.case = T, source_name_ch1),"OutdoorExposure", "control")))) %>% #extracting the treatment drug/mode of exposure from source_name_ch1 column
  dplyr::mutate(Treatment=paste(treatment_drug,treatment_time,sep="_")) %>%
  dplyr::mutate(Disease="nonasthma") %>%
  dplyr::mutate_if(is.character,as.factor) %>%
  dplyr::select(-one_of(cols)) # remove original columns
detach("package:dplyr")
pheno$ScanDate_Group <- sapply(strsplit(as.character(protocolData(raw.data)$dates), "T| "), function(x) {x[[1]]}) #including scandategroup to pheno dataframe
pheno$ScanDate_Group <- as.factor(pheno$ScanDate_Group) #converting to factor
pData(raw.data) <- pheno # update pData of rawdata with pheno
row.names(pData(raw.data)) <- sampleNames(protocolData(raw.data)) #assign rownames
pheno$Filename <- rownames(pData(raw.data)) #include filename columnn in pheno, contains details of celfile names
```

Assigning colours based on scan date group for graphing purposes.

```{r color, eval=TRUE, echo=TRUE, message=FALSE, warning=FALSE}
colours=c("#1B9E77", "#D95F02", "#7570B3", "#E7298A", "#66A61E", "#E6AB02", "#A6761D", "#666666", "#8DD3C7", "#FFFFB3", "#BEBADA", "#FB8072", "#80B1D3", "#FDB462", "#B3DE69", "#FCCDE5", "#D9D9D9", "#BC80BD", "#CCEBC5", "#FFED6F") # first 8 colour names derived from Dark2, and last 12 names from Set3
varuse="ScanDate_Group"
i=nlevels(pheno[,varuse])
colour_list <- colours[1:i]
names(colour_list) <- levels(pheno[,varuse]) 
```

## Results

Now that the required data is cleansed, the next step will be to analyse it. Firstly, a brief quality control analysis is shown. This will be followed by differential expression analysis and visualization of results. 

### Quality Control Analysis

The major QC steps include density plots, RNA degradation plots, spatial plots, MA plots, Relative logarithmic expression (RLE), Normalized unscaled standard error (NUSE) and Principal component analysis plots (PCA). In this project, we will analyse the raw intensity plots, outlier detection, RNA degradation and PCA in detail. 

First, let us begin by displaying the summary of phenotype variables and the sample size for different groups.

```{r pheno_check_raw, eval=T, warning=F,results="asis",echo=T}
# show the phenotype data
pandoc.table(head(pData(raw.data),27), split.tables=Inf,caption="Modified phenotype file")

# show the groups of interest
avail_group=c("Tissue","Disease","Treatment")[c("Tissue","Disease","Treatment")%in%names(pheno)]
res=as.data.frame(table(pheno[,avail_group]))
names(res) <- c(avail_group,"Count")
pandoc.table(res[which(res$Count>0),], split.tables=Inf, caption="Sample size in different tissue and disease/treatment groups")

# show samples in different batch
if ("ScanDate_Group"%in%names(pheno)) {
  res=as.data.frame(table(droplevels(pheno[,"ScanDate_Group"])))
  names(res) <- c("ScanDate_Group","Count")
  pandoc.table(res, split.tables=Inf, caption="Sample size in different batch")
} else {cat("No scan date information.")}
```


#### Outlier detection and raw probe intensity plots 
The outlier detection is studied by computing the Kolmogorov-Smirnov statistic (Ka) between the log-intensity expression of one array and the pooled array (array with Ka beyond the upper whisker is defined as an outlier).  The log2 transformed intensity values are plotted as a boxplot and density curve. The requirement for the boxplot testing is that the normalized intensity distributions of all samples are expected to have similar scales, i.e., the position and width of the boxes should be similar with zero outliers. The requirement for density curves is also the same, the curves should be in similar curves and sizes and it is known that samples with similar curves can create issues.


```{r}
subsamp <- function(x,num=20000, seed=123) {
  set.seed(seed)
  subsample=num # if number of probes are >20000, randomly select 20000 probes for plot or compute
  if (nrow(x)>subsample) {
    ss  = sample(nrow(x), subsample)
    Mss = x[ss,,drop=FALSE]
  } else {
    ss  = TRUE
    Mss = x
  }
  Mss
}
outlier_KS_func = function(exprs) { # matrix (row: probe intensities/RLE values etc., col: array (e.g. sample))
  fx = ecdf(as.vector(exprs)) # get empirical cumulative distribution function of the data
  KS=suppressWarnings(apply(exprs, 2, function(v)ks.test(v, y = fx, alternative="two.sided")$statistic))
  stats = stats::fivenum(KS, na.rm = TRUE) # Tukey's five number summary (minimum, lower-hinge, median, upper-hinge, maximum)
  iqr = diff(stats[c(2, 4)]) # lagged difference between the lower-hinge and upper-hinge
  coef = 1.5
  th = (stats[4] + coef * iqr)
  outlierfun <- list(threshold = th, stats=KS, outlier = sum(KS > th))
  return(outlierfun$outlier)
}

plot_func <- function(Mss,outlier,ylab) {
  # use * to mark the outliers in boxplot
  array_name <- shortname_func(colnames(Mss))
  outlier <- shortname_func(outlier)
  array_name[array_name%in%outlier] <- paste0("*",outlier)
  # boxplot raw intensity by array
  ylim = quantile(Mss, probs = c(0.01, 0.99), na.rm=TRUE) # create range of y-axsis
  # create data frame for plot
  df <- data.frame(
    sample_id=rep(colnames(Mss),each=nrow(Mss)),
    values=as.numeric(Mss),
    scandate=rep(pData(raw.data)[,varuse],each=nrow(Mss)) # for color
  )
  cols <- colour_list
  boxplot <- ggplot(df, aes(sample_id,values,fill=scandate)) + geom_boxplot(outlier.colour=NA) +
           coord_flip() + theme_bw() +
           ylim(ylim) +
           scale_x_discrete(labels=array_name) +
           ylab(ylab) +
           scale_fill_manual(varuse,values=cols) +
           theme(axis.title.y=element_blank())
  densityplot <- ggplot(df,aes(x=values,colour=scandate)) + geom_line(aes(group=sample_id),stat="density") +
                 theme_bw() +
                 xlab("Raw Probe Intensities") +
                 ylab("Density") +
                 scale_color_manual(varuse,values=cols)
  return(list(boxplot, densityplot))
}

```


```{r}
#Random selection of 20000 probes
ss  <- sample(nrow(exprs(raw.data)), 20000)
Mss <- log2(exprs(raw.data)[ss,,drop=FALSE]) #use log2 transformed raw probe intensity
outlier <- outlier_KS_func(Mss)
cat("The number of outliers are ", outlier)
plot_func(Mss=Mss, outlier=outlier, ylab="Raw Probe Intensities")
```

#### RNA Digestion


Overall RNA quality can be assessed by RNA degradation plots. Each probe sets consists of about 11-20 probes and this plot is used to show the average intensity of each probe across all probe sets from 5' to 3' end. This step requires R package affy that outputs the probes in each probe set matrix, while this function is not implemented in the oligo package. Since RNA degradation starts at the 5' end, it is expected that probe intensities are lower at 5' when compared to 3' end.


```{r RNAdegaffy_utility, eval=TRUE, echo=TRUE}
# The rawall_func function obtains all files in the data directory with full path
rawall_func <- function() {
    raw_fn=list.files(path=paste0(datadir,"/",geo_id,"/data"))
    # obtain supplementary data with path
    paste0(datadir,"/",geo_id,"/data/",raw_fn[which(raw_fn%in%basename(as.character(pheno.raw$supplementary_file)))]) # only select cel files from datadir
  }
# The RNAdegaffy_func function compute mean PM intensities in each probe position following 5' to 3' order. Adopted from the AffyRNAdeg function in the affy package but include a step that randomly selects 20,000 probe sets.
RNAdegaffy_func <- function(data){ # input a list of probe set matrix with rows as probe ids and columns as samples
  {
    names <- colnames(data[[1]])
    probe.set.size <- function(x) {
      size <- dim(x)[1]
      return(size)
    }
    max.num <- sapply(data, probe.set.size) # get the number of probes in each probe set
    tab <- (table(max.num)) # summarize the frequencies of probe numbers in probe sets
    ord <- order(-as.numeric(tab)) # order the frequency from large to small
    K <- as.numeric(names(tab))[ord[1]] # K is the number of probes appearing in most probe sets
    data <- data[max.num == K] # select data of probe sets only have K number of probes
  }
  
  subsample=20000
  if (length(data)>subsample) { # randomly select 10000 probe sets
    set.seed(12345)
    ss = sample(length(data),subsample)
    data = data[ss,drop=FALSE]
  }
  N <- length(data) # number of probe sets
  n <- dim(data[[1]])[2] # number of samples
  
  # create two matrices: number of samples * number of probes representing a probe set
  mns <- matrix(nrow = n, ncol = K) # create matrix for mean values
  sds <- mns # create matrix for sds values
  get.row <- function(x, i = 1) {return(x[i, ])} # function to get each row (i.e. probe id, i) from one probe set x (i.e. probe list[[x]])
  rowstack <- function(x, i = 1) {return(t(sapply(x, get.row, i)))} # function to combine the rows obtained using get.row (pms across samples by probe sets) to get a table (row: samples column: probe sets) and transpose the table (row: probe sets, column: samples)
  for (i in 1:K) { # get probe id (position) from 1 to K from each probe set
    data.stack <- rowstack(data, i) # get the probe pm values in a specific probe position across all samples from each probe set (rows are samples and columns are probe sets)
    if(dim(data[[1]])[2]==1) data.stack <- t(data.stack)
    mns[, i] <- colMeans(data.stack) # get the mean values at one probe position across all probe sets
    sds[, i] <- apply(data.stack, 2, sd) # get the sd values at one probe position across all probe sets
  }
    
  mns.orig <- mns # store the original mns data matrix
  mn <- mns[, 1] # select values in the first probe position
  mns <- sweep(mns, 1, mn) # adjust for the intensity at the first probe position
  mns <- mns/(sds/sqrt(N)) # adjust for standard error
  lm.stats <- function(x) {
    index <- 0:(length(x) - 1)
    ans <- summary(lm(x ~ index))$coefficients[2, c(1, 4)] # use linear model fit the relationship between intensity and probe position
    return(ans)
  }
  stats <- apply(mns, 1, lm.stats)
  answer <- list(N, names, mns.orig, sds/sqrt(N), stats[1,], stats[2, ])
  names(answer) <- c("N", "sample.names", "means.by.number","ses", "slope", "pvalue")
  return(answer)
}
# The RNAdeg_func function generates RNA degradation plots
# return a logical variable whether this array type can be read by affy.
RNAdeg_func <- function() {
  # 1. Read in raw data as an AffyBatch object
  library(affy) # for Affymetrix microarray-specific QC analysis
  raw.data.affy <- read.affybatch(rawall_func(),compress=T)
  # 2. Obtain a list of probe sets with a matrix of oligos (probes) by samples as an input and compute statistics of the mean PM intensities from 5' to 3' probe positions. 
  PM_list <- affy::pm(raw.data.affy,LIST=T) 
  PM_list <- lapply(PM_list,log2)
  raw.data.rnadeg <- RNAdegaffy_func(PM_list) # Compute mean PM intensity for probes following 5' to 3' order.
  # 3. Plot 5' to 3' mean PM intensity
  status.cols <- unlist(lapply(pData(raw.data)[,varuse],function(x)colour_list[x])) # colour list to corresponding scan date list
  plotAffyRNAdeg(raw.data.rnadeg,cols=status.cols)
  legend("topleft",legend=names(colour_list),fill=colour_list,cex=0.6)
  detach("package:affy", unload=TRUE) # detach the affy package
}
```

```{r affy_exclude, eval=T, echo=T}
# arrays not in affy package
affy_exclude = c('hta.2.0', 'pd.hugene.2.0.st')
```


```{r RNAdeg_plot, eval=T, echo=TRUE, message=F, warning=F}
if (platform=="Affymetrix"&suppldata) {
  if (any(sapply(affy_exclude, function(y){grepl(y, annotation(raw.data))}))) {cat("The affy package is not designed for this array type.\n")} else {RNAdeg_func()}
}
```

From the plot, we can interpret that the probe intensities are lower at the 5' end of a probe set when compared to the 3' end as RNA degradation starts from the 5' end of a molecule. RNA that are too degraded shows a very high slope from 5' to 3'. 

#### Principal Component Analysis (PCA)

PCA demonstrates information of the expression dataset in a reduced number of dimensions. In this step we compute PCs and the variance is explained by the first 10 PCs. PCA plots are generated using the first two principle components colored by known factors (e.g. treatment/disease conditions, tissue, donors and scan dates), visualizing similarities between arrays and these similarities' correlation to batch effects.

```{r pca_utility, eval=T, echo=T}
raw.data.pca <- na.omit(log2(exprs(raw.data)))
 
  # As scale function divides by the variance, the probe with the expression sd=0 across samples must be removed.
  sd <- apply(raw.data.pca,1,sd)
  raw.data.pca <- raw.data.pca[!sd==0,]
  # compute pcs
  pca <- prcomp(t(raw.data.pca), retx = TRUE, center = TRUE, scale = TRUE)
  pc <- data.frame(pca$x)
  # compute variance explained by each PC
  vars <- pca$sdev^2
  pcs <- t(pc)
  pvars <- vars*100.0/sum(vars) # proportion of variance (%) explained by each PC
  cumsum_pvars <- cumsum(pvars) # Cumulative Proportion of Variance (%)
  if (nrow(pcs)>10) {nres <- 10} else {nres=nrow(pcs)} # select top 10 PCs if number of PCs >10
  res <- data.frame(rownames(pcs),pvars,cumsum_pvars)[1:nres,]
  names(res) <- c("PC","Proportion of Variance (%)","Cumulative Proportion of Variance (%)")
pandoc.table(res,split.tables=Inf, caption="Variance explained")

# The pcaplot_func creates plots for pc1 and pc2
pcaplot_func <- function(pc, group_var,legend) { # group_var: column name for a specific group; legend: legend name
  group=pData(raw.data)[which(row.names(pData(raw.data))%in%row.names(pc)),group_var]
  df <- data.frame(
    PC1=pc$PC1,
    PC2=pc$PC2,
    group=group
  )
  i=length(levels(group))
  group_col <- colours[1:i]
  names(group_col) <- levels(pData(raw.data)[,group_var]) # colour to corresponding group for plot
  ggplot(df,aes(PC1,PC2,color=group)) + geom_point() +
    theme_bw() +
    scale_color_manual(legend,values=group_col,na.value="grey")
}
# The pca_func function generates multiple pca plots for scan date, disease, treatment, and Donor
pca_func <- function(pc) {
  group_vars=c("ScanDate_Group", "Disease", "Treatment", "Tissue", "Donor")
  legends=c("ScanDate_Group", "Disease", "Treatment", "Tissue", "Donor")
  idx_exist=c(1:length(group_vars))[group_vars%in%names(pData(raw.data))] # obtain index of existing variables
  plot_list=list() # store plots in a list
  for (i in idx_exist) {
    group_var=group_vars[i]
    legend=legends[i]
    nlevel=nlevels(pData(raw.data)[,group_var]) # levels of the variable
    if (group_var=="ScanDate_Group"|(nlevel>=2&nlevel<=20)) {
      plot_list[[group_var]]=pcaplot_func(pc, group_var=group_var,legend=legend)
    }
  }
  return(plot_list)
}
plot_list=pca_func(pc=pc)
for (i in plot_list) {print(i)}
```

It is clear that the scan date groups, donors and treatment form clusters in the plot. The effects of donors in the results were hence normalized and fixed during the differential expression analysis. 

#### QC Summary

QC status is set to pass or fail depending upon the number of outliers detected. The outlier detection is based on the outliers detected in raw probe intensity, MA, NUSE, RLE and spatial plots. By default, samples detected as an outlier more than twice are assigned to 0 otherwise to 1. Only the raw probe intensity was coded in detail for this project. A more comprehensive QC report which includes the outliers for all other plots are discussed in detail in the QC_GSE34607.Rmd file at the Github repository (performed from RAVED tool). 
 
As we can see from the outlier detection plot above, the number of outliers detected are 0, the boxplot and density plot curves have similar shapes/sizes and the RNA degration plot confirms low intensity at 5' than 3'. Thus, according to ArrayQualityMetrics guidelines, QC analysis passed. The QC Pass status column is added to the "Pheno" table. 

```{r files1, eval=TRUE, echo=TRUE, message=FALSE, warning=FALSE}
pheno$QC_PASS <- rep("Pass")
pandoc.table(head(pheno), style="grid", split.cells = Inf, caption="QC Summary with Phenotypes") #displaying header of pheno as pandoc table
```

### Differential Expression Analysis

An exhaustive differential expression analysis was performed using RAVED against all possible comparisons, including, indoor 4h vs control 4h, indoor 10h vs control 10h,, indoor 24h vs control 24h, outdoor 4h vs control 4h, outdoor 10h vs control 10h and outdoor 24h vs control 24h. This helped in the generation of Gene differential expression result files in CSV format. The pathway analysis was also a part of this pipeline and thereby helped in the generation of pathway results in CSV format. The detailed RMD files that were used to prepare these CSV files is also availably publicly at the same repositry under the folder "RAVED-QC&DE".

The generated .csv files for gene expression and pathway analysis from RAVED is loaded and stored in corresponding variables. 

```{r genefiles, eval=TRUE, echo=TRUE, message=FALSE, warning=FALSE}
indoorcontrol4h <- read.csv("GSE34607_BEAS-2B_nonasthma_IndoorExposure_4h_vs_control_4h.csv") #indoorvscontro4h gene report
indoorcontrol10h <- read.csv("GSE34607_BEAS-2B_nonasthma_IndoorExposure_10h_vs_control_10h.csv") #indoorvscontro10h gene report
indoorcontrol24h <- read.csv("GSE34607_BEAS-2B_nonasthma_IndoorExposure_24h_vs_control_24h.csv") #indoorvscontro24h gene results
outdoorcontrol4h <- read.csv("GSE34607_BEAS-2B_nonasthma_OutdoorExposure_4h_vs_control_4h.csv") #outdoorvscontro4h gene report
outdoorcontrol10h <- read.csv("GSE34607_BEAS-2B_nonasthma_OutdoorExposure_10h_vs_control_10h.csv") #outdoorvscontro4=10h gene report
outdoorcontrol24h <- read.csv("GSE34607_BEAS-2B_nonasthma_OutdoorExposure_24h_vs_control_24h.csv") #outdoorvscontro24h gene report
```

```{r pathwayfiles, eval=TRUE, echo=TRUE, message=FALSE, warning=FALSE}
indoorcontrol4h_pathway <- read.csv("GSE34607_BEAS-2B_nonasthma_IndoorExposure_4h_vs_control_4h_fgsea_results.csv") #indoorvscontrol4h pathway report
indoorcontrol10h_pathway <- read.csv("GSE34607_BEAS-2B_nonasthma_IndoorExposure_10h_vs_control_10h_fgsea_results.csv") #indoorvscontrol4=10h pathway report
indoorcontrol24h_pathway <- read.csv("GSE34607_BEAS-2B_nonasthma_IndoorExposure_24h_vs_control_24h_fgsea_results.csv") #indoorvscontrol24h pathway report
outdoorcontrol4h_pathway <- read.csv("GSE34607_BEAS-2B_nonasthma_OutdoorExposure_4h_vs_control_4h_fgsea_results.csv") #outdoorvscontrol4h pathway report
outdoorcontrol10h_pathway <- read.csv("GSE34607_BEAS-2B_nonasthma_OutdoorExposure_10h_vs_control_10h_fgsea_results.csv") #outdoorvscontrol4=10h pathway report
outdoorcontrol24h_pathway <- read.csv("GSE34607_BEAS-2B_nonasthma_OutdoorExposure_24h_vs_control_24h_fgsea_results.csv") #outdoorvscontrol24h pathway report
```

#### Extracting only the main pathways 
The main pathways from the overall files are extracted because there are very similar pathways in the data frame.
The final output is a dataframe that contains 9 columns: 

* pathway - name of the pathway;
* pval - an enrichment p-value;
* padj - an adjusted p-value using Benjamini-Hochberg approach;
* ES - enrichment score;
* NES - enrichment score normalized to mean enrichment of random samples of the same size;
* nMoreExtreme - a number of times a random gene set had a more extreme enrichment score value;
* size - the number of pathway genes in the gene list.
* main_pathway - main pathway
* leadingEdge - genes that drive the enrichment

```{r mainpathway, eval=TRUE, echo=TRUE, message=FALSE, warning=FALSE}
library(dplyr)
indoorcontrol4h_mainpathway <- indoorcontrol4h_pathway %>% filter(main_pathway == "main")
indoorcontrol10h_mainpathway <- indoorcontrol10h_pathway %>% filter(main_pathway == "main")
indoorcontrol24h_mainpathway <- indoorcontrol24h_pathway %>% filter(main_pathway == "main")
outdoorcontrol4h_mainpathway <- outdoorcontrol4h_pathway %>% filter(main_pathway == "main")
outdoorcontrol10h_mainpathway <- outdoorcontrol10h_pathway %>% filter(main_pathway == "main")
outdoorcontrol24h_mainpathway <- outdoorcontrol24h_pathway %>% filter(main_pathway == "main")
detach("package:dplyr")
```


#### Significant genes and pathways

To study the number of significantly expressed genes and pathways, we will exploit the adjusted P values for every treatment time and method. By norms, a P value of less than 0.05 is considered significant. We will hence use this as the bar to filter the data. But before that, let us take a look at the gene data through volcano plots, a common way of summarizing gene expression results, to see if there any significant genes present.  

```{r volcanoplot, eval=TRUE, echo=TRUE, message=FALSE, warning=FALSE}
volcanoplotfunc <- function(df,name){ #defining volcanoplot function
  df$sig <- rep("Insignificant", nrow(df))
  df$sig[which(df$adj.P.Val<0.05)] <- "Significant" #assigning significance column is pval is < 0.05
  volcanoplot <- ggplot(df, aes(x = logFC, y = -log10(adj.P.Val), color = sig)) + 
    geom_point() +
    theme_bw() +
    ggtitle(name) +
    xlab("LogFC")+
    ylab("-Log10(q-value)") +
    scale_color_manual(values = c("black", "red")) +
    theme(legend.position = "none")
  return(volcanoplot)
}
#functioncall for all conditions
ind4hvp <- volcanoplotfunc(indoorcontrol4h, "Indoor vs Control 4h")
ind10hvp <- volcanoplotfunc(indoorcontrol10h, "Indoor vs Control 10h")
ind24hvp <- volcanoplotfunc(indoorcontrol24h, "Indoor vs Control 24h")
out4hvp <- volcanoplotfunc(outdoorcontrol4h, "Outdoor vs Control 4h")
out10hvp <- volcanoplotfunc(outdoorcontrol10h, "Outdoor vs Control 10h")
out24hvp <- volcanoplotfunc(outdoorcontrol24h, "Outdoor vs Control 24h")

ggarrange(ind4hvp, ind10hvp, ind24hvp, out4hvp, out10hvp, out24hvp, ncol = 3, nrow=2) #using ggarrange to display all conditions together
```

The red dots indicate that the genes are significant and black indicates that they are insignificant. As we can see from the volcano plots, except for the 24th hour sample (both indoor and outdoor), all the other samples have significantly expressed genes. Now let us calculate how many genes are significantly expressed. We will also take a look at the number of significantly changed pathways. The results are summarized in a pandoc table.

```{r significantpathway, eval=TRUE, echo=TRUE, message=FALSE, warning=FALSE}
col1<- c("Indoor vs Control - 4h", "Indoor vs Control - 10h", "Indoor vs Control - 24h", "Outdoor vs Control - 4h","Outdoor vs Control - 10h", "Outdoor vs Control - 24h") #list containing the six conditions to be compared
col2 <- c(sum(indoorcontrol4h$adj.P.Val  < 0.05), sum(indoorcontrol10h$adj.P.Val < 0.05), sum(indoorcontrol24h$adj.P.Val < 0.05), sum(outdoorcontrol4h$adj.P.Val  < 0.05), sum(outdoorcontrol10h$adj.P.Val  < 0.05), sum(outdoorcontrol24h$adj.P.Val  < 0.05)) #finding the sum of all significant values of genes for six conditions and storing it in a variable
col3 <- c(sum(indoorcontrol4h_mainpathway$padj  < 0.05), sum(indoorcontrol10h_mainpathway$padj < 0.05), sum(indoorcontrol24h_mainpathway$padj < 0.05), sum(outdoorcontrol4h_mainpathway$padj  < 0.05), sum(outdoorcontrol10h_mainpathway$padj  < 0.05), sum(outdoorcontrol24h_mainpathway$padj  < 0.05)) #finding the sum of all significant values of pathways for six conditions and storing it in a variable
tab <- data.frame(col1, col2, col3) #creating dataframe with condition, number of significant genes and pathways
names(tab) <- c("Comparison", "NO. of significant genes", "NO. of significant main pathways") #assigniing colnames
pandoc.table(tab, style="grid", split.cells = 20, caption="Overview of significant genes and pathways") #displaying as pandoc table
```

### Genes of interest and heirarchial clustering of samples under various conditions

For this step, we are going to try and subset the data frames so that the resultant data frame only contains the required gene symbols. The genes, SERPINB2, CXCL1, ,CXCL6, CXCL8, CYP1A1, CYP1B1, TIPARP and IL6 are found to be significantly altered during the study and the official results can be found in GEO. The gene SERPINB2 is involved in blood coagulation; CXCL6, CXCL1, CXCL8 and IL6 are involved in inflammation; TIPARP, CY1A1 and CY1B1 are involved in xenobiotic metabolizing enzymes and hence chosen for further analysis. A function is created to extract all the relevant genes from the whole data frames. This is followed by summarizing the logFC and QVal values of the genes for all treatment methods and time in a table. 

```{r gene selection, eval=TRUE, echo=TRUE, message=FALSE, warning=FALSE}
library(dplyr)
genefunc <- function(genedf) { #function definition to select genes of interest
genedf <- genedf %>% 
  filter(SYMBOL == "SERPINB2"| SYMBOL == "CXCL1"|SYMBOL == "CXCL6"|SYMBOL == "IL6"|SYMBOL == "CYP1A1"|SYMBOL == "CYP1B1"|SYMBOL == "IL8" | SYMBOL == "CXCL8" | SYMBOL == "TIPARP") %>%
  group_by(SYMBOL) %>% 
    arrange(adj.P.Val) %>% 
    slice(1)
genedf$adj.P.Val <- formatC(genedf$adj.P.Val, format = "e", digits = 2)
genedf$logFC <- round(genedf$logFC, digits = 2)
return(genedf)
}
#function call for all six conditions to extract genes of interest
sel_gene_Ind4h <- genefunc(indoorcontrol4h)
sel_gene_Ind10h <- genefunc(indoorcontrol10h)
sel_gene_Ind24h <- genefunc(indoorcontrol24h)
sel_gene_Out4h <- genefunc(outdoorcontrol4h)
sel_gene_Out10h <- genefunc(outdoorcontrol10h)
sel_gene_Out24h <- genefunc(outdoorcontrol24h)

DE_Req_Genes <- data.frame(sel_gene_Ind4h$SYMBOL,sel_gene_Ind4h$ID, sel_gene_Ind4h$logFC, sel_gene_Ind4h$adj.P.Val, sel_gene_Ind10h$logFC, sel_gene_Ind10h$adj.P.Val, sel_gene_Ind24h$logFC, sel_gene_Ind24h$adj.P.Val, sel_gene_Out4h$logFC, sel_gene_Out4h$adj.P.Val, sel_gene_Out10h$logFC, sel_gene_Out10h$adj.P.Val, sel_gene_Out24h$logFC, sel_gene_Out24h$adj.P.Val) #dataframe with pval and logFC value of selected genes across six conditions

names(DE_Req_Genes) <- c("Gene Symbol", "GeneID", "Indoor4h.logFC", "Indoor4h.QVal","Indoor10h.logFC", "Indoor10h.QVal", "Indoor24h.logFC", "Indoor24h.QVal", "Outdoor4h.logFC", "Outdoor4h.QVal", "Outdoor10h.logFC", "Outdoor10h.QVal", "Outdoor24h.logFC", "Outdoor24h.QVal") #assigning colnames
ft <- DE_Req_Genes %>% as_flextable() #converting the dataframe to flextable
ft <- set_caption(ft,"Summary of genes of interest") #table caption
ft %>% span_header() #spanning table header to make it more intuitive

```

In this step, we will generate a heat map for the top significantly expressed genes. In order to proceed with generating the heat map data, the genes with no symbols should first be removed. The data frame is grouped by symbol and arranged by adj.P.Val and abs(logFC). Then, the most significantly expressed gene (lowest P Value) is sliced to generate a new dataframe.

```{r genearrangement, eval=T, echo=TRUE, message=FALSE, warning=FALSE}
library(dplyr)
#dataframe function to clean gene symbols with duplicates and NA values
genearrangefunc <- function(genedf) {
  genedf %>%
    dplyr::filter(SYMBOL!="NA") %>% #removing genes assigned as NA
    dplyr::filter(!is.na(SYMBOL)) %>% #removing genes with empty values
    dplyr::group_by(SYMBOL) %>% #grouping by gene symbol
    dplyr::arrange(adj.P.Val, -abs(logFC)) %>% #arranging by P-Val and logFC
    dplyr::filter(row_number()==1) %>% #slicing the first row of every group
    data.frame()
}
#function call for six conditions to fextract genes with low P-Val
indoorcontrol4h.arranged <- genearrangefunc(indoorcontrol4h)
indoorcontrol10h.arranged <- genearrangefunc(indoorcontrol10h)
indoorcontrol24h.arranged <- genearrangefunc(indoorcontrol24h)
outdoorcontrol4h.arranged <- genearrangefunc(outdoorcontrol4h)
outdoorcontrol10h.arranged <- genearrangefunc(outdoorcontrol10h)
outdoorcontrol24h.arranged <- genearrangefunc(outdoorcontrol24h)

```

The top genes are selected by the P value and logFC. The general norm is to select 20-50 genes for clustering with 2.5 fold change in logFC. Trial and error gives this results when logFC is greater than or equal to 1.3. Hence, we are filtering the genes with adj.P.Val < 0.05 and logFC >= 1.3. From the volcano plots, it is evident that the indoor vs control - 4h and 10h have the most significantly expressed genes. We will select the unique significantly expressed genes from these two comparisons for further analysis.

```{r genesubsetting, eval=T, echo=TRUE, message=FALSE, warning=FALSE}
#only genes with pval <0.05 and logFC >=1.3 are extracted from the grouped genes
indoorcontrol4h.gene <- indoorcontrol4h.arranged$SYMBOL[indoorcontrol4h.arranged$adj.P.Val<0.05&abs(indoorcontrol4h.arranged$logFC)>=1.3]
indoorcontrol10h.gene <- indoorcontrol10h.arranged$SYMBOL[indoorcontrol10h.arranged$adj.P.Val<0.05&abs(indoorcontrol10h.arranged$logFC)>=1.3]
indoorcontrol24h.gene <- indoorcontrol24h.arranged$SYMBOL[indoorcontrol24h.arranged$adj.P.Val<0.05&abs(indoorcontrol24h.arranged$logFC)>=1.3]
outdoorcontrol4h.gene <- outdoorcontrol4h.arranged$SYMBOL[outdoorcontrol4h.arranged$adj.P.Val<0.05&abs(outdoorcontrol4h.arranged$logFC)>=1.3]
outdoorcontrol10h.gene <- outdoorcontrol10h.arranged$SYMBOL[outdoorcontrol10h$adj.P.Val<0.05&abs(outdoorcontrol10h$logFC)>=1.3]
outdoorcontrol24h.gene <- outdoorcontrol24h.arranged$SYMBOL[outdoorcontrol24h$adj.P.Val<0.05&abs(outdoorcontrol24h$logFC)>=1.3]
sel.gene <- unique(c(indoorcontrol4h.gene, indoorcontrol10h.gene)) #selecting unique values in indvscon4h and indvscon10h has they have the most number of significant genes
```

For the selected genes, corresponding probe IDs are extracted.

```{r eval=T}
# get probe id
genedata.sel <- rbind(indoorcontrol4h.arranged, indoorcontrol10h.arranged) %>% #bind indvscon - 4h and 10h results
  dplyr::filter(SYMBOL%in%sel.gene) %>% #extract the unique values
  dplyr::group_by(SYMBOL) %>% #groupby gene symbol
  dplyr::arrange(adj.P.Val, -abs(logFC)) %>% #arrange by pval and logFC
  dplyr::filter(row_number()==1) %>% #slice the first gene of the group
  dplyr::select(ID, SYMBOL) %>% #select gene ID and corresponding symbol
  data.frame() %>%
  unique()
```

The normalized raw data expression is used as the heat map data. The expression data of the selected genes are stored in a variable "data.rma.sel". The row and column names are assigned as per symbols and treatment methods. 

```{r eval=T}
data.rma.sel <- exprs(raw.data.rma) #extracting the expression of normalized raw data
data.rma.sel <- data.rma.sel[row.names(data.rma.sel)%in%genedata.sel$ID,] #storing the expression data of selected genes
row.names(data.rma.sel) <- unname(sapply(row.names(data.rma.sel), function(x){genedata.sel$SYMBOL[genedata.sel$ID==x]})) #assigning rownames for expression values from selected genes
colnames(data.rma.sel) <- unname(sapply(colnames(data.rma.sel), function(x){pheno$Treatment[pheno$Filename==x]})) #assigning colnames for expression values from filename in pheno dataframe
```

The heat map is created using the gplot library. A dendrogram is added at the top to distinguish how well the groups are clustered between treatment and treatment time. 

```{r eval=T}
status=factor(pheno$Treatment) #converting treatment to status
color.status = colours[c(1:nlevels(status))] 
names(color.status) = levels(status) #assigning and naming colours
color.status.list <- unlist(lapply(colnames(data.rma.sel),function(x){color.status[x]}))
colnames(data.rma.sel) <- gsub("Exposure","",colnames(data.rma.sel)) #shortening filename
#heatmap function
heatmap.2(data.rma.sel, scale = "none", col=viridis(256, option="B"),
          ColSideColors=color.status.list,
          labCol=colnames(data.rma.sel), labRow=row.names(data.rma.sel), 
          trace = "none", 
          density.info = "none", 
          main="Gene Heat Map",
          key=TRUE, keysize=1.5,key.xlab="Gene Expression Values")
legend("bottomleft",legend=names(color.status),fill=color.status,cex=0.6)
```

### Gene-set Enrichment Analysis

The selected genes are strongly associated with the below pathways (commonly altered in allergy). These pathways are selected from all comparisons for pathway studies.
1. KEGG_CYTOKINE_CYTOKINE_RECEPTOR_INTERACTION
2. KEGG_CHEMOKINE_SIGNALING_PATHWAY
3. KEGG_EPITHELIAL_CELL_SIGNALING_IN_HELICOBACTER_PYLORI_INFECTION
4. REACTOME_G_ALPHA_I_SIGNALLING_EVENTS
5. REACTOME_SIGNALING_BY_INTERLEUKINS
6. REACTOME_GPCR_LIGAND_BINDING
7. REACTOME_CYTOCHROME_P450_ARRANGED_BY_SUBSTRATE_TYPE
8. KEGG_METABOLISM_OF_XENOBIOTICS_BY_CYTOCHROME_P450
9. KEGG_DNA_REPLICATION
10. REACTOME_ACTIVATION_OF_THE_PRE_REPLICATIVE_COMPLEX
11. REACTOME_CELL_CYCLE_CHECKPOINTS


```{r select pathway function, eval=TRUE, echo=TRUE, message=FALSE, warning=FALSE}
pathwayfunc <- function(df) { #function to extract the required pathways
df <- df %>% filter(pathway == "KEGG_CYTOKINE_CYTOKINE_RECEPTOR_INTERACTION"| pathway == "KEGG_CHEMOKINE_SIGNALING_PATHWAY"| pathway == "KEGG_EPITHELIAL_CELL_SIGNALING_IN_HELICOBACTER_PYLORI_INFECTION"| pathway == "REACTOME_G_ALPHA_I_SIGNALLING_EVENTS"|pathway == "REACTOME_SIGNALING_BY_INTERLEUKINS"|pathway == "REACTOME_GPCR_LIGAND_BINDING"|pathway == "REACTOME_CYTOCHROME_P450_ARRANGED_BY_SUBSTRATE_TYPE"|
pathway == "KEGG_METABOLISM_OF_XENOBIOTICS_BY_CYTOCHROME_P450"|pathway == "KEGG_DNA_REPLICATION"| pathway == "REACTOME_ACTIVATION_OF_THE_PRE_REPLICATIVE_COMPLEX"| pathway == "REACTOME_CELL_CYCLE_CHECKPOINTS")
df$padj <- formatC(df$padj, format = "e", digits = 2) #converting to scientific format
df$NES <- round(df$NES, digits = 2) #rounding off to two digits
return(df)
}
sel_pathway_Ind4h <- pathwayfunc(indoorcontrol4h_pathway) #extracting the required pathways for indoor vs control 4h
sel_pathway_Ind10h <- pathwayfunc(indoorcontrol10h_pathway) #extracting the required pathways for indoor vs control 10h
sel_pathway_Ind24h <- pathwayfunc(indoorcontrol24h_pathway) #extracting the required pathways for indoor vs control 24h
sel_pathway_Out4h <- pathwayfunc(outdoorcontrol4h_pathway) #extracting the required pathways for outdoor vs control 4h
sel_pathway_Out10h <- pathwayfunc(outdoorcontrol10h_pathway) #extracting the required pathways for outdoor vs control 10h
sel_pathway_Out24h <- pathwayfunc(outdoorcontrol24h_pathway) #extracting the required pathways for outdoor vs control 24h
```

#### NES vs Pathway Table
For the selected pathways, the NES and padj values of all different combinations are pulled into a flextable for comparison. 

```{r NES-Pathway table, eval=TRUE, echo=TRUE, message=FALSE, warning=FALSE}
FGSEA_Req_Pathway <- data.frame(sel_pathway_Ind4h$pathway, sel_pathway_Ind4h$padj, sel_pathway_Ind4h$NES,sel_pathway_Ind10h$padj, sel_pathway_Ind10h$NES,sel_pathway_Ind24h$padj, sel_pathway_Ind24h$NES, sel_pathway_Out4h$padj, sel_pathway_Out4h$NES,sel_pathway_Out10h$padj, sel_pathway_Out10h$NES, sel_pathway_Out24h$padj, sel_pathway_Out24h$NES) #extracting padj and NES values of all conditions of the pathways to a new dataframe

names(FGSEA_Req_Pathway) <- c("Pathway", "Indoor4h.padj", "Indoor4h.NES","Indoor10h.padj", "Indoor10h.NES", "Indoor24h.padj", "Indoor24h.NES", "Outdoor4h.padj", "Outdoor4h.NES", "Outdoor10h.padj", "Outdoor10h.NES", "Outdoor24h.padj", "Outdoor24h.NES") #assigning colnames

FGSEA_Req_Pathway_Table <- FGSEA_Req_Pathway %>% as_flextable() #extracting the dataframe to a flex table
FGSEA_Req_Pathway_Table <- set_caption(FGSEA_Req_Pathway_Table,"Summary of pathways of interest") #assigning caption
FGSEA_Req_Pathway_Table %>% span_header() #spanning the flex table over header to merge the treatment headers
```

#### NES vs Pathway Barplot
For the selected pathways, the NES and padj values of all different combinations are plotted into a barplot for more detailed comparison. The significantly altered pathways are marked with * in the barplot.

```{r plot relevant pathways, eval=TRUE, echo=TRUE, message=FALSE, warning=FALSE}
pathwaylabels = c("K: DNA Replication", "K:Cytokine-Cytokine Receptor Interaction", "K:Chemokine Signalling Pathway", "K:Epithelial Cell Signalling (in Helicobacter pylori infection)", "R:G-Alpha1 Signalling Events", "R:Signalling by Interleukins", "R:GPCR Ligand Binding", "R: Activation of Pre-Replicative Complex", "R: Cell Cycle Checkpoints", "R:Cytochrome P450 (arranged by susbtrate type)", "K:Metabolism of Xenobiotics by Cytochrome P450" ) #the required pathway names are simplified and stored in a label variable

df.nes <- data.frame(Pathway = pathwaylabels, Indoor4h = FGSEA_Req_Pathway$Indoor4h.NES, Outdoor4h = FGSEA_Req_Pathway$Outdoor4h.NES, Indoor10h = FGSEA_Req_Pathway$Indoor10h.NES, Outdoor10h = FGSEA_Req_Pathway$Outdoor10h.NES, Indoor24h = FGSEA_Req_Pathway$Indoor24h.NES, Outdoor24h = FGSEA_Req_Pathway$Outdoor24h.NES) #NES values of the corresponding pathways are extracted
long <- df.nes %>% gather("Condition", "NES", -Pathway) #the dataframe is converted to long form

df.padj <- data.frame(Pathway = pathwaylabels, Indoor4h = as.numeric(FGSEA_Req_Pathway$Indoor4h.padj), Outdoor4h = as.numeric(FGSEA_Req_Pathway$Outdoor4h.padj), Indoor10h = as.numeric(FGSEA_Req_Pathway$Indoor10h.padj), Outdoor10h = as.numeric(FGSEA_Req_Pathway$Outdoor10h.padj), Indoor24h = as.numeric(FGSEA_Req_Pathway$Indoor24h.padj), Outdoor24h = as.numeric(FGSEA_Req_Pathway$Outdoor24h.padj))  #padj values of the corresponding pathways are extracted and stored in a dataframe
long_padj <- df.padj %>% gather("Condition", "PValue", -Pathway) #the padj dataframe is converted to long form
long$Condition <- c(rep("Indoor", 11), rep("Outdoor", 11), rep("Indoor", 11), rep("Outdoor", 11), rep("Indoor", 11), rep("Outdoor", 11)) #converting the condition variable to a more readable format
long$adjpval <- long_padj$PValue #padj value is added to long 
long$significance <- ifelse(long_padj$PValue < 0.05, "*", NA) #significance is assigned * to pathways corresponding with p values < 0.05
long$time <- c(rep("4h", 22), rep("10h", 22), rep("24h", 22)) #time column is included

pathway_barplot <- ggplot(long, aes(x = Pathway, y = NES, fill = Condition)) +
        geom_bar(width=0.8, position="dodge", stat="identity") +
        facet_wrap(~factor(time, levels=c('4h','10h','24h'))) +
        coord_flip() +
        xlab("KEGG (K) & Reactome (R) Pathways") +
        ylab("Normalized Enrichment Score") +
        theme_bw()+
        theme(
          axis.text.y=element_text(size=8),
          axis.text.x=element_text(size=8),
          axis.title.y=element_text(size=11),
          axis.title.x=element_text(size=11)) 
        
pathway_barplot + geom_text(label=long$significance, size=4, position = position_dodge(width = 1)) + ggtitle("Changes in Enrichment Score under\n Various Exposure Conditions")  #including text to mark the significance and title
```



## Conclusion

Overall, this dataset's study contributed to a better understanding of how air contaminants affect gene transcription in airway cells. More genes have their expression values changed, showing that indoor PM10 affects more genes than outdoor PM10. This is demonstrated by the fact that over 70 genes are considerably affected in indoor exposure but only about 20 genes are significantly altered in outdoor exposure. With 79 and 76 substantially differentially expressed genes, respectively, the indoor 4h and indoor 10h samples are highly comparable. Indoor 4h and 10h expression changes are quite similar, as evidenced by heatmap clustering, which shows that many 4h and 10h genes are grouped together and have distinct expression patterns. The results also show that no major genes are changed in samples incubated for 24 hours in both indoor and outdoor environments. This could be because cells are no longer responsive to PM10 after 24 hours of exposure. These findings can therefore aid in the design of future in-vitro PM10 exposure trials. Although cell lines are useful for studying the pathways that cause PM10 exposure, they are not identical to human samples. In the future, studies of PM10 exposure should focus on human samples.

The gene expression results also confirm that indoor PM10 induced a blood coagulation gene SERPINB2 and inflammatory genes like CXCL6, CXCL1, IL6, IL8.Outdoor PM10 induced xenobiotic metabolizing enzymes CYP1A1, CYP1B1 and TIPARP PM10.  These results also falls in line with the results published in GEO for the study GEO34607. With this, we conclude that PM10 can actually induce inflammation of cells. These genes are commonly altered under exposure to inflammatory agents (like cigarette smoke particles) suggesting that exposure to PM10 particles can have the same effects of cigarette exposure thereby leading to inflammatory response in airway cells.

When cell lines are exposed to PM10 particles indoors, the effects are more pronounced and distinct from what occurs when they are exposed outside. This shows that interior pollution is worse than outside pollution, and that those who spend most of their time indoors, including students, require attention. Air filters in the school, regular change of class rooms for pupils, and ventilation are all possible remedies. Furthermore, recommending that students with asthma and other respiratory illnesses take preventative medicine might assist to mitigate the impacts of PM on their respiratory health.


## References

1. Annesi-Maesano, I., Baiz, N., Banerjee, S., Rudnai, P., Rive, S., & Group, on behalf of the S. (2013). Indoor Air Quality and Sources in Schools and Related Health Effects. Http://Dx.Doi.Org/10.1080/10937404.2013.853609, 16(8), 491–550. https://doi.org/10.1080/10937404.2013.853609 
2. Athanasios Valavanidis, Konstantinos Fiotakis, & Thomais Vlachogianni (2008). Airborne particulate matter and human health: toxicological assessment and importance of size and composition of particles for oxidative damage and carcinogenic mechanisms. Journal of Environmental Science and Health. Part C, Environmental Carcinogenesis & Ecotoxicology Reviews, 26(4), 339–362. https://doi.org/10.1080/10590500802494538
3. Breysse, P. N., Diette, G. B., Matsui, E. C., Butz, A. M., Hansel, N. N., & McCormack, M. C. (2010). Indoor Air Pollution and Asthma in Children. Proceedings of the American Thoracic Society, 7(2), 102. https://doi.org/10.1513/PATS.200908-083RM
4. de Gennaro, G., Dambruoso, P. R., Loiotile, A. D., di Gilio, A., Giungato, P., Tutino, M., Marzocca, A., Mazzone, A., Palmisani, J., & Porcelli, F. (2014). Indoor air quality in schools. Environmental Chemistry Letters 2014 12:4, 12(4), 467–482. https://doi.org/10.1007/S10311-014-0470-6
5. GitHub - HimesGroup/raved: Reproducible Analysis and Validation of Expression Data. (n.d.). Retrieved December 7, 2021, from https://github.com/HimesGroup/raved 
5. Indoor Particulate Matter | US EPA. (n.d.). Retrieved December 7, 2021, from https://www.epa.gov/indoor-air-quality-iaq/indoor-particulate-matter
6. Jovanović, M., Vučićević, B., Turanjanin, V., Živković, M., & Spasojević, V. (2014). Investigation of indoor and outdoor air quality of the classrooms at a school in Serbia. Energy, 77, 42–48. https://doi.org/10.1016/J.ENERGY.2014.03.080
7. Kan, M., Shumyatcher, M., Diwadkar, A., Soliman, G., & Himes, B. E. (2018). Integration of Transcriptomic Data Identifies Global and Cell-Specific Asthma-Related Gene Expression Signatures. AMIA Annual Symposium Proceedings, 2018, 1338. /pmc/articles/PMC6371257/ 
8. GEO Accession viewer. (n.d.). Retrieved November 4, 2021, from https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE34607
9. UtahAir - Particulate Matter. Retrieved December 7, 2021, from https://health.utah.gov/utahair/pollutants/PM/#gsc.tab=0 
10. World Health Organizatio.. Air Pollution. Retrieved December 7, 2021, from https://www.who.int/health-topics/air-pollution#tab=tab_1 

##Session Info

```{r, eval = TRUE}
sessionInfo()
```
