---
title: "Characterization of Castleman Disease Treatment Response"
author: "Sheila Pierson"
output: 
  html_document:
    toc: false 
    depth: 3 
    theme: paper 
    highlight: tango
---
```{r set-options, echo=FALSE, cache=FALSE}
options(width = 400)
```  
***
Use this template to complete your project throughout the course. Your Final Project presentation in class will be based on the contents of this document. Replace the title/name and text below with your own, but leave the headers.

### Overview
Castleman disease is a rare inflammatory disorder. The most-deadly subtype (idiopathic multicentric Castleman disease (iMCD)) has a 35% mortality rate. Siltuximab is the only FDA-approved drug to treat iMCD, and nearly 60% of patients do not respond to siltuximab. Using data obtained from the siltuximab trial and from the first-ever Castleman disease natural history registry, ACCLERATE, this project aims to identify the characteristics of siltuximab responders vs. non-responders.  
https://github.com/skpierson/BMIN503_Final_Project

### Introduction 
Castleman disease consists of a group of three heterogeneous diseases, characterized by common lymph node features.  The three subtypes include unicentric Castleman disease (UCD), HHV8-associated multicentric (HHV8+MCD), and idiopathic multicentric Castleman disease (iMCD). UCD is often curative (95%) by lymph node excision. Caused by the HHV8 virus, HHV8+MCD is well controlled with rituximab and anti-viral medications.  However, iMCD represents the most deadly subtype, with 35% 5-year mortality rate.  Though there is one FDA-approved drug for iMCD (siltuximab), only ~30-40% patients respond positively. In order to appropriately treat patients, it is crucial to understand why a subset of patients responds to siltuximab while others do not. 

Understanding the patient population and characteristics of treatment responders vs. non-responders requires a multidisciplinary approach. As its name suggests, iMCD is idiopathic and understanding the cause requires a multidisciplinary approuch. Proteomics can be used to characterize the protein state of patients in flare compared to remission and to understand if there are proteomic changes specific to any certain subgroup. The study of gene mutations will help to understand if iMCD is caused by an underlying genetic disorder. Whole genome or whole exome sequencing can be used to determine if Castleman disease is caused by a somatic mutation.  In addition to determining the underlying cause, is crucial to understand the patient's clinical presentation. We currently have available a rich set of clinical data from the siltuximab trial and are actively collecting data in the first ever patient registry.  First, the siltuximab-trial data will be used to develop a decision tree to identify responders vs. non-responders.  The decision tree algorithm will then be validated by data collected in the ACCELERATE registry. Bill LaCava advised on the development of decision tree algorithm - specifically he provided advice on restricting to certain data collection points.  In addition, regression models will be run on the ACCELERATE data in order to test the association between additional predictors and the response outcome.  Rebecca Hubbard advised on the development of regression models. Specifically, we discussed the non-standard time-points and data collection and how to develop a mixed-effects model. Using heatmaps, clinical, laboratory, and pathological features of Castleman disease patients will be graphically presented. Jason Moore advised on developing heatmaps looking at characteristics on one axis, patients on the other axis, and using hierarchical clustering to try and identify subgroups. 

### Methods
In the first paragraph, describe the data used and general methodological approach. Subsequently, incorporate full R code necessary to retrieve and clean data, and perform analysis. Be sure to include a description of code so that others (including your future self) can understand what you are doing and why. 


### Results
Describe your results and include relevant tables, plots, and code/comments used to obtain them. End with a brief conclusion of your findings related to the question you set out to address. You can include references if you'd like, but this is not required.
