---
title: "BMIN503/EPID600 Vaccination Trends and PBMC Gene Expression Responses to Ebola Vaccination"
author: "David Mai"
output: 
  html_document:
    toc: false 
    depth: 3 
    theme: paper 
    highlight: tango
---
```{r set-options, echo=FALSE, cache=FALSE}
options(width = 400)
```  
***

### Overview

The invention of vaccines has arguably changed the course of human history by drastically decreasing the incidence of many infectious diseases, saving countless lives. However, recent years have seen an increase in reported cases of diseases thought to have been effectively eliminated since the advent and implementation of vaccination. This project seeks to analyze historic and recent trends between vaccination coverage and disease incidence across various infectious diseases and explore the genetic profile of immune cells as a response against vaccination using Ebola as a case study to highlight factors associated with immunity.

Follow the link [here][repository] to my final project GitHub repository.

<!-- Links -->
[repository]: https://github.com/davemai/BMIN503_Final_Project

### Introduction

The first reported vaccine in the western world was invented by Edward Jenner, who observed that milkmaids did not suffer the severe disfiguring seen by others during the contraction of smallpox. It turned out that milkmaids contracted cowpox from contact with the cowâ€™s udders, leading to a milder form of the pox that provided protection against smallpox. Since then, the concept of vaccination has been applied to many different diseases such as measles, mumps, and polio, effectively eliminating them from the western world. However, recent trends show an increase in parents refusing to vaccinate their children due to religious reasons or the belief that vaccines cause impairment. This trend coincides with the resurfacing of diseases thought to be long eliminated since the implementation of comprehensive vaccination policies.

This problem can be best approached with interdisciplinary expertise from clinicians, public health experts, as well as health surveillance officials. However, stakeholders in social economics, vaccine developers, and public health workers would benefit from increased understanding about why particular diseases are resurfacing. I am using datasets provided from the WHO and unicef for vaccination coverage and disease incidence by country over time (1980s - 2018). Their datasets also cover various vaccine antigens on a per country basis. Analysis of these datasets will lead to a greater understanding of how vaccines contributed to the precipitous decline in once-dreaded diseases, and hopefully insight into their recent recurrence. Additionally, I take Ebola vaccination data (sequencing and microwell data of PBMCs) as a case study to highlight what genetic factors are upregulated upon vaccination and identify key differences that lead to effective immunity.


### Methods
In the first paragraph, describe the data used and general methodological approach. Subsequently, incorporate full R code necessary to retrieve and clean data, and perform analysis. Be sure to include a description of code so that others (including your future self) can understand what you are doing and why. 

This project retrieves longitudinal data from WHO about vaccine coverage for various infectious diseases as well as the incidence of those disease on a per country basis up to 2018. We will reformat and parse through this data so that we can plot longitudinal trends about worldwide coverage and incidence of each disease, and then break down into specific country trends to analyze whether a particular behavior is significant or consistent across all countries. We will also take microarray data from a subject infected with ebola to investigate what factors are significantly or differentially upregulated upon ebola pathogenesis. 

Package and library installation:
```{r, eval = TRUE, message = FALSE}

library(tidyverse)
library(dplyr)
library(ggplot2)
library(ggpubr)
library(rlist)
library(rapport)
library(tmap)
theme_set(theme_pubr())

```

Import datasets:
```{r, eval = TRUE, message = FALSE}

# Dataset on vaccination coverage
coverage_series <- read.csv(url("https://raw.githubusercontent.com/davemai/BMIN503_Final_Project/master/data/coverage_series.csv"), header = TRUE)
colnames(coverage_series)[which(names(coverage_series) == "Percent_covrage")] <- "Percent_coverage"

# Datasets on various disease incidences
crs_incidence <- read.csv(url("https://raw.githubusercontent.com/davemai/BMIN503_Final_Project/master/data/crs_incidence.csv"), header = TRUE)
diptheria_incidence <- read.csv(url("https://raw.githubusercontent.com/davemai/BMIN503_Final_Project/master/data/diptheria_incidence.csv"), header = TRUE)
japenc_incidence <- read.csv(url("https://raw.githubusercontent.com/davemai/BMIN503_Final_Project/master/data/japenc_incidence.csv"), header = TRUE)
measles_incidence <- read.csv(url("https://raw.githubusercontent.com/davemai/BMIN503_Final_Project/master/data/measles_incidence.csv"), header = TRUE)
mumps_incidence <- read.csv(url("https://raw.githubusercontent.com/davemai/BMIN503_Final_Project/master/data/mumps_incidence.csv"), header = TRUE)
ntetanus_incidence <- read.csv(url("https://raw.githubusercontent.com/davemai/BMIN503_Final_Project/master/data/ntetanus_incidence.csv"), header = TRUE)
pertussis_incidence <- read.csv(url("https://raw.githubusercontent.com/davemai/BMIN503_Final_Project/master/data/pertussis_incidence.csv"), header = TRUE)
polio_incidence <- read.csv(url("https://raw.githubusercontent.com/davemai/BMIN503_Final_Project/master/data/polio_incidence.csv"), header = TRUE)
rubella_incidence <- read.csv(url("https://raw.githubusercontent.com/davemai/BMIN503_Final_Project/master/data/rubella_incidence.csv"), header = TRUE)
ttetanus_incidence <- read.csv(url("https://raw.githubusercontent.com/davemai/BMIN503_Final_Project/master/data/ttetanus_incidence.csv"), header = TRUE)
yfever_incidence <- read.csv(url("https://raw.githubusercontent.com/davemai/BMIN503_Final_Project/master/data/yfever_incidence.csv"), header = TRUE)
regional_global_incidence <- read.csv(url("https://raw.githubusercontent.com/davemai/BMIN503_Final_Project/master/data/regional_global_incidence.csv"), header = TRUE)

# Datasets on Ebola factors
# https://www.ncbi.nlm.nih.gov/pmc/articles/PMC5011782/
# https://www.ncbi.nlm.nih.gov/pmc/articles/PMC5497219/

```

Clean datasets for analysis:
```{r, eval = TRUE, message = FALSE}

# Remove extraneous columns for graphing
  # CRS
crs_incidence_trimmed <- select(crs_incidence, c("X2018", "X2017", "X2016", "X2015", "X2014", "X2013", "X2012", "X2011", "X2010", "X2009", "X2008", "X2007", "X2006", "X2005", "X2004", "X2003", "X2002", "X2001", "X2000", "X1999", "X1998"))
years_crs <- str_remove(colnames(crs_incidence_trimmed), "X")

  # Diptheria
dip_incidence_trimmed <- select(diptheria_incidence, c("X2018", "X2017", "X2016", "X2015", "X2014", "X2013", "X2012", "X2011", "X2010", "X2009", "X2008", "X2007", "X2006", "X2005", "X2004", "X2003", "X2002", "X2001", "X2000", "X1999", "X1998", "X1997", "X1996", "X1995", "X1994", "X1993", "X1992", "X1991", "X1990", "X1989", "X1988", "X1987", "X1986", "X1985", "X1984", "X1983", "X1982", "X1981", "X1980"))
years_dip <- str_remove(colnames(dip_incidence_trimmed), "X")

  # Japanese Encephalitis
japenc_incidence_trimmed <- select(japenc_incidence, c("X2018", "X2017", "X2016", "X2015", "X2014", "X2013", "X2012", "X2011", "X2010", "X2009", "X2008", "X2007", "X2006"))
years_japenc <- str_remove(colnames(japenc_incidence_trimmed), "X")

  # Measles
measles_incidence_trimmed <- select(measles_incidence, c("X2018", "X2017", "X2016", "X2015", "X2014", "X2013", "X2012", "X2011", "X2010", "X2009", "X2008", "X2007", "X2006", "X2005", "X2004", "X2003", "X2002", "X2001", "X2000", "X1999", "X1998", "X1997", "X1996", "X1995", "X1994", "X1993", "X1992", "X1991", "X1990", "X1989", "X1988", "X1987", "X1986", "X1985", "X1984", "X1983", "X1982", "X1981", "X1980"))
years_measles <- str_remove(colnames(measles_incidence_trimmed), "X")

  # Mumps
mumps_incidence_trimmed <- select(mumps_incidence, c("X2018", "X2017", "X2016", "X2015", "X2014", "X2013", "X2012", "X2011", "X2010", "X2009", "X2008", "X2007", "X2006", "X2005", "X2004", "X2003", "X2002", "X2001", "X2000", "X1999", "X1998"))
years_mumps <- str_remove(colnames(mumps_incidence_trimmed), "X")

  # Neonatal Tetanus
ntet_incidence_trimmed <- select(ntetanus_incidence, c("X2018", "X2017", "X2016", "X2015", "X2014", "X2013", "X2012", "X2011", "X2010", "X2009", "X2008", "X2007", "X2006", "X2005", "X2004", "X2003", "X2002", "X2001", "X2000", "X1999", "X1998", "X1997", "X1996", "X1995", "X1994", "X1993", "X1992", "X1991", "X1990", "X1989", "X1988", "X1987", "X1986", "X1985", "X1984", "X1983", "X1982", "X1981", "X1980"))
years_ntet <- str_remove(colnames(ntet_incidence_trimmed), "X")

  # Pertussis
pertussis_incidence_trimmed <- select(pertussis_incidence, c("X2018", "X2017", "X2016", "X2015", "X2014", "X2013", "X2012", "X2011", "X2010", "X2009", "X2008", "X2007", "X2006", "X2005", "X2004", "X2003", "X2002", "X2001", "X2000", "X1999", "X1998", "X1997", "X1996", "X1995", "X1994", "X1993", "X1992", "X1991", "X1990", "X1989", "X1988", "X1987", "X1986", "X1985", "X1984", "X1983", "X1982", "X1981", "X1980"))
years_pertussis <- str_remove(colnames(pertussis_incidence_trimmed), "X")

  # Polio
polio_incidence_trimmed <- select(polio_incidence, c("X2018", "X2017", "X2016", "X2015", "X2014", "X2013", "X2012", "X2011", "X2010", "X2009", "X2008", "X2007", "X2006", "X2005", "X2004", "X2003", "X2002", "X2001", "X2000", "X1999", "X1998", "X1997", "X1996", "X1995", "X1994", "X1993", "X1992", "X1991", "X1990", "X1989", "X1988", "X1987", "X1986", "X1985", "X1984", "X1983", "X1982", "X1981", "X1980"))
years_polio <- str_remove(colnames(polio_incidence_trimmed), "X")

  # Rubella
rubella_incidence_trimmed <- select(rubella_incidence, c("X2018", "X2017", "X2016", "X2015", "X2014", "X2013", "X2012", "X2011", "X2010", "X2009", "X2008", "X2007", "X2006", "X2005", "X2004", "X2003", "X2002", "X2001", "X2000", "X1999", "X1998"))
years_rubella <- str_remove(colnames(rubella_incidence_trimmed), "X")

  # TTetanus
ttet_incidence_trimmed <- select(ttetanus_incidence, c("X2018", "X2017", "X2016", "X2015", "X2014", "X2013", "X2012", "X2011", "X2010", "X2009", "X2008", "X2007", "X2006", "X2005", "X2004", "X2003", "X2002", "X2001", "X2000", "X1999", "X1998", "X1997", "X1996", "X1995", "X1994", "X1993", "X1992", "X1991", "X1990", "X1989", "X1988", "X1987", "X1986", "X1985", "X1984", "X1983", "X1982", "X1981", "X1980"))
years_ttet <- str_remove(colnames(ttet_incidence_trimmed), "X")

  # Yellow Fever
yfever_incidence_trimmed <- select(yfever_incidence, c("X2018", "X2017", "X2016", "X2015", "X2014", "X2013", "X2012", "X2011", "X2010", "X2009", "X2008", "X2007", "X2006", "X2005", "X2004", "X2003", "X2002", "X2001", "X2000", "X1999", "X1998", "X1997", "X1996", "X1995", "X1994", "X1993", "X1992", "X1991", "X1990", "X1989", "X1988", "X1987", "X1986", "X1985", "X1984", "X1983", "X1982", "X1981", "X1980"))
years_yfever <- str_remove(colnames(yfever_incidence_trimmed), "X")


# Reformat dataframe into columns by year and total incidence
  # CRS
total_crs_incidence_by_year <- data.frame("Year" = years_crs, "Incidence" = colSums(crs_incidence_trimmed, na.rm = TRUE))

  # Diptheria
total_dip_incidence_by_year <- data.frame("Year" = years_dip, "Incidence" = colSums(dip_incidence_trimmed, na.rm = TRUE))

  # Japanese Encephalitis
total_japenc_incidence_by_year <- data.frame("Year" = years_japenc, "Incidence" = colSums(japenc_incidence_trimmed, na.rm = TRUE))

  # Measles
total_measles_incidence_by_year <- data.frame("Year" = years_measles, "Incidence" = colSums(measles_incidence_trimmed, na.rm = TRUE))

  # Mumps
total_mumps_incidence_by_year <- data.frame("Year" = years_mumps, "Incidence" = colSums(mumps_incidence_trimmed, na.rm = TRUE))

  # Neonatal Tetanus
total_ntet_incidence_by_year <- data.frame("Year" = years_ntet, "Incidence" = colSums(ntet_incidence_trimmed, na.rm = TRUE))

  # Pertussis
total_pertussis_incidence_by_year <- data.frame("Year" = years_pertussis, "Incidence" = colSums(pertussis_incidence_trimmed, na.rm = TRUE))

  # Polio
total_polio_incidence_by_year <- data.frame("Year" = years_polio, "Incidence" = colSums(polio_incidence_trimmed, na.rm = TRUE))

  # Rubella
total_rubella_incidence_by_year <- data.frame("Year" = years_rubella, "Incidence" = colSums(rubella_incidence_trimmed, na.rm = TRUE))

  # TTetanus
total_ttet_incidence_by_year <- data.frame("Year" = years_ttet, "Incidence" = colSums(ttet_incidence_trimmed, na.rm = TRUE))

  # Yellow Fever
total_yfever_incidence_by_year <- data.frame("Year" = years_yfever, "Incidence" = colSums(yfever_incidence_trimmed, na.rm = TRUE))

```

### Results
Describe your results and include relevant tables, plots, and code/comments used to obtain them. End with a brief conclusion of your findings related to the question you set out to address. You can include references if you'd like, but this is not required.

**As expected, worldwide incidence of each of these infectious days have declined to stably low levels as vaccination coverage has increased and maintained a steadily high coverage rate. However, upon taking a closer look at specific countries, we see that there are actually ones that have seen an increased incidence, owing to a coverage rate that may not be sufficiently high.**

```{r, eval = TRUE, message = FALSE}

# Wrapper function for title fitting
wrapper <- function(x, ...) 
{
  paste(strwrap(x, ...), collapse = "\n")
}

# Plots of reported worldwide incidence over time
  # CRS
crs <- ggplot(total_crs_incidence_by_year, aes(Year, Incidence)) +
  geom_line(size = 1, linetype = 2) +
  geom_point(shape = 4, size = 2) +
  scale_color_manual(values = c("black", "gray40")) +
  theme_bw() +
  geom_smooth() +
  ggtitle(wrapper(paste("Reported worldwide incidence of congenital rubella syndrome from ", min(years_crs), "to ", max(years_crs)), width = 35)) +
theme(plot.title = element_text(hjust = 0),
      axis.text.x = element_text(angle = -90, hjust = 0, vjust = 0))

  # Diptheria
dip <- ggplot(total_dip_incidence_by_year, aes(Year, Incidence)) +
  geom_line(size = 1, linetype = 2) +
  geom_point(shape = 4, size = 2) +
  scale_color_manual(values = c("black", "gray40")) +
  theme_bw() +
  geom_smooth() +
  ggtitle(wrapper(paste("Reported worldwide incidence of diptheria from ", min(years_dip), "to ", max(years_dip)), width = 35)) +
theme(plot.title = element_text(hjust = 0),
      axis.text.x = element_text(angle = -90, hjust = 0, vjust = 0))

  # Japanese Encephalitis
jap <- ggplot(total_japenc_incidence_by_year, aes(Year, Incidence)) +
  geom_line(size = 1, linetype = 2) +
  geom_point(shape = 4, size = 2) +
  scale_color_manual(values = c("black", "gray40")) +
  theme_bw() +
  geom_smooth() +
  ggtitle(wrapper(paste("Reported worldwide incidence of japanese encephalitis from ", min(years_japenc), "to ", max(years_japenc)), width = 35)) +
theme(plot.title = element_text(hjust = 0),
      axis.text.x = element_text(angle = -90, hjust = 0, vjust = 0))

  # Measles
mea <- ggplot(total_measles_incidence_by_year, aes(Year, Incidence)) +
  geom_line(size = 1, linetype = 2) +
  geom_point(shape = 4, size = 2) +
  scale_color_manual(values = c("black", "gray40")) +
  theme_bw() +
  geom_smooth() +
  ggtitle(wrapper(paste("Reported worldwide incidence of measles from ", min(years_measles), "to ", max(years_measles)), width = 35)) +
theme(plot.title = element_text(hjust = 0),
      axis.text.x = element_text(angle = -90, hjust = 0, vjust = 0))

  # Mumps
mum <- ggplot(total_mumps_incidence_by_year, aes(Year, Incidence)) +
  geom_line(size = 1, linetype = 2) +
  geom_point(shape = 4, size = 2) +
  scale_color_manual(values = c("black", "gray40")) +
  theme_bw() +
  geom_smooth() +
  ggtitle(wrapper(paste("Reported worldwide incidence of mumps from ", min(years_mumps), "to ", max(years_mumps)), width = 35)) +
theme(plot.title = element_text(hjust = 0),
      axis.text.x = element_text(angle = -90, hjust = 0, vjust = 0))

  # Neonatal Tetanus
neo <- ggplot(total_ntet_incidence_by_year, aes(Year, Incidence)) +
  geom_line(size = 1, linetype = 2) +
  geom_point(shape = 4, size = 2) +
  scale_color_manual(values = c("black", "gray40")) +
  theme_bw() +
  geom_smooth() +
  ggtitle(wrapper(paste("Reported worldwide incidence of neonatal tetanus from ", min(years_ntet), "to ", max(years_ntet)), width = 35)) +
theme(plot.title = element_text(hjust = 0),
      axis.text.x = element_text(angle = -90, hjust = 0, vjust = 0))

  # Pertussis
per <- ggplot(total_pertussis_incidence_by_year, aes(Year, Incidence)) +
  geom_line(size = 1, linetype = 2) +
  geom_point(shape = 4, size = 2) +
  scale_color_manual(values = c("black", "gray40")) +
  theme_bw() +
  geom_smooth() +
  ggtitle(wrapper(paste("Reported worldwide incidence of pertussis from ", min(years_pertussis), "to ", max(years_pertussis)), width = 35)) +
theme(plot.title = element_text(hjust = 0),
      axis.text.x = element_text(angle = -90, hjust = 0, vjust = 0))

  # Polio
pol <- ggplot(total_polio_incidence_by_year, aes(Year, Incidence)) +
  geom_line(size = 1, linetype = 2) +
  geom_point(shape = 4, size = 2) +
  scale_color_manual(values = c("black", "gray40")) +
  theme_bw() +
  geom_smooth() +
  ggtitle(wrapper(paste("Reported worldwide incidence of polio from ", min(years_polio), "to ", max(years_polio)), width = 35)) +
theme(plot.title = element_text(hjust = 0),
      axis.text.x = element_text(angle = -90, hjust = 0, vjust = 0))

  # Rubella
rub <- ggplot(total_rubella_incidence_by_year, aes(Year, Incidence)) +
  geom_line(size = 1, linetype = 2) +
  geom_point(shape = 4, size = 2) +
  scale_color_manual(values = c("black", "gray40")) +
  theme_bw() +
  geom_smooth() +
  ggtitle(wrapper(paste("Reported worldwide incidence of rubella from ", min(years_rubella), "to ", max(years_rubella)), width = 35)) +
theme(plot.title = element_text(hjust = 0),
      axis.text.x = element_text(angle = -90, hjust = 0, vjust = 0))

  # TTetanus
tte <- ggplot(total_ttet_incidence_by_year, aes(Year, Incidence)) +
  geom_line(size = 1, linetype = 2) +
  geom_point(shape = 4, size = 2) +
  scale_color_manual(values = c("black", "gray40")) +
  theme_bw() +
  geom_smooth() +
  ggtitle(wrapper(paste("Reported worldwide incidence of ttetanus from ", min(years_ttet), "to ", max(years_ttet)), width = 35)) +
theme(plot.title = element_text(hjust = 0),
      axis.text.x = element_text(angle = -90, hjust = 0, vjust = 0))

  # Yellow Fever
yel <- ggplot(total_yfever_incidence_by_year, aes(Year, Incidence)) +
  geom_line(size = 1, linetype = 2) +
  geom_point(shape = 4, size = 2) +
  scale_color_manual(values = c("black", "gray40")) +
  theme_bw() +
  geom_smooth() +
  ggtitle(wrapper(paste("Reported worldwide incidence of yellow fever from ", min(years_yfever), "to ", max(years_yfever)), width = 35)) +
theme(plot.title = element_text(hjust = 0),
      axis.text.x = element_text(angle = -90, hjust = 0, vjust = 0))

figure1 <- ggarrange(per, dip, tte
                    # labels = c("A", "B", "C"),
                    # ncol = 1, nrow = 3
                    )

figure2 <- ggarrange(mea, mum, jap, crs
                    # labels = c("A", "B", "C"),
                    # ncol = 1, nrow = 3
                    )

figure3 <- ggarrange(pol, rub, neo, yel
                    # labels = c("A", "B", "C"),
                    # ncol = 1, nrow = 3
                    )

figure1; figure2; figure3;

```

```{r, eval = TRUE, message = FALSE}

# Focus on coverage of DTP1 and DTP3 (for diphtheria, tetanus, and pertussis)
DTP1_coverage <- filter(coverage_series, Vaccine == "DTP1")
DTP3_coverage <- filter(coverage_series, Vaccine == "DTP3")

# Focus on coverage of MCV1 and MCV2 (for measles)
MCV1_coverage <- filter(coverage_series, Vaccine == "MCV1")
MCV2_coverage <- filter(coverage_series, Vaccine == "MCV2")

# Focus on coverage of Pol3 (for polio)
Pol3_coverage <- filter(coverage_series, Vaccine == "Pol3")

# DTP1 coverage per country over time
for (country in levels(DTP1_coverage$Cname)) {
  cur <- filter(DTP1_coverage, Cname == country)
  if (is_empty(cur$Year)) {
    next
  }
  # Plots of reported coverage by country over time
  print(
    ggplot(cur, aes(Year, Percent_coverage)) +
      geom_line(size = 1, linetype = 2) +
      geom_point(shape = 4, size = 2) +
      scale_color_manual(values = c("black", "gray40")) +
      theme_bw() +
      # geom_smooth() +
      ggtitle(wrapper(paste("Reported coverage of DTP1 from ", min(cur$Year), "to ", max(cur$Year), "in ", country), width = 35)) +
    theme(plot.title = element_text(hjust = 0),
          axis.text.x = element_text(angle = -90, hjust = 0, vjust = 0))
  )
}

```

```{r, eval = TRUE, message = FALSE}

# DTP3 coverage per country over time
for (country in levels(DTP3_coverage$Cname)) {
  cur <- filter(DTP3_coverage, Cname == country)
  if (is_empty(cur$Year)) {
    next
  }
  # Plots of reported coverage by country over time
  print(
    ggplot(cur, aes(Year, Percent_coverage)) +
      geom_line(size = 1, linetype = 2) +
      geom_point(shape = 4, size = 2) +
      scale_color_manual(values = c("black", "gray40")) +
      theme_bw() +
      # geom_smooth() +
      ggtitle(wrapper(paste("Reported coverage of DTP3 from ", min(cur$Year), "to ", max(cur$Year), "in ", country), width = 35)) +
    theme(plot.title = element_text(hjust = 0),
          axis.text.x = element_text(angle = -90, hjust = 0, vjust = 0))
  )
}

```

```{r, eval = TRUE, message = FALSE}

# Herd immunity thresholds for diphtheria and pertussis
immunity_thresholds <- data.frame(
  Disease = c("Diphtheria", "Pertussis", "Measles", "Mumps", "Polio", "Rubella", "Smallpox", "Influenza", "Ebola"),
  Reproduction_number = c("6-7", "12-17", "12-18", "4-7", "2-15", "6-7", "5-7", "1.4-4", "1.5-2.5"),
  Threshold_percentage = c(85, 94, 94, 86, 93, 85, 85, 75, NA)
)

head(immunity_thresholds)

```

```{r, eval = TRUE, message = FALSE}

# Number of countries (out of 194 total) that are at or above the immunity threshold percentage

# Diphtheria
counts_dip <- list()
start_year <- 1980
i <- 1
while (start_year != 2019) {
  cur_year_counts <- filter(DTP3_coverage, Year == start_year)
  coverages <- cur_year_counts$Percent_coverage
  coverage_count <- 0
  for (coverage in coverages) {
    if (coverage >= immunity_thresholds$Threshold_percentage[which(immunity_thresholds$Disease == "Diphtheria")]) {
      coverage_count <- coverage_count + 1
    }
  }
  counts_dip[[i]] <- coverage_count
  i <- i + 1
  start_year <- start_year + 1
}

# Pertussis
counts_per <- list()
start_year <- 1980
i <- 1
while (start_year != 2019) {
  cur_year_counts <- filter(DTP3_coverage, Year == start_year)
  coverages <- cur_year_counts$Percent_coverage
  coverage_count <- 0
  for (coverage in coverages) {
    if (coverage >= immunity_thresholds$Threshold_percentage[which(immunity_thresholds$Disease == "Pertussis")]) {
      coverage_count <- coverage_count + 1
    }
  }
  counts_per[[i]] <- coverage_count
  i <- i + 1
  start_year <- start_year + 1
}

# Number of countries that reach the vaccine coverage threshold for herd immunity for Diphtheria over time
covered_thresh_dip <- data.frame(
  Year = 1980:2018,
  Num_countries_covered = unlist(counts_dip)
)

ggplot(covered_thresh_dip, aes(Year, Num_countries_covered)) +
  geom_line(size = 1, linetype = 2) +
  geom_point(shape = 4, size = 2) +
  scale_color_manual(values = c("black", "gray40")) +
  theme_bw() +
  ggtitle(wrapper("Number of countries at coverage threshold for diphtheria herd immunity from 1980 to 2018", width = 65)) +
theme(plot.title = element_text(hjust = 0),
      axis.text.x = element_text(angle = -90, hjust = 0, vjust = 0)) +
  geom_hline(yintercept = 194, linetype = "dashed", color = "red", size = 1.5)

# Number of countries that reach the vaccine coverage threshold for herd immunity for Pertussis over time
covered_thresh_per <- data.frame(
  Year = 1980:2018,
  Num_countries_covered = unlist(counts_per)
)

ggplot(covered_thresh_per, aes(Year, Num_countries_covered)) +
  geom_line(size = 1, linetype = 2) +
  geom_point(shape = 4, size = 2) +
  scale_color_manual(values = c("black", "gray40")) +
  theme_bw() +
  ggtitle(wrapper("Number of countries at coverage threshold for pertussis herd immunity from 1980 to 2018", width = 65)) +
theme(plot.title = element_text(hjust = 0),
      axis.text.x = element_text(angle = -90, hjust = 0, vjust = 0)) +
  geom_hline(yintercept = 194, linetype = "dashed", color = "red", size = 1.5)

```