---
title: "Characterization of Castleman Disease Treatment Response"
author: "Sheila Pierson"
output: 
  html_document:
    toc: false 
    depth: 3 
    theme: paper 
    highlight: tango
---
```{r set-options, echo=FALSE, cache=FALSE}
options(width = 400)
```  
***

### Overview
Castleman disease is a rare inflammatory disorder. The most-deadly subtype (idiopathic multicentric Castleman disease (iMCD)) has a 35% mortality rate. Siltuximab is the only FDA-approved drug to treat iMCD, and nearly 60% of patients do not respond to siltuximab. Using data obtained from the siltuximab trial and from the first-ever Castleman disease natural history registry, ACCLERATE, this project aims to identify the characteristics of siltuximab responders vs. non-responders.  
https://github.com/skpierson/BMIN503_Final_Project

### Introduction 
Castleman disease consists of a group of three heterogeneous diseases, characterized by common lymph node features.  The three subtypes include unicentric Castleman disease (UCD), HHV8-associated multicentric (HHV8+MCD), and idiopathic multicentric Castleman disease (iMCD). UCD is often curative (95% survival) by lymph node excision. Caused by the HHV8 virus, HHV8+MCD is well controlled with rituximab and anti-viral medications.  However, iMCD represents the most deadly subtype, with 35% 5-year mortality rate.  Though there is one FDA-approved drug for iMCD (siltuximab), only ~30-40% patients respond positively (van Rhee et al, Lancet, 2014, Fajgenbaum et al, Blood 2017).  In 2017, the first set of iMCD diagnostic criteria was established, and recently, two clinical subtypes of iMCD have been described - TAFRO (thrombocytopenia, anasarca, fibrosis of bone marrow, renal failure, and organomgaly) patients and NOS (not otherwise specialized) (preveioulsy identified as IPL (idiopathic plasmacytic lymphadenopathy)) patients. It is unclear if these patients respond differently to IL-6 blockade treatment.  In order to appropriately treat patients, it is crucial to understand why a subset of patients responds to siltuximab while others do not. 

Understanding the patient population and characteristics of treatment responders vs. non-responders requires a multidisciplinary approach. As its name suggests, iMCD is idiopathic and understanding the cause requires a multidisciplinary approuch. Proteomics can be used to characterize the protein state of patients in flare compared to remission and to understand if there are proteomic changes specific to any certain subgroup. The study of gene mutations will help to understand if iMCD is caused by an underlying genetic disorder. Whole genome or whole exome sequencing can be used to determine if Castleman disease is caused by a somatic mutation.  In addition to determining the underlying cause, it is crucial to understand the patient's clinical presentation. We currently have available a rich set of clinical data from the siltuximab trial and are actively collecting data in the first ever patient registry.  
  To begin characterizing the Castleman disease population, clinical and demographic characteristics will be summarized for all patients with data in the ACCELERATE registry. As advised by Jason Moore, heatmaps can be used as a visualization tool. A heatmap of clinical features will be generated and clustering will be performed on the patients. The primary aim of this project is to identify characteristics of responders to siltuximab. To do so, analyses will be performed to determine if any of the minor criteria from the iMCD diagnostic criteria (Fajgenbaum et al., Blood, 2017) are predictive of response to siltuximab. Rebecca Hubbard advised on the development of regression models and restricting to baseline timepoints or developing mixed effects to account for time. In addition, Bill LaCava advised on the development of decision tree and random forest models - specifically he provided advice on restricting to certain data collection points (minor diagnostic criteria). Using these methods, I will attempt to identify predictors and test a prediction model using the ACCELERATE data. 

### Methods
Two sets of data are being used for this project.The first set includes data from the ACCELERATE registry. ACCELERATE is the first-ever natural history registry of Castleman disease. Patient enrollment has been ongoing for 1 year and clinical, laboratory, treatment, and treatment response data has been entered for approximately 40 patients.  The second set includes the data from the siltuximab clinical trial ([siltuximab](https://clinicaltrials.gov/ct2/show/NCT01024036), a placebo controlled trial. This trial evaluated the efficacy of siltuximab in treatment of Castleman disease.  Patients were randomized to placebo or siltuximab plus best supportive care. Of 79 patients who met inclusion criteria, 53 were randomized to siltuximab. Of the 53, 18 (34%) responded to treatment.  In this project, I am first cleaning the ACCELERATTE and siltuximab datasets in order to perform analyses.  I will then describe the ACCELERATE cohort.  Finally, I will evaluate the characteristics of the siltuximab responders vs. non-responders and attempt to identify predictors for response. I will run univariate regression models, logistic regression models, and will develop a decision tree and random forest model to identify predictors of response.  I will try to validate the random forest using the ACCEERATE.  I will identify patients treated with siltuximab in the ACCELERATE dataset and will use that data to validate the decision tree developed with the siltuximab data. 

```{r eval=TRUE, warning=FALSE, message=FALSE, tidy=TRUE}
library(tidyverse)
library(reshape2)
library(eeptools)

#ACCELERATE DATA CLEANING
#Import all datasets needed for analysis
#eligibility
reg_elig <-read.csv(file="/Users/sheilapierson/Desktop/Data Science/Final Datasets/eligibility.csv", header=TRUE)
#demographics
reg_dem <-read.csv(file="/Users/sheilapierson/Desktop/Data Science/Final Datasets/demographic.csv", header=TRUE)
#clinical
reg_clin <-read.csv(file="/Users/sheilapierson/Desktop/Data Science/Final Datasets/clinical.csv", header=TRUE)
#pathology
reg_path <-read.csv(file="/Users/sheilapierson/Desktop/Data Science/Final Datasets/pathology.csv", header=TRUE)
#labs
reg_lab <-read.csv(file="/Users/sheilapierson/Desktop/Data Science/Final Datasets/labs.csv", header=TRUE)
#medications by regimen
reg_med <-read.csv(file="/Users/sheilapierson/Desktop/Data Science/Final Datasets/medications_by_regimen.csv", header=TRUE)
#regimen response
reg_response <-read.csv(file="/Users/sheilapierson/Desktop/Data Science/Final Datasets/Regresponse.csv", header=TRUE)
#lab test type id
lab_type <-read.csv(file="/Users/sheilapierson/Desktop/Data Science/Final Datasets/labtype.csv", header=TRUE)


#From reg_elig and reg_dem select the appropriate background vairables and merge datasets
#elig
reg_elig1<-reg_elig[,c(1,4,10,15)]
colnames(reg_elig1)[3] <- "diagnosis"
colnames(reg_elig1)[4] <- "diagnosisdate"
#dem
reg_dem1<-reg_dem[,c(1,6:8)]
names(reg_dem1)<-c("guid","dob","sex","race")
#merge on ids available in the eligibilityset
eligdem<-left_join(reg_elig1,reg_dem1,by="guid")
#change variables to factors
eligdem<-eligdem%>%
  mutate(diagnosis=factor(diagnosis, levels=c(2,3,4),labels=c("UCD","HHV8+","iMCD")))%>%
  mutate(sex=factor(sex, levels=c(1,2),labels=c("Male","Female")))%>%
  mutate(race=factor(race, levels=c(1, 5, 8, 9, 15, 16, 18),labels=c("American Indian", "Japanese","Other Asian","Black", "White","Other","Refuse")))
#Calculate age based on date of birth
eligdem$age <- as.numeric(as.Date(as.character(eligdem$diagnosisdate), format="%m/%d/%y")-
  as.Date(as.character(eligdem$dob), format="%Y-%m-%d"))/365.25
eligdem<-eligdem[,c(1:4,8,6,7)]

#From reg_clin filter out observations that are more than +/-30 days away from diagnosis date
#find difference between diagnosis date and clinical date
reg_clin$date_diff <- as.Date(as.character(reg_clin$diagnosisdate), format="%d-%b-%y")-
  as.Date(as.character(reg_clin$date_of_clinical_note), format="%d-%b-%y")
#select observations within 30 days on either end
reg_clin_diag<-reg_clin%>%
  filter((date_diff >=0 & date_diff<30)|(date_diff<=0 & date_diff> -30))
#subset to variables for heatmap
clin_hm<- reg_clin_diag[c(1,4,6,9:27)]
#recode the hepato, spleno, and lymphad data
clin_hm1 <-clin_hm %>% 
  mutate(hepatomegaly=ifelse(hepatomegaly %in% c(0,1), 1, ifelse(hepatomegaly %in% c(2,3),0, ifelse(hepatomegaly==4,2,NA))))%>%
  mutate(splenomegaly=ifelse(splenomegaly %in% c(0,1), 1, ifelse(splenomegaly %in% c(2,3),0, ifelse(splenomegaly==4,2,NA))))%>%
  mutate(fluid_retention=ifelse(fluid_retention %in% c(0,1), 1, ifelse(fluid_retention %in% c(2,3),0, ifelse(fluid_retention==4,2,NA))))%>%
  mutate(lymphadenopathy=ifelse(lymphadenopathy %in% c(0,1), 1, ifelse(lymphadenopathy %in% c(2,3),0, ifelse(lymphadenopathy==4,2,NA))))
#change all NA (2) to -1 for easier sorting
notassessed <- as.data.frame(sapply(clin_hm1[,4:22], function(x) ifelse(x==2,-1,x)))
#bind back to data
together<-cbind(clin_hm1[,1:3],notassessed[,1:19])
convert <-together %>%
  mutate_if(is.numeric,as.factor)
#melt the dataframe
melted <- melt(convert, id=(c("guid", "patient_id", "diagnosis")))
max1<-mutate(melted, numeric = as.numeric(value))
#select maximum observation (e.g. if 1 is maximum then the patient was "yes" to that feature within the 30 days, if 0 is maximum then the patient was "no" to that feature within 30 dyas, and if -1 is maximum then the patietn was "not assessed"" for htat feature within 30 days)
selectmax <- max1 %>% 
  group_by(patient_id, variable) %>%
  filter(numeric == max(numeric)) 
#keep only variables necessary to merge into full set
merge<-selectmax[c(1,4,5)]
#remove duplicates so each patient is left with one "max" value for a feature
dedup <-merge[!duplicated(merge), ]
#turn back to wide for merge with full set
data_wide <- spread(dedup, variable, value)
#bind to eligibility, demographics dataset 
elig_dem_clin<-left_join(eligdem,data_wide, by="guid")


#From labs filter out the labs within 30 days after diagnosis or up to 365 days prior to diagnosis. This is done in an attempt to include as much data as possible and under assumption that prior to diagnosis values are more accurate becuase patients may be treated shortly thereafter, altering their lab values. This is not ideal, but with small numbers, I want to incldue as much as possible
#find difference between diagnosis date and labtest date
reg_lab$date_diff <- as.Date(as.character(reg_lab$diagnosisdate), format="%d-%b-%y")-
  as.Date(as.character(reg_lab$lab_test_date), format="%d-%b-%y")
reg_lab$abs<-as.numeric(abs(reg_lab$date_diff))
#select those within 365 days prior and 30 days after
reg_lab_30<-reg_lab%>%
  group_by(guid, lab_test_type_id)%>%
  filter((abs==min(abs)))%>%
  filter((date_diff >=0 & date_diff<30)|(date_diff<=0 & date_diff> -365))%>%
  filter(date_diff==min(date_diff))
#merge in test name and unit
lab_name<-left_join(reg_lab_30,lab_type,by=c("lab_test_type_id"="id"))
lab_form<-lab_name[,c(1,7,16)]
lab_form<-lab_form[!duplicated(lab_form),]
lab_wide <- spread(lab_form, lab_test_name, value)
#convert to same units used in siltuximab
#change names
names(lab_wide)<-c("guid","alb","crp","egfr", "esr","gammaglob", "hgb", "igg","platelet","proteinuria")
lab_convert<-lab_wide%>%
  mutate(alb=alb*10)%>%
    mutate(hgb=hgb*10)%>%
    mutate(igg=igg/100)%>%
  mutate(proteinuria=factor(proteinuria, levels=c(1,0),labels=c("Present","Absent")))
#merge with elig, dem, clin data
demo_clinchar<-left_join(elig_dem_clin,lab_convert, by="guid")

#From reg_path, select the variables and merge into the clinical dataset
reg_path1<-reg_path[,c(1,8:21)]
#recode the histopath subtype to factor
reg_path1<-mutate(reg_path1,histopathological_subtype=factor(histopathological_subtype, levels=c(1, 2, 3 ,4, 5, 6),labels=c("Hyaline Vascular","Hypervascular","Mixed", "Plasmacytic","Not Documented","Plasmablastic")))
#merge data to full set
all_char<-left_join(demo_clinchar,reg_path1,by="guid")
#select only patients with path report to report characteristics
all_char_sub <- subset(all_char, !is.na(histopathological_subtype))

#From reg_med, select the regimens that included siltuximab and assign those regimen IDs a "siltuximab" variable
reg_siltux<-reg_med %>%
  group_by (regimen_id)%>%#group by regimen and assign a siltuximab variable for all observations of a regimen containing siltuximab
  mutate(siltuximab=ifelse(medication_name %in% "SILTUXIMAB", 1,NA))%>%
  group_by (patient_id, regimen_id)%>%
  fill(siltuximab)%>%
  group_by (patient_id, regimen_id)%>%
  fill(siltuximab, .direction="up")%>%
  mutate(siltuximab=ifelse(is.na(siltuximab),2,siltuximab))
#select necessary observations and delete duplicates
reg_siltux1<-reg_siltux[,c(2,3,17)]
reg_siltux2<-reg_siltux1[!duplicated(reg_siltux1),]
#From reg_response, merge in best clinical responses to regimens by reg id
reg_response1<-reg_response[,c(1:3,10)]
siltux_resp<-left_join(reg_siltux2,reg_response1,by= c("regimen_id" = "id"))
#assign CR (complete) or PR (partial) to response and all others to non-response/unknown
siltux_resp_1<-siltux_resp%>%
  mutate(response=ifelse(regimen_best_clinical_response %in% c("Complete Response (CR)","Partial Response (PR)"), 1, 2))%>%
  mutate(response=factor(response,levels=c(1,2), labels=c("Responder", "NonResponder")))%>%
  droplevels()
ordered_resp <- siltux_resp_1[order(siltux_resp_1$patient_id.x, siltux_resp_1$siltuximab, siltux_resp_1$response),]
sil_resp_fin <- ordered_resp[!duplicated(ordered_resp$guid),]
#if regimen does not include siltuximab then assign response to NA
sil_resp_fin<-sil_resp_fin%>%
  mutate(response=ifelse(siltuximab==1,response,NA))%>%
  mutate(response=factor(response,levels=c(1,2), labels=c("Responder", "NonResponder")))
#keep only necessary levels
sil_resp_fin<-sil_resp_fin[,c(3,4,7)]
#merge response data with clinical data
accelerate<-left_join(all_char_sub,sil_resp_fin, by="guid")
#siltuximab patients only
accel_siltux<-accelerate%>%
  filter(!is.na(response))

#SILTUXIMAB TRIAL DATA CLEANING
#import Siltuximab Trial files for cleaning
#screening
screening <-read.csv(file="/Users/sheilapierson/Desktop/Data Science/Final Datasets/adsl - Subject-level.csv", header=TRUE)
#response
efficacy <-read.csv(file="/Users/sheilapierson/Desktop/Data Science/Final Datasets/adef - efficacy template.csv", header=TRUE)
#clinical
clinical <-read.csv(file="/Users/sheilapierson/Desktop/Data Science/Final Datasets/addc - Disease Characteristics.csv",header=TRUE, na.strings=c("","NA"))
#labs
labs <-read.csv(file="/Users/sheilapierson/Desktop/Data Science/Final Datasets/adlb - Laboratory Test Results.csv", header=TRUE)

#Format the datafiles needed for the siltuximab dataset
#From SCREENING, filter for eligible patients and select needed columns
eligible <-screening%>%
  filter(ARM=="CNTO328 + BSC" | ARM=="PLACEBO + BSC")%>%
  dplyr::select(USUBJID,SUBJID, AGE,SEX,RACE,ARM,TRTP,TRTPN,HISTYPER)%>%
  droplevels()

#From EFFICACY, filter for the parameter "Durable tumor and symptomatic response during blinded treatment period by independent review" and select the needed columns
response<-efficacy%>%
  filter(PARAM=="Durable tumor and symptomatic response during blinded treatment period by independent review")%>%
  dplyr::select(USUBJID,PARAM,AVALC)%>%
  mutate(RESP=ifelse(AVALC %in% c("CR","PR"), 1, 2))%>%
  mutate(RESP=factor(RESP, levels=c(1, 2), labels=c("Responder", "NonResponder")))%>%
  droplevels()


#From CLINICAL, filter for visit that occurred at "CYCLE 1 DAY 1" (baseline), select necessary variables
clin_screen<-clinical%>%
  filter(VISIT=="CYCLE 1 DAY 1")%>%
  dplyr::select(USUBJID,DCTEST,DCTESTCD,DCCAT,DCSCAT,DCORRES,
         DCSTRESC,PARAMCD)
#Look at the clinical parameters
#drop variables that will not be involved in the model
#assign DCTEST levels that are wanted because there is corresponding data in validation set
WantedLevels<-c(1,11,15:17, 22,25,27,35:36,52,56:58)
clin_var<-filter(clin_screen,as.integer(DCTEST) %in% WantedLevels)
#restrict to variables to spread to wide 
clin_var_to_wide<-clin_var[c(1,3,7)]
clin_wide<-spread(clin_var_to_wide,"DCTESTCD","DCSTRESC")
#Turn all clinical values into 2-level factors
clin_wide<-droplevels(clin_wide)

#From LABS, filter for visit that occurred at "CYCLE 1 DAY 1", select necessary variables
lab_screen<-labs%>%
  filter(VISIT=="CYCLE 1 DAY 1" | VISIT=="SCREENING")%>%
  dplyr::select(USUBJID,LBTEST,LBTESTCD,LBORRES,LBORRESU,LBSTRESC,
         LBSTRESN,LBSTRESU,VISIT, LBDT, LBDY, LBTM,LBTPT,LBTPTNUM)
#Select LABS that are also measured in validated set and that will be included in decision tree
WantedLevels<-c(3,12,23,35, 41,62,67)
lab_var<-filter(lab_screen,as.integer(LBTEST) %in% WantedLevels)
#Select one lab per patient (currently CRP is collected at 3 timepoints on cycle 1 day 1, exclude "2HR" and "4HR" measurements; also need to select for screening if missing cycle 1 day 1) 

one_lab<-lab_var%>%
  group_by(USUBJID,LBTEST)%>%
  filter(LBTPT!="2HR")%>%
  filter(LBTPT!="4HR")%>%
  group_by(USUBJID, LBTEST)%>%
  fill(LBSTRESN, LBSTRESU)%>%
  filter(VISIT!="SCREENING")

#change the </> labs and replace the N version with character
one_lab$LBSTRESC<-gsub("<","",one_lab$LBSTRESC)
one_lab$LBSTRESC<-gsub(">","",one_lab$LBSTRESC)
one_lab$LBSTRESN <- with( one_lab, as.numeric(ifelse( is.na(LBSTRESN), LBSTRESC, LBSTRESN )))

#restrict to variables to spread to wide - spread labs
lab_var_to_wide<-one_lab[c(1,3,7)]
lab_wide_value<-spread(lab_var_to_wide,"LBTESTCD","LBSTRESN")
#rename  columns to add unit ids
names(lab_wide_value) <- c("USUBJID","ALB_g_L","CRP_mg_L","ESR_mm_hr","HGB_g_L", "IGG_g_L","PLT_k_uL","PROT_g_L")

#Merge the 4 siltuximab trial sets (screening, efficacy, clinical, and lab)
#join the units and labs dataframe
elig_resp<-left_join(eligible,response,by="USUBJID")
addclin<-left_join(elig_resp,clin_wide,by="USUBJID")
alldata<-left_join(addclin,lab_wide_value,by="USUBJID")
#restrict to siltuximab arm (trtpn=2)
siltuximab<-filter(alldata,TRTPN==2)

```

### Results
```{r eval=TRUE, warning=FALSE}
#ACCELERATE SUMMARY OF COHORT
#Describe demographics
#all
summary(accelerate$age)
summary(accelerate$sex)
summary(accelerate$race)
#by group
#age
accelerate%>%
  group_by(diagnosis)%>%
  summarise( age=mean(age, na.rm=TRUE))
accelerate%>%
  group_by(diagnosis)%>%
  summarise( age=median(age, na.rm=TRUE))
accelerate%>%
  group_by(diagnosis)%>%
  summarise( age=min(age, na.rm=TRUE))
accelerate%>%
  group_by(diagnosis)%>%
  summarise( age=max(age, na.rm=TRUE))
#sex
accelerate%>%
  group_by(diagnosis,sex)%>%
  summarise (n = n()) %>%
  mutate(freq = n / sum(n))
#race
accelerate%>%
  group_by(diagnosis,race)%>%
  summarise (n = n()) %>%
  mutate(freq = n / sum(n))
#subtype
accelerate%>%
  group_by(diagnosis,histopathological_subtype)%>%
  summarise (n = n()) %>%
  mutate(freq = n / sum(n))
```
Overall there is an even distribution of males and females. Mean age is 35.8.  The populate is predominantly white with a hyaline vascular histopathological subtype.


```{r eval=TRUE, warning=FALSE}
#Describe CLINICAL FEATURES with a heatmap
#keep only variables necessary for hm
hmvar<-selectmax[c(2:5)]
#remove duplicates e.g. if patient had fatigue=1 multple times
dedup <-hmvar[!duplicated(hmvar), ]
#turn back to wide for HM
data_wide <- spread(dedup, variable, value)
#add rownmaes
withnames <- data.frame(data_wide[,-1], row.names=data_wide$patient_id)
clinical_data<-withnames[2:20]
names(clinical_data) <- c("Fatigue", "Malaise","Hyperhidrosis","Night sweats","Fever","Weight loss","Anorexia","Lymph node pain", "Dyspnea","Pruritis",
                     "Fluid retention","Neuropathy","Castleman skin disorders","Other skin disorders","Interstitial lymphocytic pneumonitis","Hepatomegaly",
                     "Splenomegaly","Lymphadenopathy","Arthralgia")
#is.na(clinical)<-clinical==-1
#convert dataframe to matrix
matrixhm<-data.matrix(clinical_data)
status.colors <- unlist(lapply(withnames$diagnosis, function(x){if (x==4) "#7570b3" else if (x==3) "#d95f02" else "#1b9e77"}))

#Try plotting in heatmap.2
my_palette <- colorRampPalette(c("gray", "purple",  "yellow"))(n = 3)

#Plot
lhei=c(2,12)
heatmap.2(matrixhm, 
          col=my_palette,
          dendrogram="row",
          labRow = FALSE,
          cexRow=.65,cexCol=1.2,
          margins =c(14,6),
          srtCol=45,   
          RowSideColors = status.colors,
          key=FALSE,
          trace="none",
          lhei=lhei)

#Describe laboratory features with plots by subtype
#change diagnosis to a factor

plt<-ggplot(data=accelerate, aes(diagnosis, platelet)) +
  geom_boxplot() +
  geom_jitter(size=1) +
  labs(y="Platelets, k/uL", x="Diagnosis Subtype")+
  theme(axis.title.x = element_text( size=7), axis.title.y=element_text(size=7),axis.text.x=element_text(size=7),axis.text.y=element_text(size=7))

alb<-ggplot(data=accelerate, aes(diagnosis, alb)) +
  geom_boxplot() +
  geom_jitter(size=1) +
  labs(y="Albumin, g/L", x="Diagnosis Subtype")+
theme(axis.title.x = element_text( size=7), axis.title.y=element_text(size=7),axis.text.x=element_text(size=7),axis.text.y=element_text(size=7))

crp<-ggplot(data=accelerate, aes(diagnosis, crp)) +
  geom_boxplot() +
  geom_jitter(size=1) +
  labs(y="CRP, mg/L", x="Diagnosis Subtype")+
theme(axis.title.x = element_text( size=7), axis.title.y=element_text(size=7),axis.text.x=element_text(size=7),axis.text.y=element_text(size=7))

egfr<-ggplot(data=accelerate, aes(diagnosis, egfr)) +
  geom_boxplot() +
  geom_jitter(size=1) +
  labs(y="eGFR, mL/min/1.73m2", x="Diagnosis Subtype")+
theme(axis.title.x = element_text( size=7), axis.title.y=element_text(size=7),axis.text.x=element_text(size=7),axis.text.y=element_text(size=7))

esr<-ggplot(data=accelerate, aes(diagnosis, esr)) +
  geom_boxplot() +
  geom_jitter(size=1) +
  labs(y="ESR, mm/hr", x="Diagnosis Subtype")+
theme(axis.title.x = element_text( size=7), axis.title.y=element_text(size=7),axis.text.x=element_text(size=7),axis.text.y=element_text(size=7))

gg<-ggplot(data=accelerate, aes(diagnosis, gammaglob)) +
  geom_boxplot() +
  geom_jitter(size=1) +
  labs(y="Gammaglobulin, g/L", x="Diagnosis Subtype")+
theme(axis.title.x = element_text( size=7), axis.title.y=element_text(size=7),axis.text.x=element_text(size=7),axis.text.y=element_text(size=7))

hgb<-ggplot(data=accelerate, aes(diagnosis, hgb)) +
  geom_boxplot() +
  geom_jitter(size=1) +
  labs(y="Hemoglobin, g/L", x="Diagnosis Subtype")+
  theme(axis.title.x = element_text( size=7), axis.title.y=element_text(size=7),axis.text.x=element_text(size=7),axis.text.y=element_text(size=7))

igg<-ggplot(data=accelerate, aes(diagnosis, igg)) +
  geom_boxplot() +
  geom_jitter(size=1) +
  labs(y="IgG, g/L", x="Diagnosis Subtype")+
theme(axis.title.x = element_text( size=7), axis.title.y=element_text(size=7),axis.text.x=element_text(size=7),axis.text.y=element_text(size=7))

prot<-ggplot(data=accelerate, aes(x=diagnosis, fill=proteinuria)) +
    geom_bar(position="fill")+
  scale_y_continuous(labels=scales::percent)+
  labs(y="Frequency, %", x="Proteinuria")+
theme(axis.title.x = element_text( size=7), axis.title.y=element_text(size=7),axis.text.x=element_text(size=7),axis.text.y=element_text(size=7), legend.position = "bottom", legend.text = element_text(size=8), legend.title = element_text(size=8))


library(cowplot)
plot_grid(plt, alb,crp,egfr,esr,gg,hgb,igg, prot, align = "h", nrow = 3 ,ncol=3,  rel_heights = c(1,1,1), rel_widths=c(1,1,1),scale=1, label_size = .5 )

#Describe pathology features with bar chart
#select the pathology data
accel_path<-accelerate[,c(3,36:48)]
#convert all "Yes" (mild, moderate, severe, unspecified) to 1
accel_path[,2:14][accel_path[,2:14]==2]<-1
accel_path[,2:14][accel_path[,2:14]==3]<-1
accel_path[,2:14][accel_path[,2:14]==4]<-1
accel_path[2:14] <- lapply(accel_path[2:14], factor, 
              levels=c(-1,0, 1), 
              labels = c("Not documented", "No", "Yes"))

#melt dataframe 
melt_path <- melt(accel_path, id=(c("diagnosis")))
#select only those with "yes"
melt_path1<-melt_path%>%
  filter(diagnosis!="HHV8+")
#By Facet Wrap
p<-ggplot(melt_path1) +
  geom_bar(aes(x = diagnosis, fill=value),
           position = "dodge", stat="count")+
  facet_wrap( ~variable, nrow=5, ncol=4) +
  theme(strip.text.x = element_text(size=7), legend.position = "bottom")
p

```



The clinical features heatmap shows patients cluster together by subgroup. The iMCD pateint are more symptomatic than UCD patients.  Symptoms most frequently observed are lymphadenopathy, fatigue, fluid retention, and fever.  Lab features similar show a more symptomatic iMCD cohort. ESR and CRP demonstrate inflammation, consistent with iMCD. Thrombocytopenia and thrombocytosis are evident in the iMCD subgroup, and ow hemoglobin also demonstrates anemia, all consistent with iMCD diagnostic criteria. Typical iMCD histopathological features were observed. 



```{r eval=TRUE, warning=FALSE}
#COMPARISON OF COHORTS - compare siltuximab cohort (N=53) with participants in ACCELERATE treated with siltuximab (N=13)
#Change ACCELERATE factor levels so data can be stacked and compared 
compare_siltux<-siltuximab[,c(1,3:5,9,12,13:33)]
compare_siltux$study<-"Siltuximab CT"
compare_accel<-accel_siltux[,c(1,7,5,6,49, 51,14,16,8,12,18,10,23,9,19,11,21,24,15,13,27,28,30,32,33,34,35)]
compare_accel$study<-"ACCELERATE"
colnames(compare_accel)<-colnames(compare_siltux)

#Set factors to same level as siltux 
#assign the "not assessed values" to NA
compare_accel[compare_accel == -1] <- NA

 compare_accel[7:20] <- lapply(compare_accel[7:20], factor, 
              levels=c(1,0), 
              labels = c("Y", "N"))
 str(compare_accel)

#Stack the dataframes
stacked<-rbind(compare_siltux,compare_accel)
stacked$study <- as.factor(stacked$study)

#differences between groups
#prepare an output dataset for variables to be included  
diffvars<-stacked[,c(2:26)]
output_diff<-data.frame(matrix(nrow=25, ncol=2))
names(output_diff)=c("diffvars", "pval")
output_diff$diffvars<-colnames(diffvars) 

#Run univariate analyis
uni_diff<-stacked[,c(2:28)]
n<-length(diffvars)
for (i in 1:n){
  result<-(glm(uni_diff$study ~ uni_diff[[i+1]],family="binomial"))
  output_diff[i,2] <- coef(summary(result))[,4][2]
}
#Review P-value output
output_diff
low_diff<-filter(output_diff, pval<.05)
low_diff

#Plot the different variables: fatigue, fever, hyperhidrosis, liver, skin disease, ESR 
ggplot(data=stacked, aes(x=study, fill=FATGXR)) +
    geom_bar(position="fill")+
  scale_y_continuous(labels=scales::percent)+
  labs(y="Frequency, %", x="Fatigue")
ggplot(data=stacked, aes(x=study, fill=HYPERHID)) +
    geom_bar(position="fill")+
  scale_y_continuous(labels=scales::percent)+
  labs(y="Frequency, %", x="Hyperhidrosis")
ggplot(data=stacked, aes(x=study, fill=LIVER)) +
    geom_bar(position="fill")+
  scale_y_continuous(labels=scales::percent)+
  labs(y="Frequency, %", x="Hepatomegaly")
ggplot(data=stacked, aes(x=study, fill=SKINDIS)) +
    geom_bar(position="fill")+
  scale_y_continuous(labels=scales::percent)+
  labs(y="Frequency, %", x="Skin Disorders")
ggplot(data=stacked, aes(x=study, fill=FEVER)) +
    geom_bar(position="fill")+
  scale_y_continuous(labels=scales::percent)+
  labs(y="Frequency, %", x="Fever")
ggplot(stacked, aes(study, ESR_mm_hr)) +
    geom_boxplot()


```



These data compare the ACCELERATE cohort at the time of diagnosis with the siltuximab cohort at the time of study screening.  Data show that there are differences between the cohorts. The ACCELERATE cohort is more symptomatic, with significantly more fever, fatigue, hyperhidrosis, hepatomegaly, and skin disease.  This isn't unexpected. The siltuximab cohort enrolled newly diagnosed patients or patients who had been diagnosed and previously. Patients had to be 0-2 on the ECOG scale to be enrolled in the clinical trial, excluding the more sick patients. Whereas it is to be expected that patients may be most sick at the time of diagnosis.



```{r eval=TRUE, warning=FALSE}
#SILTUXIMAB SUMMARY AND UNIVARIATE ANALYSIS
summary(siltuximab)

#Run univariate analyses on all continous and categorical variables with >1 level
#prepare an output dataset for variables to be included  
vars<-siltuximab[,c(3:5,9,16:17,19:21,23:24,27:32)]
output<-data.frame(matrix(nrow=17, ncol=2))
names(output)=c("var", "pval")
output$var<-colnames(vars) 

#run univariate analyis
uni<-siltuximab[,c(3:5,9,12,16:17,19:21,23:24,27:32)]
n<-length(vars)
for (i in 1:n){
  result<-(glm(uni$RESP ~ uni[[i+1]],family=binomial()))
  output[i,2] <- coef(summary(result))[,4][2]
}
#review p-value output
output
lowp<-filter(output, pval<.05)
lowp

#prepare graphical output of the significant variables
alb<-ggplot(siltuximab, aes(ALB_g_L, fill = RESP)) +
  geom_density()+
  labs(y="Density",x="Albumin g/L")+  
  theme(axis.title.x = element_text( size=7), axis.title.y=element_text(size=7),axis.text.x=element_text(size=7),axis.text.y=element_text(size=7), legend.position = "bottom", legend.text = element_text(size=8), legend.title = element_text(size=8))

hgb<-ggplot(siltuximab, aes(HGB_g_L, fill = RESP)) +
  geom_density()+
  labs(y="Density",x="Hemoglobin g/L")+  
  theme(axis.title.x = element_text( size=7), axis.title.y=element_text(size=7),axis.text.x=element_text(size=7),axis.text.y=element_text(size=7), legend.position = "bottom", legend.text = element_text(size=8), legend.title = element_text(size=8))

igg<-ggplot(siltuximab, aes(IGG_g_L, fill = RESP)) +
  geom_density()+
  labs(y="Density",x="IgG g/L")+  
  theme(axis.title.x = element_text( size=7), axis.title.y=element_text(size=7),axis.text.x=element_text(size=7),axis.text.y=element_text(size=7), legend.position = "bottom", legend.text = element_text(size=8), legend.title = element_text(size=8))

esr<-ggplot(siltuximab, aes(ESR_mm_hr, fill = RESP)) +
  geom_density()+
  labs(y="Density",x="ESR mm/hr")+  
  theme(axis.title.x = element_text( size=7), axis.title.y=element_text(size=7),axis.text.x=element_text(size=7),axis.text.y=element_text(size=7), legend.position = "bottom", legend.text = element_text(size=8), legend.title = element_text(size=8))

library(cowplot)
plot_grid(alb,hgb,igg,esr ,align = "h", nrow = 2 ,ncol=2,  rel_heights = c(1,1), rel_widths=c(1,1),scale=1)

```



These data show that albumin, hemoglobin, density, and ESR are significantly associated with response.  Responders show lower albumin (hypoalbuminemia), lower hemoglobin (anemia), higher IgG, and higher ESR (inflammation). Clinical features were systematically missing for 10 patients, which likely affected analyses. 

Logistic Regression Models
```{r eval=TRUE, warning=FALSE}
#LOGISTIC REGRESSION MODEL
#logistic regression model with all variables applied
summary((glm(RESP~FEVER++LIVER+SPLEEN+FLUIDRET+SKINDIS+IGG_g_L+ ALB_g_L+ HGB_g_L+PLT_k_uL+ESR_mm_hr +CRP_mg_L, data=siltuximab, family=binomial())))

#logisitic regression model with only the significant variables
summary((glm(RESP~ IGG_g_L+HGB_g_L+ALB_g_L+ESR_mm_hr, data=siltuximab, family=binomial())))

```
There were no significant associations when all iMCD diagnostic criteria at baseline were put into a logistic regression model. When restricting to only the significant variables, there were still no significant associations. 

Decision Tree 
```{r eval=TRUE, warning=FALSE}
#build decision tree with all demographic, clinical, and lab variables
library(party)
library(rpart)
library(rpart.plot)
library(tree)
tree <- rpart(RESP ~CRP_mg_L+ALB_g_L+FEVER+
              IGG_g_L+FEVER+FLUIDRET+
              LIVER+SPLEEN+SKINDIS+ESR_mm_hr+HGB_g_L+
              PLT_k_uL, data = siltuximab, method = "class")
tree
rpart.plot(x=tree)


#using complete data
#complete cases
complete<-siltuximab[complete.cases(siltuximab), ]
tree <- rpart(RESP ~CRP_mg_L+ALB_g_L+FEVER+
              IGG_g_L+FEVER+FLUIDRET+
              LIVER+SPLEEN+SKINDIS+ESR_mm_hr+HGB_g_L
              +PLT_k_uL, data = complete, method = "class")
tree
rpart.plot(x=tree)

```


When looking at all cases, high levels of IgG is the variable that splits responders from non-responders. This is of interest because high levels of IgG are associated with iMCD-NOS/iMCD-IPL patients and low/normal levels of IgG are associated with iMCD-TAFRO patients.  It is still unclear if iMCD-TAFRO and iMCD-NOS patients respond differently to siltuximab. When restricting to complete cases, anemia is the variable that splits responders from non-responders.  

```{r eval=TRUE, warning=FALSE}
#build a random forest model on the complete data labs only
library(randomForest)
#complete cases of labs, minus esr (so many missing)
labs<-siltuximab[,c(1,12,27,28,29,30,31,32)]
complab<-labs[complete.cases(labs), ]
siltux.rf <- randomForest(RESP~ ALB_g_L+
                            IGG_g_L+HGB_g_L+PLT_k_uL+CRP_mg_L,
                           data=complab, ntree=75, importance=TRUE)
siltux.rf
siltux.rf$importance

#ACCELERATE data to test model with
acceleratelab<-accel_siltux[,c(1,51,27,28,32,33,34)]
names(acceleratelab)<-c("USUBJID","RESP","ALB_g_L","CRP_mg_L","HGB_g_L","IGG_g_L","PLT_k_uL")

#Prediction model
rf.top.pred <- predict(siltux.rf, complab, type="prob")
rf.top.pred 
#Validation
obs_outputs <- vector(mode="numeric", length=13)
    pred_outputs.rf <- vector(mode="numeric", length=13)
    obs_outputs[1:13] <- c(1,1,1,1,1,2.2,2,2,1,2,1,2)
    #train/test
    rf <- randomForest(RESP~ALB_g_L+CRP_mg_L+PLT_k_uL+
                            IGG_g_L+HGB_g_L, data=complab, ntree=75)
    rf.pred.curr <- predict(rf, newdata=acceleratelab, type="prob") 
    pred_outputs.rf <- rf.pred.curr[,2]
pred_outputs.rf
#ROC Curves
library(pROC)
plot.roc(complab$RESP, rf.top.pred[,2], ci=TRUE, main="Random Forest")
plot.roc(obs_outputs, pred_outputs.rf, ci=TRUE, col="darkgreen", add=TRUE) 
auc(complab$RESP, rf.top.pred[,2])
auc(obs_outputs, pred_outputs.rf)


```



The random forest model again identified IgG as the most predictive variable, with the highest mean decrease GINI value. AUC for the test data is .75.  However, the dataset is too small to interpret too much from this data.  

```

#CONCLUSION

Overall, clinical characteristics of the ACCELERATE cohort demonstrate the heterogeneity of Castleman disease. iMCD patients typically demonstrate a greater number of clinical abnormalities and a wider range of laboratory abnormalities. 

The main aim of this project was to identifiy predictors of response to siltuximab. These analyses continually identified IgG as a potential predictor of response.  In the univariate analysis of iMCD diagnostic clinical and laboratory variabels on siltuximab response (siltuximab dataset), albumin (low), hemoglobin (low), IgG (high), and ESR (high) as measured at baseline were found to be significantly associated with response to siltuximab.  However, when applied to a logistic regression, no variables were significant.  When considering all data, the decision tree model again identified high levels of IgG as a predictor of response. Similarly, IgG demonstrated the greatest mean decrease gini value in a random forest model.  The validation of the model on test data from the ACCELERATE dataset in not terribly interpretable because of the low N. In addition, random forest model was trained and validated on the same data, resulting in overfitting. 

Though this is a small dataset, the continual identification of IgG as a predictor is worth further pursuit. Particularly because of the association of IgG levels with clinical subtype of iMCD (TAFRO vs. iMCD-NOS).  As we are currently collecting data for the ACCELERATE registry, we expect that we will be able to repeat these analyses on the ACCELERATE dataset in the future. We will look at additional variables and plan to use ACCELERATE to train, validate, and test models of response prediction. 
