---
title: "BMIN503/EPID600 Project Template"
author: "Heather Wachtel"
output: 
  html_document:
    toc: false 
    depth: 3 
    theme: paper 
    highlight: tango
---
```{r set-options, echo=FALSE, cache=FALSE}
options(width = 400)
```  
***
Use this template to complete your project throughout the course. Your Final Project presentation will be based on the contents of this document. Replace the title/name above and text below with your own, but keep the headers.

### Overview
Give a brief a description of your project and its goal(s), what data you are using to complete it, and what three faculty/staff in different fields you have spoken to about your project with a brief summary of what you learned from each person. Include a link to your final project GitHub repository.

Succinate dehydrogenase (SDH) germline pathogenic variants have been implicated in several human cancers.  The goal of this study is to characterize SDHx mutations in a large cohort of subjects undergoing multigene panel testing of cancer predisposition genes, and to determine the risks of specific cancer types associated with SDHx germline pathogenic variants.

Faculty:  
1. Katherine Nathanson, Basser Professor, Div. of Genetics and Translational Medicine, PSOM  
2. Ben Voight, Assoc. Professor, Genetics/Systems Pharmacology and Translational Therapeutics, PSOM  
3. Eric Polley, Asst. Professor, Biostatistics, Mayo Clinic 


### Introduction 

Succinate dehydrogenase (SDH) is a mitochondrial enzyme which plays an essential role in energy production in oxygenated cells.[1, 2] SDHA, SDHB, SDHC, SDHD, SDHAF1, and SDHAF2 are the genes encoding the four subunits of SDH.[1, 3] Germline SDHx pathogenic variants (PVs) have been best characterized in pheochromocytoma/paraganglioma (PCC/PGL), neuroendocrine tumors in which pseudohypoxia appears to drive tumorigenesis.[2, 4, 5] SDHx mutations have also been identified in gastrointestinal stromal tumors (GIST), renal cell carcinomas, thyroid cancers, and sarcomas.[1, 2]  

SDH functions as a tumor suppressor and succinate, its substrate, functions as an oncometabolite at high levels.[6, 7] The latter may play a role in metabolic alteration, epithelial-to-mesenchymal transition, and angiogenesis.[7, 8]  Given the fundamental impact of these pathways on differentiation and development, SDHx mutations have broad implications for human disease.  To date, no comprehensive analysis has been performed to assess for SDHx pathogenic variants in common cancers, and SDHx penetrance is incompletely understood. It has been posited that there is a link between SDHB and breast cancer, however there is no empiric evidence to support this hypothesis. 

The goals of this study are therefore twofold: firstly, to determine the prevalence of SDHx germline PVs in a large cohort of patients undergoing multigene panel testing of cancer predisposition genes, and secondly to estimate risks of breast cancer and other common cancer types associated with germline SDHx, as compared with reference populations.

### Methods

# Study cohort
The study cohort consists of subjects undergoing multigene panel testing of cancer predisposition genes between March 2012 and 6/30/2019, at a single diagnostic laboratory (Ambry Genetics). Clinical histories were obtained from clinician-completed test requisition forms (TRFs) and included clinical documentation.  Gene-specific PVs and variants likely pathogenic (VLPs) of SDHx subunits (SDHA, SDHAF2, SDHB, SDHC, and SDHD) were assessed.  

# Definitions
A masked ‘positive’ was defined as an individual with a null allele (frameshift, nonsense, canonical splice and other alterations such as missense that have been previously classified as PV/VLP) at a heterozygote ratio of 30% or greater, reflecting that >99% of calls with heterozygote ratios over 30% confirmed on Sanger sequencing.  Gross deletions were excluded.

# Statistical analysis
Descriptive statistics were performed.  Continuous variables were presented as mean with standard deviation (SD) for normally distributed variables, and median with interquartile range (IQR) for non-normally distributed variables.  Logistic regression was performed to evaluate the association between SDHx PVs and common human cancers.  A p-value of 0.05 was considered statistically significant. 

# Case-control analysis
Case-control analysis was performed to estimate gene-specific risks of cancer.  Non-Finnish European subjects from the publicly available Genome Aggregation Database (gnomAD) were used as reference controls for Caucasian study subjects.[9] Case-control analysis was performed by pooling PVs at the gene level. Subjects with concomitant PV/VLP in SDHx and other hereditary cancer predisposition genes were excluded from analysis (n=74). Odds ratios and 95% confidence intervals were estimated for subjects with a personal cancer history. 

### Results
Describe your results and include relevant tables, plots, and code/comments used to obtain them. End with a brief conclusion of your findings related to the question you set out to address. You can include references if you'd like, but this is not required.

### References

1.	Dalla Pozza E, Dando I, Pacchiana R, et al (2020) Regulation of succinate dehydrogenase and role of succinate in cancer. Semin Cell Dev Biol 98:4-14.:10.1016/j.semcdb.2019.1004.1013. Epub 2019 May 1011.
2.	Gill AJ (2018) Succinate dehydrogenase (SDH)-deficient neoplasia. Histopathology 72:106-116. doi: 110.1111/his.13277.
3.	Amorim-Pires D, Peixoto J, Lima J (2016) Hypoxia Pathway Mutations in Pheochromocytomas and Paragangliomas. Cytogenet Genome Res 150:227-241. doi: 210.1159/000457479. Epub 000452017 Feb 000457424.
4.	Kluckova K, Tennant DA (2018) Metabolic implications of hypoxia and pseudohypoxia in pheochromocytoma and paraganglioma. Cell Tissue Res 372:367-378. doi: 310.1007/s00441-00018-02801-00446. Epub 02018 Feb 00415.
5.	Dahia PL, Ross KN, Wright ME, et al (2005) A HIF1alpha regulatory loop links hypoxia and mitochondrial signals in pheochromocytomas. PLoS genetics 1:72-80
6.	Dando I, Pozza ED, Ambrosini G, et al (2019) Oncometabolites in cancer aggressiveness and tumour repopulation. Biol Rev Camb Philos Soc 94:1530-1546. doi: 1510.1111/brv.12513. Epub 12019 Apr 12510.
7.	Chen L, Liu T, Zhang S, et al (2014) Succinate dehydrogenase subunit B inhibits the AMPK-HIF-1alpha pathway in human ovarian cancer in vitro. J Ovarian Res 7:115.:10.1186/s13048-13014-10115-13041.
8.	Aspuria PP, Lunt SY, Varemo L, et al (2014) Succinate dehydrogenase inhibition leads to epithelial-mesenchymal transition and reprogrammed carbon metabolism. Cancer Metab 2:21.:10.1186/2049-3002-1182-1121. eCollection 2014.
9.	Karczewski KJ, Francioli LC, Tiao G, et al (2019) Variation across 141,456 human exomes and genomes reveals the spectrum of loss-of-function intolerance across human protein-coding genes. bioRxiv:531210

```{r, eval = TRUE}
library(dplyr)
library(ggplot2)
library(table1)
library(forcats)
```

```{r, eval = TRUE}
#Read in masked data files
masked_demographic <- read.csv("Data/SDHx_Masked&ReportedDeID_052820_Proband_Demographic.csv", header = TRUE)
masked_clinical <- read.csv("Data/SDHx_Masked&ReportedDeID_051420_Proband_Clinical.csv", header = TRUE)
masked_family <- read.csv("Data/SDHx_Masked&ReportedDeID_051420_Family_Clinical.csv", header = TRUE)

#Merge masked files for demographic and clinical data
masked <- left_join(masked_demographic, masked_clinical, by = "Project.ID")

#Read in reported data files
reported_demographic <- read.csv("Data/SDHx_Negative CasesDeID_051420_Proband_Demographic.csv", header = TRUE)
reported_clinical <- read.csv("Data/SDHx_Negative CasesDeID_051420_Proband_Clinical.csv", header = TRUE)
reported_family <- read.csv("Data/SDHx_Negative CasesDeID_051420_Family_Clinical.csv", header = TRUE)

#Merge reported files for demographic and clinical data
reported <- left_join(reported_demographic, reported_clinical, by = "Project.ID")

#Merge masked and reported datasets
c <- full_join(masked, reported, all=TRUE)
head(c)

#select columns and convert to factors
c2 <- c%>%
  select(X1st.degree.FHx.with.PGL.and.or.PCC, Test.Name.Condensed, Panel.Version, Genes.Tested, Ethnicity.Condensed,  Gender, Overall.Classification, Age...Year.Only, SDHA, SDHB, SDHC, SDHD, SDHAF2, SDHx.Zygosity, SDHx.Classification, SDHx.Gene, SDHx.C_variant.1, SDHx.P_variant.1, Zygosity, Classification, Gene, C_variant.1, P_variant.1, Classification.2, Gene.2, Classification.3, Gene.3, Classification.4, Gene.4, Classification.5, Gene.5, Cohort.x, Cohort.y, Cancer.Category1, Cancer.Site1, Cancer.Category2, Cancer.Site2, Cancer.Category3, Cancer.Site3, Cancer.Category4, Cancer.Site4, Cancer.Category5, Cancer.Site5, Cancer.Category6, Cancer.Site6, More_than_1_result, Details_more_than_1_result, Multiple.mutations.same.gene., Multiple_VUS_same_gene) %>%
  mutate(Gender= as.factor(Gender)) %>%
  mutate(Age= as.factor(Age...Year.Only)) %>%
  mutate(SDHx.Gene = as.factor (SDHx.Gene)) %>%
  mutate(Gene_1 = as.factor(Gene)) %>%
  mutate(SDH.Mut = ifelse(SDHx.Classification=="Mutation"|SDHx.Classification=="VLP" |SDHx.Classification =="Unclassified", 1, 0))

#consolidate levels for key variables
c2$Gender_Binary= fct_collapse(c2$Gender, 
    Female = c("F", "Female", "F/M"), 
    Male = c("M", "M/F"))
c2$Ethnicity_Race = fct_collapse(c2$Ethnicity.Condensed, 
    Ashkenazi_Jewish = c("ashkenazi jewish", "Ashkenazi Jewish"),
    Asian = c("asian", "Asian", "middle eastern", "Middle Eastern"),
    Black = c("african american/black", "African American/Black"),
    Caucasian = c("caucasian", "Caucasian"),
    Hispanic = c("hispanic", "Hispanic"),
    Other = c("mixed ethnicity", "mixed Ethnicity", "Mixed Ethnicity", "other", "Other", "Unknown", "unknown", "native american", "Native American"))

#characterize unique SDHx mutations and their frequencies
Supp_Table_1 <- c2 %>%
  count(SDHx.Gene, SDHx.C_variant.1, SDHx.P_variant.1) %>%
  group_by(SDHx.Gene) %>%
  arrange(desc(n), .by_group=TRUE)

print(Supp_Table_1)

table(c2$Test.Name.Condensed)
table(c2$Genes.Tested)
table(c2$Gender)
table(c2$Ethnicity.Condensed)
table(c2$Ethnicity_Race)
table(c2$SDHx.Classification)
table(c2$SDH.Mut)
table(c2$SDHx_Gene)
table(c2$Gene_1)

```
