---
title: "Retinal Bipolar Single Cell Analysis"
author: "Nichoals Dana"
output: 
  html_document:
    toc: false 
    depth: 3 
    theme: paper 
    highlight: tango
---
```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE,warning = FALSE,message = FALSE)
```
***
### Overview

The goal of this analysis is to take publicly available retinal single cell sequencing data and correctly identify cell types using R, the R package Seurat, and known applicable cell markers. The data to be analyzed will be the 25,000 retinal bipolar cells from the 2017 August paper, "COMPREHENSIVE CLASSIFICATION OF RETINAL BIPOLAR NEURONS BY SINGLE-CELL TRANSCRIPTOMICS" by Shekhar et al. (https://www.ncbi.nlm.nih.gov/pmc/articles/PMC5003425/) data publicly available at https://portals.broadinstitute.org/single_cell/study/retinal-bipolar-neuron-drop-seq. The proper identification of these cells will lead to a greater understanding of the roles each cell type preforms in the eye and could also help to identify novel gene markers for the cell types analyzed. These potential novel markers may help future studies successfully categorize retinal cells.

To properly understand and manipulate the data three UPenn faculty members were consulted. Dr. Dwight Stambolian, from the department of Ophthalmology and Human Genetics, was consulted to provided ideas and information on how the result of the analysis should appear and which graphics would be most useful to show any results. Dr. Mingyao Li, from the department of Biostatistics and Epidemiology was consulted to gain a better understanding of single cell sequencing data and what programs would be best suited to analyze such data. Dr. Randy J. Zauhar, Professor of Chemistry and Biochemistry, was able to give instruction on how to check the quality of a sequencing run. With the guidance provided by the above listed persons, the direction of the project was made clear.

Final GitHub Repository: https://github.com/NicholasDana/BMIN503_Final_Project

### Introduction 

The ability to correctly identify cell types plays a critical role when interpreting single cell sequencing data. Single cell sequencing grants insight into the role each cell may play in a cell population based on the genes expressed within each individual cell. As opposed to tissue sequencing which is limited to providing information between two distinct tissue types, such as a control tissue and a diseased tissue, single cell sequencing can provide answers to which genes in which cell are responsible for a manifested behavior. Due to this increased resolution of sequencing data, it has become critical to identify which cells are which and establish control expression profiles for cell types of interest.

The issue presented by this project requires knowledge on; how to interact with and manipulate single cell sequencing data, known biological markers of human retinal bipolar cells, how to preform and interpret statistical queries, and the role observed expressed genes preform. All these required disciplines make solving this issue of cell type identification a complicated process which requires continued re-analysis and interpretations with multiple cell populations. The guidance from the previously mentioned UPenn faculty members will hopefully allow this project to successfully identify cell types and possibly novel cell markers.

### Methods

The data to be analyzed is the sequencing data from the 25,000 retinal bipolar cells. The data comes in the form of an expression matrix, listing the expression data from each of the 25,000 cells for each gene observed. The data is stated to already be medium normalized and log transformed. Originally the data was analyzed by Shekhar et al. in R with thier own disclosed code and function, however this analysis will utilize the R package Seurat which was created specifically to interpret single cell data.

```{r eval=TRUE,results='hide'}
library(Seurat)
library(dplyr)
```

## Quality Control

The analysis starts with the creation of a Seurat object, which needs to recieve the expression data matrix. The Seurat object has numerous meta data and functions which are generated from the data upon creation. By plotting the number of genes and number of unique molecular identifiers (UMI) we can better understand the distribution of cells and use this information to apply potential filters. We can also look at the ident of the Seurat object to list all the cells and see any cell groupings that may be  present in the data.

```{r eval=TRUE}
setwd("~/BMIN503_Final_Project/")
shekhar.data=read.table(paste0("Retinal Bipolar Neurons/exp_matrix.txt"),header = TRUE,row.names = 1)
shekhar=CreateSeuratObject(raw.data = shekhar.data)
head(shekhar@ident)
table(shekhar@ident)
VlnPlot(object = shekhar,features.plot = c("nGene"))
VlnPlot(object = shekhar,features.plot = c("nUMI"))
```

We can see that there are 6 groupings of cells unevenly distributed, possibly indicating the cells were sequenced in 6 batches, Bipolar1 through Bipolar6. Looking at the number of genes and unique molecular identifiers we can see some outlyer cells which may be cell doubets, two different cells which were sequenced together as a single cell, that we may decide to filter out. Each point on the graph represents an individual cell.

The mitochondial genes are also commonly used to filter out unwanted cells. By identifying all expressed mitochondrial genes contained in the expression matrix, and then comparing the percentage of mitochondrial expression to all other genes, we can determine how much of a cell is composed of mitochondrial genes.

```{r eval=TRUE}
shekhar.mito.genes=grep(pattern = "mt-",x = rownames(x = shekhar@raw.data),value = TRUE)
shekhar.mito.genes
length(shekhar.mito.genes)
shekhar.mito=colSums(shekhar@raw.data[shekhar.mito.genes,])/colSums(shekhar@raw.data)
shekhar=AddMetaData(object = shekhar,metadata = shekhar.mito,col.name = "percent.mito")
VlnPlot(object = shekhar,features.plot = c("percent.mito"))
```

There are a total of 20 mitochondrial genes listed above. Based on the distribution of cells by mitochondrial percentage, it is observed that the bulk of cells tend to have around 12-15% mitochondrial content. It may be a good idea to filter out cells with more than 30% mitochondrial content as a means to reduce the number of cell doublets in the data, unless it is anticipated that a rare cell type may have high mitochondrial content. To better understand the data quality we can plot cells by multiple variables, and observe any obvious outlyers to help set potential filter levels.

```{r eval=TRUE}
GenePlot(object = shekhar,gene1 = "nUMI",gene2 = "percent.mito")
GenePlot(object = shekhar,gene1="nUMI",gene2 = "nGene")
GenePlot(object = shekhar,gene1 = "percent.mito",gene2 = "nGene")
```

Using the above information as a qualitative measure of the data, cell filters were set for the number of genes, the number of unique molecular identifiers, and the percentage of mitochondrial content. Any cell with more than 3000 expressed genes, 30% mitochondrial gene content, or 950 unique molecular identifiers were filtered out of the data. To see the effect of these filters, the previous plots will be reproduced on the filtered data.

```{r eval=TRUE}
shekhar=FilterCells(object = shekhar,subset.names = c("nGene","percent.mito","nUMI"),low.thresholds = c(0,-Inf,0),high.thresholds = c(3000,0.3,950))
VlnPlot(object = shekhar,features.plot = c("nGene","nUMI","percent.mito"),nCol = 3,ident.include = shekhar@ident)
par(mfrow=c(1,3))
GenePlot(object = shekhar,gene1 = "nUMI",gene2 = "percent.mito")
GenePlot(object = shekhar,gene1="nUMI",gene2 = "nGene")
GenePlot(object = shekhar,gene1 = "percent.mito",gene2 = "nGene")
```

Genes that show high variablity between the cells are now identified by calculating the average expression and dispersion for each gene, a list of these genes is stored within the Seurat object @var.genes. These genes will be important as a good biological marker should show relatively high expression in only the cell type of interest. We would expect a good marker to have a low average expression and high variability across all cells, while having a high average expression and low variability in the cells it is expected to mark. After looking the mean variability plot, the cutoff parameters were set to mark visual outlyers on the plot. With the current parameters there are 347 variable genes identified, which could be potential cell markers. 

```{r eval=TRUE}
shekhar=FindVariableGenes(object = shekhar,mean.function = ExpMean,dispersion.function = LogVMR,x.low.cutoff = 0.0125,x.high.cutoff = 1.75,y.cutoff = 1.5)
str(shekhar@var.genes)
```

The dataset may also contain variation from unintresting sources, such as batch effects and biological information such as mitochondrial expression. Removing these sources of variation can allow for greater accuracy in determining intresting cell type gene expression variation. This is accomplished by constructing linear models to predict and scale the gene expression based only on the variables we find interesting. The effect of the number of molecular identifiers and mitochondrial content will be mitigated from the dataset. Normally the data would be normalized before scaling the data, however the data was stated to be medium normalized and log transformed, so this step was omitted.

```{r eval=TRUE}
shekhar=ScaleData(object = shekhar,vars.to.regress = c("nUMI","percent.mito"))
```

## Principle Component Analysis

Principle component analysis will be preformed to better understand the variance in this multivariate dataset. A total of 30 principle components will be calculated for the scaled dataset using the previous 347 highly variable genes. Each PC calculated will effectively represent an extra layer of data for a set of correlated genes from the variable genes list. The strength of the correlated genes included in the analysis will ultimately decide how the cells are eventually clustered together, so determining how many PCs to include will have a large effect on the data.

```{r eval=TRUE}
shekhar=RunPCA(object = shekhar,pc.genes = shekhar@var.genes,do.print = FALSE, pcs.print = 1:5,genes.print = 5,pcs.compute = 30)
```

The principle components of the analysis will be visualized and reviewed using 4 methods. First by Viewing the cells and genes that define each principle component. 

```{r eval=TRUE}
PrintPCA(object = shekhar,pcs.print = 1:15, genes.print=10, use.full = FALSE)
```

By plotting the top genes assiociated with principle components.

```{r eval=TRUE}
VizPCA(object = shekhar,pcs.use = 1:9)
VizPCA(object = shekhar,pcs.use = 10:18)
VizPCA(object = shekhar,pcs.use = 19:27)
```

By plotting the output of principle components against each other. 

```{r eval=TRUE}
PCAPlot(object = shekhar, dim.1=1,dim.2=2)
PCAPlot(object = shekhar, dim.1=1,dim.2=15)
PCAPlot(object = shekhar, dim.1=1,dim.2=30)
```

Finally the principle components will be visualized by use of heatmaps.

```{r eval=TRUE}
shekhar=ProjectPCA(object = shekhar,do.print = FALSE)
#PCHeatmap(object = shekhar,pc.use = 1,cells.use = 500,do.balanced = TRUE,label.columns = FALSE)
PCHeatmap(object = shekhar,pc.use = 1:9,cells.use = 400,do.balanced = TRUE,label.columns = FALSE,use.full = FALSE)
PCHeatmap(object = shekhar,pc.use = 10:18,cells.use = 400,do.balanced = TRUE,label.columns = FALSE,use.full = FALSE)
PCHeatmap(object = shekhar,pc.use = 19:27,cells.use = 400,do.balanced = TRUE,label.columns = FALSE,use.full = FALSE)
```

As determining the true dimensionality of the dataset and how many PCs to include has a great influence on the clustering and identification of cell types, many approaches to analyze each PC must be considered. Jack Straw plots allow for the visualization of gene p-values within the plotted PC. By comparing the distribution of p-values between the 30 PCs computed, a more informed decision on which PCs to include downstream can be made. Significant PCs should have distribution curves above the dashed line, or null distribution. A plot of the standard deviation for each PC will also be reviewed, to determine if there is any prominent feature that can be used to determine a cut off PC.

```{r eval=TRUE}
shekhar=JackStraw(object = shekhar,num.replicate = 100,do.print = FALSE,num.pc = 30)
JackStrawPlot(object = shekhar,PCs = 1:30,nCol = 5)
PCElbowPlot(object = shekhar,num.pc = 30)
```

Comparing the distribution of p-values for each PC, the last significant PC appears to be PC 21, with PC 22 and onward beginning to touch and fall under the null distribution. Also comparing the standard deviations seems to show that beyond PC 19-23 the change in standard deviation becomes miniscule. As the other methods previously implemented did not present an imediately apparent significant PC cut-off point, the analysis will continue using 21 PCs.

## Cell Clustering and Identification

Cell clustering is implemented in Seurat by a shared nearest neighbor (SNN) modularity optimization based algorithm. This algorithm is capable of finding clusters of varying shapes, sizes, and densities in high dimensional data. In short it would compute a similarity matrix based on the imputed PCs, determine core points, and form clusters based on the core points. It is expected that cells of the same cell type would cluster together. To visualize the clusters t-distributed Stochastic neighbor Embedding (tSNE) dimensionality reduction will be implemented, this is a machine learning algorithm, non-linear dimensionality redution technique that is capable of embedding high dimensional data into two or three dimensions which can then be plotted.

```{r eval=TRUE}
shekhar=FindClusters(object = shekhar,reduction.type = "pca",dims.use = 1:21,resolution = 0.6,print.output = 0,save.SNN = TRUE)
shekhar=RunTSNE(object = shekhar,dims.use = 1:21,do.fast = TRUE)
PrintFindClustersParams(object = shekhar)
TSNEPlot(object = shekhar,do.label = T)
```

For each of the clusters generated, differentially expressed genes are identified to determine potential markers.

```{r eval=TRUE}
library(dplyr)
shekhar.markers=FindAllMarkers(object = shekhar,only.pos = TRUE,min.pct = 0.25,thresh.use = 0.25)
shekhar.markers %>% group_by(cluster) %>% top_n(3,avg_logFC)
```

These markers will be compared to the markers identified by the Shekhar et. al. paper to identify clusters that exhibit the same gene expressions. 

### Results
Describe your results and include relevant tables, plots, and code/comments used to obtain them. End with a brief conclusion of your findings related to the question you set out to address. You can include references if you'd like, but this is not required.

The goal of this project is to correctly identify the cell types of the 25,000 bipolar cells presented by Shekhar et al. The identification of the cells will be done by comparing the top differentially expressed genes, or markers, for each cluster to known bipolar cell markers. The accuracy of this cell identification process will be checked by comparison to the clusters and differentially expressed genes of the same clusters found by Shekhar et al. As they originally analyzed this data, this analysis would be expected to at least partially match cell clusters, even though the methods used in each analysis are different than one another.

The shekhar paper utilized three different principle component analysis and clustering tenchinques/analysis to arrive at three different clustering possibility for the data. These analyses lead to the formation of 26, 31, and 29 clusters, of which the Louvain-Jaccard generated clustering of 26 was selected as the best representation. The analysis presented in this project yeilded 18 clusters. Shekhar et al. was able to identify 14 bipolar cell cluster amoung their 26 by checking for the expression of Prkca, which is indicitive of bipolar rod cell types, and Scgn, which is indicitive of bipolar cone cell types.

```{r eval=TRUE}
VlnPlot(object = shekhar,features.plot = c("Prkca")) #Bipolar Rod Cells
FeaturePlot(object = shekhar, features.plot = c("Prkca"), cols.use = c("grey", "blue"), reduction.use = "tsne")
VlnPlot(object = shekhar,features.plot = c("Scgn"))  #Bipolar Cone Cells
FeaturePlot(object = shekhar, features.plot = c("Scgn"), cols.use = c("grey", "blue"), reduction.use = "tsne")

VlnPlot(object = shekhar,features.plot = c("Isl1"))  #Bipolar On Cone Cells
FeaturePlot(object = shekhar, features.plot = c("Isl1"), cols.use = c("grey", "blue"), reduction.use = "tsne")
VlnPlot(object = shekhar,features.plot = c("Grm6"))  #Bipolar Off Cone Cells
FeaturePlot(object = shekhar, features.plot = c("Grm6"), cols.use = c("grey", "blue"), reduction.use = "tsne")
```

The above plots visualize the expression value of each of the 4 mentioned genes with the cells categorized into one of the 18 clusters, or by indicating the location of the cell on the tSNE plot. From these specific gene expressions we can conclude that: clusters 1, 2, and 4 correspond to rod bipolar cells; clusters  3, 5, 6, 7, and 14 correspond to ON-cone bipolar cells; and clusters 8, 9, 10, 11, and 12 correspond to OFF-cone bipolar cells. This identifies 13 total bipolar cell clusters to known markers. Cluster 16 may also be identifiable as ON-cone bipolar cells, as the known ON-cone bipolar markers show decent expression in those cells. Including cluster 16 that brings the total identified bipolar clusters from this analysis to 14, matching those completed by the literature analysis. 

```{r eval=TRUE}

top10=shekhar.markers %>% group_by(cluster) %>% top_n(10,avg_logFC)
DoHeatmap(object = shekhar, genes.use = top10$gene,slim.col.label = TRUE,remove.key = TRUE)
```





