---
title: "BMIN503/EPID600 Project Template"
author: "Taha Iqbal"
output: 
  html_document:
    toc: false 
    depth: 3 
    theme: paper 
    highlight: tango
---



### Overview
Analyze GWAS Autosomal Datasets and determine genes/variants associated with AD and their chromosomal locations. Use the FUMA platform to 
annotate and map these associations to their perspective genes and visualize GWAS results. Compare results with that of well known associations and determine the role of most significant genes and their pathways.

Use ADC10 study --> Use SNPTEST to generate GWAS Summary Output --> Use Rscript to clean and generate plots --> Use FUMA to annotate and get functional genomic information



### Introduction 
Alzheimerâ€™s Disease is a progressive neurodegenerative disease associated with cognitive decline. 13% of people over age 65 and 45% of people over age 85 are estimated to have AD. AD can be classified into 2 categories, Late-onset AD (95%) and Early-onset AD (5~). EOAD is known to have a significant genomic component with APP, PSEN1, PSEN2 as main disease-causing genes. For LOAD one of the most significant genetic risk factors are the different alleles of APOE. Gene polymorphisms such as e4 and e2 are implicated. 

In this project I plan to find significantly associated variants with AD and use functional genomics to annotate their locations, determine the role and pathways implicated of these genes. 

The data used is ADC10 which is from the dataset of ADC1-10. This NIA ADC cohort includes subjects evaluated by the clinical and neuropathology core. Data collection is coordinated by the National Alzheimer's Coordinating Center (NAAC). NAAC coordinates collection of phenotypes from ADCs, cleans all data, coordinates implementation of Alzheimer's definitions cases and controls, and collection of samples. The cohort contains 3,311 autopsy-confirmed cases and 2,889 clinically-confirmed disease cases, and 247 cognitively normal elders (CNEs) with complete neuropathology data > 60 years old at death, and 3,687 living CNEs.


### Methods
Input Data: 
  - ADC 10: Data from NIA ADC cohort which includes subjects evaluated by the clinical and neuropathology core. Data collection is coordinated by the National Alzheimer's Coordinating Center (NAAC). NAAC coordinates collection of phenotypes from ADCs, cleans all data, coordinates implementation of Alzheimer's definitions cases and controls, and collection of samples. 

Tools:
  - SNPTESTv2.5.6: Generate GWAS summary results using ADC10 dataset
  - R: To clean data, add rsid annotations, and generate Manhattan and QQ plots
  - FUMA: A web platform to functionally annotate and get in-depth genomic results of variants 


Data is initially prepared using SNPTEST, a tool to analyze SNP association in GWAS.

SNPTEST Commands:

 $snptest 
-data genfile/chr*.gen.gz Covar/adc10.covar.snptest.txt 
-missing_code NA,-9 
-hwe 
-frequentist 1 
-method score 
-pheno status 
-cov_names pc1 pc2 pc3 
-exclude_samples_where omit=1 
-o result/snptest/model1/chr*.out

**Commands for SNPTEST Can be seen in greater detail in Alzheimers_GWAS.pptx**


RCode:

```{r, eval=TRUE}
BiocManager::install('CMplot')
BiocManager::install('SNPlocs.Hsapiens.dbSNP144.GRCh37')

library("CMplot")
library(SNPlocs.Hsapiens.dbSNP144.GRCh37)
library(vroom)
library(plyr)
library(purrr)
library("readr")

#For Annotation, Retrieve all annotated SNPS from SNPlocs.Hsapiens.dbSNP144.GRCh37 package:
all_snps <- SNPlocs.Hsapiens.dbSNP144.GRCh37


#Loop through annotated SNPS and assign to dataframe per chromosome
final_snp_df = data.frame()

for (i in 1:22){
  # vector output
  model <- snpsBySeqname(all_snps, as.character(i))
  df <- data.frame(c(i),c(model@ranges@pos), c(model@elementMetadata@listData[["RefSNP_id"]]))
  #print(paste0(model@ranges@pos)
  
  # add vector to a dataframe
  final_snp_df <- rbind(final_snp_df,df)
  
  print(paste0("Done loading SNPS for chr: ", i))
}


#Rename columns
colnames(final_snp_df) <- c('Chromosome','Position', 'RefSNP_id')
#Combine chr + pos into new column to be used for merge with GWAS summary statistics data later:
final_snp_df$rsid <- paste0(final_snp_df$Chromosome, ":", final_snp_df$Position)
#Select only needed columns (Can remove Chromosome now):
final_snp_df <- subset(final_snp_df, select = c("rsid", "RefSNP_id", "Position"))



#Load in GWAS summary statistics:

#Load in data of chr1-22
list_of_files <- list.files(path = "/Users/tahai/Documents/GWAS/chr/", recursive = TRUE,
                            pattern = "\\.output$", 
                            full.names = TRUE)

#Load in files from list_of_files vector, choose specific columns, rename them, and assign rsid column as character because default is Data/time
data_all = map_df(list_of_files, ~vroom(.x, col_select = c("rsid", Chromosome="alternate_ids", Position="position", pvalue="frequentist_add_pvalue"), col_types = c(rsid = 'c')))

#Clean data: Remove NAs
data_all <- na.omit(data_all)

#Merge SNP annotations and GWAS df by rsid:
final_df <- merge(final_snp_df, data_all, by="rsid")

#Assign Genes vector to be highlighted in Manhattan Plot
GENES <- final_df[final_df[,6] < (0.05 / nrow(final_df)), 2]

#Reorder / Drop unwanted columns 
final_df <- final_df[, c("rsid", "Chromosome", "Position.x", "pvalue")]
colnames(final_df) <- c("rsid", "Chromosome", "Position", "pvalue")


#Sort Data and export top 50 SNPS:
sorted_df <- final_df[order(final_df$pvalue),]
df_export <- head(sorted_df,50,)

write_tsv(df_export, path="Top50Snps.tsv")

#Assign SNPS < p_value threshold to be highlighted in Manhattan_plot
SNPs <- final_df[final_df[,4] < (0.05 / nrow(final_df)), 1]


#Plots:

#QQ Plot
CMplot(final_df,plot.type="q",box=FALSE,file="jpg",memo="",dpi=300,
       conf.int=TRUE,conf.int.col=NULL,threshold.col="red",threshold.lty=2,
       file.output=TRUE,verbose=TRUE,width=5,height=5)

#ManhattanPlot highlighting RSids
CMplot(final_df, plot.type="m",multracks=TRUE,threshold=c(1e-8,1e-6),threshold.lty=c(1,2), 
       threshold.lwd=c(1,1), threshold.col=c("black","grey"), highlight.text=GENES, amplify=TRUE,bin.size=1e6,
       chr.den.col=c("darkgreen", "yellow", "red"), signal.col=c("red","green","blue"),
       signal.cex=1, file="jpg",memo="",dpi=300,file.output=TRUE,verbose=TRUE,
       highlight=SNPs, highlight.text.cex=1.4)

```



### Results
**Chr1-22 Manhattan plot with Gene annotations shown**
**QQ-Plot showing significant association shown**
**FUMA genomic annotations for Chr19 shown**

Manhattan plot indicates significant hits of SNPS with Lowest P-value of 8.4x10-54 Chr19 at around 45Mb (APOE region).

Genes involved in this region are APOE, and NECTIN2.

CTNNA2 gene on Chr 2 with a P-value of 4.02x10-24 also seen, as well as LSAMP with a p-value of 4.47X10-24 seen on Chr3.
FBH1 seen on Chr 10 with a P-value of 2.07Ex10-21 and a Snp with Rsid of rs542543824 seen on chr8 with a P-value of 2.15x10-21.



Lead SNP is rs429358 with a p-value of 8.4x10-54-. **SNP of APOE 388 T>C **

**.Tsv file showing top 50 Significant SNPS showin in Top50SNPS.tsv**


The QQ plot shows multiple significant SNPS much smaller than from expectation. Plot indicates SNPs with significant P-value deviation than expected by chance alone.


FUMA platform output for Chr19indicates interesting results:
  *Summary of SNPs and mapped genes:*
    - 16 genomic risk loci
    - 32 lead snps
    - 48 significant SNPs < 1x10-8
    - 326 candidate SNPs 
    - 209 mapped genes
  
  *Functional consequence of SNPS on gene:*
    - 42% of SNPS were in intronic regions. Second smallest P-value for this range
    - 32% of SNPS were in intergenic regions. **Most significant**
    
  

