---
title: "BMIN503/EPID600 Project Template"
author: "Katie Tsukahara"
output: 
  html_document:
    toc: false 
    depth: 3 
    theme: paper 
    highlight: tango
---

```{r set-options, echo=FALSE, cache=FALSE}
options(width = 400)
```  
***
Use this template to complete your project throughout the course. Your Final Project presentation will be based on the contents of this document. Replace the title/name above and text below with your own, but keep the headers.

### Overview
Give a brief a description of your project and its goal(s), what data you are using to complete it, and what three faculty/staff in different fields you have spoken to about your project with a brief summary of what you learned from each person. Include a link to your final project GitHub repository.


>OVERVIEW:<br>Childhood interstitial lung disease (chILD) is a heterogeneous group of rare childhood lung diseases affecting the interstitial region of the lung, leading to abnormal gas exchange and a range of morbidity and mortality. Research and development of therapies is challenging due to the low prevalence and heterogeneity of disease, thus the CHILD Research Network and Registry was established to collect data from multiple centers across the United States. Our goal is to compare patient demographics across subsets of chILD subjects by diagnosis, and to examine the diagnostic and clinical factors that contribute to diagnosis.


### Introduction 
Describe the problem addressed, its significance, and some background to motivate the problem.

Explain why your problem is interdisciplinary, what fields can contribute to its understanding, and incorporate background related to what you learned from meeting with faculty/staff.

><br><br>INTRODUCTION:  
Childhood interstitial lung disease (chILD) encompasses a variety of lung diseases that manifest via changes to the pulmonary interstitium. The pulmonary interstitium is, simply put, the lung tissue between the airspace and the pulmonary blood vessels, and excludes other lung structures such as the airways. These diseases are grouped together owing to similarities in the area of the lung affected and in clinical presentation. While both children and adults can develop interstitial lung disease (ILD), many causes of chILD are seen only in the pediatric age group. These diseases can be genetic in origin, or associated with abnormal lung development, immunologic dysfunction, rheumatologic disease, or infectious history. Some cases do not have a clear etiology. Certain diagnoses confer a poor prognosis of early lung transplant or mortality. Other diagnoses confer a more positive prognosis, including clinical improvement and resolution of disease over early childhood.
<br><br>To study this group of diseases, the Children’s Interstitial and Diffuse Lung Disease Research Network (ChILDRN) established a national registry in 2016, with participation by 13 sites across the United States. The original registry contains 436 subjects, enrolled from 2016 to 2020, with data on demographics, medical history, diagnostics, management and outcomes. With this project, we aim to examine differences in patient demographics, including age of onset and severity of presentation, among the more common chILD diagnoses. We also aim to examine the diagnostic and clinical factors that contribute to diagnosis, if possible examining the role that different diagnostic tests play in diagnosis and whether invasive testing can be avoided in these patients that are at higher risk for procedures due to their underlying pulmonary disease. The database and this subsequent project use the skills and expertise of clinicians across many fields (Pulmonology, Radiology, Pathology, Rheumatology, and others), clinical researchers, basic science researchers, and biomedical informaticians to collect, manage, and interpret data on this complex group of patients.



### Methods
Describe the data used and general methodological approach. Subsequently, incorporate full R code necessary to retrieve and clean data, and perform analysis. Be sure to include a description of code so that others (including your future self) can understand what you are doing and why. 




```{r, include=FALSE, echo = TRUE}
# loading RedCAP data
setwd("C:/Users/tsukaharak/Documents/R files/BMIN503_Final_Project")
source("CHOPChildhoodILDNati-110620BMIN_R_2020-11-27_1739.R", local = knitr::knit_global())
# or sys.source("your-script.R", envir = knitr::knit_global())

# checking for duplicate entries
length(unique(data$child_registry_id))
length(data$child_registry_id)

child <- data

```


```{r, echo = FALSE, warning = FALSE, message = FALSE}
# loading libraries for data cleaning and graphics
library(tidyr)
library(dplyr)
library(ggplot2)
library(cowplot)

```



### Results
Describe your results and include relevant tables, plots, and code/comments used to obtain them. End with a brief conclusion of your findings related to the question you set out to address. You can include references if you'd like, but this is not required.

```{r, warning = FALSE, message = FALSE, eval = TRUE}
library(maps)

head(us.cities)
DAG.df <- data.frame(
  name = c("Boston MA", "Atlanta GA", "Philadelphia PA", "Cleveland OH", "Queens NY", "Aurora CO", "Indianapolis IN", "Kansas City MO", "Chicago IL", "Ann Arbor MI", "Columbia MO", "Bronx NY", "Portland OR", "Pittsburgh PA", "San Diego CA", "Seattle WA", "Palo Alto CA", "Chapel Hill NC", "Nashville TN", "Saint Louis MO"),
  prefix = c(704,720,712,709,702,703,718,711,717,716,715,708,713,705,706,710,707,719,701,714),
  stringsAsFactors = FALSE
)

DAG.df <- DAG.df %>% 
  arrange(prefix) %>%
  left_join(us.cities, by = "name") %>%
  mutate(prefix = as.character(prefix))

# latitude and longitude for NY boroughs
# Queens 40.728, -73.795
# Bronx 40.845, -73.865
DAG.df[which(grepl("Queens",DAG.df$name)), ]$lat <- 40.7282
DAG.df[which(grepl("Queens",DAG.df$name)), ]$long <- -73.7949
DAG.df[which(grepl("Bronx",DAG.df$name)), ]$lat <- 40.8448
DAG.df[which(grepl("Bronx",DAG.df$name)), ]$long <- -73.8648
DAG.df

DAG.child <- child %>%
  select(child_registry_id) %>%
  mutate(prefix = substr(child_registry_id, 0, 3)) %>%
  count(prefix) %>%
  mutate(prefix = as.character(prefix)) %>%
  left_join(DAG.df, by = "prefix")

USA <- map_data("world") %>% filter(region=="USA")


# Create breaks for the color scale
mybreaks <- c(20,40,60,80,100)

# Build the map
ggplot() +
    geom_polygon(data = USA, aes(x=long, y = lat, group = group), fill="grey", alpha=0.4) +
    geom_point(data = DAG.child, aes(x=long, y=lat, size=n), color = "#2b8cbe", shape=20, stroke=FALSE) +
    scale_size_continuous(name="Number of Enrolled \nSubjects", trans="identity", range=c(0,12), breaks = mybreaks) +
    theme_void() +
    coord_map() + ylim(25,50) + xlim(-130, -60) +
    guides(colour = guide_legend()) +
    ggtitle("Subject Enrollment by Center") +
    theme(
      legend.position = c(0.85, 0.15),
      text = element_text(color = "#22211d"),
      plot.title = element_text(size= 16, hjust=0.1, color = "#4e4d47", margin = margin(b = -0.1, t = 0.4, l = 2, unit = "cm")),
    )

```




```{r, warning = FALSE, message = FALSE}
# enrollment statistics
summary(child$age_studyentry)
ggplot(data = child, aes(x = age_studyentry)) + 
  geom_histogram(color = "black", fill = "white", binwidth=3) +
  labs(x = "Age (in months)", title = "Age at Study Entry in chILD Registry")
summary(child$age_diagnosis)
ggplot(data = child, aes(x = age_diagnosis)) + 
  geom_histogram(color = "black", fill = "white", binwidth=3) +
  labs(x = "Age (in months)", title = "Age at Diagnosis")

```


```{r}
# race across the registry

library(dplyr)
child.race <- child %>%
  select(child_registry_id, race___1:race___6, race___1.factor:race___6.factor, diagnosis.factor) %>%
  mutate(race.sum = rowSums(select(child.race, race___1:race___6), na.rm = TRUE)) %>%
  mutate(race.text = NA)

table(child.race$race.sum)

N = nrow(child.race)
for(i in 1:N){
  if(child.race[i,"race.sum"] == 0){
    child.race[i,"race.text"] <- "Unknown"
  }else if(child.race[i,"race.sum"] == 2){
    child.race[i,"race.text"] <- "Two or more races"
  }else{
    if(child.race[i,"race___1"] == 1){
      child.race[i,"race.text"] <- "White"
    }else if(child.race[i,"race___2"] == 1){
      child.race[i,"race.text"] <- "Black"
    }else if(child.race[i,"race___3"] == 1){
      child.race[i,"race.text"] <- "Asian"
    }else if(child.race[i,"race___4"] == 1){
      child.race[i,"race.text"] <- "Native Hawaiian/Pacific Islander"
    }else if(child.race[i,"race___5"] == 1){
      child.race[i,"race.text"] <- "American Indian or Alaska Native"
    }else if(child.race[i,"race___6"] == 1){
      child.race[i,"race.text"] <- "Other"
    }else{print(i)}
  }
}
child.race <- child.race %>%
  mutate(race.text = factor(race.text, levels = c("White", "Black", "Asian", "Native Hawaiian/Pacific Islander", "American Indian or Alaska Native", "Other", "Two or more races", "Unknown")))


```



```{r}
# attempt at a Table 1
# not able to download ethnicity data
library(gtsummary)
library(gt)

table.data <- child %>%
  dplyr::select(child_registry_id, age_studyentry, age_diagnosis, patient_gender.factor,  relative_in_registry.factor, fibrosis.factor, airway.factor, home_o2.factor, mech_vent___1.factor, mech_vent___2.factor, transplant_yn.factor) 
levels(table.data$mech_vent___1.factor)=c("No","Yes")
levels(table.data$mech_vent___2.factor)=c("No","Yes")

table.data <- table.data %>%
  rename("Patient Gender" = patient_gender.factor) %>%
  rename("Relative in Registry" = relative_in_registry.factor) %>%
  rename("Pulmonary Fibrosis" = fibrosis.factor) %>%
  rename("Airway Disease" = airway.factor) %>%
  rename("Past or Current Home Supplemental Oxygen" = home_o2.factor) %>%
  rename("Past or Current Noninvasive Ventilation (>3 weeks)" = mech_vent___1.factor) %>%
  rename("Past or Current Invasive Ventilation (>3 weeks)" = mech_vent___2.factor) %>%
  rename("Lung Transplant" = transplant_yn.factor) %>%
  mutate(Death = ifelse(!is.na(child$death_age), "Yes", "No"))


table.one <- tbl_summary(select(table.data, -child_registry_id),  missing = "no") %>%
as_gt() %>%
gt::tab_header("Table 1. Summary of Clinical Characteristics of the Study Cohort") %>%
opt_align_table_header(align = "left") %>%
gt::tab_options(table.font.size = 'small')
table.one


table.data.race <- left_join(table.data, select(child.race, child_registry_id, race.text), by = "child_registry_id") %>%
  relocate(race.text, .after = "Patient Gender") %>%
  rename("Patient Race" = race.text)
  

table.two <- tbl_summary(select(table.data.race, -child_registry_id),  missing = "no") %>%
as_gt() %>%
gt::tab_header("Table 1. Summary of Clinical Characteristics of the Study Cohort") %>%
opt_align_table_header(align = "left") %>%
gt::tab_options(table.font.size = 'small')
table.two

```





```{r, warning = FALSE, message = FALSE}
# Breakdown of primary diagnoses
ggplot(data = child, aes(x = diagnosis.factor)) + 
  geom_bar() + 
  theme(axis.text.x = element_text(angle = 45, hjust = 1)) +
  labs(x = "Primary Diagnosis", y = "Number of Subjects", title = "Primary Diagnoses in chILD Registry")
```


```{r, warning = FALSE, message = FALSE}
# Analysis of gender across diagnoses - chi square test and visualization
p1chi <- chisq.test(table(child$diagnosis.factor, child$patient_gender.factor))
p1chi

ggplot(data = child, aes(x = diagnosis.factor, fill = patient_gender.factor)) + 
  geom_bar(position = "fill") + 
  theme(axis.text.x = element_text(angle = 45, hjust = 1)) +
  labs(x = "Primary Diagnosis", y = "Proportion of Total", title = "Gender by Diagnosis in chILD Registry", fill = "Gender",subtitle = paste("Chi-squared p value =", format(p1chi$p.value, digits = 3)))


```


```{r, echo = TRUE, warning = FALSE, message = FALSE}
# Analysis of race across diagnoses
# Adding the race as factor column if it has not already been added
if(!"race.text" %in% colnames(child)){
child <- left_join(child, select(child.race, child_registry_id, race.text), by = "child_registry_id")
}
head(child$race.text)

ggplot(data = child, aes(x = diagnosis.factor, fill = race.text)) + 
  geom_bar(position = "fill") + 
  theme(axis.text.x = element_text(angle = 45, hjust = 1)) +
  labs(x = "Primary Diagnosis", y = "Proportion of Total", title = "Race by Diagnosis in chILD Registry", fill = "Race")

```



```{r, echo = TRUE, warning = FALSE, message = FALSE}
# age across diagnosis

summary(data$age_diagnosis)
mean(data$age_diagnosis, na.rm = TRUE)

ggplot(data = filter(child, child$diagnosis < 10), aes(x = diagnosis.factor, y = age_diagnosis)) + 
  geom_boxplot() + 
  theme(axis.text.x = element_text(angle = 45, hjust = 1)) +
  labs(x = "Primary Diagnosis", y = "Age (in Months)", title = "Age at Diagnosis in chILD Registry")

```




```{r, warning = FALSE, message = FALSE}
# looking at age at diagnosis and diagnostic modality

child.mod <- select(child, child_registry_id:diagnosis, fibrosis.factor:diagnosis.factor)
child.mod <- child.mod %>%
  mutate(modality = ifelse(child.mod$child_dia2___1 == 1 & child.mod$child_dia2___2 == 0 & child.mod$child_dia2___3 == 0 & child.mod$child_dia2___4 == 0, "CT only", NA)) %>%
  mutate(modality = ifelse(child.mod$child_dia2___2 == 1 & child.mod$child_dia2___1 == 0 & child.mod$child_dia2___3 == 0 & child.mod$child_dia2___4 == 0, "Biopsy only", modality)) %>%
  mutate(modality = ifelse(child.mod$child_dia2___3 == 1 & child.mod$child_dia2___1 == 0 & child.mod$child_dia2___2 == 0 & child.mod$child_dia2___4 == 0, "Genetic only", modality)) %>%
  mutate(modality = ifelse(child.mod$child_dia2___2 == 0 & child.mod$child_dia2___1 == 0 & child.mod$child_dia2___3 == 0 & child.mod$child_dia2___4 == 0, "No Diagnostic Modality Listed", modality)) %>%
  mutate(modality = ifelse(child.mod$child_dia2___4 == 1 & child.mod$child_dia2___1 == 0 & child.mod$child_dia2___2 == 0 & child.mod$child_dia2___3 == 0, "Other", modality)) %>%
  mutate(modality = ifelse(is.na(modality) == "TRUE", "Multiple", modality)) %>%
  mutate(modality.factor = as.factor(modality))
child.mod <- rename(child.mod, "Modality" = modality.factor)

tbl_summary(select(child.mod, "Modality"),  missing = "no") %>%
as_gt() %>%
gt::tab_header("Table 2. How was the diagnosis made?") %>%
opt_align_table_header(align = "left") %>%
gt::tab_options(table.font.size = 'small')


ggplot(data = filter(child.mod, !is.na(diagnosis)), aes(x = diagnosis.factor, fill = Modality)) +
  geom_bar(position = position_dodge(preserve = "single")) + 
  theme(plot.title = element_text(size=14, vjust=3),
  axis.text.x = element_text(angle = 45, hjust = 1, size = 8), 
  legend.title = element_text(size = 10), 
  legend.text = element_text(size = 8)) +
  labs(x = "Primary Diagnosis", fill = "Diagnostic Modality", title = "The chILD Registry: How was the diagnosis made?")

```




```{r, warning = FALSE, message = FALSE}
child.surf <- filter(child, diagnosis.factor == "Surfactant dysfunction disorder")
print(paste0("In the chILD Registry, there are ", nrow(child.surf), " entries with a primary diagnosis of surfactant dysfunction disorder"))

ggplot(data = filter(child.mod, diagnosis.factor == "Surfactant dysfunction disorder"), aes(x = modality.factor)) +
  geom_bar() + 
  scale_x_discrete(drop=FALSE) +
  theme(axis.text.x = element_text(angle = 45, hjust = 1)) +
  labs(x = "Diagnostic Modality", title = "The chILD Registry: Diagnosis of Surfactant Dysfunction Disorders")


# colSums(select(child.surf, child_dia2___1:child_dia2___4), na.rm = TRUE)
# head(filter(child.surf, child_dia2___4 == 1) %>% select(child_registry_id:child_dia2___4)) - shows only 1 entry

library(VennDiagram)
library(magick)

venn.ct <- child.surf %>% filter(child_dia2___4 == 0) %>% select(child_registry_id, child_dia2___1) %>% filter(child_dia2___1 == 1) %>% mutate(venn.modality = "CT") %>% select(child_registry_id, venn.modality)
venn.bx <- child.surf %>% filter(child_dia2___4 == 0) %>% select(child_registry_id, child_dia2___2) %>% filter(child_dia2___2 == 1) %>% mutate(venn.modality = "Biopsy") %>% select(child_registry_id, venn.modality)
venn.gen <- child.surf %>% filter(child_dia2___4 == 0) %>% select(child_registry_id, child_dia2___3) %>% filter(child_dia2___3 == 1) %>% mutate(venn.modality = "Genetic Testing") %>% select(child_registry_id, venn.modality)
venn <- rbind(venn.ct, venn.bx, venn.gen)

venn.diagram(
  x = list(
    venn %>% filter(venn.modality == "CT") %>% select(child_registry_id) %>% unlist() , 
    venn %>% filter(venn.modality == "Biopsy")  %>% select(child_registry_id) %>% unlist() , 
    venn %>% filter(venn.modality == "Genetic Testing") %>% select(child_registry_id) %>% unlist()
    ),
  category.names = c("Chest Computed Tomography (CT)" , "Lung Biopsy" , "Genetic Testing"),
  filename = 'venn.png',
  output = TRUE ,
          imagetype="png" ,
          height = 480 , 
          width = 480 , 
          resolution = 300,
          compression = "lzw",
          lwd = 1,
          col=c("#440154ff", '#21908dff', '#fde725ff'),
          fill = c(alpha("#440154ff",0.3), alpha('#21908dff',0.3), alpha('#fde725ff',0.3)),
          cex = 0.5,
          fontfamily = "sans",
          cat.cex = 0.3,
          cat.default.pos = "outer",
          cat.pos = c(-27, 27, 120),
          cat.dist = c(0.085, -0.055, 0.055),
          cat.fontfamily = "sans",
          cat.col = c("#440154ff", '#21908dff', '#fde725ff'),
          rotation = 1
        )

img <- magick::image_read('C:/Users/tsukaharak/Documents/R files/BMIN503_Final_Project/venn.png')
plot(img) # or print(img)

```


```{r}
# will look at genetic testing - which genes tested, if rate of genetic testing has increased compared to consent date
child.surf.gen <- child %>%
  filter(diagnosis.factor == "Surfactant dysfunction disorder") %>%
  filter(genetics.factor == "Yes")
child.surf.gen <- child.surf.gen %>%
  mutate(date_genetics.date = as.Date(child.surf.gen$date_genetics))
if(!"race.text" %in% colnames(child.surf.gen)){
child.surf.gen <- left_join(child.surf.gen, select(child, child_registry_id, race.text), by = "child_registry_id")
}

ggplot(child.surf.gen, aes(x=date_genetics.date)) +
  geom_histogram(binwidth = 365) +
  ggtitle("Genetic Testing over Time in the chILD Registry") +
  ylab("Number of Subjects with Genetic Tests") +
  xlab("Approximate Time of Genetic Testing (Date-Shifted)") +
  theme(axis.text.x=element_text(angle=60, hjust=1)) +
  scale_x_date(date_breaks = "1 year", date_labels = "%Y")

gg.gen <- child.surf.gen %>%
  select(child_registry_id, type_genetic_test___1.factor:type_genetic_test___4.factor) %>%
  rename("Single Gene" = type_genetic_test___1.factor) %>%
  rename("Gene Panel" = type_genetic_test___2.factor) %>%
  rename("Whole Exome Sequencing (WES)" = type_genetic_test___3.factor) %>%
  rename("Whole Genome Sequencing (WGS)" = type_genetic_test___4.factor)
gg.gen <- tidyr::pivot_longer(gg.gen, cols = "Single Gene":"Whole Genome Sequencing (WGS)", names_to = "type_genetic_test.long")
gg.gen <- filter(gg.gen, value == "Checked")

ggplot(gg.gen, aes(x = type_genetic_test.long)) +
  geom_bar() +
  ggtitle("Type of Genetic Testing in Surfactant Dysfunction Disorder") +
  ylab("Number of Subjects with Genetic Tests") +
  xlab("Type of Genetic Testing") +
  theme(axis.text.x=element_text(angle=60, hjust=1)) 


```


```{r}

# will look at 3 earlier-onset diagnoses: NEHI, growth abnormalities, surfactant dysfunction
# clinical characteristics
# diagnostics


```


```{r}

# will look at cohort by 3 month intervals of presentation - 0-3 mo, 3-6 mo, 6-9 mo, 9-12 mo and which diagnostics contributed to diagnosis


```

Limitations - recruitment biases, availability of data, variability in data entry, i.e., when is a patient considered to have their diagnosis? This could be when they are determined to have ILD, or when they undergo BAL and discover pulmonary hemorrhage, or it could be weeks to months later when they undergo a lung biopsy to determine whether there is capillaritis. The database is fairly granular, however, many fields are unstructured data and thus present a challenge when trying to synthesize data on a large scale. There are certain areas of data collection that may be beneficial to study that have not been included, for example, administration of systemic steroids, which can greatly affect outcomes in certain disease processes. As chILD is a rapidly changing area of study, diagnostic tools and classification systems continue to evolve, which introduces increased variance in the dataset. Nevertheless, this is a large dataset on a group of diseases with significant morbidity and mortality, and it proves an invaluable tool in gaining further insight into workup and management.

Future directions - 


### Acknowledgements
We would like to thank the Children’s Interstitial and Diffuse Lung Disease Research Network (chILDRN) and Dr. Lisa Young for the opportunity to work on this project.
