---
title: "BMIN503/EPID600 Project Template"
author: "Leigh Cressman"
output: 
  html_document:
    theme: paper 
    highlight: tango
---
```{r set-options, echo=FALSE, cache=FALSE}
options(width = 400)
```  
***
Use this template to complete your project throughout the course. Your Final Project presentation will be based on the contents of this document. Replace the title/name above and text below with your own, but keep the headers.

### Overview
Give a brief a description of your project and its goal(s), what data you are using to complete it, and what three faculty/staff in different fields you have spoken to about your project with a brief summary of what you learned from each person. Include a link to your final project GitHub repository.

This analysis seeks to determine how changes in community mobility, SARS-CoV-2 variants, and vaccination relate to weekly changes in SARS-CoV-2 case incidence at the county level. County-level vaccination and case incidence data are publicly available on the CDC website. Community mobility data is captured using Google's county-level Social Mobility data. SARS-CoV-2 variant data is from a study led by Dr. Brendan Kelly and Dr. Rick Bushman in PSOM involving metagenomic viral whole genome sequencing (WGS) performed on COVID-19 tests from Penn Medicine clinical laboratories. [My final project GitHub repository can be accessed here.](https://github.com/LeighCressman/BMIN503_Final_Project)

### Introduction 
Describe the problem addressed, its significance, and some background to motivate the problem.

Explain why your problem is interdisciplinary, what fields can contribute to its understanding, and incorporate background related to what you learned from meeting with faculty/staff.

Since the start of the SARS-CoV-2 pandemic, scientists have been developing statistical models to predict various epidemiologic outcomes in different areas of the world. This analysis will contribute to that knowledge base by analyzing how changes in community mobility, vaccination, and SARS-CoV-2 variants from a Penn study are related weekly changes in case incidence in Philadelphia and the surrounding counties (PA: Bucks, Delaware, Chester, Montgomery; NJ: Burlington, Camden, Gloucester, Mercer; DE: New Castle). As Google Social Mobility data is available starting February 15, 2020, the analysis spans from then through October 2, 2021, using MMWR week as time intervals. The above-referenced study in PSOM randomly samples SARS-CoV-2 positive specimens collected at UPHS facilities with cycle threshold (Ct) values < 30. Whole genome sequencing is performed on these specimens, which have higher viral loads, to identify the variant. The results of WGS will be used as an additional explanatory variable in this analysis to indicate which SARS-CoV-2 variants were dominant in each MMWR week in the greater Philadelphia metropolitan area during the study period.  

This analysis draws upon microbiology for the identification of SARS-CoV-2 variants, and infectious disease epidemiology and biostatistics to analyze and interpret the relationship between dominant SARS-CoV-2 variants, vaccination, and change in case incidence. Geographical epidemiology will also contribute to the analysis in the form of visualizing the change in vaccinations, cases, variants, and community mobility across time in the target geographic areas. In addition, this analysis incorporates the behavioral sciences, as it seeks to identify whether Google Social Mobility data is associated with case incidence One particular difficulty I had in the initial phase of this analysis was data harmonization. I examined a plethora of datasets from the CDC, the Pennsylvania Department of Health, Google, and OpenDataPhilly before I was able to find data that covered the time spans and geographic areas of interest. My discussion with John Holmes was especially helpful for finalizing the data harmonization aspect of this analysis. He explained that if I were trying to publish this project, I would need express written permission from the CDC to link their data to other data sources. However, for a class project, it is fine to link the data to other sources without obtaining permission.



### Methods
Describe the data used and general methodological approach. Subsequently, incorporate full R code necessary to retrieve and clean data, and perform analysis. Be sure to include a description of code so that others (including your future self) can understand what you are doing and why. 


### Results
Describe your results and include relevant tables, plots, and code/comments used to obtain them. End with a brief conclusion of your findings related to the question you set out to address. You can include references if you'd like, but this is not required.
