---
title: "Transcriptomic Changes Due to Serum Amyloid A in Murine Naive CD4+ T cells"
subtitle: "BMIN 503 Final Project"
date: "December 2020"
author: "Ceire Hay"
output: 
  html_document:
    toc: false 
    depth: 3 
    theme: paper 
    highlight: tango
---
```{r set-options, echo=FALSE, cache=FALSE}
options(width = 400)
``` 

```{r global options, include = FALSE}
knitr::opts_chunk$set(echo=TRUE, eval=TRUE, warning = FALSE, message = FALSE)
``` 


# Overview
## Project Overview
Give a brief a description of your project and its goal(s), what data you are using to complete it, and what three faculty/staff in different fields you have spoken to about your project with a brief summary of what you learned from each person. Include a link to your final project GitHub repository.


- **Link to GEO:** https://www-ncbi-nlm-nih-gov.proxy.library.upenn.edu/geo/query/acc.cgi?acc=GSE132761
- **Corresponding Publication:** Lee JY, Hall JA, Kroehling L, Wu L, Najar T, Nguyen HH, Lin WY, Yeung ST, Silva HM, Li D, Hine A, Loke P, Hudesman D, Martin JC, Kenigsberg E, Merad M, Khanna KM, Littman DR. Serum Amyloid A Proteins Induce Pathogenic Th17 Cells and Promote Inflammatory Disease. Cell. 2020 Jan 9;180(1):79-91.e16. doi: 10.1016/j.cell.2019.11.026. Epub 2019 Dec 19. PMID: 31866067; PMCID: PMC7039443.


```{r, GEO_page, echo=FALSE, fig.cap=paste("GEO")}
knitr::include_graphics("C:/Users/ceire/Box Sync/Coursework/Fall 2020/BMIN503/BMIN503_Final_Project/images/GSE132761_GEO.JPG")

```

```{r, Lee_etal, echo=F, fig.cap=paste()}

```

## Analysis Overview
Using the [taffeta](https://github.com/HimesGroup/taffeta) from the Himes Lab I completed the following analysis on the lab cluster/server. Note that these pipelines are written as python scripts that ultimately generate reports as R Markdown files:

1. Download of raw fastq files from GEO
2. Preliminary quality control (QC) of raw data
3. Alignment of raw data to the reference genome
4. QC of aligned data
5. Identified differentially expressed genes (DEGs)
6. Normalization of DEGs


Following this, I preformed functional enrichment analysis in my local environment.




# Introduction 
Describe the problem addressed, its significance, and some background to motivate the problem.

Explain why your problem is interdisciplinary, what fields can contribute to its understanding, and incorporate background related to what you learned from meeting with faculty/staff.


# Methods
Describe the data used and general methodological approach. Subsequently, incorporate full R code necessary to retrieve and clean data, and perform analysis. Be sure to include a description of code so that others (including your future self) can understand what you are doing and why.

## 1. Set-up Working Environment
### 1.1. Define Variables and Directories
```{r, define_vars_directories}
#Assign the variables for GEO ID (geo_id), data directory (out_dir), or phenotype file if user defined (pheno_fn). Phenotype data is obtained from GEO (if user did not specify phenotype file)

setwd("C:/Users/ceire/Box Sync/Coursework/Fall 2020/BMIN503/BMIN503_Final_Project")
out_dir <- "C:/Users/ceire/Box Sync/Coursework/Fall 2020/BMIN503/BMIN503_Final_Project"

#SRA Accession ID
project_name <- 'SRP201470'
#GEO Accession ID
geo_id <- 'GSE132761'
#SRA Accession ID, same as project name
sra_id <- 'SRP201470'
```

### 1.2. Install and Load Required Packages
#### 1.2.1 
*Description of packages*

1. *GEOquery*: A Bioconductor package for obtaining publicly available data from The NCBI Gene Expression Omnibus (GEO)
2. *SRAdb*: A Bioconductor package for accessing the NCBI Sequence Read Archive (SRA) metadata associated with a dataset
3. *dplyr*: Part of the 'tidyverse' package. Provides a consistent set of verbs for common data manipulation challenges
4. *pander*: A CRAN package used to transform R objects to markdown format

```{r library, eval=T, echo=T, message=F, warning=F, results="hide"}
#Create a character vector that specifies the list of required packages to be installed.
req_pckgs=c("dplyr","SRAdb","pander", "GEOquery")
#Check the list of required packages against what is currently installed in R. Only packages that have not been previously installed will be installed.
installed_packages <- req_pckgs %in% rownames(installed.packages()) 
#install packages from CRAN
if (any(installed_packages == FALSE)) {
  install.packages(req_pckgs[!installed_packages])
}
#install packages from Bioconductor
if (any(installed_packages == FALSE)) {
  BiocManager::install(c(req_pckgs[!installed_packages]))
}
# Load the necessary libraries. 
# Create a character vector of all required packages 
req_lib <- c("dplyr","SRAdb","pander", "GEOquery")
#load the packages in DElib
lapply(req_lib, require, character.only=TRUE)
```

### 1.3. Obtain Raw Data
#### 1.3.1 Raw GEO Phenotype File

Obtain raw sample information including SRA ID from GEO. If pheno_fn is specified, download samples based on SRA_ID column; otherwise, download samples based on the GEO raw phenotype file.

```{r geo_download}
if (exists("geo_id")){
  if (!exists("pheno_fn")) {
    geo_fn <- paste0(out_dir, geo_id, "_series_matrix.txt.gz")
    if (!file.exists(geo_fn)) { # matrix files are alreadly downloaded
      gse <- getGEO(geo_id, destdir=out_dir, GSEMatrix = TRUE)[[1]] # download matrix file
    } else {
      gse <- getGEO(filename=geo_fn,GSEMatrix = TRUE)
    }
  }
}
```

#### 1.3.2. Examine Raw GEO Phenotype File
Show expression dataset features using gse matrix
```{r gse}
if (exists("geo_id")){
  if (!exists("pheno_fn")) {
    gse
  }
}
```

#### 1.3.3. Raw phenotype summary

Generated a summary of all the phenotypic variables for overview.

For continuous variables, show the summary table. For categorical variables, only show the first five levels of variables.

Generate a variable, suppldata (whether supplementary data are available), based on whether the column supplementary_file is none.

```{r phenoraw_summ}
if(exists("geo_id")){
  pheno.raw <- pData(phenoData(gse))
  for (x in names(pheno.raw)) {
    vec=pheno.raw[,x]
    if (!is.numeric(vec)) {
      vec <- factor(vec)
      if (nlevels(vec)>5) {res=table(droplevels(vec[vec%in%levels(vec)[1:5]]))} else {res=table(vec)}
      res=data.frame(res)
      names(res) <- c(x,"counts")
    }
    if (is.numeric(vec)){res=summary(vec)}
    pandoc.table(res, justify='left',split.tables=Inf, caption=x)
  }
}
```

#### 1.3.4. Get SRA ftp Link for Samples

If pheno_fn is not defined, download samples that have SRA id from GEO raw phenotype files, otherwise, download samples from defined phenotype file.

SRA sample ID is derived from GEO phenotype where the column name with **relation** if SRA download is needed.

```{r pheno.raw, eval=T, echo=F, message=F, warning=F}
if (exists("geo_id")){
  if (!exists("pheno_fn")) {
    relation_cols <- which(grepl("relation",names(pheno.raw))) # columns which have "relation" in column names
    relation_col=relation_cols[sapply(relation_cols, function(x){ # the column with "SRX" is where to retain the SRA ID
      relation=pheno.raw[,x]
      any(grepl("SRX", relation))
    })]
    if (length(relation_col)>1) {stop("More than two columns contain SRX information. Please check and speciy column name to the variable relation_name.")}
    relation_name=names(pheno.raw)[relation_col]
    files <- gsub(".*=","",pheno.raw[,relation_name])
  } else {
    pheno.raw <- read.table(pheno_fn, header=T, sep='\t')
    files <- pheno.raw$SRA_ID
  }
  cat("Samples for download:", paste(files,collapse=", "),"\n")
}
```

#### 1.3.5. Get fastq File Information from SRA
Raw .fastq files will be downloaded in parallele using Linux bash script

```{r sra_func, eval=F,echo=T}
# The sradownload_func downloads fastq file based on SRA id
sradownload_func <- function(file) {
  SRAdb::getSRAfile(file, sra_con,
    destDir = out_dir , fileType = 'fastq',
    srcType= 'ftp', makeDirectory = FALSE,
    method = 'libcurl', ascpCMD = NULL )
  cat("Finished downloading file:", file, "\n")
}

# The srainfo_func obtains ftp information for SRA fastq files
srainfo_func <-function(file) {
  sra_acc <- sraConvert(file, out_type=c('run'), sra_con = sra_con )
  getFASTQinfo (sra_acc$run, sra_con, srcType="ftp")
}

# The fastqinfo function can be used in absence of a GEO entry using the SRA ID
fastq_info_func <- function(SRA_ID) {
	sra_info <- sraConvert( in_acc=SRA_ID, out_type=c('experiment','run'), sra_con = sra_con )
 	#fastq_files <- arrange(sra_info,experiment,run)
 	sraFiles <- getFASTQinfo(in_acc = c(sra_info$experiment), sra_con, srcType = 'ftp' )
 	return(sraFiles)
}

# check if sample from GEO exists in SRAdb
sraconv_func <- function(files) {
  conversion <- sraConvert(files, sra_con = sra_con )
}
```

### 1.3.6. Download SRA Database 
```{r sraFiles, eval=F, echo=F, message=F, warning=F}
# download SRAmetadb.sqlite if it does not exist
if (!file.exists(paste0(out_dir, "/SRAmetadb.sqlite"))) {
  sqlfile <- getSRAdbFile()
  sra_con <- dbConnect(SQLite(),sqlfile)
} else {
  sra_con <- dbConnect(SQLite(),dbname="SRAmetadb.sqlite")
}
# check if files exist in sradb. If not, download fastq file from SRA using SRA toolkits
conversion <- sraconv_func(files)
sradb_exist <- nrow(conversion)!=0
if (!sradb_exist) {
  if (!exists("sra_id")) {stop(geo_id, " is not available in SRAdb. Please provide sra_id.")}
}
```


```{r sra_download_sradbexist, eval=sradb_exist, echo=F, message=F, warning=F}
# if study the exists in SRAdb
# obtain SRA file information
if(exists("geo_id")){
  sraFiles <- do.call(rbind, lapply(files, srainfo_func)) %>% dplyr::filter(experiment%in%files) %>% unique() # exclude the same samples that are used in other projects
  write.table(sraFiles, paste0(out_dir, project_name, "_sraFile.info"), col.names=T,row.names=F,sep="\t",quote=F )
  
} else if (exists("sra_id")){
  sraFiles <- fastq_info_func(sra_id)
  sraFiles <- sraFiles %>% dplyr::filter(experiment%in%files) %>% dplyr::select(run,submission,study,sample,experiment,ftp) %>% unique() # exclude the same samples that are used in other projects
  write.table(sraFiles, paste0(out_dir, project_name, "_sraFile.info"), col.names=T,row.names=F,sep="\t",quote=F )
}
```

```{r sra_download_sradbnotexist, eval=!sradb_exist, echo=F, message=F, warning=F}
# if the study does not exist in SRAdb
# obtain metadata from SRA website
metadata_fn <- paste0(out_dir, "/", sra_id, ".metadata.csv") 
cmd_metadata_download <- paste0("esearch -db sra -query ", sra_id, " | efetch -format runinfo | awk -F ',' 'NR==1; NR>1 {print $0 | \"grep -v Run\" }' | sed '/^$/d' > ", metadata_fn)
system(cmd_metadata_download)
# save meta data in sraFile.info
sraFiles <- read.csv(metadata_fn)
names(sraFiles) <- unname(sapply(names(sraFiles), tolower))
sraFiles <- sraFiles %>% dplyr::filter(experiment%in%files) %>% dplyr::select(run,submission,experiment) %>% unique() # exclude the same samples that are used in other projects
write.table(sraFiles, paste0(out_dir, project_name, "_sraFile.info"), col.names=T,row.names=F,sep="\t",quote=F )
```

Show first five rows of SRA information
```{r srainfo_show, eval=T, echo=F, message=F, warning=F, results="asis"}
row.names(sraFiles) <- NULL
pandoc.table(head(sraFiles,5), justify='left',split.tables=Inf, caption="SRA information")
```

```{r sra_download, eval=F, echo=F, message=F, warning=F}
# Download fastq files, disabled here
if (!file.exists(paste0(out_dir, "/SRAmetadb.sqlite"))) {
  sqlfile <- getSRAdbFile()
  sra_con <- dbConnect(SQLite(),sqlfile)
} else {
  sra_con <- dbConnect(SQLite(),dbname="SRAmetadb.sqlite")
}
system.time(sapply(files, sradownload_func))
```

### User tailored phenotype file

**This step is only used for tailoring raw phenotype file from GEO. This step requires mannual inspection.**

Raw phenotypic variables are not informative (e.g. description, characteristics_ch1 and source_name_ch1) and not created in a consice way. Select useful phenotype variables and manually modify them using a standard format including the following columns: Sample, GEO_ID, Donor (required column for treatment comparison), Disease, Treatment, and any other information if provided such as Age, Gender, ERCC spike-in, Sequencing index, Sample preparation kit.

The example codes for this section are based on phenotype data from GSE52778 (SRA study ID: SRP033351).

```{r pheno, eval=T, echo=F, message=F, warning=F}
#list the location of the fastq files on the server
fastqpath <- "/projects/RNASeq/SRP201470/SRP201470_SRAdownload/"

if (!exists("pheno_fn")) {
  if(exists("geo_id")){
    library(dplyr)
    pheno.raw$files <- files # Note that file column should be selected in the code below
    cols=c("title","geo_accession","organism_ch1","characteristics_ch1", "characteristics_ch1.1", "characteristics_ch1.2","description", "description.1", "files", "type")
    pheno <- pheno.raw %>%
      dplyr::select(cols) %>%
      dplyr::mutate(SRA_ID=files) %>%
      dplyr::mutate(GEO_ID=geo_accession) %>%
      dplyr::mutate(Donor=description.1) %>%
      dplyr::mutate(Donor=gsub("-","_", Donor)) %>% 
      dplyr::mutate(Tissue=gsub("cell typ: ", "", characteristics_ch1)) %>% #replace all instances of 'cell typ:' with nothing
      dplyr::mutate(Tissue=gsub("\\+ ", "_", Tissue)) %>% 
      dplyr::mutate(Tissue=gsub(" ", "", Tissue)) %>%
      dplyr::mutate(Treatment=gsub(" treated_.*","",title)) %>% #eliminate ' treated_' and everything that follows it from character string
      dplyr::mutate(Treatment=gsub("\\s*\\([^\\)]+\\)", "", Treatment)) %>%
      dplyr::mutate(Treatment=gsub("-b", "b", Treatment)) %>%
      dplyr::mutate(Treatment=gsub(" \\+ ", "_", Treatment)) %>% 
      dplyr::mutate(Time=gsub("time [()]h[()]: ", "", characteristics_ch1.1)) %>% #eliminate 'time (h)' from character string
      dplyr::mutate(Status=paste(Treatment,Time, sep="_"))%>% #create the "Status" column by including data from the 'Treatment' column first and 'Time' column, separating the terms with an underscore
      dplyr::mutate(ERCC_Mix=gsub("SRA","-", type)) %>% #no ERCC data so the column entries are "-" bc the variable is required by pipeline script
      dplyr::mutate(Disease=gsub("SRA","-", type)) %>% #see above
      dplyr::mutate_if(is.character,as.factor) %>%
      dplyr::select(-one_of(cols)) # remove original columns
    pheno <- merge(unique(sraFiles[,c("run","experiment")]), pheno, by.x="experiment", by.y="SRA_ID", all.y=T) # add sample name (run column from SRA database)
    names(pheno)[names(pheno)%in%c("run")] <- c("Sample")
    names(pheno)[names(pheno)%in%c("experiment")] <- c("SRA_ID")
    pheno$R1 <- paste(fastqpath, pheno$Sample, sep="")#  the R1 column by including fastqpath 1st with sample name
    pheno$R1 <- gsub("$","_1.fastq.gz", pheno$R1) #substitute the end of the string with "_1.fastq.gz"
    pheno$R2 <- paste(fastqpath, pheno$Sample, sep="")
    pheno$R2 <- gsub("$","_2.fastq.gz", pheno$R2) #substitute the end of the string with "_1.fastq.gz"
    write.table(pheno, paste0(out_dir, geo_id, "_withoutQC.txt"),col.names=T,row.names=F,sep="\t",quote=F)
  }
} else {pheno.raw <- read.table(pheno_fn, header=T, sep="\t")}
```

Show the summary of phenotype variables and the sample size for different groups

```{r pheno_summ, eval=T, echo=F, message=F, warning=F, results="asis"}
# show the first five rows
if(exists("pheno_fn")|exists("geo_id")){
  pandoc.table(head(pheno,5), split.tables=Inf,caption="Show the first 5 rows of the modified phenotype file")
  # show the groups of interest
  avail_group=c("Tissue","Disease","Treatment")[c("Tissue","Disease","Treatment")%in%names(pheno)]
  res=as.data.frame(table(pheno[,avail_group]))
  names(res) <- c(avail_group,"Count")
  pandoc.table(res, split.tables=Inf, caption="Sample size in different tissue and disease/treatment groups")
 }
```

### Results
Describe your results and include relevant tables, plots, and code/comments used to obtain them. End with a brief conclusion of your findings related to the question you set out to address. You can include references if you'd like, but this is not required.
```{r}

```

