---
title: "An Exploratory Analysis of Molecular Testing in Lung Cancer" 
subtitle: "BMIN503/EPID600 Final Project"
author: "Anthony O. Martella"
format:
  html:
    theme: slate
    toc: true
    toc-location: left
    code-fold: true
    code-tools: 
      source: true
      toggle: true
    code-copy: true
    code-overflow: scroll
editor: visual
number-sections: true
embed-resources: true
editor_options:
  chunk_output_type: inline
execute:
  warning: false
  message: false
---

------------------------------------------------------------------------

## Overview {#sec-overview}

This project seeks to explore next-generation gene sequencing (NGS) testing rates, modalities (tissue, plasma, or both), and receipt of guideline concordant care among 552 patients with stage IV mNSq NSCLC using electronic health record (EHR)-extracted data. Receipt of guideline-concordant care will be looked at in relation to test modality, comprehensiveness, availability of results prior to 1L therapy, and other demographic characteristics. Secondary analyses will examine testing turn-around-time (TAT) of results by test modality, time-to-treatment (TTT) differences by test modality, and most commonly identified alterations. The contents of this project are available on [my GitHub](https://github.com/amartella1/BMIN503_Final_Project).

I have discussed my final project with Charu Aggarwal, MD, MPH and Melina Marmarelis, MD, MSCE. Dr. Aggarwal is the Leslye M. Heisler Associate Professor for Lung Cancer Excellence at the University of Pennsylvania Perelman School of Medicine, Director of the Program in Precision Oncology at the Penn Center for Cancer Care Innovation, and a thoracic medical oncologist at the Abramson Cancer Center. Dr. Marmarelis is an Assistant Professor of Medicine and Co-Director of the Molecular Tumor Board at the University of Pennsylvania, and a thoracic medical oncologist at the Abramson Cancer Center.

## Introduction {#sec-introduction}

Lung cancer is one of the leading causes of cancer death in the United States (US) and internationally. With approximately 2,206,771 people diagnosed worldwide in 2020, lung cancer is also one of the most frequently diagnosed. Non-small cell lung cancer (NSCLC) comprises 80-85% of cases -- roughly 70% with non-squamous histology and 30% squamous histology. Over the past few decades, 5-year survival rates for individuals with stage IV metastatic NSCLC have increased from less than 5% to greater than 25%. This increased survival rate is in part attributed to advancements in the field of precision medicine, which have allowed for the development of targeted (or personalized) therapies in lung cancer care. Delivery of appropriate targeted therapies (at the right time) has been shown to improve patient outcomes, including overall survival rate.

To date, the majority of FDA-approved targeted therapies for lung cancer are indicated in the treatment of metastatic non-squamous (mNSq) NSCLC. Although, research has begun demonstrating targeted therapy may have beneficial effects at earlier stages. To determine whether someone might benefit from targeted therapy necessitates comprehensive next-generation gene sequencing (NGS) -- also referred to as molecular testing -- which can be performed on a tumor tissue sample or plasma (blood) sample. The current National Comprehensive Cancer Network (NCCN) guidelines define comprehensiveness as the detection of alterations in seven genes (*EGFR, ALK, BRAF, ROS1, MET, RET,* and *NTRK*). Specific alterations in these seven genes each have FDA-approved targeted therapies for the treatment of mNSq NSCLC in the first-line (1L) setting. The FDA has also recently approved use of targeted therapy in second-line (2L) treatment for *KRAS G12C* and *ErbB2* alterations.

The NCCN guidelines recommend comprehensive NGS testing be completed for all patients with mNSq NSCLC prior to initiating 1L treatment. The survival benefit from targeted therapy can be forfeited if it is not delivered at the most appropriate time. For example, research has shown that administering targeted therapy for patients with an EGFR alteration can be toxic if they have already received 1L immunotherapy. This paradigm illustrates the need for a quick turn-around-time (TAT) of results before 1L treatment.

Aside from NGS, single-gene or "hot-spot" testing, limited gene panels, fusion transcript, and fluorescence in situ hybridization (FISH) analysis are other forms of molecular testing in lung cancer. Traditional approaches to molecular testing involve tissue testing that can take anywhere from 2 to 12 weeks to result. While some medical institutions conduct NGS testing in-house, more commonly tissue samples are sent out to 3^rd^ party companies. A recurrent limitation of tissue testing is that there is not always enough tissue to perform sequencing -- quality/quantity not sufficient (QNS), which can sometimes lead to a re-biopsy, further delaying treatment for an already vulnerable population.

Due to evolving technology and increased high-performance computing abilities, plasma-based comprehensive NGS testing, also referred to as liquid biopsy, is made possible by genotyping circulating-tumor DNA (ct-DNA). Results for plasma-based NGS are often available within 7-10 days. While tissue testing remains a gold standard, research has demonstrated that incorporating plasma NGS testing into standard of care leads to a significant increase in detection of clinically actionable alterations. In addition, plasma NGS testing is associated decreased time to treatment.

Despite the known benefits of comprehensive molecular testing, testing rates have not caught up to speed for a variety of reasons. The MYLUNG (Molecularly Informed Lung Cancer Treatment in a Community Cancer Network) Consortiumâ„¢ provides real-world biomarker testing patterns and in a recent study found that out of 10,333 patients with nSq NSCLC only 53% of patients (n = 5494) had undergone NGS testing. Testing rates tend to be significantly lower for patients seen in community cancer centers compared to those seen in academic medical practices. Furthermore, there are known disparities of NGS testing rates by race. Differences up to over 10 percentage-points have been observed in testing rates such that Black individuals received comprehensive testing less than White individuals.

This project seeks to explore next-generation gene sequencing (NGS) testing rates, modalities (tissue, plasma, or both), and receipt of guideline concordant care among patients with stage IV mNSq NSCLC. Receipt of guideline-concordant care will be looked at in relation to test modality, comprehensiveness, availability of results prior to 1L therapy, and other demographic characteristics. Secondary analyses will examine testing turn-around-time (TAT) of results by test modality, time-to-treatment (TTT) differences by test modality, and most commonly identified alterations.

## Methods {#sec-methods}

This study employs the use of electronic health record (EHR)- extracted data from 552 adult patients with newly diagnosed stage IV mNSq NSCLC within the years 2019-2022. Patients included were seen in an outpatient cancer center at either an academic medical institution (PCAM) or two affiliated community practices. RStudio R version 4.3.1 was used to complete all analyses. Variables were selected based on known relevance. A copy of the data dictionary for the variables included in the analyses is available in [my GitHub repository](https://github.com/amartella1/BMIN503_Final_Project).

The `gt_summary` package was used to generate descriptive tables and perform various statistical analyses. To investigate differences among groups, a combination of tests was applied as appropriate including Fisher's exact, Pearson's Chi-squared, Wilcoxon and Kruskal-Wallis rank sum. Bivariate tests are conducted based on the findings from the summary tables to further explore the relationships among variables. A multivariate regression model is created in attempts to determine whether test modality and/or other variables are predictors for receipt of guideline concordant care. Variance inflation factor (VIF) analyses are performed to assess multicollinearity of predictor variables. Odds ratios and confidence intervals for the model are applied.

To assess turn-around-time (TAT) of results by test modality, the number of days between collection date and result date is calculated for tissue testing alone and plasma testing alone. Summary statistics are generated and a Wilcoxon signed rank test with continuity correction is performed. For time-to-treatment (TTT), the day difference between comprehensive result date and 1L treatment start date is calculated and a Kruskal-Wallis rank sum test is applied to look at differences by testing modality.

A more detailed description of methods with coding is included below.

I begin by loading the required packages to perform the analyses.

```{r}
#Packages loaded - use install.packages() if not already on your system

library(tidyverse)
library(tidymodels)
tidymodels_prefer()
library(readr)
library(dplyr)
library(gtsummary)
library(lubridate)
library(ggplot2)
library(scales)
library(glmnet)
library(car)
```

I load the three datasets and use `bind_rows` to merge them. The result is a master data frame, `data_all` containing 552 observations of 134 variables.

```{r}
#Load the datasets
df.2019 <- read_csv("Clean_2019_Cohort_AOM.csv")
df.2020 <- read_csv("Clean_2020_Cohort_AOM.csv")
df.2021.2022 <- read_csv("Clean_2021-2022_Cohort_AOM.csv")

#Merge the datasets using bind_rows
data_all <- bind_rows(df.2019, df.2020, df.2021.2022)

```

Given the large number of variables, I create a new data frame, `df.1` that I use as a working data frame throughout. I use `dplyr::mutate` to clean and transform the data. In some instances I use subset data frames below for ease but this is not required.

```{r}
#Data Cleaning and Transforming

#Create new factors using mutate

df.1 <- data_all |>
  mutate(race = factor(case_when(
    race == "Pacific Island" | race == "Hispanic Latino/White" | race == "Other/Unknown" | race == "Multiple Races" ~ "Other/Unknown",
    TRUE ~ race
  ))) |>
  mutate(sex = factor(sex, levels = c("F", "M"),
                      labels = c("Female", "Male"))) |>
  mutate(guide_conc_care = factor(guide_conc_care, levels = c("Yes", "No"),
                                  labels = c("Yes", "No"))) |>
  mutate(markers_1L = factor(markers, levels = c("Yes", "No"),
                          labels = c("Yes", "No"))) |>
  mutate(smoke = factor(smoke, levels = c("Former", "Current", "Never"),
                        labels = c("Former", "Current", "Never"))) |>
  mutate(test_modality = factor(samp_type, levels = c("Tissue", "Plasma", "TP"),
                                labels = c("Tissue", "Plasma", "Tissue+Plasma"))) |>
  mutate(alteration_di = factor(alteration_di, levels = c("Yes", "No"),
                                labels = c("Yes", "No"))) |>
  mutate(alteration_tpb = factor(alteration_tpb, levels = c("Tissue", "Plasma", "Both"),
                                 labels = c("Tissue", "Plasma", "Both"))) |>
  mutate(tx_prescribed = factor(tx_prescribed, levels = c("Yes", "No"),
                                labels = c("Yes", "No"))) |>
  mutate(tx_type = factor(tx_type, levels = c("Chemo", "ChemoIO", "IO", "TKI", "NONE"),
                          labels = c("Chemo", "ChemoIO", "IO", "TKI", "NONE"))) |>
  mutate(comprehensive_testing = factor(if_else(tier == "Tier 1" | tier == "Tier 2", "Yes", "No"))) |>
  mutate(tx_site = gsub("VF", "Valley Forge", tx_site),
         tx_site = gsub("CCH", "Chester County", tx_site)) |>
  mutate(tx_site = factor(tx_site, levels = c("PCAM", "Chester County", "Valley Forge"),
                          labels = c("PCAM", "Chester County", "Valley Forge"))) |>
  mutate(ngs_ttr_col = as.numeric(ngs_ttr_col)) |>
  mutate(pb_ttr_col = as.numeric(pb_ttr_col)) |>
  mutate(dx_date = mdy(dx_date)) |>
  mutate(tx_1l_start = mdy(tx_1l_start)) |>
  mutate(comprehensiveness_date = mdy(comprehensiveness_date)) |>
  mutate(day.diff.compdate.tx1l = as.numeric(difftime(ymd(tx_1l_start),
                                                      (comprehensiveness_date),
                                                      units = "days")))
```

## Results {#sec-results}

### Demographics & Testing Characteristics

552 patients (ages 22-93) with newly-diagnosed stage IV mNSq NSCLC were included in the final analyses (Table 1). Among these patients, 91% received comprehensive molecular testing as defined by the NCCN guidelines and 84% had testing completed prior to initiation of 1L therapy. 66% of patients had both tissue and plasma-based testing, 24% with tissue testing only, and 11% with plasma testing only.

```{r}
#Descriptive Tables: Overview Demographics & Testing (Table 1)

#Select demographic variables into a new data frame
df.demo <- df.1 |>                      
  select(sex, age, dx_date, race, smoke, tx_site, comprehensive_testing, markers_1L, test_modality)

#Create a demographics table summary
tbl.demo <- tbl_summary(df.demo, missing = "no",
                        label = list(sex ~ "Sex",
                                     age ~ "Age at Diagnosis",
                                     dx_date ~ "Diagnosis Date",
                                     race ~ "Race",
                                     smoke ~ "Smoking Status",
                                     tx_site ~ "Treatment Site",
                                     comprehensive_testing ~ "Comprehensive Testing",
                                     markers_1L ~ "Markers tested prior to 1L therapy",
                                     test_modality ~ "Test Modality")) |>
                        modify_caption("**Table 1. Demographics & Testing**") |>
                        bold_labels()
  
tbl.demo

```

An overview of testing patterns by treatment site is included below.

```{r}
#Set theme for all plots 
ggthemr::ggthemr('flat dark', type = 'outer') 

#Visualizations: Testing modality x Treatment site
df.demo |>
  filter(!is.na(test_modality) & !is.na(tx_site)) |>
  ggplot(aes(x = tx_site, fill = test_modality)) +
  geom_bar(position = "fill") +
  scale_y_continuous(labels = scales::percent_format(scale = 100, accuracy = 1)) +
  theme(axis.text = element_text(size = 12),
        axis.title.x.bottom = element_text(size = 10)) +
  labs(title = "Testing Modality by Treatment Site",
       x = "Treatment Site",
       y = "",
       fill = "Test Modality")
```

Combined test modality, treatment site, and smoking status were significantly associated with receipt of comprehensive testing (Table 2). 98% (n = 42) of patients that did not have comprehensive testing (n = 48) had tissue testing only. In contrast to previous studies, significant differences in comprehensive testing rates by race were not observed in this sample. No significant differences found by sex or age at diagnosis.

```{r}
#Analyses: Differences in Comprehensive Testing Rates (Table 2)

#Are there significant differences in comprehensive testing rates by test modality, tx_site, smoking status, race, age, or sex? 

tbl.comp.testing <- df.1 |>
  select(comprehensive_testing, test_modality, tx_site, smoke, race, age, sex) |>
  tbl_summary(by = comprehensive_testing,
              missing = "no",
              label = list(test_modality ~ "Combined Test Modality",
                           tx_site ~ "Treatment Site",
                           smoke ~ "Smoking Status",
                           race ~ "Race",
                           age ~ "Age at Diagnosis",
                           sex ~ "Sex")) |>
  modify_caption("**Table 2. Comprehensive Testing**") |>
  modify_spanning_header(c("stat_1", "stat_2") ~ "Comprehensive Testing") |>
  add_p() |>
  bold_labels()
tbl.comp.testing

```

```{r}
#Visualizations: Demographics & Testing 

#Comprehensive testing x Race 
df.demo |>
  filter(!is.na(comprehensive_testing) & !is.na(race)) |>
  ggplot(aes(x = race, fill = comprehensive_testing)) +
  geom_bar(position = "fill") +
  scale_y_continuous(labels = scales::percent_format(scale = 100, accuracy = 1)) +
  theme(axis.text = element_text(size = 9, angle = 45, hjust = 1),
        axis.title.x.bottom = element_text(size = 10),
        legend.text = element_text(size = 10)) +
  labs(title = "Comprehensive Testing by Race",
       x = "Patient Race",
       y = "",
       fill = "Comp. Testing")
```

Next, differences by testing modality (tissue, plasma, or tissue+plasma) are explored in relation to comprehensiveness, having markers tested prior to 1L therapy, receiving guideline-concordant care, smoking status, and having an alteration detected (Table 3). Six patients were excluded from these analyses due to missing or unknown testing modality. Significant differences by testing modality were observed in all of the aforementioned variables, with the exception of having an alteration detected.

```{r}
#Analyses: Differences by Testing Modality (Table 3)

df.test.char <- df.1 |>
  select(comprehensive_testing, markers_1L, test_modality, guide_conc_care, smoke, alteration_di)

#Summary Table by Testing Modality
tbl.test.char <- df.test.char |>
  tbl_summary(by = test_modality,
              missing = "no",
              label = list(
                comprehensive_testing ~ "Comprehensive Testing",
                markers_1L ~ "Markers tested prior to 1L therapy",
                guide_conc_care ~ "Received Guideline-Concordant Care",
                smoke ~ "Smoking Status",
                alteration_di ~ "Alteration Detected"
              )) |>
  add_p() |>
  add_overall() |>
  bold_labels() |>
  modify_caption("**Table 3. Differences by Test Modality**")
tbl.test.char

```

100% of patients who had plasma testing (alone or in combination with tissue) had testing that is considered comprehensive according to NCCN recommendations. Testing was less likely to be comprehensive for patients who had tissue testing only.

```{r}
#Visualizations: Comprehensive testing x Test modality

df.demo |>
  filter(!is.na(comprehensive_testing) & !is.na(test_modality)) |>
  ggplot(aes(x = test_modality, fill = comprehensive_testing)) +
  geom_bar(position = "fill") +
  scale_y_continuous(labels = scales::percent_format(scale = 100, accuracy = 1)) +
  theme(axis.text = element_text(size = 12),
        axis.title.x.bottom = element_text(size = 10)) +
  labs(title = "Comprehensive Testing by Test Modality",
       x = "Test Modality",
       y = "",
       fill = "Comp. Testing")
```

85% of patients had comprehensive testing completed prior to 1L treatment. Among those patients, 90% had both tissue and plasma testing, 82% plasma-only, and 73% tissue-only. These differences were statistically significant and indicate incorporation of plasma testing can help ensure comprehensive testing is completed prior to 1L treatment.

```{r}
#Visualizations: Markers tested prior to 1L x Test Modality
df.demo |>
  filter(!is.na(markers_1L) & !is.na(test_modality)) |>
  ggplot(aes(x = test_modality, fill = markers_1L)) +
  geom_bar(position = "fill") +
  scale_y_continuous(labels = scales::percent_format(scale = 100, accuracy = 1)) +
  theme(axis.text = element_text(size = 12),
        axis.title.x.bottom = element_text(size = 10)) +
  labs(title = "Markers Tested Prior to 1L Therapy",
       x = "Test Modality",
       y = "",
       fill = "Markers Prior 1L")
```

Turn-around-time (TAT) of results was significantly different between tissue testing versus plasma testing. The average TAT for tissue NGS was 24.75 days and 8.23 days for plasma NGS.

```{r}
#Analyses: TAT - Turn around time of results by test modality 

#Are there significant differences in turn around time (TAT) of results tissue ngs v plasma?  ngs_ttr_col & pb_ttr_col -- days from collection to report date 

#Tissue Testing
summary(df.1$ngs_ttr_col, na.rm = TRUE)

#Plasma Testing
summary(df.1$pb_ttr_col, na.rm = TRUE)

#Test for significant differences 
wilcoxon.rank.test <- wilcox.test(df.1$ngs_ttr_col, df.1$pb_ttr_col, paired = TRUE)
wilcoxon.rank.test

df.1 |>
  filter(!is.na(ngs_ttr_col) & !is.na(pb_ttr_col)) |>
  ggplot(aes(x = ngs_ttr_col, y = pb_ttr_col)) +
  scale_x_continuous(limits = c(1, 75)) +
  geom_point() +
  geom_smooth(method = "lm")

```

For time-to-treatment (TTT), no significant differences by test modality were observed in number of days from comprehensive result date to treatment 1L start date.

```{r}
#Analyses: TTT - Time to Treatment

#Are there sig. differences between groups/testing modality in # of days to treatment start? -- Comprehensive date to date 1L start

df.ttt <- df.1 |>
  select(test_modality, day.diff.compdate.tx1l, markers_1L) |>
  filter(markers_1L == "Yes")

#Table Summary
tbl.ttt <- df.ttt |>
  tbl_summary(by = test_modality, include = day.diff.compdate.tx1l,
              missing = "no") |>
  add_p() |>
  modify_caption("Day Difference b/w Comprehensive Result Date & 1L Treatment Start")
tbl.ttt

```

### Treatment & Guideline-Concordant Care

Overall, 84% of patients received guideline-concordant care (Table 4). Treatment types included chemotherapy, immunotherapy (IO), chemo+IO, tyrosine kinase inhibitor (TKI)/targeted therapy, or none. Reasons for no treatment could include poor performance status, patient preference, hospice referral or physician recommendation. 88% of patients who had a clinically-actionable alteration went on to receive targeted therapy.

```{r}
#Overview Table: Treatment & Guideline-Concordant Care

#Table 4
tbl.guidelines.overview <- df.1 |>
  select(tx_site, tx_type, guide_conc_care, tx_prescribed) |>
  tbl_summary(missing = "no",
              label = list(
                tx_site ~ "Treatment Site",
                tx_type ~ "Treatment Type",
                guide_conc_care ~ "Received Guideline-Concordant Care",
                tx_prescribed ~ "Targeted therapy prescribed (if alteration detected)"
                )) |>
  modify_caption("**Table 4. Treatment Characteristics**") |>
  bold_labels()
tbl.guidelines.overview


```

Comprehensive testing, markers tested prior to 1L, and test modality had the strongest associations with receipt of guideline-concordant care. Bivariate analyses that further explore these relationships are described below.

```{r}
#Analyses: Differences by Guideline-Concordant Care (Table 5)

#subset data frame
df.guidelines <- df.1 |>
  select(guide_conc_care, comprehensive_testing, markers_1L, test_modality, alteration_di, alteration_tpb, tx_site, tx_type, sex, race, smoke) 

#Table summary (Table 5)
tbl.guidelines <- df.guidelines |>
  tbl_summary(by = guide_conc_care,
              missing = "no",
              label = list(comprehensive_testing ~ "Comprehensive Testing",
                           markers_1L ~ "Markers tested prior to 1L therapy",
                           test_modality ~ "Combined Test Modality",
                           alteration_di ~ "Alteration Detected",
                           alteration_tpb ~ "Alteration detected by tissue/plasma/both",
                           tx_site ~ "Treatment Site",
                           tx_type ~ "Treatment Type",
                           sex ~ "Sex",
                           race ~ "Race",
                           smoke ~ "Smoking Status"
                           )) |>
  modify_caption("**Table 5. Guideline-Concordant Care**") |>
  modify_spanning_header(c("stat_1", "stat_2") ~ "Guideline-Concordant Care") |>
  bold_labels() |>
  add_p()
tbl.guidelines
```

```{r}
#Analyses: Predictors of Guideline-Concordant Care (Bivariate Analyses)

#subset frame
clean.df.guidelines <- df.guidelines |>
  select(guide_conc_care, comprehensive_testing, markers_1L, test_modality, alteration_di, race, sex, smoke, tx_type, tx_site)
```

**Test 1.** Comprehensive molecular testing is highly associated with receipt of guideline-concordant care.

```{r}
#Comprehensive testing x Guideline-Concordant Care
df.guidelines |>
  filter(!is.na(comprehensive_testing) & !is.na(guide_conc_care)) |>
  ggplot(aes(x = comprehensive_testing, fill = guide_conc_care)) +
  scale_y_continuous(labels = scales::percent_format(scale = 100, accuracy = 1)) +
  geom_bar(position = "fill") +
  theme(axis.text = element_text(size = 13),
        axis.title.x.bottom = element_text(size = 12)) +
  labs(title = "Comprehensive Testing x Guideline-Concordant Care",
       x = "Comprehensive Testing",
       y = "",
       fill = "Guide-Conc Care")
```

```{r}
#Bivariate Analysis: Comprehensive testing x Guideline-Concordant-Care

chisq.test(table(clean.df.guidelines$guide_conc_care, clean.df.guidelines$comprehensive_testing))

test.1 <- summary(glm(guide_conc_care ~ comprehensive_testing,
                      data = clean.df.guidelines,
                      family = "binomial"))
test.1

exp(coef(glm(guide_conc_care ~ comprehensive_testing,   #Odds ratio
             data = clean.df.guidelines, 
             family = binomial())))
```

**Test 2.** Having comprehensive biomarker testing prior to 1L therapy is highly associated with receipt of guideline-concordant care.

```{r}
#Markers prior to 1L x Guideline-Concordant Care

df.guidelines |>
  filter(!is.na(markers_1L) & !is.na(guide_conc_care)) |>
  ggplot(aes(x = markers_1L, fill = guide_conc_care)) +
  geom_bar(position = "fill") +
  scale_y_continuous(labels = scales::percent_format(scale = 100, accuracy = 1)) +
  theme(axis.text = element_text(size = 12),
        axis.title.x.bottom = element_text(size = 12)) +
  labs(title = "Markers Prior to 1L x Guideline-Concordant Care",
       x = "Markers Tested Prior to 1L Treatment",
       y = "",
       fill = "Guide-Conc Care")
```

```{r}
#Bivariate Analysis: Markers prior to 1L x Guideline Concordant-Care

chisq.test(table(clean.df.guidelines$guide_conc_care,
                 clean.df.guidelines$markers_1L))

test.2 <- summary(glm(guide_conc_care ~ markers_1L,
                      data = clean.df.guidelines, 
                      family = "binomial"))
test.2

exp(coef(glm(guide_conc_care ~ markers_1L,
             data = clean.df.guidelines,
             family = binomial())))
```

**Test 3.** Delivery of TKIs (targeted therapies) was most significantly associated with receipt of guideline-concordant care.

```{r}
#Treatment type x Guideline-Concordant Care

df.guidelines |>
  filter(!is.na(tx_type) & !is.na(guide_conc_care) & !tx_type == "NONE") |>
  ggplot(aes(x = tx_type, fill = guide_conc_care)) +
  geom_bar(position = "fill") +
  scale_y_continuous(labels = scales::percent_format(scale = 100, accuracy = 1)) +
  theme(axis.text = element_text(size = 12),
        axis.title.x.bottom = element_text(size = 12)) +
  labs(title = "Treatment Type x Guideline-Concordant Care",
       x = "Treatment Type",
       y = "",
       fill = "Guide-Conc Care")
```

```{r}
#Bivariate Analysis: Treatment type x Guideline-Concordant Care

chisq.test(table(clean.df.guidelines$guide_conc_care,
                 clean.df.guidelines$tx_type))

test.3 <- summary(glm(guide_conc_care ~ tx_type,
                      data = clean.df.guidelines,
                      family = "binomial"))
test.3

exp(coef(glm(guide_conc_care ~ tx_type,
             data = clean.df.guidelines, family = binomial())))
```

**Test 4.** Patients who received plasma or tissue+plasma testing were more likely to receive guideline-concordant care than those with tissue-testing only.

```{r}
#Guideline-Concordant Care x Test Modality 
df.guidelines |>
  filter(!is.na(test_modality) & !is.na(guide_conc_care)) |>
  ggplot(aes(x = test_modality, fill = guide_conc_care)) +
  scale_y_continuous(labels = scales::percent_format(scale = 100, accuracy = 1)) +
  geom_bar(position = "fill") +
  theme(axis.text = element_text(size = 12),
        axis.title.x.bottom = element_text(size = 12)) +
  labs(title = "Test Modality x Guideline-Concordant Care",
       x = "Test Modality",
       y = "",
       fill = "Guide-Conc Care")
```

```{r}
#Bivariate Analysis: Guideline-Concordant Care x Test Modality
chisq.test(table(clean.df.guidelines$guide_conc_care,
                 clean.df.guidelines$test_modality))

test.4 <- summary(glm(guide_conc_care ~ test_modality, 
                      data = clean.df.guidelines, 
                      family = "binomial"))
test.4

exp(coef(glm(guide_conc_care ~ test_modality,      #Odds ratio
             data = clean.df.guidelines, 
             family = binomial()))) 
```

**Test 5.** Detection of a clinically-actionable alteration was associated with receipt of guideline-concordant care.

```{r}
#Bivariate Analysis: Guideline-Concordant Care x Alteration detected

chisq.test(table(clean.df.guidelines$guide_conc_care,
                 clean.df.guidelines$alteration_di))

test.5 <- summary(glm(guide_conc_care ~ alteration_di,
                      data = clean.df.guidelines, family = "binomial"))
test.5

exp(coef(glm(guide_conc_care ~ alteration_di,
             data = clean.df.guidelines,
             family = binomial())))

```

**Test 6.** Smoking status was associated with receipt of guideline-concordant care.

```{r}
#Bivariate Analysis: Guideline-Concordant Care x Smoking Status
chisq.test(table(clean.df.guidelines$guide_conc_care,
                 clean.df.guidelines$smoke))

test.6 <- summary(glm(guide_conc_care ~ smoke,
                      data = clean.df.guidelines,
                      family = "binomial"))
test.6

exp(coef(glm(guide_conc_care ~ smoke,
             data = clean.df.guidelines, 
             family = binomial())))
```

### Multivariate Regression Model

A multivariate regression model for predictors of guideline-concordant care was created using the variables with significant associations in Table 5. After combining these variables, markers tested prior to 1L was the biggest predictor of receiving guideline-concordant treatment, followed by receipt of targeted therapy (or TKIs). These two predictors were the only ones with statistical significance after accounting for the other variables. The variance inflation factor (VIF) analysis for multicollinearity yielded no correlations strong enough to imply significant collinearity among the variables in the model (VIF values \<5).

```{r}
#Analyses: Predictors of Guideline-Concordant Care (Multivariate GLM)

#Multivariable Regression GLM #w/ significant variables from Table 5 

glm.guidelines <- glm(guide_conc_care ~ comprehensive_testing + markers_1L + smoke + test_modality + alteration_di + tx_type + tx_site,
                      data = clean.df.guidelines, family = "binomial")
summary(glm.guidelines)

```

```{r}
#Multivariable Regression GLM - Variance inflation factor

vif(glm.guidelines)    #Calculate variance inflation factor (VIF)

```

```{r}
#Multivariable Regression GLM - odds ratios

exp(coef(glm.guidelines))   #Odds ratios
```

```{r}
#Multivariable Regression GLM - Confidence intervals

confint(glm.guidelines) #Confidence intervals for fit

exp(cbind(OR = coef(glm.guidelines), CI = confint(glm.guidelines))) #OR and 95% CI
```

Lastly, I wanted to find out what the most commonly identified alterations were. *KRAS G12C* (with targeted therapies approved in second-line treatment) was the most frequent. *EGFR L858R (Exon 21), EGFR Exon 19 deletion*, *BRAF V600E, EM4L-ALK fusion*, and *MET Exon 14 Skipping* were the most commonly identified alterations -- all of which have FDA-approved targeted therapies for 1L treatment.

```{r}
#Top 10 Actionable Alterations Detected
genes <- str_split(df.1$alteration_gene, pattern = ",")
genes_vector <- unlist(genes)
genes_vector_no_na <- genes_vector[!(genes_vector) %in% "N/A"]
gene_frequencies <- table(genes_vector_no_na)
gene_frequencies_df <- as.data.frame(gene_frequencies)

top_genes <- gene_frequencies_df |>
  arrange(desc(Freq)) |>
  head(10)
top_genes

#Plot frequencies

ggplot(data = top_genes,
       mapping = aes(x = reorder(genes_vector_no_na, Freq), Freq)) + 
  geom_bar(stat = "identity") + 
  coord_flip() +
    labs(title = "Top 10 Most Commonly Identified Alterations",
       x = "",
       y = "Frequency")

```

## Discussion

The results from this exploratory analysis of molecular testing overwhelmingly show evidence for incorporating liquid biopsy (plasma NGS) to help guide clinical decision-making in the treatment of patients newly-diagnosed stage IV mNsq NSCLC. Given its comprehensiveness, quick turn-around-time of results, and likelihood association with having markers tested prior to 1L, plasma NGS testing has the potential to improve timely access to molecularly-informed, guideline-concordant care \-- inevitably yielding better outcomes for an already vulnerable patient population.

### Limitations

It is worth noting that the rates of comprehensiveness and testing prior to 1L demonstrated in this study are elevated in comparison to many other real-world clinical practices, in which lack of comprehensive testing remains prevalent. The increased rates observed can be attributed to multiple health-system initiatives geared toward increasing comprehensive molecular genotyping prior to 1L treatment. It is also possible that some of the findings were due to the inherent nature of the characteristics observed (i.e., smoking status may have been associated with receipt of guideline-concordant care due to the large percentage of smokers who are diagnosed with lung cancer).

### Future Directions

While clinically advantageous, comprehensive plasma NGS testing can be costly. Few studies have been done in the United States to explore cost analyses. To better determine the practicality of incorporating liquid biopsy into routine care for patients with NSCLC, cost analysis studies should be conducted in the future. Implementation studies for ways to increase comprehensive molecular testing prior to 1L should be pursued. As targeted therapies continue to be developed and the existing knowledge-base expands, longitudinal studies could assess clinical outcomes over time for patients who receive targeted therapy.

### Concluding Remarks

Using EHR-data, this study provides real-world findings that are translatable to clinical practice. With the continued evolution of precision medicine, the number of targeted therapies and their approved uses will increase, obviating the need for comprehensive molecular testing prior to 1L therapy. Without the proper testing, the proper treatment cannot be delivered, and patients may lose the survival benefits of targeted therapies.

## References

1\. Lung Cancer - Non-Small Cell - Statistics. Cancer.Net. Published June 25, 2012. Accessed December 7, 2023. https://www.cancer.net/cancer-types/lung-cancer-non-small-cell/statistics

2\. Siegel RL, Miller KD, Fuchs HE, Jemal A. Cancer statistics, 2022. *CA: A Cancer Journal for Clinicians*. 2022;72(1):7-33. doi:10.3322/caac.21708

3\. Bruno DS, Hess LM, Li X, Su EW, Zhu YE, Patel M. Racial disparities in biomarker testing and clinical trial enrollment in non-small cell lung cancer (NSCLC). *JCO*. 2021;39(15_suppl):9005-9005. doi:10.1200/JCO.2021.39.15_suppl.9005

4\. McNulty R. Inequalities Within and Between Practices Contribute to NGS Testing Disparities in NSCLC. Published June 15, 2023. Accessed December 7, 2023. https://www.ajmc.com/view/inequalities-within-and-between-practices-contribute-to-ngs-testing-disparities-in-nsclc

5\. Aggarwal C, Thompson JC, Black TA, et al. Clinical Implications of Plasma-Based Genotyping With the Delivery of Personalized Therapy in Metastatic Non--Small Cell Lung Cancer. *JAMA Oncology*. 2019;5(2):173-180. doi:10.1001/jamaoncol.2018.4305

6\. Aggarwal C, Marmarelis ME, Hwang WT, et al. Incorporation of plasma-based next-generation sequencing to improve guideline-concordant molecular testing in patients with newly diagnosed metastatic nonsquamous non-small cell lung cancer. *JCO*. 2021;39(28_suppl):14-14. doi:10.1200/JCO.2020.39.28_suppl.14

7\. Aggarwal C, Marmarelis ME, Hwang WT, et al. Association Between Availability of Molecular Genotyping Results and Overall Survival in Patients With Advanced Nonsquamous Non--Small-Cell Lung Cancer. *JCO Precis Oncol*. 2023;(7):e2300191. doi:10.1200/PO.23.00191

8\. Thompson JC, Aggarwal C, Wong J, et al. Plasma Genotyping at the Time of Diagnostic Tissue Biopsy Decreases Time-to-Treatment in Patients With Advanced NSCLC---Results From a Prospective Pilot Study. *JTO Clin Res Rep*. 2022;3(4):100301. doi:10.1016/j.jtocrr.2022.100301

9\. Roberts TJ, Kehl KL, Brooks GA, et al. Practice-Level Variation in Molecular Testing and Use of Targeted Therapy for Patients With Non-Small Cell Lung Cancer and Colorectal Cancer. *JAMA Netw Open*. 2023;6(4):e2310809. doi:10.1001/jamanetworkopen.2023.10809

10\. Waterhouse DM, Tseng WY, Espirito JL, Robert NJ. Understanding Contemporary Molecular Biomarker Testing Rates and Trends for Metastatic NSCLC Among Community Oncologists. *Clin Lung Cancer*. 2021;22(6):e901-e910. doi:10.1016/j.cllc.2021.05.006

11\. Hess L, Krein P, Haldane D, Han Y, Sireci A. Biomarker Testing for Patients with Advanced/Metastatic Non-squamous Non-small Cell Lung Cancer (NSCLC) in the USA, 2015-2021. *JTO Clinical and Research Reports*. 2022;3:100336. doi:10.1016/j.jtocrr.2022.100336
