---
title: "An Exploratory Analysis of Molecular Testing in Lung Cancer" 
subtitle: "BMIN503/EPID600 Final Project"
author: "Anthony O. Martella"
format: html
editor: visual
number-sections: true
embed-resources: true
editor_options:
  chunk_output_type: inline
execute:
  warning: false
  message: false
---

------------------------------------------------------------------------

### Overview {#sec-overview}

This project seeks to explore next-generation gene sequencing (NGS) testing rates and patterns, testing modalities, and receipt of guideline concordant care among 552 patients with stage IV mNSq NSCLC using electronic health record (EHR)-extracted data. A primary aim of this project is to identify if there are any predictors of receiving guideline concordant care, particularly in relation to test modality, comprehensiveness, prior to 1L, and other demographic characteristics. Here is a link to the working draft of [my GitHub project](https://github.com/amartella1/BMIN503_Final_Project).

I have discussed my final project with Charu Aggarwal, MD, MPH and Melina Marmarelis, MD, MSCE. Dr. Aggarwal is the Leslye M. Heisler Associate Professor for Lung Cancer Excellence at the University of Pennsylvania Perelman School of Medicine, Director of the Program in Precision Oncology at the Penn Center for Cancer Care Innovation, and a thoracic medical oncologist at the Abramson Cancer Center. Dr. Marmarelis is an Assistant Professor of Medicine and Co-Director of the Molecular Tumor Board at the University of Pennsylvania, and a thoracic medical oncologist at the Abramson Cancer Center.

### Introduction {#sec-introduction}

Lung cancer is one of the leading causes of cancer death in the United States (US) and internationally. With approximately 2,206,771 people diagnosed worldwide in 2020, lung cancer is also one of the most frequently diagnosed. Non-small cell lung cancer (NSCLC) comprises 80-85% of cases -- roughly 70% with non-squamous histology and 30% squamous histology. Over the past few decades, 5-year survival rates for individuals with stage IV metastatic NSCLC have increased from less than 5% to greater than 25%. This increased survival rate is in part attributed to advancements in the field of precision medicine, which have allowed for the development of targeted (or personalized) therapies in lung cancer care. Delivery of appropriate targeted therapies (at the right time) has been shown to improve patient outcomes, including overall survival rate.

To date, the majority of FDA-approved targeted therapies for lung cancer are indicated in the treatment of metastatic non-squamous (mNSq) NSCLC. Although, research has begun demonstrating targeted therapy may have beneficial effects at earlier stages. To determine whether someone might benefit from targeted therapy necessitates comprehensive next-generation gene sequencing (NGS) -- also referred to as molecular testing -- which can be performed on a tumor tissue sample or plasma (blood) sample. The current National Comprehensive Cancer Network (NCCN) guidelines define comprehensiveness as the detection of alterations in seven genes (*EGFR, ALK, BRAF, ROS1, MET, RET,* and *NTRK*). Specific alterations in these seven genes each have FDA-approved targeted therapies for the treatment of mNSq NSCLC in the first-line (1L) setting. The FDA has also recently approved use of targeted therapy in second-line (2L) treatment for *KRAS G12C* and *ErbB2* alterations.

The NCCN guidelines recommend comprehensive NGS testing be completed for all patients with mNSq NSCLC prior to initiating 1L treatment. The survival benefit from targeted therapy can be forfeited if it is not delivered at the most appropriate time. For example, research has shown that administering targeted therapy for patients with an EGFR alteration can be toxic if they have already received 1L immunotherapy. This paradigm illustrates the need for a quick turn-around-time (TAT) of results before 1L treatment.

Aside from NGS, single-gene or "hot-spot" testing, limited gene panels, fusion transcript, and fluorescence in situ hybridization (FISH) analysis are other forms of molecular testing in lung cancer. Traditional approaches to molecular testing involve tissue testing that can take anywhere from 4 to 12 weeks to result. While some medical institutions conduct NGS testing in-house, more commonly tissue samples are sent out to 3^rd^ party companies. A recurrent limitation of tissue testing is that there is not always enough tissue to perform sequencing -- quality/quantity not sufficient (QNS), which can sometimes lead to a re-biopsy, further delaying treatment for an already vulnerable population.

Due to evolving technology and increased high-performance computing abilities, plasma-based comprehensive NGS testing, also referred to as liquid biopsy, is made possible by genotyping circulating-tumor DNA (ct-DNA). Results for plasma-based NGS are often available within 7-10 days. While tissue testing remains a gold standard, research has demonstrated that incorporating plasma NGS testing into standard of care leads to a significant increase in receipt of  the detection of clinically actionable alterations. In addition, plasma NGS testing is associated decreased time to treatment.

Despite the known benefits of comprehensive molecular testing, testing rates have not caught up to speed for a variety of reasons.  The MYLUNG (Molecularly Informed Lung Cancer Treatment in a Community Cancer Network) Consortium™ provides real-world biomarker testing patterns and in a recent study found that out of 10,333 patients with nSq NSCLC only 53% of patients (n = 5494) had undergone NGS testing. Testing rates tend to be significantly lower for patients seen in community cancer centers compared to those seen in academic medical practices. Furthermore, there are known disparities of NGS testing rates by race. Differences up to over 10 percentage-points have been observed in testing rates such that Black individuals received comprehensive testing less than White individuals.    

This project seeks to explore NGS testing rates and patterns, testing modalities, and receipt of guideline concordant care among patients with stage IV mNSq NSCLC. A primary aim of this project is to identify if there are any predictors of receiving guideline concordant care, particularly in relation to test modality, comprehensiveness, prior to 1L, and other demographic characteristics. Secondary analyses will examine testing TAT of results by test modality, testing differences by race, treatment site, predictors of having a clinically actionable alteration detected, and most commonly detected alterations.  

### Methods {#sec-methods}

This study employs electronic health record (EHR)- extracted data from 552 patients with newly diagnosed mNSq NSCLC within the years 2019-2022. Patients included were seen in an outpatient cancer center at either an academic medical institution or two affiliated community practices.

```{r}
#Packages loaded
#| label: load-packages
library(tidyverse)
library(tidymodels)
library(readr)
library(dplyr)
library(gtsummary)
library(lubridate)
library(ggplot2)
library(cowplot)
```

```{r}
#Load the datasets
df.2019 <- read_csv("Clean_2019_Cohort_AOM.csv")
df.2020 <- read_csv("Clean_2020_Cohort_AOM.csv")
df.2021.2022 <- read_csv("Clean_2021-2022_Cohort_AOM.csv")

#Merge the datasets using bind_rows
data_all <- bind_rows(df.2019, df.2020, df.2021.2022)
```

```{r}
#Data Cleaning and Transforming

#Create new factors using mutate

df.1 <- data_all |>
  mutate(race = factor(case_when(
    race == "Pacific Island" | race == "Hispanic Latino/White" | race == "Other/Unknown" | race == "Multiple Races" ~ "Other/Unknown",
    TRUE ~ race
  ))) |>
  mutate(sex = factor(sex, levels = c("F", "M"),
                      labels = c("Female", "Male"))) |>
  mutate(guide_conc_care = factor(guide_conc_care, levels = c("Yes", "No"),
                                  labels = c("Yes", "No"))) |>
  mutate(markers_1L = factor(markers, levels = c("Yes", "No"),
                          labels = c("Yes", "No"))) |>
  mutate(smoke = factor(smoke, levels = c("Former", "Current", "Never"),
                        labels = c("Former", "Current", "Never"))) |>
  mutate(test_modality = factor(samp_type, levels = c("Tissue", "Plasma", "TP"),
                                labels = c("Tissue", "Plasma", "Tissue+Plasma"))) |>
  mutate(alteration_di = factor(alteration_di, levels = c("Yes", "No"),
                                labels = c("Yes", "No"))) |>
  mutate(alteration_tpb = factor(alteration_tpb, levels = c("Tissue", "Plasma", "Both", "N/A"),
                                 labels = c("Tissue", "Plasma", "Both", "N/A"))) |>
  mutate(tx_prescribed = factor(tx_prescribed, levels = c("Yes", "No"),
                                labels = c("Yes", "No"))) |>
  mutate(tx_type = factor(tx_type, levels = c("Chemo", "ChemoIO", "IO", "TKI", "NONE"),
                          labels = c("Chemo", "ChemoIO", "IO", "TKI", "NONE"))) |>
  mutate(comprehensive_testing = factor(if_else(tier == "Tier 1" | tier == "Tier 2", "Yes", "No"))) |>
  mutate(tx_site = gsub("VF", "Valley Forge", tx_site),
         tx_site = gsub("CCH", "Chester County", tx_site)) |>
  mutate(tx_site = factor(tx_site, levels = c("PCAM", "Chester County", "Valley Forge"),
                          labels = c("PCAM", "Chester County", "Valley Forge"))) |>
  mutate(ngs_ttr_col = as.numeric(ngs_ttr_col)) |>
  mutate(pb_ttr_col = as.numeric(pb_ttr_col)) |>
  mutate(dx_date = mdy(dx_date)) |>
  mutate(tx_1l_start = mdy(tx_1l_start)) |>
  mutate(comprehensiveness_date = mdy(comprehensiveness_date)) |>
  mutate(day.diff.dxdate.tx1l = as.numeric(difftime(ymd(tx_1l_start),
                                                    (dx_date),
                                                    units = "days"))) |>
  mutate(day.diff.compdate.tx1l = as.numeric(difftime(ymd(tx_1l_start),
                                                      (comprehensiveness_date),
                                                      units = "days")))
```

### Results (*in progress*) {#sec-results}

Describe your results and include relevant tables, plots, and code/comments used to obtain them. End with a brief conclusion of your findings related to the question you set out to address. You can include references if you'd like, but this is not required.

```{r}
#Descriptive Tables: Demographics & Testing

#Select demographic variables into a new data frame
df.demo <- df.1 |>                      
  select(sex, age, dx_date, race, smoke, tx_site, comprehensive_testing, markers_1L, test_modality)

#Create a demographics table summary
tbl.demo <- tbl_summary(df.demo, missing = "no",
                        label = list(sex ~ "Sex",
                                     age ~ "Age at Diagnosis",
                                     dx_date ~ "Diagnosis Date",
                                     race ~ "Race",
                                     smoke ~ "Smoking Status",
                                     tx_site ~ "Treatment Site",
                                     comprehensive_testing ~ "Comprehensive Testing",
                                     markers_1L ~ "Markers tested prior to 1L therapy",
                                     test_modality ~ "Combined Test Modality")) |>
                        bold_labels()
tbl.demo

```

```{r}
#Analyses: Differences in Comprehensive Testing Rates
#Are there significant differences in comprehensive testing rates by race and tx_site? (run analyses and plot)
tbl.comp.testing <- df.1 |>
  select(comprehensive_testing, test_modality, tx_site, smoke, race, age, sex) |>
  tbl_summary(by = comprehensive_testing,
              missing = "no") |>
  add_p() |>
  add_significance_stars()
tbl.comp.testing

#Test modality, tx site, and smoking status are significantly associated with receiving comprehensive testing - how to assess differences among sub-groups? 
```

```{r}
#Descriptive Table: Differences by Testing Modality 

df.test.char <- df.1 |>
  select(comprehensive_testing, markers_1L, test_modality, guide_conc_care, smoke, alteration_di)

#Summary Table by Testing Modality -- *need to clean up labels!
tbl.test.char <- df.test.char |>
  tbl_summary(by = test_modality,
              missing = "no",
              label = list(
                comprehensive_testing ~ "Comprehensive Testing",
                markers_1L ~ "Markers tested prior to 1L therapy",
                guide_conc_care ~ "Received Guideline-Concordant Care",
                smoke ~ "Smoking Status",
                alteration_di ~ "Alteration Detected"
              )) |>
  add_p() |>
  add_overall() |>
  add_significance_stars()
tbl.test.char


#Result comment: Patients who underwent plasma or plasma+tissue testing had comprehensive testing. Those who just had tissue alone did not? 

#Sig relationship b/w having markers tested prior to 1L and whether testing done on T/P/TP.
#Appears to be a relationship with receiving guideline concordant care and testing modality - how to explore this further??
```

```{r}
#Analyses: TAT - Turn around time of results 

#Are there significant differences in turn around time (TAT) of results tissue ngs v plasma? (analyze and plot)
#ngs_ttr_col  #pb_ttr_col -- days from collection to report date 

median(df.1$ngs_ttr_col, na.rm = TRUE)
median(df.1$pb_ttr_col, na.rm = TRUE)

wilcoxon.rank.test <- wilcox.test(df.1$ngs_ttr_col, df.1$pb_ttr_col, paired = TRUE)
wilcoxon.rank.test

```

```{r}
#Analyses: TTT - Time to Treatment

#Are there sig. differences between groups/testing modality in # of days to treatment start?

#Comprehensive date to date 1L start

#Is testing modality a predictor of # of days to 1L treatment?? Double check how to do this

df.ttt <- df.1 |>
  select(test_modality, day.diff.dxdate.tx1l, day.diff.compdate.tx1l)

#Table Summary
tbl.ttt <- df.ttt |>
  tbl_summary(by = test_modality,
              missing = "no") |>
  add_p()
tbl.ttt

```

```{r}
#Analyses: Treatment & Guideline-Concordant Care

#Overview table
tbl.guidelines.overview <- df.1 |>
  select(tx_site, tx_type, guide_conc_care, tx_prescribed) |>
  tbl_summary(missing = "no",
              label = list(
                tx_site ~ "Treatment Site",
                tx_type ~ "Treatment Type",
                guide_conc_care ~ "Received Guideline-Concordant Care",
                tx_prescribed ~ "Received Targeted-Therapy (if alteration detected)")) |>
  bold_labels()
tbl.guidelines.overview

#subset data frame
df.guidelines <- df.1 |>
  select(guide_conc_care, comprehensive_testing, markers_1L, test_modality, alteration_di, alteration_tpb, tx_site, tx_type, age, sex, race, smoke)

#Differences by guideline care concordance 
tbl.guidelines <- df.guidelines |>
  tbl_summary(by = guide_conc_care,
              missing = "no") |>
  add_p() |>
  add_significance_stars()
tbl.guidelines

```

```{r}
#Analyses: Predictors of Guideline Concordant Care

#Is testing modality a predictor of receiving guideline concordant care?
test.1 <- summary(glm(guide_conc_care ~ test_modality, data = df.guidelines, family = "binomial"))
test.1

#Is having moleculars tested prior to 1L a predictor of receiving guideline concordant care?
test.2 <- summary(glm(guide_conc_care ~ markers_1L, data = df.guidelines, family = "binomial"))
test.2

#Is smoking status a predictor of receiving guideline concordant care?
test.3 <- summary(glm(guide_conc_care ~ smoke, data = df.guidelines, family = "binomial"))
test.3

#Is having comprehensive testing a predictor of receiving guideline concordant care?
test.4 <- summary(glm(guide_conc_care ~ comprehensive_testing, data = df.guidelines, family = "binomial"))
test.4

#Multivariable Regression GLM

glm.guidelines <- glm(guide_conc_care ~ sex + age + race + smoke + test_modality + markers_1L + comprehensive_testing + alteration_di + alteration_tpb,
                      data = df.guidelines, family = "binomial")
summary(glm.guidelines)

```

```{r}
#Top 10 Actionable Alterations Detected
genes <- str_split(df.1$alteration_gene, pattern = ",")
genes_vector <- unlist(genes)
genes_vector_no_na <- genes_vector[!(genes_vector) %in% "N/A"]
gene_frequencies <- table(genes_vector_no_na)
gene_frequencies_df <- as.data.frame(gene_frequencies)

top_genes <- gene_frequencies_df |>
  arrange(desc(Freq)) |>
  head(10)
top_genes

#Plot frequencies / barplot? 
```

```{r}
#Analyses: Predictors for alteration detected? 'alteration_di'

#Is testing modality a predictor of alteration detected?
alt.1 <- summary(glm(alteration_di ~ test_modality, data = df.guidelines, family = "binomial"))
alt.1

#Is smoking status a predictor of alteration detected?
alt.2 <- summary(glm(alteration_di ~ smoke, data = df.guidelines, family = "binomial"))
alt.2
```

```{r}
#Comprehensive testing x treatment site

df.demo |>
  filter(!is.na(comprehensive_testing) & !is.na(tx_site)) |>
  ggplot(aes(x = tx_site, fill = comprehensive_testing)) +
  geom_bar(position = "fill")
```

```{r}
#Testing modalities x treatment site
df.demo |>
  filter(!is.na(test_modality) & !is.na(tx_site)) |>
  ggplot(aes(x = tx_site, fill = test_modality)) +
  geom_bar(position = "fill")
```

```{r}
#Comprehensive testing x markers_1L by treatment site
#Facet Grid - comprehensive testing (levels y/n) (filter yeses) x markers prior to 1L, by tx_site
```

```{r}
#Guideline concordance x Test Modality 
df.guidelines |>
  filter(!is.na(test_modality) & !is.na(guide_conc_care)) |>
  ggplot(aes(x = test_modality, fill = guide_conc_care)) +
  geom_bar(position = "fill")
```

```{r}
#Comp testing and guideline concordant care - corr plot? / scatter plot?

```

```{r}
#Alteration_di x alteration_tpb

```

```{r}
#Plot for frequency of alterations detected 

#Manhattan plot for genes tested?

#Create plots of increases over time 
```

### Discussion

### References
