---
title: "Investigating the role of rs199347 in Parkinson's disease and other neurological disorders"
subtitle: Marc Carceles-Cordon
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

## INTRODUCTION

Parkinsonâ€™s disease (PD) is the second most common neurodegenerative disorder worldwide affecting an estimated 2-3% of the population >60 years of age [1], and the fastest growing neurological disorder [2]. The hallmark pathological finding of PD is the aggregation of insoluble alpha-synuclein (aSyn) in the dopaminergic neurons of the substantia nigra pars compacta (SNpc) during the initial motor stages of the disease [2, 3]. The resulting decrease of dopamine release in the striatum perturbs basal ganglia circuitry and results in cardinal motor manifestations that have been, for several decades, the target of pharmacologic dopamine replacement strategies. Despite achieving a reduction in motor symptoms and an overall improvement in the quality of life of patients, there are currently no disease-modifying therapies for PD [4], and while disease course is heterogeneous [5], most patients progress to exhibit extensive motor symptoms as well as non-motor symptoms such as cognitive dysfunction during their lifetime [6-8]. 

The genetic architecture of PD is complex and heterogeneous, ranging from rare DNA variants that are highly penetrant (and cause familial PD) to common variants that individually exert small increases in lifetime risk but that collectively cause a significant amount of risk (and are usually identified in sporadic PD). Over 20 rare variants have been identified thus far that cause familial PD, including but not limited to PARK7, LRRK2 or GBA [12, 13]. At the genome-wide level, GWAS studies have identified over 90 common variant risk factors for PD by performing large sample unbiased comparisons of PD and healthy controls [14]. These studies have been a success in terms of the number of GWAS hits found, but in many cases have failed to elucidate the downstream mechanistic implications of these discoveries, underlining the timely necessity of harnessing these gene variants to understand molecular mechanisms of disease. 

In this context, a recently linked GWAS-derived chromosome 7 PD risk locus [sentinel single-nucleotide polymorphism (SNP) rs199347] has been linked to the non-coding region of GPNMB, which encodes the glycoprotein nonmetastatic melanoma protein B (GPNMB) [15]. rs199347 has been reported to be a risk SNP for PD, with the G allele conferring a protective effect [14, 16] associating with lower GPNMB expression [17, 18]. In a recently published paper [15], it has been confirmed that the rs199347 haplotype is associated with threefold higher GPNMB expression in both PD and normal brains, with the highest expression being in the caudate region. Interestingly, GPNMB has also been reportedly elevated in other neurological disorders, such as ALS [19]. In the CNS, GPNMB is expressed mainly by microglial cells and seems to be part of the DAM (disease associated microglia) phenotype expressed in neuroinflammatory circumstances and particularly enriched in neurodegenerative disorders [20]. 

Due to this growing evidence, I hypothesize that GPNMB expression could be linked to various major neurological disorders - particularly PD, for which there is a growing body of evidence. I also hypothesize that GPNMB can be linked to clinical outcomes in the disorders where it is differentially expressed. 

## METHODS

In order to explore these hypotheses, I will use the INDD database, which contains longitudinal clinical data, genotype, and biomarker information from >10000 neurological patients followed at the University of Pennsylvania. In particular, the dataset that I will be using will contain the following variables: 

- INDDID. A unique identification code for each patient.  

- rs199347 haplotype (AA, BA, BB). All patients selected for the downstream analysis will be genotyped for the rs199347 haplotype. A is the risk allele conferring a higher GPNMB expression while B is the protective allele conferring lower GPNMB expression.  

- GlobalDx. A variable that combines the primary neurological diagnosis of the patient. 

- GlobalAgeOnset. A variable that contains the initial age of onset of the primary neurological diagnosis. 

### Cleaning up the dataset

I first will begin by loading the required packages, reading in and cleaning up the dataset in order to group by major neurological disease diagnosis. 
```{r cars}
library(dplyr)
library(readxl)
library(ggplot2)
library(tidyr)

df <- read_excel("~/Desktop/Lab/Chen-Plotkin/GPNMB_Cognition/INDD_GPNMB_AdvancedClinicalInfo.xlsx")
df %>% count(rs199347) #There are 1677 "NAs" that I will drop in order to keep only the cases for which we have genotype info
df <- subset(df, rs199347 == "AA" | rs199347 == "BA" | rs199347 == "BB") #Filtering for only AA, BA, and BB GPNMB genotypes
global_dx_list <- df %>% count(GlobalDx) #Creating a list of all diagnostic labels from GlobalDx variable 
neuropath_dx_list <- df %>% count(NPDx) #Idem for Neuropath dx 

#Cleaning the initial dataset to group by major disease DX 
df$GlobalDx <- sub("^(Alz).*", "AD", df$GlobalDx)
df$GlobalDx <- sub("^(Amy).*", "ALS", df$GlobalDx)
df$GlobalDx <- sub("Normal", "NC", df$GlobalDx)
df$GlobalDx <- sub("^(Park).*", "PD", df$GlobalDx)
df$GlobalDx <- sub("^(bvFTD).*", "FTD", df$GlobalDx)
df$GlobalDx <- sub("^(Mild).*", "MCI", df$GlobalDx)
df$GlobalDx <- sub("Multiple Clinical Dx", "Unspecified", df$GlobalDx)
df$GlobalDx <- sub("Dementia with Lewy Bodies", "DLB", df$GlobalDx)
df$GlobalDx <- sub("FTD-NOS", "FTD", df$GlobalDx)
df$GlobalDx <- sub("^(FTLD).*", "FTD", df$GlobalDx)
df$GlobalDx <- sub("Progressive supranuclear palsy", "PSP", df$GlobalDx)
df$GlobalDx <- sub("Corticobasal syndrome", "CBS", df$GlobalDx)
df$GlobalDx <- sub("^(PPA).*", "PPA", df$GlobalDx)
df <- df%>% drop_na(GlobalDx)
df.dx <- filter(df, GlobalDx == "AD" | GlobalDx == "ALS" | GlobalDx == "NC" | GlobalDx == "PD" | GlobalDx == "FTD" | 
                  GlobalDx == "PSP" | GlobalDx == "MCI" | GlobalDx == "Unspecified" | GlobalDx == "DLB" | GlobalDx == "PSP" | 
                  GlobalDx == "CBS" | GlobalDx == "PPA")
df.dx <- df.dx %>% select(INDDID, rs199347, GlobalDx)

library('gt')
df.dx %>% count(GlobalDx) %>% gt() 
```
The result (df.dx) is a dataframe containing the clinical information and GPNMB haplotype of >5000 patients. The printed list represents the patients grouped by major neurological disorder, where "NC" are healthy individuals, and "Unspecified" are individuals with not enough information regarding disease diagnosis. 

### Plotting the initial dataframe

I will now plot a graphical representation of the cleaned dataframe. 

```{r}
color = grDevices::colors()[grep('gr(a|e)y', grDevices::colors(), invert = T)]

p.genotypes <- ggplot(df.dx, aes(x= rs199347, fill=rs199347)) +
  geom_bar(position = "dodge") +
  scale_fill_hue(c = 40) +
  theme(legend.position="none") +
  geom_text(aes(label = ..count..), stat="count", position= position_dodge(width = 1),
            vjust = -0.5, size = 3.5)
p.genotypes 

p.diseases <- ggplot(df.dx, aes(x=GlobalDx, fill=GlobalDx)) + 
  geom_bar(position = "dodge") +
  scale_fill_brewer(palette="Spectral") + #Choosing a color palette that allows for 11 distinct categories
  theme(legend.position="none") +
  geom_text(aes(label = ..count..), stat="count", position= position_dodge(width = 1),
            vjust = -0.5, size = 3.5)
p.diseases 
```

"p.genotypes" is a count of each haplotype (AA, BA, or BB) for the rs199347 SNP within our cleaned dataframe. "p.diseases" is a count of each major neurological disorder (plus NCs, and unspecified) in our cleaned dataframe. 

## Can the rs199347 haplotype predict disease category? 

The first question I am interested in answering based off the cleaned dataframe is whether the rs199344 haplotype status is associated with any major neurological disorders when compared to healthy controls. In particular, based on existing literature I predict that the rs199347 AA haplotype, which correlates with higher GPNMB expression levels in the brain, should correlate with a PD diagnosis. It is unknown, howwever, if that same haplotype can also correlate with other neurological disease diagnoses. 

### Grouping the data by disease category and counting how many individuals classify per haplotype

```{r}
df.dx2 <- df.dx[,-c(1)] #Cleaning the dataframe to exclude the identification of patients (INDDID)
grouped_data <- group_by(df.dx2, GlobalDx)
count.all <- summarize(grouped_data, AA = sum(rs199347 == "AA"), BA = sum(rs199347 == "BA"), BB = sum(rs199347 == "BB"))
df.count.all <- data.frame(count.all)
counts.g <- tidyr::gather(df.count.all, haplotype, value, -GlobalDx) #New dataframe in which each disorder type has been subset per haplotype
counts.g %>% gt()
```

We now have a dataframe (counts.g) in which each disease category has been subset per rs199347 haplotype, and we will subsequently visualize this result.

```{r}
p.counts <- ggplot(counts.g, aes(x = GlobalDx, y = value, fill = haplotype)) +
  geom_bar(stat = "identity", position = "dodge") +
  labs(title = "rs199347 Haplotype per Major Neurological Disorder", x = "Major Neurological Disorder", y = "Count") + 
  scale_fill_manual(values = c('#7CAE00', '#C99800', '#F8766D')) +
  geom_text(aes(label=value), vjust = -0.5, size = 2.5, position = position_dodge(width = 1))
p.counts
```

"p.counts" is a bar plot showing the number of individuals per haplotype per disease category. As we can see, the major number of individuals carrying the AA haplotype for this particular SNP are located within the PD category. However, it is also true that this particular dataset is enriched by PD patients. We are interested in investigating the percentage of individuals carrying each haplotype per disease category.

```{r}
df.count.all <- df.count.all %>%
  mutate("AA_Percentage" := AA / (AA + BA + BB) * 100,
         "BA_Percentage" := BA / (AA + BA + BB) * 100, 
         "BB_Percentage" := BB / (AA + BA + BB) * 100,
         "Total" := AA + BA +BB)
print(df.count.all)
```

df.count.all is a dataframe that contains the percentages of each haplotype per disease category. We can see that the risk allele (AA), which correlates with higher GPNMB expression in the brain, is significantly enriched in our dataset for PD diagnosis. It is also particularly enriched in FTD, DLB and PSP. 

### Statistical analysis of rs199347 haplotype per disease category

Now I want to study which risk haplotypes are statistically significantly different when compared to healthy controls. To do so, I have decided to use a two proportions Z-test comparing an observed proportion (for each particular disorder) to an expected one (NCs).

```{r}
my.list <- df.count.all %>%    #Create a function that performs a prop test on each row of the dataframe on the AA haplotype
    rowwise %>% 
    mutate(prop_out = list(prop.test(c(AA, 91), n = c(Total, 343)))) %>% 
    ungroup
proptest_vector <- unlist(my.list$prop_out, use.names = FALSE)[seq(3, nrow(my.list) * 10, by = 10)]
proptest_vector #Contains the pvalues of each prop test 
df.count.all$p.value <- proptest_vector
df.count.AA <- df.count.all %>% select(GlobalDx, AA, BA, BB, AA_Percentage, BA_Percentage, BB_Percentage, p.value) #Create a cleaned dataframe containing the information of interest
stars.pval <- function(x){
  stars <- c("***", "**", "*", "n.s.")
  var <- c(0, 0.01, 0.05, 0.10, 1)
  i <- findInterval(x, var, left.open = T, rightmost.closed = T)
  stars[i]
}
df.count.AA <- transform(df.count.AA, stars = stars.pval(df.count.AA[[8]])) #Add significance to the table as a new column
df.count.AA <- arrange(df.count.AA, p.value)

df.count.AA %>% gt() %>% 
  fmt_number(
    columns = c(AA_Percentage, BA_Percentage, BB_Percentage), 
    decimals = 2
  ) %>%
    tab_header(
    title = "Haplotype Count per Major Neurological Disorder",
    subtitle = "Z-test paired comparison for the risk (AA) haplotype"
    ) %>%
  gtsave("tab1.png", expand = 10)
```

df.count.AA contains the statistical p.value (and significance asterisks) of each paired comparison Z-test performed against the expected observations (HCs). At a first glance, we can see that as expected PD is the most significant result, but unexpectedly individuals carrying a AA haplotype are also enriched for other major neurological disorders such as PSP, FTD, and ALS. We can see this more clearly if we group the results by disease. 

```{r}
df.count.AA.table1 <- df.count.AA %>% select(GlobalDx, AA_Percentage, BA_Percentage, BB_Percentage, p.value, stars)
final.df.haplotypes <- as.data.frame(t(df.count.AA.table1))
colnames(final.df.haplotypes) <- final.df.haplotypes[1,]
final.df.haplotypes <- final.df.haplotypes[-c(1),]
print(final.df.haplotypes)
```

### Graphical representation of rs199347 haplotype per disease category

I will now plot the previous results. 

```{r}
df.count.AA$GlobalDx <- factor(df.count.AA$GlobalDx,
                               levels = df.count.AA$GlobalDx[order(df.count.AA$p.value, decreasing = TRUE)])  

p.diseases.AA <- ggplot(df.count.AA, aes(x = GlobalDx, y = AA_Percentage, fill = GlobalDx)) +
  geom_bar(stat = "identity", position = "dodge") +
  labs(title = "GPNMB AA Haplotype Percentage per Major Neurological Disorder", x = "Major Neurological Disorder", y = "%") + 
  scale_fill_brewer(palette="Spectral") +
  geom_text(aes(label=stars), vjust = -0.5, size = 4, color="red")
p.diseases.AA
```

The 'p.diseases.AA' graph represents the percentages of individuals carrying the risk alle per disease category. Each barplot is labelled with the significance level of the previous statistic test. On the left side we have the least significantly different result, which is of course represented by the NCs (which are the expected proportion). On the right side of the graph we can see the disorders in which a the risk haplotype represents a significantly different proportion than expected, with the most different being PD. 

## Does GPNMB AA haplotype correlate with cognitive status? 

Having seen that the risk allele for rs199347 correlates with the diagnosis of various neurological disorders that impair cognition, my next question is to investigate whether the AA haplotype correlates with impaired cognitive status. 

```{r}
df.onset <- select(df, INDDID, GlobalAgeOnset, CognitiveAgeOnset, MotorAgeOnset, CognitiveDx)
df.onset <- df.onset[,-c(3,4)] 
```

I have decided against using CognitiveAgeOnset because it does not provide enough information. Since in PD the diagnosis is usually made based on motor symptoms, it would also be redundant to use MotorAgeOnset. I will use the variable GlobalAgeOnset, which provides a summary of when the disease was first diagnosed. 

```{r}
df.cognitive <- merge(df.dx, df.onset, by="INDDID")
df.cognitive <- df.cognitive %>% drop_na(CognitiveDx)

p.cognitive <- ggplot(df.cognitive, aes(factor(CognitiveDx),fill = rs199347)) +
  geom_bar(stat="count", position = "dodge") + 
  scale_fill_brewer(palette = "Set1") +
  labs(title = "Cognitive Status per GPNMB Haplotype", x = "Cognitive Status", y = "Count") +
  geom_text(aes(label = ..count..), stat="count", position= position_dodge(width = 1),
            vjust = -0.5, size = 3.5) +
  theme(plot.title = element_text(hjust = 0.5))
p.cognitive 
```

In this case, 'p.cognitive' shows that the AA haplotype is not particularly enriched in demented patients when compared to normal or MCI patients. This may be because other diseases (such as AD) that impair cognition are not particularly enriched for the AA haplotype based on our previous results.

## Does rs199347 haplotype correlate with earlier disease onset in PD? 

Having seen that individuals carrying the risk allele for SNP rs199347 are more at risk for being diagnosed with PD when compared to controls, I now want to observe whether the haplotype has also an impact on the clinical course of disease. Based on the information provided in my initial dataframe, I have decided to study the impact of rs199347 haplotype on age of disease onset.  

```{r}
df.clinical <- merge(df.dx, df.onset, by="INDDID")
df.clinical <- df.clinical %>% drop_na(GlobalAgeOnset)
head(df.clinical) #We will do some statistical analysis on this dataset
ggplot(df.clinical, aes(x = GlobalAgeOnset, y=GlobalDx, color=rs199347)) +
  geom_point() 
```

The previous scatterplot shows the cases per age of onset and haplotype for each major neurological disorder. 

First, I will analyze if age of onset has a significant effect on the probability of belonging to each haplotype (3-way categorical variable) for Parkinson's Disease.

```{r}
df.PD <- select(df.clinical, rs199347, GlobalDx, GlobalAgeOnset)
df.PD <- df.PD %>% filter(GlobalDx == "PD") #Filter dataframe only for PD 
contingency_table <- table(df.PD$GlobalAgeOnset, df.PD$rs199347)
chisq.test(contingency_table)
```

The chi-squared test performed on the contingency table has determined that there is a significant association between the age of onset and the haplotype. 

```{r}
ggplot(data = df.PD, aes(x = factor(rs199347), y = GlobalAgeOnset)) +
  geom_violin(fill = "lightyellow", color = "sienna4") +
  geom_boxplot(width = 0.1, color = "sienna4", alpha = 0.2)
normalized = (df.PD$GlobalAgeOnset-min(df.PD$GlobalAgeOnset)) / (max(df.PD$GlobalAgeOnset)-min(df.PD$GlobalAgeOnset))
df.PD$GlobalAgeOnsetNormalized <- normalized
summary((glm(GlobalAgeOnsetNormalized ~ rs199347 , data = df.PD, family = binomial())))
```

According to the glm model, there is a statistically significant association between AA haplotype and the age of onset of disease. 

Next, I will investigate whether there is a statistically significant difference between haplotypes using a one-way ANOVA.

```{r}
df.PD <- df.PD %>% 
  mutate(rs199347 = factor(rs199347, levels = c("AA", "BA", "BB"), labels = c("AA", "BA", "BB")))
levels(df.PD$rs199347)
group_by(df.PD, rs199347) %>%
  summarise(
    count = n(),
    mean = mean(GlobalAgeOnset, na.rm = TRUE),
    sd = sd(GlobalAgeOnset, na.rm = TRUE)
  )
res.aov <- aov(GlobalAgeOnset ~ rs199347, data = df.PD)
summary(res.aov) 
```

According to the one-way ANOVA, there is a highly significant difference in the age of onset of PD based on GPNMB haplotype. Let's represent the results graphically.

```{r}
library(viridis)
library(hrbrthemes)
df.PD %>%
  ggplot( aes(x=rs199347, y=GlobalAgeOnset, fill=rs199347)) +
    geom_boxplot() +
    scale_fill_viridis(discrete = TRUE, alpha=0.6) +
    #geom_jitter(color="black", size=0.4, alpha=0.9) +
    theme_ipsum() +
    theme(
      legend.position="none",
      plot.title = element_text(size=11)
    ) +
    ggtitle("Age of Onset of PD according to GPNMB Haplotype") +
    xlab("")
```

## DISCUSSION

From this study, I have seen that there is a statistically significant association between the AA haplotype for rs199347 and several neurological disorders. The most statistically significant association has been found in PD, as it was expected based on the existing literature. In positive correlation with the existing literature, the AA haplotype was also statistically enriched in our database for ALS. DLB, a closely disorder associated with PD, was also statistically enriched for the AA haplotype, and these results can be explained by the fact that DLB also contains alpha-synuclein inclusions as a pathological hallmark. rs199347 is associated with a higher GPNMB level, and GPNMB has been reported to act as a mediator of alpha-synuclein internalization. Finally, the fact that the rs199347 risk haplotype is associated with PSP and FTD, two other neurodegenerative disorders, opens the door for future studies investigating whether GPNMB is elevated in the plasma or CSF of patients with these disorders. 

Next, I turned to investigating whether the rs199347 haplotype could be related not only with the diagnosis, but also with the clinical aggressiveness of one of these neurological disorders. Since GPNMB has been mainly studied in PD, and PD was the most statistically significant result on the initial analysis, I next turned to investigating whether the rs199347 haplotype would correlate with a more agressive form of disease. My results have shown that, as expected, a higher GPNMB level (represented by both AA and BA haplotypes) correlate with an earlier age of onset for PD. This result is in positive correlation with the literature and identify GPNMB not only as a potential biomarker but also a potential therapeutical target for PD. 

### REFERENCES

1.	James, S.L., et al., Global, regional, and national incidence, prevalence, and years lived with disability for 354 diseases and injuries for 195 countries and territories, 1990â€“2017: a systematic analysis for the Global Burden of Disease Study 2017. The Lancet, 2018. 392(10159): p. 1789-1858.
2.	Poewe, W., et al., Parkinson disease. Nature Reviews Disease Primers, 2017. 3(1): p. 17013.
3.	Spillantini, M.G., et al., Î±-Synuclein in Lewy bodies. Nature, 1997. 388(6645): p. 839-840.
4.	Elkouzi, A., et al., Emerging therapies in Parkinson disease â€” repurposed drugs and new approaches. Nature Reviews Neurology, 2019. 15(4): p. 204-223.
5.	Greenland, J.C., C.H. Williams-Gray, and R.A. Barker, The clinical heterogeneity of Parkinson's disease and its therapeutic implications. European Journal of Neuroscience, 2019. 49(3): p. 328-338.
6.	Coelho, M. and J.J. Ferreira, Late-stage Parkinson disease. Nature Reviews Neurology, 2012. 8(8): p. 435-442.
7.	Hely, M.A., et al., The Sydney multicenter study of Parkinson's disease: The inevitability of dementia at 20 years. Movement Disorders, 2008. 23(6): p. 837-844.
8.	Halliday, G.M. and H. McCann, The progression of pathology in Parkinson's disease. Annals of the New York Academy of Sciences, 2010. 1184(1): p. 188-195.
9.	Braak, H., et al., Stages in the development of Parkinsonâ€™s disease-related pathology. Cell and Tissue Research, 2004. 318(1): p. 121-134.
10.	Hawkes, C.H., K. Del Tredici, and H. Braak, Parkinson's disease: a dual-hit hypothesis. Neuropathology and Applied Neurobiology, 2007. 33(6): p. 599-614.
11.	Snead, D. and D. Eliezer, Alpha-Synuclein Function and Dysfunction on Cellular Membranes. Experimental Neurobiology, 2014. 23(4): p. 292-313.
12.	Blauwendraat, C., M.A. Nalls, and A.B. Singleton, The genetic architecture of Parkinson's disease. The Lancet Neurology, 2020. 19(2): p. 170-178.
13.	VÃ¡zquez-VÃ©lez, G.E. and H.Y. Zoghbi, Parkinson's Disease Genetics and Pathophysiology. Annual Review of Neuroscience, 2021. 44(1): p. 87-108.
14.	Nalls, M.A., et al., Identification of novel risk loci, causal insights, and heritable risk for Parkinson's disease: a meta-analysis of genome-wide association studies. The Lancet Neurology, 2019. 18(12): p. 1091-1102.
15.	Diaz-Ortiz, M.E., et al., GPNMB confers risk for Parkinsonâ€™s disease through interaction with Î±-synuclein. Science. 377(6608): p. eabk0637.
16.	Chang, D., et al., A meta-analysis of genome-wide association studies identifies 17 new Parkinson's disease risk loci. Nature Genetics, 2017. 49(10): p. 1511-1516.
17.	Murthy, M.N., et al., Increased brain expression of GPNMB is associated with genome wide significant risk for Parkinsonâ€™s disease on chromosome 7p15.3. neurogenetics, 2017. 18(3): p. 121-133.
18.	Ardlie, K.G., et al., The Genotype-Tissue Expression (GTEx) pilot analysis: Multitissue gene regulation in humans. Science, 2015. 348(6235): p. 648-660.
19.	Henderson, M.X., et al., Spread of Î±-synuclein pathology through the brain connectome is modulated by selective vulnerability and predicted by network analysis. Nature Neuroscience, 2019. 22(8): p. 1248-1257.
20.	Shimozawa, A., et al., Propagation of pathological Î±-synuclein in marmoset brain. Acta Neuropathologica Communications, 2017. 5(1).




