---
title: "BMIN503 Final Project"
author: "Junchao Shen"
date: "12/10/2021"
output: html_document
---

```{r setup , include=TRUE,message = FALSE}
setwd('/Users/junchaoshen/UPenn BMI')
library(data.table)
library(dendextend)
library(stabs)
library(lars)
library(GGally)
library(mlbench) # for cross validation
library(caret)
library(ggplot2)
library(caret) # LR model training
library(dplyr) # data manipulation
library(ROCR) #for plotting ROC
library(e1071)
library(ggpubr) # for plotting
library(tidyverse) #for data manipulation
library(caret) #for cross validation
library(ROCR) #for ROC
library(pROC) #for generating ROC curves
library(lme4) # for LME model
library(survival) # for survivalan alysis
library(survminer) #for survival analysis
library(ggfortify) # for plotting 
library(gridExtra)
library(grid)
library(sjPlot)
library(sjmisc)
library(nlme)
library(lme4)
library(plyr)
library(knitr)
library(kableExtra)
options(warn = -1)
``` 


## Background and Introduction

BMIN 503


### Overview

Large-scale proteomic analysis identified both diagnostic and prognostic plasma biomarkers for Parkinson's Disease 

### Introduction 

Background: Parkinson's Disease (PD) is the second most common neurodegenerative disease affecting over 2 million people in the United States. Patients with Parkinson's Disease demonstrated a series of cognitive, motor and psychotic symptoms. However, up to date, there still weren't any cliniaclly validated blood based diagnostic test that can quickly help clinicians to confirm the diagnosis of PD. The fact that different Parkinson's patients displayed highly heterogenous sypmtoms raised further challenges to the development of such tests. Therefore, the goal of this project is to identify any potential biomarkers associated with the diagnosis and prognosis of PD, and based on that trained a classifier that can 1) differente PD patients from non-PD patients 2) different PD patients with a faster progression rate from slowly progressing PD patients with high performance.

Goal: This project aims to conduct a large-scale proteomic analysis to screen potential plasma proteins that can A) differentiate Parkinson's patients from normal healthy controls B) differentiate fast cognitive progressors from slow cognitive progressors based on change in MoCA (Montreal Cognitive Assessment) score change over time C) differentiate fast motor progressors from slow motor progressors based on change in UPDRS-III (Unified Parkinson's Disease Rating Scale) score over time

Faculty and Staff Mentor: 

Alice Chen-Plotkin Professor of Neurology

Marijan Posavi Senior Research Scientist


### Methods
Data Source: Datasets used for this project were downloaded from NIH Parkinson's Disease Biomarker Program database. A total of 319 PD + Normal Control Subjects were enrolled in this study through two different cohorts - University of Texas Southwestern School of Medicine and Penn State Hershey School of Medicine. A total of ~1100 plasma proteins were measured in all 319 subjects through a high through-put proteomic platform known as SomaScan. Clinical and demographic information for these patients were also collected at each cohort.

Analysis Plan:

A. Cross Section Analysis

Data Cleaning and Outlier Detection Since those samples were collected from two different cohorts, it would be helpful to start with a principle component analysis (PCA) to detect any potential batch effect. Besides, PCA analysis can help us quickly identify if there's any significant outliers so that we can remove outliers before formal analysis

Differential Expression Analysis Compare the proteins expression level between normal control and subjects diagnosed with Parkinson's Disease. Benjamini-Hochberg method was used to adjust for multiple comparison. Proteins that pass the adjusted p-value of 0.05 were considered as proteins that were significantly associated with the phenotype of Parkinson's Disease. Fold change was calculated using log2(Control-PD). Vocalno plot was used to visualize the difference in protein expression between PD and control group.

3). Gene Set Enrichment Analysis (GSEA)

After identifying plasma proteins that differentiate PD from control, we created a dictionary to map protein names into their corresponding gene symbols. Using KEGG pathway files, GESA analysis was conducted to identify any potential underlying mechanisms associated with the proteins identified in the 2) and their associated genes.

4). Classifcation Analysis

Using the top 3 proteins identified through the differential expression analysis, we next trained a logistic-regression based classification model, validated with 5-fold cross validation over 50 iterations, to understand how well the top 3 proteins we identified - BSP, OMD and ApoM - together with age and sex, can help differentiate PD subjects from control subjects.

B) Longitudinal Analysis (Cognitive Function)

Linear-mixed effects model was used to evaluate the effect of the top protein biomarkers identified in (A) on change in MoCA score over time, after age and sex were adjusted for.

Model: MoCA ~ time x top proteins + age + sex

C) Longitudinal Analysis (Motor Function)

Linear-mixed efefcts model was used to evaluate the effect of the top protein biomarekrs indentified in (A) on change in UPDRS-III score over time, after age and sex were adjusted for

Model: UPDRS ~ time x top proteins + age + sex


### Results

Differential Expression Analysis ended up identifying a total of 47 proteins that are differentially expressed between PD and normal. With only 8 proteins, the model already start to overfitting as you can see that the model with 8 proteins + age showed an AUC score of 0.78 on training cohort but only less than 0.6 on training cohort. However, it is also possible that there existed a significant batch effect when proteins were measured in samples in two different batches. Clustering analysis clearly showed the presence of co-linearlities between proteins, therefore it might be helpful to use some kind of feature selection algorithms to further reduce the dimensions before training the model. It is very interesting that KEGG pathway enrichment analysis showed that chemokine signaling seemed to be most significantly enriched. However, downstream manupilative experiments are needed to understand the biological meaning of that pathway. It is also very interesting that a lot of proteins showed up in the DE analysis were involved in the inflammatory reaction, like C-reactive protein. This might suggested that PD patients have a higher systemic inflammation level compared to normal control, which is consistent with the wide observation that microglia activation level in higher in the CNS of PD patients than normal control. 

Furthermore, generalized linear mixed effect model with random intercept was used to study the effect of top 20 proteins on the long-term clinical outcomes (both on cognitive fucntion and on motor function). Out of the top 20 proteins, FCN1 and BSP seemed to be potential prognostic biomarkers that predict cognitve function. Patients who are in top tertile of BSP measurements developed cognitive function decline much more rapidly compared to patients in the lower two tertiles. Patients who are in the first tertile (the lowest tertile of FCN1 measuremnts) seem to have a faster cognitive function decline than those who are in the first two tertiles. Out of top 20 proteins, IFGBP.2 level seems to be correlated with the incidents of developing dementia according to the Cox Proportional Hazard model. Specifically, patients in the fourth quratile of IGFBP.2 measurements are 7 times more likely to develop dementia compared to those in the lowest quartile. 

In addition, longitudinal motor function analysis suggested that NCAM.120 might be a potential biomaker for motor function over time. Patients who have lower NCAM 120 seemed to have a slower motor function progression (not very clear difference though). Also, since it looks like patients with different baseline NCAM 120 measurements have different baseline MoCA score as well, it might be useful for us to include baseline MoCA as a covariant in the LME model as well in the future.

The limit of this study is that this is an entirely analytic project, therefore, no causal relationship can be inferred. A quick screen at the top 20 proteins didn't identify any significant pQTL effect, which made the Mendelian Randomization (MR) analysis planned in the beginning impossible to conduct. Also, since Somalogic is a new proteomic platform, it remains unknown how reliable the protein measurements are. Therefore, it might be interesting to use ELISA essay to validate some of the proteomic measurements to see how valid they are.  




# Quick Summary of Replication Cohort

```{r preprocessing, echo=TRUE}
# Read the proteomic dataset
setwd('/Users/junchaoshen/UPenn BMI')
df = read.csv('Protein Demographic.csv',skip = 6)
kable(df[1:5,], "html") %>% kable_styling("striped") %>% scroll_box(width = "100%")
# Let's extract the columns that include proteine expression only
protein = df[9:1313]
# Let's first normaliz the dataset using log10 transformation
protein.log = log10(protein)
# Let's first do some quality control analysis to identify potential outliers
dat.pca <- prcomp(protein.log,cor=F)
protein.log$Site = df$Site
autoplot(dat.pca)
# Based on the PCA plot,looks like there's one significant outlier. let's remove that outlier before further analysis
protein.log = protein.log[-309,]
df = df[-309,]
#Now, we can add clinical variables back to the pre-processed protein values
protein.log$GUID = df$GUID
protein.log$Group = df$Group
protein.log$Site = df$Site
protein.log$Age = df$Age
protein.log$LDOPA = df$LDOPADoseTm
protein.log$UPDRS = df$UPDRS
```
#  Differential Expression Analysis

Now, let's move on to compare the proteomic expression level difference betwen normal control and Parkinson's patients

```{r DE_Analysis, echo=TRUE,message = FALSE}
setwd('/Users/junchaoshen/UPenn BMI')
library(GEOquery)
library(oligo)
library(limma)
library(viridis)
library(GEOquery)


design <- model.matrix(~ -1 + factor(protein.log$Group))
colnames(design) <- levels(factor(protein.log$Group))
fit <- lmFit(t(protein.log[,1:1305]), design)
fit.contrast <- makeContrasts(diff = Control - PD, 
                                  levels = design)
fit2 <- contrasts.fit(fit, fit.contrast)
fit2 <- eBayes(fit2)
exposure_results <- topTable(fit2, coef = "diff", adjust = "BH", num = Inf)
exposure_results[exposure_results$sig == 'significant',]
library(ggplot2)
library(ggrepel)
# Assign (in)significant genes
exposure_results$sig <- rep("insignificant", nrow(exposure_results))
exposure_results$sig[which(exposure_results$adj.P.Val<0.05)] <- "significant"
v1 = ggplot(exposure_results, aes(x = logFC, y = -log10(adj.P.Val), color = sig)) + 
  geom_point() +
  theme_bw() +
  ggtitle("Volcano plot") +
  xlab("LogFC")+
  ylab("-Log10(q-value)") +
  scale_color_manual(values = c("black", "red")) +
  theme(legend.position = "none")

v1+geom_text_repel(data=head(exposure_results, 40), aes(label=rownames(exposure_results)[1:40]))

dict = read.csv('Protein Demographic.csv',skip = 1)

dict1 = t(dict[4:5,])[-1:-8,]
rownames(dict1) = NULL
colnames(dict1) = c('SYMBOL','Protein.Name')
exposure_results$Protein.Name = rownames(exposure_results)
protein_gene = merge(exposure_results,dict1,by = 'Protein.Name')
DE_res = protein_gene
write.table(exposure_results, "./Protein_DE_results.txt", row.names = F, 
            col.names = T, quote = F, sep = "\t")

df_boxplot1 = data.frame(
  expression = protein.log[,'BSP'],
  status = protein.log$Group)

ggplot(df_boxplot1, aes(x = status, y = expression)) +
  geom_boxplot(outlier.colour = NA, color = "grey18", fill = "lightblue") +
  stat_boxplot(geom = "errorbar", color = "grey18") +
  geom_jitter(size = 1, position = position_jitter(width = 0.3)) +
  ggtitle("BSP Expression Level between PD and Control") +
  xlab(" ") +
  ylab("BSP Expression Level (RFU)") +
  theme_bw() +
  theme(legend.position = "none")

df_boxplot2 = data.frame(
  expression = protein.log[,'OMD'],
  status = protein.log$Group)

ggplot(df_boxplot2, aes(x = status, y = expression)) +
  geom_boxplot(outlier.colour = NA, color = "grey18", fill = "lightblue") +
  stat_boxplot(geom = "errorbar", color = "grey18") +
  geom_jitter(size = 1, position = position_jitter(width = 0.3)) +
  ggtitle("OMD Expression Level between PD and Control") +
  xlab(" ") +
  ylab("OMD Expression Level (RFU)") +
  theme_bw() +
  theme(legend.position = "none")

df_boxplot3 = data.frame(
  expression = protein.log[,'ApoM'],
  status = protein.log$Group)

ggplot(df_boxplot3, aes(x = status, y = expression)) +
  geom_boxplot(outlier.colour = NA, color = "grey18", fill = "lightblue") +
  stat_boxplot(geom = "errorbar", color = "grey18") +
  geom_jitter(size = 1, position = position_jitter(width = 0.3)) +
  ggtitle("ApoM Expression Level between PD and Control") +
  xlab(" ") +
  ylab("ApoM Expression Level (RFU)") +
  theme_bw() +
  theme(legend.position = "none")

library(gplots)
# keep top 200 genes
top_results <- subset(exposure_results,sig =='significant')
top.eset <- protein.log[colnames(protein.log) %in% row.names(top_results)]
status.colors <- unlist(lapply(protein.log$Group, 
                               function (x) {if (x == "PD") "red"  
                                 else "navy"}))
heatmap.2(t(top.eset), col = viridis(256, option = "B"),
          trace = "none", keysize = 1.5, key.title = NA,
          ColSideColors = status.colors)
legend("topright", legend = c("PD","Control"), fill = c("red", "navy")) 

library(fgsea)
kegg <- gmtPathways("./c2.cp.kegg.v7.4.symbols.gmt")
reactome <- gmtPathways("./c2.cp.reactome.v7.4.symbols.gmt")
pathways <- c(kegg, reactome)
head(DE_res)

library(dplyr)
DE_res <- DE_res %>%
  filter(!is.na(SYMBOL)) %>%
  group_by(SYMBOL) %>%
  arrange(P.Value) %>%
  filter(row_number() == 1) %>%
  data.frame()

gene_list <- DE_res$t
names(gene_list) <- DE_res$SYMBOL
gene_list <- sort(gene_list, decreasing = T)
head(gene_list, 5)


options(warn=-1)
fgseaRes <- fgsea(pathways = pathways, stats = gene_list, minSize = 15, 
                 maxSize = 500, nperm = 10000, gseaParam = 1)
fgseaRes <- fgseaRes[order(fgseaRes$pval), ]

collapsedPathways <- collapsePathways(fgseaRes = fgseaRes[padj < 0.05], 
                                      pathways = pathways, stats = gene_list)
mainPathways <- fgseaRes[pathway %in% collapsedPathways$mainPathways] # keep results of independent pathways

top_pathway <- mainPathways %>%
  filter(padj < 0.05 & abs(NES) > 1) %>%
  arrange(NES)

top_pathway$pathway <- factor(top_pathway$pathway, levels = top_pathway$pathway)
ggplot(top_pathway, aes(y = NES, x = pathway)) +
  geom_bar(width = 0.8, position = position_dodge(width = 0.8), stat = "identity", fill = "blue") + 
  coord_flip() +
  theme_bw() +
  theme(
    axis.title.y = element_blank()
  )

plotEnrichment(pathway = pathways[["KEGG_CHEMOKINE_SIGNALING_PATHWAY"]], stats = gene_list) + 
  labs(title="KEGG_CHEMOKINE_SIGNALING_PATHWAY")

top_pathway %>% 
  filter(pathway=="KEGG_CHEMOKINE_SIGNALING_PATHWAY") %>% 
  dplyr::select(leadingEdge) %>% 
  unlist() %>% 
  unname()
         

```

#  Classification Model of the Disease (PD vs Normal)

Let's train the model using top 8 proteins + clinical variates like age + sex +
disease duration

```{r Classification, echo=TRUE,message = FALSE}
setwd('/Users/junchaoshen/UPenn BMI')
library(caret)
library(e1071)
library(precrec)
library(ROCit)

utsw = protein.log[protein.log$Site == 'UTSW',]
penn = protein.log[protein.log$Site == 'PennState',]


### Let's train the model using 4 proteins that passed the p value threshold of 0.05 on both cohorts
train_control <- trainControl(method = "repeatedcv", number = 5, repeats = 50)
model <- train(as.factor(Group) ~
                        Age+
                        ApoM+
                        BSP+
                        OMD+
                        NCAM.120+
                        IDS+
                        IGFBP.2+
                        FCN1,
                        method = "glm",
                        data = utsw,
                        na.action=na.omit,
                        family = binomial)

# Let's take a look at the final model
summary(model)

# Plot ROC curve with AUC score
options(repr.plot.width=8, repr.plot.height=8)
predict_train1 <- predict(model, newdata = utsw, type= "prob")
predict_test1 <- predict(model, newdata = penn, type= "prob")
precrec_obj1 <- rocit(class = utsw$Group, score = predict_train1$PD)
precrec_obj2 <- rocit(class = penn$Group, score = predict_test1$PD)
plot(precrec_obj1,col = 'Red')
plot(precrec_obj2,col = 'Blue')

summary(precrec_obj1)
summary(precrec_obj2)
```

# Longitudinal Analysis

MoCA ~ Top 20 proetins quartile X Days from Baseline + Age + Sex + Disease Duration

```{r Cognitive, echo=TRUE,message = FALSE}
library(sjPlot)
library(sjmisc)
library(nlme)
library(lme4)

Longitudinal = read.csv('UTSW MoCA Longitudinal Final.csv')
kable(Longitudinal[1:5,], "html") %>% kable_styling("striped") %>% scroll_box(width = "100%")

Longitudinal$BSP.Tertile <- with(Longitudinal, factor(
                            findInterval( BSP, c(-Inf,
                               quantile(BSP, probs=c(0.33,0.66)), Inf)), 
                            labels=c("First Tertile","Second Tertile","Third Tertile")))

# Train LME model to see how fasat and slow progressors differ
MoCA1 = lmer(MoCATotal ~ BSP.Tertile*YearstoVisit + AgeAtSample.x + Sex + as.numeric(DiseaseDuration)+ (1 | GUID), data=Longitudinal, REML=F)

data(efc)
theme_set(theme_sjplot())

p5 = plot_model(MoCA1, type = "pred", terms = c("YearstoVisit", "BSP.Tertile"),
           title = " Cognitive Change \n (UTSW + Penn State Cohort)",
           axis.title = c("YearstoVisit","Cognition(MoCA)"),
           axis.labels = NULL, legend.title = "BSP Tertile",
           wrap.title = 50, wrap.labels = 25, axis.lim = NULL, grid.breaks = NULL,
           ci.lvl = 0.9, se = NULL, colors = "Set2", order.terms = c(3,1,2)) + scale_y_continuous(limits = c(20, 30), breaks = c(20,22,24,26,28,30)) + theme_classic() +
  theme(legend.position = c(0.25, 0.2), legend.title = element_text(face = "bold", size = 16), legend.text = element_text(size = 20),
        title = element_text(size = 23),
        axis.title.x = element_text(size = 25, margin = margin(t = 10, r = 0, b = 0, l = 0)),
        axis.title.y = element_text(size = 25, margin = margin(t = 0, r = 12, b = 0, l = 0)),
        axis.text = element_text(size = 25, color = "black"), axis.line = element_line(size = 1), axis.ticks = element_line(size = 1))

p5

Longitudinal$OMD.Tertile <- with(Longitudinal, factor(
                            findInterval( OMD, c(-Inf,
                               quantile(OMD, probs=c(0.33,0.66)), Inf)), 
                            labels=c("First Tertile","Second Tertile","Third Tertile")))

# Train LME model to see how fasat and slow progressors differ
MoCA = lmer(MoCATotal ~ OMD.Tertile*YearstoVisit + AgeAtSample.x + Sex + as.numeric(DiseaseDuration)+ (1 | GUID), data=Longitudinal, REML=F)

data(efc)
theme_set(theme_sjplot())

Longitudinal$FCN1.Tertile <- with(Longitudinal, factor(
                            findInterval( FCN1, c(-Inf,
                               quantile(FCN1, probs=c(0.33,0.66)), Inf)), 
                              labels=c("First Tertile","Second Tertile","Third Tertile")))

# Train LME model to see how fasat and slow progressors differ
MoCA = lmer(MoCATotal ~ FCN1.Tertile*YearstoVisit + AgeAtSample.x + Sex + as.numeric(DiseaseDuration)+ (1 | GUID), data=Longitudinal, REML=F)

data(efc)
theme_set(theme_sjplot())

p5 = plot_model(MoCA, type = "pred", terms = c("YearstoVisit", "FCN1.Tertile"),
           title = " Cognitive Change \n (UTSW + Penn State Cohort)",
           axis.title = c("YearstoVisit","Cognition(MoCA)"),
           axis.labels = NULL, legend.title = "FCN1 Tertile",
           wrap.title = 50, wrap.labels = 25, axis.lim = NULL, grid.breaks = NULL,
           ci.lvl = 0.9, se = NULL, colors = "Set2", order.terms = c(3,1,2)) + scale_y_continuous(limits = c(20, 30), breaks = c(20,22,24,26,28,30)) + theme_classic() +
  theme(legend.position = c(0.25, 0.2), legend.title = element_text(face = "bold", size = 16), legend.text = element_text(size = 20),
        title = element_text(size = 23),
        axis.title.x = element_text(size = 25, margin = margin(t = 10, r = 0, b = 0, l = 0)),
        axis.title.y = element_text(size = 25, margin = margin(t = 0, r = 12, b = 0, l = 0)),
        axis.text = element_text(size = 25, color = "black"), axis.line = element_line(size = 1), axis.ticks = element_line(size = 1))

p5

Longitudinal$CRP.Tertile <- with(Longitudinal, factor(
                            findInterval( CRP, c(-Inf,
                               quantile(CRP, probs=c(0.33,0.66)), Inf)), 
                            labels=c("First Quartile","Second Quartile","Third Quartile")))


# Train LME model to see how fasat and slow progressors differ
MoCA = lmer(MoCATotal ~ CRP.Tertile*YearstoVisit + AgeAtSample.x + Sex + as.numeric(DiseaseDuration)+ (1 | GUID), data=Longitudinal, REML=F)

data(efc)
theme_set(theme_sjplot())

p5 = plot_model(MoCA, type = "pred", terms = c("YearstoVisit", "CRP.Tertile"),
           title = " Cognitive Change \n (UTSW + Penn State Cohort)",
           axis.title = c("YearstoVisit","Cognition(MoCA)"),
           axis.labels = NULL, legend.title = "CRP Tertile",
           wrap.title = 50, wrap.labels = 25, axis.lim = NULL, grid.breaks = NULL,
           ci.lvl = 0.9, se = NULL, colors = "Set2", order.terms = c(3,1,2)) + scale_y_continuous(limits = c(20, 30), breaks = c(20,22,24,26,28,30)) + theme_classic() +
  theme(legend.position = c(0.25, 0.17), legend.title = element_text(face = "bold", size = 16), legend.text = element_text(size = 20),
        title = element_text(size = 23),
        axis.title.x = element_text(size = 25, margin = margin(t = 10, r = 0, b = 0, l = 0)),
        axis.title.y = element_text(size = 25, margin = margin(t = 0, r = 12, b = 0, l = 0)),
        axis.text = element_text(size = 25, color = "black"), axis.line = element_line(size = 1), axis.ticks = element_line(size = 1))

```

# Survival Analysis 

Clinical Conversion into Dementia ~ Top 20 protein teritile + Age + Sex + Disease Duration

```{r Survival, echo=TRUE,message = FALSE}
setwd('/Users/junchaoshen/UPenn BMI')
survival.upenn = read.csv('Survival Analysis Replication.csv')
kable(survival.upenn[1:5,], "html") %>% kable_styling("striped") %>% scroll_box(width = "100%")
survival.upenn$YearsVisitToSample = survival.upenn$DaysVisitToSample/365

# Create survival object
survive_object = Surv(survival.upenn$YearsVisitToSample, survival.upenn$Event)

# Equally binning patients into quartiles based on their blood MIA concentration
# Convert the MIA factor variables into factor variables
survival.upenn$IGFBP.2_Factor <- with(survival.upenn, factor(
                            findInterval( IGFBP.2, c(-Inf,
                               quantile(IGFBP.2, probs=c(0.25,0.5,0.75)), Inf)), 
                            labels=c("First Quartile","Second Quartile","Third Quartile","Fourth Quartile")
      ))
survival.upenn$IGFBP.2_Factor = as.factor(survival.upenn$IGFBP.2_Factor)

IGFBP.2_Factor = coxph(survive_object ~ AgeAtSample + Sex + DieseaseDuration + IGFBP.2_Factor,data = survival.upenn)

# Take a look at the summary of Cox model
summary(IGFBP.2_Factor)

coxCurve_IGFBP.2 <- ggadjustedcurves(IGFBP.2_Factor, data = survival.upenn, method = "average", variable = "IGFBP.2_Factor", legend.title = "Plasma IGFBP.2 level Quartile") + 
  labs(title = "Clinical Conversion by IGFBP.2 subgroup \n (UPenn Cohort)", x = "Time (Years)", y = "% Cognitively Normal")  + 
  theme_classic() +
  theme(legend.position = c(0.35, 0.3), legend.title = element_text(face = "bold", size = 18), legend.text = element_text(size = 18),
        title = element_text(size = 18),
        axis.title.x = element_text(size = 20, margin = margin(t = 10, r = 0, b = 0, l = 0)),
        axis.title.y = element_text(size = 20, margin = margin(t = 0, r = 12, b = 0, l = 0)),
        axis.text = element_text(size = 20, color = "black"), axis.line = element_line(size = 1), axis.ticks = element_line(size = 1))
coxCurve_IGFBP.2

fit_coxph_MIAcog.forest <- ggforest(IGFBP.2_Factor , data = survival.upenn, main = "Hazard Ratios", 
                          cpositions = c(0.02, 0.23, 0.4), fontsize = 1, refLabel = "reference", noDigits = 2)
fit_coxph_MIAcog.forest
```

# Longitudinal Motor Function Analaysis 

UPDRS III (Motor Function) ~ Age + Sex + Disease Duration

```{r Motor, echo=TRUE,message = FALSE}
setwd('/Users/junchaoshen/UPenn BMI')
UTSW.UPDRS = read.csv('UTSW UPDRS Longitudinal Final.csv')
UTSW.UPDRS.merge = merge(UTSW.UPDRS,df, by = 'GUID',all.x = TRUE)
kable(UTSW.UPDRS.merge[1:5,], "html") %>% kable_styling("striped") %>% scroll_box(width = "100%")
UTSW.UPDRS.merge$NCAM.120.Tertile <- with(Longitudinal, factor(
                            findInterval( NCAM.120, c(-Inf,
                               quantile(NCAM.120, probs=c(0.33,0.66)), Inf)), 
                            labels=c("First Tertile","Second Tertile","Third Tertile")))

# Train LME model to see how fasat and slow progressors differ
UPDRS = lmer(UPDRS3 ~ NCAM.120.Tertile*YearstoVisit + AgeAtTest + Sex + as.numeric(DiseaseDuration)+ (1 | GUID), data=UTSW.UPDRS.merge, REML=F)

data(efc)
theme_set(theme_sjplot())

p5 = plot_model(UPDRS, type = "pred", terms = c("YearstoVisit", "NCAM.120.Tertile"),
           title = " Cognitive Change \n (UTSW + Penn State Cohort)",
           axis.title = c("YearstoVisit","Cognition(MoCA)"),
           axis.labels = NULL, legend.title = "NCAM.120 Tertile",
           wrap.title = 50, wrap.labels = 25, axis.lim = NULL, grid.breaks = NULL,
           ci.lvl = 0.9, se = NULL, colors = "Set2", order.terms = c(3,1,2)) + scale_y_continuous(limits = c(0, 30), breaks = c(0,10,20,30,40)) + theme_classic() +
  theme(legend.position = c(0.25, 0.2), legend.title = element_text(face = "bold", size = 16), legend.text = element_text(size = 20),
        title = element_text(size = 23),
        axis.title.x = element_text(size = 25, margin = margin(t = 10, r = 0, b = 0, l = 0)),
        axis.title.y = element_text(size = 25, margin = margin(t = 0, r = 12, b = 0, l = 0)),
        axis.text = element_text(size = 25, color = "black"), axis.line = element_line(size = 1), axis.ticks = element_line(size = 1))

p5

```

