# BMIN503/EPID600 Final Project


This repository contains:

-gesualdiJ_503_Final.Rmd
    -R Markdown file containing code, plots, and descriptions of the processes outlined below (Methods and Results)
    -This file is still incomplete. All of the code for the methods section (eg accessing, cleaning, normalization to prepare for Exploratory and DGE analysis) is completed but needs further comments     / descriptions and work on the Blurton dataset. The code and associated figures for the results section is not ready. 

-4 Datasets accessed via the National Center for Biotechnology Information's Gene Expression Omnibus (GEO) Data Repository
    -GSE89189_RAW(1).tar: Abud, IPSC Microglia monocultured or cultured with neurons (archs4)
    -GSE139194_RAW.tar: Svoboda, IPSC Microglia either transplanted into the brain of humanized mice or cultured in vitro (archs4) (unable to upload)
    -GSE99047_HumanMicrogliaBrainCounts.txt: Eggen, Reference data set: ExVivo microglia sequenced directly following surgical resection/isolation from adult patients (_tsv -> EnsDb)
    -GSE133432_Hasselmann_2019_BulkSeqCounts.tsv: Blurton Jones, IPSC microglia transplanted into the brain of humanized mice
    

-2 .xlsx files containing data previously generated by colleagues in either thesis lab or that of a collaborator
    -KJS_iMG_MDM_TMM_Normalized.xlsx: Bulk sequencing data generated in my lab group for a previous publication. These data were previously cleaned and pre-processed for differential gene expression     analysis in a comparable manner to the manipulation of the other datasets seen in the .Rmd File. Therefore, the code involved was excluded for brevity
    -Human_RNA-seq_data_Gosselin_Fetal_ExVivo.xlsx: Gosselin, Reference data set. ExVivo microglia sequenced directly following surgical resection/isolation from adult patients (.xlsx straight to        cleaning) 

-Various scripts (.R) that were used to build or test different parts of the methodology / results

-A Note: This project is limited in its scope, but could serve as a starting point for a more comprehensive literature review and meta-analysis of different approaches used to generate IPSC-microglia including various coculture methods and integration into three dimensional culture systems such as neuronal organoids or spheroids. 

Overview:

My project will be a limited meta-analysis of existing bulk RNAseq data sets detailing the differentiation of induced-pluripotent stem cells (IPSCs) into microglia, a key population of resident immune cells in the central nervous system (CNS). IPSC-based modeling is the future of microglia biology in vitro, thus there are many competing standards - eg different IPSC-derived microglia produced by different groups. My analysis will compare the transcriptomes of these competing microglia to a "Gold Standard" or reference transcriptomic dataset generated using microglia isolated from post-mortem biopsies from human donors. 


Introduction:

Microglia have long been challenging to model in-vitro using conventional approaches such as immortalized cell lines or primary culture because their morphology, function, and (more recently appreciated) transcriptome rapidly deteriorate when cultured in isolation. To address this issue, numerous groups have developed microglial cell lines derived from IPSCs through supplementation of cultures with various factors normally present in the micro-encironment of the CNS. The goal of this project is to interrogte the merit of the various published IPSC-derived microglia lines. 

Addressing this problem involves suject matter knowledge of both immunology and neuroscience, as well as various bioinformatic techniques. This analysis will include only a subset of published microglial lines, prioritizing those models that myself and Drs Schaletzki and Su considered influential. The reference transcriptomes - one from adult donors and one generated from analysis of fetal tissue - add rigor to this project because they represent the closest that we can realistically hope to get to the in vivo gene expression of microglia. Finally, this analysis will be useful to microglia biologists because while many transcriptomic data sets exist, there have been few if any comprehensive meta-analyses of these data and still fewer attempts to harmonize their insights. This project aims to make progress in that direction. 

Methods:

* Gosselin 2017 reference (.xlsx from johannes) as reference for ex-vivo fetal microglia, Bart J L Eggen 2017 as reference for ex-vivo adult microglia (eg from geo)
    * geo: https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE99074 <- Eggen 2017
        * need to disambiguate sample labels for Eggen 2017 with the stupid donor info file, going to be a pain in the ass
    * Use ready-processed data from Johannes, but have act 1 of the final project .RMD include all the work of cleaning / filtering and pre-processing this other reference dataset, for the exercise (eg basically just copy script 4 from diy into this project lol)
        * then, for the other microgia that you include later, you will still have to do that same cleaning, but just leave that boring mess out of the final project file I feel like. you can do an accessory .Rmd that shows you cleaning all the other datasets but it would be cumbersome to include all of them in the main file 

* Compare to our IPSC, IPSC in the abud et al paper we always do, IPSC in the yijing / svoboda transplant paper, IPSC in johannes / blurton jones transplant paper (includes geos for abud!!), include an organoid data set if we want later
    * Abud 2017 Geo: GSE89189
        * https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE89189 
    * Geo Blurton jones 2019 et al: GSE133432
        * https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE133432 
    * Svoboda 2019 geo: GSE139194
        * https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE139194 
        
Cleaned data using either base R or dplyr functions. U


Results




This repository contains templates for your final written report and GitHub repository. Follow the instructions below to clone this repository, and then turn in your final project's code via a pull request to this repository.


1. To start, **fork** this BMIN503_Final_Project repository.
1. **Clone** the forked repository to your computer.
1. Modify the files provided, add your own, and **commit** changes to complete your final project.
1. **Push**/sync the changes up to your GitHub account.
1. Create a **pull request** on this, the original BMIN503_Final_Project, repository to turn in your final project.


Follow the instructions [here][forking] if you are unsure what the above steps mean.

DUE DATE FOR FINAL VERSION: 12/09/22 11:59PM. This is a hard deadline. Turn in whatever you have by this date.


<!-- Links -->
[forking]: https://guides.github.com/activities/forking/

