# BMIN503/EPID600 Final Project

This repository includes files for analyzing TCGA RNA sequencing data for hepatocellular carcinoma to assess gene expression of interferon target gene sets. This project was generated by Caitlin Foley while working in Andy Minn's laboratory at University of Pennsylvania for the BMIN503 course.

TCGA_LIHC_RNAseq_normalized_counts_DESeq2_vst.txt: A truncated dataframe (40 samples) for purpose of code check.  Normalized data were generated by Jingya Qui from Andy Minn's laboratory at University of Pennsylvania.  HTseq Counts data from Genomic Data Commons portal (https://portal.gdc.cancer.gov/) was downloaded 4/10/2019, then normalized using DESeq2 and VST.

gdc_manifest_LIHC.txt: manifest used to download files from GDC and functions as key to covert the sample names (derived from GDC filename) to standard ID to merge with patient metadata

ISG.RS.txt: List of genes by HGNC name in interferon-target resistance signature (Weichselbaum, et al. PNAS, 2009)
IFNG.GS.txt: List of genes by HGNC name in interferon-target immunotherapy response signatures (Benci et al. Cell, 2019)

metadata_Liu2018.xlsx: HCC patient metadata (tumor grade, survival, demographics, etc) that was compiled from TCGA JSON files by Liu et al and published in Cell in 2018.

Final_project_template.Rmd: takes TCGA normalized RNA count dataframe, converts rownames to patient UUID to correlated with metadata file.  Examines ISG.RS and IFNG.GS gene sets and genereates heat maps of these gene sets.  Divdes cohort into high vs low gene set signatures and plots survival based on this.
